{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3be52c6f-6647-4d93-b0bf-3e05e87917e1",
   "metadata": {},
   "source": [
    "# Reprogramming Recepies\n",
    "\n",
    "Auth: [Joshua Pickard](jpic@umich.edu)\n",
    "\n",
    "Date: August 26, 2024"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f959f08-1c73-41f8-90b1-565f03fd6d8e",
   "metadata": {},
   "source": [
    "# Day 8\n",
    "\n",
    "Generatting the perturbation files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "baf3a364-df74-4aac-981d-2088dec6a2b2",
   "metadata": {},
   "source": [
    "## Set up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f12a82bf-b175-4cf9-b4ee-432fd13a0842",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import scanpy as sp\n",
    "import os\n",
    "\n",
    "def iterate_perturb_counts(adata, tf_list, scalar_list):\n",
    "    \"\"\"\n",
    "    Applies perturbations to the expression data of specified transcription factors across multiple scalars \n",
    "    and stores the resulting AnnData objects in a dictionary.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Iterates over a list of scalar values.\n",
    "    2. For each scalar, creates a copy of the AnnData object to preserve the original data.\n",
    "    3. Applies the `perturb_counts` function to scale the expression data of genes listed in `tf_list` by\n",
    "       the maximum gene expression of each cell and the current scalar.\n",
    "    4. Stores the perturbed AnnData object in a dictionary with the scalar as the key.\n",
    "\n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols (transcription factors) to be perturbed.\n",
    "    scalar_list (list): A list of scalar values for scaling the gene expression.\n",
    "    adata (AnnData): The AnnData object containing gene expression data (cells x genes).\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary where keys are scalar values and values are the corresponding perturbed AnnData objects.\n",
    "    \"\"\"\n",
    "    \n",
    "    adata_dict = {}\n",
    "    \n",
    "    for scalar in scalar_list:\n",
    "        # Create a copy of the AnnData object for each scalar value\n",
    "        adata_temp = adata.copy()\n",
    "        \n",
    "        # Apply perturb_counts to the copied AnnData object\n",
    "        perturbed_adata = perturb_counts(adata_temp, tf_list, scalar)\n",
    "        \n",
    "        # Store the perturbed AnnData object in the dictionary with scalar as the key\n",
    "        adata_dict[scalar] = perturbed_adata\n",
    "    \n",
    "    return adata_dict\n",
    "\n",
    "def perturb_counts(adata, tf_list, scalar): \n",
    "    \"\"\"\n",
    "    Applies a perturbation to the expression data of specific genes in an AnnData object.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Computes the maximum gene expression level for each cell.\n",
    "    2. Applies a scaling operation to the expression levels of genes listed in `tf_list`.\n",
    "       - Each entry of these genes in the matrix is multiplied by the maximum expression level \n",
    "         of its respective cell and a specified scalar value.\n",
    "    3. Updates the AnnData object with new columns:\n",
    "       - 'scaled': A boolean column indicating whether each gene is in the `tf_list`.\n",
    "       - 'scaled_by': Contains the scaling factor used for each gene (the product of the maximum \n",
    "         expression level of each cell and the scalar), or `1` if the gene was not in `tf_list`.\n",
    "    \n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols to be perturbed.\n",
    "    scalar (float): The scalar value used to scale the expression levels.\n",
    "    adata (AnnData): The AnnData object containing gene expression data.\n",
    "\n",
    "    Returns:\n",
    "    AnnData: The updated AnnData object with applied perturbations and new columns.\n",
    "    \"\"\"\n",
    "\n",
    "    # Create a boolean mask for genes in tf_list\n",
    "    gene_mask = adata.var['gene_symbol'].isin(tf_list)\n",
    "    \n",
    "    # Save the original state of the parameter objects, in case some tfs do not translate (failsafe)\n",
    "    original_X = adata.X.copy()\n",
    "    original_gene_mask = gene_mask.copy()\n",
    "    \n",
    "    # Compute maximum expression level of each cell\n",
    "    max_exp = np.max(adata.X, axis=1)\n",
    "\n",
    "    \"\"\"This is new today. v \"\"\"\n",
    "    # Raise an error if any of the gene names in tf_list do not match column names (we will manually update these in adata):\n",
    "    missing_genes = [gene for gene in tf_list if gene not in adata.var['gene_symbol'].values]\n",
    "    \n",
    "    if missing_genes:\n",
    "        # Restore original parameter objects\n",
    "        adata.X = original_X\n",
    "        gene_mask = original_gene_mask\n",
    "        raise ValueError(f\"Genes {missing_genes} not found in anndata object\")\n",
    "\n",
    "    else:    \n",
    "        \n",
    "        # Apply the scaling operation to the specified genes\n",
    "        adata.X[:, gene_mask] = max_exp * scalar\n",
    "        \n",
    "        # Add/Update 'scaled' column in var\n",
    "        adata.var['scaled'] = gene_mask\n",
    "        \n",
    "        # Add/Update 'scaled_by' column in var\n",
    "        adata.var['scaled_by'] = scalar  # Default value for genes not in tf_list\n",
    "\n",
    "        # Add/Update 'scaled_by' column in var\n",
    "        adata.obs['U'] = scalar  # Default value for genes not in tf_list\n",
    "    # adata.var = \n",
    "    return adata\n",
    "\n",
    "def validateTFs(TFs, adata):\n",
    "    adata_gene_list = adata.var['gene_symbol'].values.tolist()\n",
    "    for TF in TFs:\n",
    "        if TF not in adata_gene_list:\n",
    "            return False\n",
    "    return True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20c14b1d-f9db-444f-86ef-2c6f92953081",
   "metadata": {},
   "source": [
    "## Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4b6cdc59-7ace-4463-a93d-3851adffca25",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load reprogramming recipes\n",
    "df = pd.read_csv('data/recipe_table_9_6_2024.csv')\n",
    "df.head()\n",
    "\n",
    "# Load firboblast source cells\n",
    "DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/unperturbed\"\n",
    "FILE = \"fibroblast.h5ad\"\n",
    "adata = sp.read_h5ad(os.path.join(DATAPATH, FILE))\n",
    "adata_gene_list = adata.var['gene_symbol'].values.tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c916ecb8-cd29-487e-a9a2-38b34902a260",
   "metadata": {},
   "source": [
    "## Make Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1356c8c2-aa1f-47d1-b140-4d79e9f52b8d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/ASCL1_POU3F2_MYT1L_LMX1A_FOXA2.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/SOX2.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/NEUROG2_SOX11_ISL1_LHX3.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/SOX2_HMGA2.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/ASCL1_PAX6.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/SOX2_PAX6.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/PTF1A.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/SOX2_ASCL1_NEUROG2.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/ASCL1_SOX2.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/FOXM1_SOX2_MYC_SALL4_STAT6.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/AR_SOX2_SMAD3_MYC_JUN_WT1_TAL1_SPI1_RUNX1.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/POU3F2_SOX2_BACH1_AR_PBX1_NANOG.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/MSI1_NEUROG2_MBD2.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/GATA4_MEF2C_TBX5_MESP1_MYOCD.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/GATA4_MEF2C_TBX5_ESRRG_MESP1_ZFPM2_MYOCD.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/ATF5_PROX1_FOXA2_FOXA3_HNF4A.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/PDX1_PAX4_MAFA.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/NFE2_MAFG_MAFK.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/SPI1_IRF8_BATF3.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/MYC_KLF4_SOX9.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/MYOD1_MYCL.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/PRDM16_CEBPB.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/CRX_RAX_NEUROD1.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/MITF_SOX10_PAX3.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/SOX18.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/CRX_RAX_OTX2_NEUROD1.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/NFKB1_LEF1.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/EMX2_HNF1B_HNF4A_PAX8.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/MYOCD.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/ETV2.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/PPARG.h5ad already exist: continue!\n",
      "['HNF4A', 'FOXA3', 'GATA6', 'CDX2']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    file created\n",
      "['MYOCD', 'GATA6', 'MEF2C']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    file created\n",
      "['SNAI2', 'EYA1', 'SIX1']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    file created\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'float' object has no attribute 'split'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 5\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;28mlen\u001b[39m(df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTFs\u001b[39m\u001b[38;5;124m'\u001b[39m])):\n\u001b[1;32m      4\u001b[0m     val \u001b[38;5;241m=\u001b[39m df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTFs\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39miloc[i]\n\u001b[0;32m----> 5\u001b[0m     TFs \u001b[38;5;241m=\u001b[39m \u001b[43mval\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msplit\u001b[49m()\n\u001b[1;32m      6\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m validateTFs(TFs, adata):\n\u001b[1;32m      7\u001b[0m \n\u001b[1;32m      8\u001b[0m         \u001b[38;5;66;03m# Join the TFs list into a string for the filename\u001b[39;00m\n\u001b[1;32m      9\u001b[0m         TFs_str \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(TFs)\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'float' object has no attribute 'split'"
     ]
    }
   ],
   "source": [
    "output_directory = \"/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed\"\n",
    "scalars = [0.5, 0.75, 1.001]\n",
    "for i in range(len(df['TFs'])):\n",
    "    val = df['TFs'].iloc[i]\n",
    "    TFs = val.split()\n",
    "    if validateTFs(TFs, adata):\n",
    "\n",
    "        # Join the TFs list into a string for the filename\n",
    "        TFs_str = \"_\".join(TFs)\n",
    "        \n",
    "        # Generate the file path for saving\n",
    "        file_name = f\"{TFs_str}.h5ad\"\n",
    "        output_path = os.path.join(output_directory, file_name)\n",
    "\n",
    "        # Check if the file already exists\n",
    "        if os.path.exists(output_path):\n",
    "            print(output_path + \" already exist: continue!\")\n",
    "            continue\n",
    "        \n",
    "        print(TFs)\n",
    "        adataDict = iterate_perturb_counts(adata.copy(), TFs, scalars)\n",
    "\n",
    "        # Concatenate all AnnData objects along the observations axis\n",
    "        concatenated_adata = ad.concat(list(adataDict.values()), axis=0)\n",
    "\n",
    "        # This is the line was added today to fix the bug\n",
    "        concatenated_adata.var = adata.var.copy()\n",
    "\n",
    "        # Save reprogramming metadata into the concatenated_adata.obs table\n",
    "        concatenated_adata.obs['Source_cells'] = df['Source cells'].iloc[i]\n",
    "        concatenated_adata.obs['Target_cells'] = df['Target cells'].iloc[i]\n",
    "        concatenated_adata.obs['Treatment'] = df['Treatment'].iloc[i]\n",
    "        concatenated_adata.obs['Species'] = df['Species'].iloc[i]\n",
    "        concatenated_adata.obs['Cell_Transplantation'] = df['Cell Transplantation'].iloc[i]\n",
    "        concatenated_adata.obs['Published_Year'] = df['Published Year'].iloc[i]\n",
    "        concatenated_adata.obs['PMID'] = df['PMID'].iloc[i]\n",
    "        \n",
    "        # Save the concatenated AnnData object to the file\n",
    "        concatenated_adata.write_h5ad(output_path)\n",
    "        print(\"    file created\")\n",
    "\n",
    "print('All recipes complete!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ec7ad249-5d22-488f-8ae6-551b8f1018cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                                     ASCL1 LXMLA NR4A2\n",
       "1                           POU3F2 ASCL1 MYT1L NEUROD11\n",
       "2                        ASCL1 POU3F2 MYT1L LMX1A FOXA2\n",
       "3                                                  SOX2\n",
       "4                    ASCL1 NEUROG2 SOX2 NR4A2 PITX3 P53\n",
       "5                               NEUROG2 SOX11 ISL1 LHX3\n",
       "6                                            SOX2 HMGA2\n",
       "7                                            ASCL1 PAX6\n",
       "8                           POU3F2 ASCL1 MYT1L NEUROD11\n",
       "9     ASCL1 ISL1 NEUROD11 POU3F2 MNX1 LHX3 MYT1L NEU...\n",
       "10                                  SOX2 GATA3 NEUROD11\n",
       "11                                            SOX2 PAX6\n",
       "12                                                PTF1A\n",
       "13        SOX2; SOX2 PAX6; SOX2 LMX1A; SOX2 LMX1A FOXA2\n",
       "14                                   SOX2 ASCL1 NEUROG2\n",
       "15                                           ASCL1 SOX2\n",
       "16                           FOXM1 SOX2 MYC SALL4 STAT6\n",
       "17            AR SOX2 SMAD3 MYC JUN WT1 TAL1 SPI1 RUNX1\n",
       "18                      POU3F2 SOX2 BACH1 AR PBX1 NANOG\n",
       "19     ASCL1  PHOX2B  AP-2A  GATA3  HAND2  NR4A2 PHOX2A\n",
       "20                                    MSI1 NEUROG2 MBD2\n",
       "21                         GATA4 MEF2C TBX5 MESP1 MYOCD\n",
       "22             GATA4 MEF2C TBX5 ESRRG MESP1 ZFPM2 MYOCD\n",
       "23     HNF1A HNF4A ONECUT1 ATF5 PROX1 CEBPA; MYC + TP53\n",
       "24            FOXA3 HNF1A  HNF4A + SV40 Large T Antigen\n",
       "25    HNF1A FOXA1 FOXA3; HNF1A FOXA1 HNF4A; HNF1A FO...\n",
       "26                         ATF5 PROX1 FOXA2 FOXA3 HNF4A\n",
       "27                               OCT4 FOXA2 HNF1A GATA3\n",
       "28                                       PDX1 PAX4 MAFA\n",
       "29                                       NFE2 MAFG MAFK\n",
       "30                                      SPI1 IRF8 BATF3\n",
       "31                                        MYC KLF4 SOX9\n",
       "32    POU5F1 OCT6; POU3F4 MYCL; POU3F4 MYC; POU3F4 MYCN\n",
       "33                            RUNX2  OCT4  OSTERIX MYCL\n",
       "34                                   OCT4  OSTERIX MYCL\n",
       "35                                           MYOD1 MYCL\n",
       "36                                         PRDM16 CEBPB\n",
       "37                                     CRX  RAX NEUROD1\n",
       "38                                     MITF  SOX10 PAX3\n",
       "39                                                SOX18\n",
       "40                               CRX  RAX  OTX2 NEUROD1\n",
       "41                                           NFKB1 LEF1\n",
       "42                                                 OCT4\n",
       "43                              EMX2  HNF1B  HNF4A PAX8\n",
       "44                                                MYOCD\n",
       "45                                                 ETV2\n",
       "46                                                PPARG\n",
       "47                             HNF4A  FOXA3  GATA6 CDX2\n",
       "48                                   MYOCD  GATA6 MEF2C\n",
       "49                                     SNAI2  EYA1 SIX1\n",
       "50                                                  NaN\n",
       "51                                                  NaN\n",
       "52                                                  NaN\n",
       "53                                                  NaN\n",
       "Name: TFs, dtype: object"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['TFs']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70852292-74b3-4f3b-a531-c05d61d2e847",
   "metadata": {},
   "source": [
    "## Debug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0f74faa7-0c9a-459f-9e12-44b893f5d97b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 38151 × 58870\n",
       "    obs: 'organ_tissue', 'method', 'donor', 'anatomical_information', 'n_counts_UMIs', 'n_genes', 'cell_ontology_class', 'free_annotation', 'manually_annotated', 'compartment', 'gender'\n",
       "    var: 'gene_symbol', 'feature_type', 'ensemblid', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
       "    uns: '_scvi', '_training_mode', 'dendrogram_cell_type_tissue', 'dendrogram_computational_compartment_assignment', 'dendrogram_consensus_prediction', 'dendrogram_tissue_cell_type', 'donor_colors', 'donor_method_colors', 'hvg', 'method_colors', 'neighbors', 'organ_tissue_colors', 'sex_colors', 'tissue_colors', 'umap'\n",
       "    obsm: 'X_pca', 'X_scvi', 'X_scvi_umap', 'X_umap'\n",
       "    layers: 'decontXcounts', 'raw_counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d3262a3a-fe70-4f49-9cde-9620bb5b3420",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>ensemblid</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>False</td>\n",
       "      <td>6.398244e-05</td>\n",
       "      <td>0.835044</td>\n",
       "      <td>-0.573947</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.005574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>False</td>\n",
       "      <td>2.274395e-03</td>\n",
       "      <td>2.442280</td>\n",
       "      <td>0.533203</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.031731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR6859-1</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>False</td>\n",
       "      <td>6.175251e-05</td>\n",
       "      <td>1.295335</td>\n",
       "      <td>-0.256874</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.005634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>False</td>\n",
       "      <td>1.372886e-04</td>\n",
       "      <td>2.656352</td>\n",
       "      <td>0.680668</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.008041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND6</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>False</td>\n",
       "      <td>9.634841e-01</td>\n",
       "      <td>2.466404</td>\n",
       "      <td>0.154140</td>\n",
       "      <td>0.590065</td>\n",
       "      <td>0.741395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TE</th>\n",
       "      <td>MT-TE</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.600667e-01</td>\n",
       "      <td>1.603787</td>\n",
       "      <td>-0.044396</td>\n",
       "      <td>0.083929</td>\n",
       "      <td>0.301820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CYB</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>False</td>\n",
       "      <td>4.367693e+00</td>\n",
       "      <td>4.765751</td>\n",
       "      <td>-0.499747</td>\n",
       "      <td>3.874830</td>\n",
       "      <td>1.104192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TT</th>\n",
       "      <td>MT-TT</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>False</td>\n",
       "      <td>6.573967e-02</td>\n",
       "      <td>0.624316</td>\n",
       "      <td>-0.719108</td>\n",
       "      <td>0.040580</td>\n",
       "      <td>0.186848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TP</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>True</td>\n",
       "      <td>4.385102e-01</td>\n",
       "      <td>2.705558</td>\n",
       "      <td>1.115482</td>\n",
       "      <td>0.204806</td>\n",
       "      <td>0.501265</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58870 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_symbol     feature_type          ensemblid  highly_variable  \\\n",
       "DDX11L1          DDX11L1  Gene Expression  ENSG00000223972.5            False   \n",
       "WASH7P            WASH7P  Gene Expression  ENSG00000227232.5            False   \n",
       "MIR6859-1      MIR6859-1  Gene Expression  ENSG00000278267.1            False   \n",
       "MIR1302-2HG  MIR1302-2HG  Gene Expression  ENSG00000243485.5            False   \n",
       "MIR1302-2      MIR1302-2  Gene Expression  ENSG00000284332.1            False   \n",
       "...                  ...              ...                ...              ...   \n",
       "MT-ND6            MT-ND6  Gene Expression  ENSG00000198695.2            False   \n",
       "MT-TE              MT-TE  Gene Expression  ENSG00000210194.1            False   \n",
       "MT-CYB            MT-CYB  Gene Expression  ENSG00000198727.2            False   \n",
       "MT-TT              MT-TT  Gene Expression  ENSG00000210195.2            False   \n",
       "MT-TP              MT-TP  Gene Expression  ENSG00000210196.2             True   \n",
       "\n",
       "                    means  dispersions  dispersions_norm      mean       std  \n",
       "DDX11L1      6.398244e-05     0.835044         -0.573947  0.000039  0.005574  \n",
       "WASH7P       2.274395e-03     2.442280          0.533203  0.001080  0.031731  \n",
       "MIR6859-1    6.175251e-05     1.295335         -0.256874  0.000033  0.005634  \n",
       "MIR1302-2HG  1.372886e-04     2.656352          0.680668  0.000048  0.008041  \n",
       "MIR1302-2    1.000000e-12          NaN          0.000000  0.000000  1.000000  \n",
       "...                   ...          ...               ...       ...       ...  \n",
       "MT-ND6       9.634841e-01     2.466404          0.154140  0.590065  0.741395  \n",
       "MT-TE        1.600667e-01     1.603787         -0.044396  0.083929  0.301820  \n",
       "MT-CYB       4.367693e+00     4.765751         -0.499747  3.874830  1.104192  \n",
       "MT-TT        6.573967e-02     0.624316         -0.719108  0.040580  0.186848  \n",
       "MT-TP        4.385102e-01     2.705558          1.115482  0.204806  0.501265  \n",
       "\n",
       "[58870 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ea0b0577-e58e-4857-8e11-22ea03a51569",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n"
     ]
    }
   ],
   "source": [
    "padata = perturb_counts(adata, ['DDX11L1'], 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b0302620-1304-496e-b753-23cb9c2cdf4e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>ensemblid</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>scaled</th>\n",
       "      <th>scaled_by</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>False</td>\n",
       "      <td>6.398244e-05</td>\n",
       "      <td>0.835044</td>\n",
       "      <td>-0.573947</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.005574</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>False</td>\n",
       "      <td>2.274395e-03</td>\n",
       "      <td>2.442280</td>\n",
       "      <td>0.533203</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.031731</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR6859-1</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>False</td>\n",
       "      <td>6.175251e-05</td>\n",
       "      <td>1.295335</td>\n",
       "      <td>-0.256874</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.005634</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>False</td>\n",
       "      <td>1.372886e-04</td>\n",
       "      <td>2.656352</td>\n",
       "      <td>0.680668</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.008041</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND6</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>False</td>\n",
       "      <td>9.634841e-01</td>\n",
       "      <td>2.466404</td>\n",
       "      <td>0.154140</td>\n",
       "      <td>0.590065</td>\n",
       "      <td>0.741395</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TE</th>\n",
       "      <td>MT-TE</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.600667e-01</td>\n",
       "      <td>1.603787</td>\n",
       "      <td>-0.044396</td>\n",
       "      <td>0.083929</td>\n",
       "      <td>0.301820</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CYB</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>False</td>\n",
       "      <td>4.367693e+00</td>\n",
       "      <td>4.765751</td>\n",
       "      <td>-0.499747</td>\n",
       "      <td>3.874830</td>\n",
       "      <td>1.104192</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TT</th>\n",
       "      <td>MT-TT</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>False</td>\n",
       "      <td>6.573967e-02</td>\n",
       "      <td>0.624316</td>\n",
       "      <td>-0.719108</td>\n",
       "      <td>0.040580</td>\n",
       "      <td>0.186848</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TP</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>True</td>\n",
       "      <td>4.385102e-01</td>\n",
       "      <td>2.705558</td>\n",
       "      <td>1.115482</td>\n",
       "      <td>0.204806</td>\n",
       "      <td>0.501265</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58870 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_symbol     feature_type          ensemblid  highly_variable  \\\n",
       "DDX11L1          DDX11L1  Gene Expression  ENSG00000223972.5            False   \n",
       "WASH7P            WASH7P  Gene Expression  ENSG00000227232.5            False   \n",
       "MIR6859-1      MIR6859-1  Gene Expression  ENSG00000278267.1            False   \n",
       "MIR1302-2HG  MIR1302-2HG  Gene Expression  ENSG00000243485.5            False   \n",
       "MIR1302-2      MIR1302-2  Gene Expression  ENSG00000284332.1            False   \n",
       "...                  ...              ...                ...              ...   \n",
       "MT-ND6            MT-ND6  Gene Expression  ENSG00000198695.2            False   \n",
       "MT-TE              MT-TE  Gene Expression  ENSG00000210194.1            False   \n",
       "MT-CYB            MT-CYB  Gene Expression  ENSG00000198727.2            False   \n",
       "MT-TT              MT-TT  Gene Expression  ENSG00000210195.2            False   \n",
       "MT-TP              MT-TP  Gene Expression  ENSG00000210196.2             True   \n",
       "\n",
       "                    means  dispersions  dispersions_norm      mean       std  \\\n",
       "DDX11L1      6.398244e-05     0.835044         -0.573947  0.000039  0.005574   \n",
       "WASH7P       2.274395e-03     2.442280          0.533203  0.001080  0.031731   \n",
       "MIR6859-1    6.175251e-05     1.295335         -0.256874  0.000033  0.005634   \n",
       "MIR1302-2HG  1.372886e-04     2.656352          0.680668  0.000048  0.008041   \n",
       "MIR1302-2    1.000000e-12          NaN          0.000000  0.000000  1.000000   \n",
       "...                   ...          ...               ...       ...       ...   \n",
       "MT-ND6       9.634841e-01     2.466404          0.154140  0.590065  0.741395   \n",
       "MT-TE        1.600667e-01     1.603787         -0.044396  0.083929  0.301820   \n",
       "MT-CYB       4.367693e+00     4.765751         -0.499747  3.874830  1.104192   \n",
       "MT-TT        6.573967e-02     0.624316         -0.719108  0.040580  0.186848   \n",
       "MT-TP        4.385102e-01     2.705558          1.115482  0.204806  0.501265   \n",
       "\n",
       "             scaled  scaled_by  \n",
       "DDX11L1        True          1  \n",
       "WASH7P        False          1  \n",
       "MIR6859-1     False          1  \n",
       "MIR1302-2HG   False          1  \n",
       "MIR1302-2     False          1  \n",
       "...             ...        ...  \n",
       "MT-ND6        False          1  \n",
       "MT-TE         False          1  \n",
       "MT-CYB        False          1  \n",
       "MT-TT         False          1  \n",
       "MT-TP         False          1  \n",
       "\n",
       "[58870 rows x 11 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concatenated_adata = ad.concat([adata, padata], axis=0)\n",
    "concatenated_adata.var = adata.var.copy()\n",
    "concatenated_adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "da19cf08-e416-405a-82af-d6b7125ce829",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>ensemblid</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>scaled</th>\n",
       "      <th>scaled_by</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>False</td>\n",
       "      <td>6.398244e-05</td>\n",
       "      <td>0.835044</td>\n",
       "      <td>-0.573947</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.005574</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>False</td>\n",
       "      <td>2.274395e-03</td>\n",
       "      <td>2.442280</td>\n",
       "      <td>0.533203</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.031731</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR6859-1</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>False</td>\n",
       "      <td>6.175251e-05</td>\n",
       "      <td>1.295335</td>\n",
       "      <td>-0.256874</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.005634</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>False</td>\n",
       "      <td>1.372886e-04</td>\n",
       "      <td>2.656352</td>\n",
       "      <td>0.680668</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.008041</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND6</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>False</td>\n",
       "      <td>9.634841e-01</td>\n",
       "      <td>2.466404</td>\n",
       "      <td>0.154140</td>\n",
       "      <td>0.590065</td>\n",
       "      <td>0.741395</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TE</th>\n",
       "      <td>MT-TE</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.600667e-01</td>\n",
       "      <td>1.603787</td>\n",
       "      <td>-0.044396</td>\n",
       "      <td>0.083929</td>\n",
       "      <td>0.301820</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CYB</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>False</td>\n",
       "      <td>4.367693e+00</td>\n",
       "      <td>4.765751</td>\n",
       "      <td>-0.499747</td>\n",
       "      <td>3.874830</td>\n",
       "      <td>1.104192</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TT</th>\n",
       "      <td>MT-TT</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>False</td>\n",
       "      <td>6.573967e-02</td>\n",
       "      <td>0.624316</td>\n",
       "      <td>-0.719108</td>\n",
       "      <td>0.040580</td>\n",
       "      <td>0.186848</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TP</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>True</td>\n",
       "      <td>4.385102e-01</td>\n",
       "      <td>2.705558</td>\n",
       "      <td>1.115482</td>\n",
       "      <td>0.204806</td>\n",
       "      <td>0.501265</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58870 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_symbol     feature_type          ensemblid  highly_variable  \\\n",
       "DDX11L1          DDX11L1  Gene Expression  ENSG00000223972.5            False   \n",
       "WASH7P            WASH7P  Gene Expression  ENSG00000227232.5            False   \n",
       "MIR6859-1      MIR6859-1  Gene Expression  ENSG00000278267.1            False   \n",
       "MIR1302-2HG  MIR1302-2HG  Gene Expression  ENSG00000243485.5            False   \n",
       "MIR1302-2      MIR1302-2  Gene Expression  ENSG00000284332.1            False   \n",
       "...                  ...              ...                ...              ...   \n",
       "MT-ND6            MT-ND6  Gene Expression  ENSG00000198695.2            False   \n",
       "MT-TE              MT-TE  Gene Expression  ENSG00000210194.1            False   \n",
       "MT-CYB            MT-CYB  Gene Expression  ENSG00000198727.2            False   \n",
       "MT-TT              MT-TT  Gene Expression  ENSG00000210195.2            False   \n",
       "MT-TP              MT-TP  Gene Expression  ENSG00000210196.2             True   \n",
       "\n",
       "                    means  dispersions  dispersions_norm      mean       std  \\\n",
       "DDX11L1      6.398244e-05     0.835044         -0.573947  0.000039  0.005574   \n",
       "WASH7P       2.274395e-03     2.442280          0.533203  0.001080  0.031731   \n",
       "MIR6859-1    6.175251e-05     1.295335         -0.256874  0.000033  0.005634   \n",
       "MIR1302-2HG  1.372886e-04     2.656352          0.680668  0.000048  0.008041   \n",
       "MIR1302-2    1.000000e-12          NaN          0.000000  0.000000  1.000000   \n",
       "...                   ...          ...               ...       ...       ...   \n",
       "MT-ND6       9.634841e-01     2.466404          0.154140  0.590065  0.741395   \n",
       "MT-TE        1.600667e-01     1.603787         -0.044396  0.083929  0.301820   \n",
       "MT-CYB       4.367693e+00     4.765751         -0.499747  3.874830  1.104192   \n",
       "MT-TT        6.573967e-02     0.624316         -0.719108  0.040580  0.186848   \n",
       "MT-TP        4.385102e-01     2.705558          1.115482  0.204806  0.501265   \n",
       "\n",
       "             scaled  scaled_by  \n",
       "DDX11L1        True          1  \n",
       "WASH7P        False          1  \n",
       "MIR6859-1     False          1  \n",
       "MIR1302-2HG   False          1  \n",
       "MIR1302-2     False          1  \n",
       "...             ...        ...  \n",
       "MT-ND6        False          1  \n",
       "MT-TE         False          1  \n",
       "MT-CYB        False          1  \n",
       "MT-TT         False          1  \n",
       "MT-TP         False          1  \n",
       "\n",
       "[58870 rows x 11 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8fa9cac1-342f-4f92-8ac2-94f88199cf64",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR6859-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND6</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CYB</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TP</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58870 rows × 0 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [DDX11L1, WASH7P, MIR6859-1, MIR1302-2HG, MIR1302-2, FAM138A, OR4G4P, OR4G11P, OR4F5, AL627309.1, AL627309.3, CICP27, AL627309.6, AL627309.7, AL627309.2, AL627309.5, RNU6-1100P, AL627309.4, FO538757.1, WASH9P, MIR6859-2, AP006222.1, RPL23AP24, AL732372.1, AL732372.2, WBP1LP7, OR4F29, CICP7, AL732372.3, RF00026, AL669831.3, AC114498.1, MTND1P23, MTND2P28, MTCO1P12, AC114498.2, MTCO2P12, MTATP8P1, MTATP6P1, MTCO3P12, WBP1LP6, OR4F16, CICP3, AL669831.1, RNU6-1199P, AL669831.2, AL669831.5, AL669831.7, AL669831.4, FAM87B, LINC01128, LINC00115, FAM41C, TUBB8P11, AL669831.6, AL645608.6, AL645608.2, AL645608.4, LINC02593, SAMD11, NOC2L, KLHL17, PLEKHN1, PERM1, AL645608.7, HES4, ISG15, AL645608.3, AL645608.1, AGRN, AL645608.5, AL390719.1, AL645608.8, RNF223, C1orf159, AL390719.3, LINC01342, MIR200B, MIR200A, MIR429, AL390719.2, TTLL10-AS1, TTLL10, TNFRSF18, TNFRSF4, SDF4, B3GALT6, C1QTNF12, AL162741.1, UBE2J2, LINC01786, SCNN1D, ACAP3, MIR6726, PUSL1, INTS11, MIR6727, AL139287.1, CPTP, TAS1R3, ...]\n",
       "\n",
       "[58870 rows x 0 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concatenated_adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "947cdb03-0f13-4071-aeba-f3519dffb083",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>ensemblid</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>scaled</th>\n",
       "      <th>scaled_by</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>False</td>\n",
       "      <td>6.398244e-05</td>\n",
       "      <td>0.835044</td>\n",
       "      <td>-0.573947</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.005574</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>False</td>\n",
       "      <td>2.274395e-03</td>\n",
       "      <td>2.442280</td>\n",
       "      <td>0.533203</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.031731</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR6859-1</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>False</td>\n",
       "      <td>6.175251e-05</td>\n",
       "      <td>1.295335</td>\n",
       "      <td>-0.256874</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.005634</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>False</td>\n",
       "      <td>1.372886e-04</td>\n",
       "      <td>2.656352</td>\n",
       "      <td>0.680668</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.008041</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND6</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>False</td>\n",
       "      <td>9.634841e-01</td>\n",
       "      <td>2.466404</td>\n",
       "      <td>0.154140</td>\n",
       "      <td>0.590065</td>\n",
       "      <td>0.741395</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TE</th>\n",
       "      <td>MT-TE</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.600667e-01</td>\n",
       "      <td>1.603787</td>\n",
       "      <td>-0.044396</td>\n",
       "      <td>0.083929</td>\n",
       "      <td>0.301820</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CYB</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>False</td>\n",
       "      <td>4.367693e+00</td>\n",
       "      <td>4.765751</td>\n",
       "      <td>-0.499747</td>\n",
       "      <td>3.874830</td>\n",
       "      <td>1.104192</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TT</th>\n",
       "      <td>MT-TT</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>False</td>\n",
       "      <td>6.573967e-02</td>\n",
       "      <td>0.624316</td>\n",
       "      <td>-0.719108</td>\n",
       "      <td>0.040580</td>\n",
       "      <td>0.186848</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TP</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>True</td>\n",
       "      <td>4.385102e-01</td>\n",
       "      <td>2.705558</td>\n",
       "      <td>1.115482</td>\n",
       "      <td>0.204806</td>\n",
       "      <td>0.501265</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58870 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_symbol     feature_type          ensemblid  highly_variable  \\\n",
       "DDX11L1          DDX11L1  Gene Expression  ENSG00000223972.5            False   \n",
       "WASH7P            WASH7P  Gene Expression  ENSG00000227232.5            False   \n",
       "MIR6859-1      MIR6859-1  Gene Expression  ENSG00000278267.1            False   \n",
       "MIR1302-2HG  MIR1302-2HG  Gene Expression  ENSG00000243485.5            False   \n",
       "MIR1302-2      MIR1302-2  Gene Expression  ENSG00000284332.1            False   \n",
       "...                  ...              ...                ...              ...   \n",
       "MT-ND6            MT-ND6  Gene Expression  ENSG00000198695.2            False   \n",
       "MT-TE              MT-TE  Gene Expression  ENSG00000210194.1            False   \n",
       "MT-CYB            MT-CYB  Gene Expression  ENSG00000198727.2            False   \n",
       "MT-TT              MT-TT  Gene Expression  ENSG00000210195.2            False   \n",
       "MT-TP              MT-TP  Gene Expression  ENSG00000210196.2             True   \n",
       "\n",
       "                    means  dispersions  dispersions_norm      mean       std  \\\n",
       "DDX11L1      6.398244e-05     0.835044         -0.573947  0.000039  0.005574   \n",
       "WASH7P       2.274395e-03     2.442280          0.533203  0.001080  0.031731   \n",
       "MIR6859-1    6.175251e-05     1.295335         -0.256874  0.000033  0.005634   \n",
       "MIR1302-2HG  1.372886e-04     2.656352          0.680668  0.000048  0.008041   \n",
       "MIR1302-2    1.000000e-12          NaN          0.000000  0.000000  1.000000   \n",
       "...                   ...          ...               ...       ...       ...   \n",
       "MT-ND6       9.634841e-01     2.466404          0.154140  0.590065  0.741395   \n",
       "MT-TE        1.600667e-01     1.603787         -0.044396  0.083929  0.301820   \n",
       "MT-CYB       4.367693e+00     4.765751         -0.499747  3.874830  1.104192   \n",
       "MT-TT        6.573967e-02     0.624316         -0.719108  0.040580  0.186848   \n",
       "MT-TP        4.385102e-01     2.705558          1.115482  0.204806  0.501265   \n",
       "\n",
       "             scaled  scaled_by  \n",
       "DDX11L1        True          1  \n",
       "WASH7P        False          1  \n",
       "MIR6859-1     False          1  \n",
       "MIR1302-2HG   False          1  \n",
       "MIR1302-2     False          1  \n",
       "...             ...        ...  \n",
       "MT-ND6        False          1  \n",
       "MT-TE         False          1  \n",
       "MT-CYB        False          1  \n",
       "MT-TT         False          1  \n",
       "MT-TP         False          1  \n",
       "\n",
       "[58870 rows x 11 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "padata.var"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "307ff3c0-6c4e-4066-a49a-5128654c479a",
   "metadata": {},
   "source": [
    "# Day 7\n",
    "\n",
    "Nat says the Reproramming recipes should work"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65f8e30f-ccc2-43be-9689-c88e4b40528f",
   "metadata": {},
   "source": [
    "## Set up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ccd84ae6-11de-4c43-a83c-91a42d5d8bf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import scanpy as sp\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "41f3efa3-47a6-4805-9a63-b73b999f2ad0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def iterate_perturb_counts(adata, tf_list, scalar_list):\n",
    "    \"\"\"\n",
    "    Applies perturbations to the expression data of specified transcription factors across multiple scalars \n",
    "    and stores the resulting AnnData objects in a dictionary.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Iterates over a list of scalar values.\n",
    "    2. For each scalar, creates a copy of the AnnData object to preserve the original data.\n",
    "    3. Applies the `perturb_counts` function to scale the expression data of genes listed in `tf_list` by\n",
    "       the maximum gene expression of each cell and the current scalar.\n",
    "    4. Stores the perturbed AnnData object in a dictionary with the scalar as the key.\n",
    "\n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols (transcription factors) to be perturbed.\n",
    "    scalar_list (list): A list of scalar values for scaling the gene expression.\n",
    "    adata (AnnData): The AnnData object containing gene expression data (cells x genes).\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary where keys are scalar values and values are the corresponding perturbed AnnData objects.\n",
    "    \"\"\"\n",
    "    \n",
    "    adata_dict = {}\n",
    "    \n",
    "    for scalar in scalar_list:\n",
    "        # Create a copy of the AnnData object for each scalar value\n",
    "        adata_temp = adata.copy()\n",
    "        \n",
    "        # Apply perturb_counts to the copied AnnData object\n",
    "        perturbed_adata = perturb_counts(adata_temp, tf_list, scalar)\n",
    "        \n",
    "        # Store the perturbed AnnData object in the dictionary with scalar as the key\n",
    "        adata_dict[scalar] = perturbed_adata\n",
    "    \n",
    "    return adata_dict\n",
    "\n",
    "def perturb_counts(adata, tf_list, scalar): \n",
    "    \"\"\"\n",
    "    Applies a perturbation to the expression data of specific genes in an AnnData object.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Computes the maximum gene expression level for each cell.\n",
    "    2. Applies a scaling operation to the expression levels of genes listed in `tf_list`.\n",
    "       - Each entry of these genes in the matrix is multiplied by the maximum expression level \n",
    "         of its respective cell and a specified scalar value.\n",
    "    3. Updates the AnnData object with new columns:\n",
    "       - 'scaled': A boolean column indicating whether each gene is in the `tf_list`.\n",
    "       - 'scaled_by': Contains the scaling factor used for each gene (the product of the maximum \n",
    "         expression level of each cell and the scalar), or `1` if the gene was not in `tf_list`.\n",
    "    \n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols to be perturbed.\n",
    "    scalar (float): The scalar value used to scale the expression levels.\n",
    "    adata (AnnData): The AnnData object containing gene expression data.\n",
    "\n",
    "    Returns:\n",
    "    AnnData: The updated AnnData object with applied perturbations and new columns.\n",
    "    \"\"\"\n",
    "\n",
    "    # Create a boolean mask for genes in tf_list\n",
    "    gene_mask = adata.var['gene_symbol'].isin(tf_list)\n",
    "    \n",
    "    # Save the original state of the parameter objects, in case some tfs do not translate (failsafe)\n",
    "    original_X = adata.X.copy()\n",
    "    original_gene_mask = gene_mask.copy()\n",
    "    \n",
    "    # Compute maximum expression level of each cell\n",
    "    max_exp = np.max(adata.X, axis=1)\n",
    "\n",
    "    \"\"\"This is new today. v \"\"\"\n",
    "    # Raise an error if any of the gene names in tf_list do not match column names (we will manually update these in adata):\n",
    "    missing_genes = [gene for gene in tf_list if gene not in adata.var['gene_symbol'].values]\n",
    "    \n",
    "    if missing_genes:\n",
    "        # Restore original parameter objects\n",
    "        adata.X = original_X\n",
    "        gene_mask = original_gene_mask\n",
    "        raise ValueError(f\"Genes {missing_genes} not found in anndata object\")\n",
    "\n",
    "    else:    \n",
    "        \n",
    "        # Apply the scaling operation to the specified genes\n",
    "        adata.X[:, gene_mask] = max_exp * scalar\n",
    "        \n",
    "        # Add/Update 'scaled' column in var\n",
    "        adata.var['scaled'] = gene_mask\n",
    "        \n",
    "        # Add/Update 'scaled_by' column in var\n",
    "        adata.var['scaled_by'] = scalar  # Default value for genes not in tf_list\n",
    "\n",
    "        # Add/Update 'scaled_by' column in var\n",
    "        adata.obs['U'] = scalar  # Default value for genes not in tf_list\n",
    "    # adata.var = \n",
    "    return adata\n",
    "\n",
    "def validateTFs(TFs, adata):\n",
    "    adata_gene_list = adata.var['gene_symbol'].values.tolist()\n",
    "    for TF in TFs:\n",
    "        if TF not in adata_gene_list:\n",
    "            return False\n",
    "    return True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a1d11c2-7719-4bd4-8e5f-d698130b706f",
   "metadata": {},
   "source": [
    "## Validate Recipes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "689990b6-9b23-4fcb-ac90-a73c80351ac6",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/jpic/\"\n",
    "FILE = \"fibroblast.h5ad\"\n",
    "adata = sp.read_h5ad(os.path.join(DATAPATH, FILE))\n",
    "adata_gene_list = adata.var['gene_symbol'].values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4b0ddeca-2845-4d6f-808c-9948a3a2bdb2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source cells</th>\n",
       "      <th>Source</th>\n",
       "      <th>Target cells</th>\n",
       "      <th>Target</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>TFs</th>\n",
       "      <th>Species</th>\n",
       "      <th>Cell Transplantation</th>\n",
       "      <th>Published Year</th>\n",
       "      <th>PMID</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "      <th>Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Embryonic Fibroblasts and Adult Skin Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Dopamine Neurons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ASCL1, LXMLA, and NURR1</td>\n",
       "      <td>ASCL1 LXMLA NR4A2</td>\n",
       "      <td>Both</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>2172534_27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Fetal Fibroblasts and Postnatal Foreskin Fibro...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neuronal Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BRN2, ASCL1, MYT1L, and NEUROD1</td>\n",
       "      <td>POU3F2 ASCL1 MYT1L NEUROD11</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>21617644_28</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Embryonic Fibroblasts and Postnatal Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neuronal Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ASCL1, BRN2, MYT1L, LMX1A, and FOXA2</td>\n",
       "      <td>ASCL1 POU3F2 MYT1L LMX1A FOXA2</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>21646515_29</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Embryonic Fibroblasts and Fetal Foreskin Fibro...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neural Stem Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SOX2</td>\n",
       "      <td>SOX2</td>\n",
       "      <td>Both</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2012.0</td>\n",
       "      <td>22683203_31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Fibroblasts (IMR90 Cells)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Dopaminergic Neurons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MASH1, NGN2, SOX2, NURR1, and PITX3 + A Domina...</td>\n",
       "      <td>ASCL1 NEUROG2 SOX2 NR4A2 PITX3 P53</td>\n",
       "      <td>Human</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>25129808_214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        Source cells  Source  \\\n",
       "0   Embryonic Fibroblasts and Adult Skin Fibroblasts     NaN   \n",
       "1  Fetal Fibroblasts and Postnatal Foreskin Fibro...     NaN   \n",
       "2    Embryonic Fibroblasts and Postnatal Fibroblasts     NaN   \n",
       "3  Embryonic Fibroblasts and Fetal Foreskin Fibro...     NaN   \n",
       "4                          Fibroblasts (IMR90 Cells)     NaN   \n",
       "\n",
       "           Target cells  Target  \\\n",
       "0      Dopamine Neurons     NaN   \n",
       "1        Neuronal Cells     NaN   \n",
       "2        Neuronal Cells     NaN   \n",
       "3     Neural Stem Cells     NaN   \n",
       "4  Dopaminergic Neurons     NaN   \n",
       "\n",
       "                                           Treatment  \\\n",
       "0                            ASCL1, LXMLA, and NURR1   \n",
       "1                    BRN2, ASCL1, MYT1L, and NEUROD1   \n",
       "2               ASCL1, BRN2, MYT1L, LMX1A, and FOXA2   \n",
       "3                                               SOX2   \n",
       "4  MASH1, NGN2, SOX2, NURR1, and PITX3 + A Domina...   \n",
       "\n",
       "                                  TFs Species Cell Transplantation  \\\n",
       "0                   ASCL1 LXMLA NR4A2    Both                  Yes   \n",
       "1         POU3F2 ASCL1 MYT1L NEUROD11   Human                  NaN   \n",
       "2      ASCL1 POU3F2 MYT1L LMX1A FOXA2   Human                  NaN   \n",
       "3                                SOX2    Both                  Yes   \n",
       "4  ASCL1 NEUROG2 SOX2 NR4A2 PITX3 P53   Human                  Yes   \n",
       "\n",
       "   Published Year          PMID  Unnamed: 10 Notes  \n",
       "0          2011.0    2172534_27          NaN   NaN  \n",
       "1          2011.0   21617644_28          NaN   NaN  \n",
       "2          2011.0   21646515_29          NaN   NaN  \n",
       "3          2012.0   22683203_31          NaN   NaN  \n",
       "4          2014.0  25129808_214          NaN   NaN  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('data/recipe_table_9_6_2024.csv')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "59882187-4366-40f5-9941-fe56e29bf1f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of times True was printed: 34\n",
      "Number of times True rate: 0.6296296296296297\n"
     ]
    }
   ],
   "source": [
    "true_count = 0\n",
    "\n",
    "for i in range(df.shape[0]):\n",
    "    try:\n",
    "        TFs = df['TFs'].iloc[i].split()\n",
    "        result = validateTFs(TFs, adata)\n",
    "        # print(result)\n",
    "        if result == True:\n",
    "            true_count += 1\n",
    "    except:\n",
    "        continue\n",
    "\n",
    "print(f'Number of times True was printed: {true_count}')\n",
    "print(f'Number of times True rate: {true_count/df.shape[0]}')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b66a7c3-ce82-45a1-87ee-28d9e5f1797a",
   "metadata": {},
   "source": [
    "## Make the perturbation files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "44c7a764-13f5-4f4c-b380-e315b7883a2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>ensemblid</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>False</td>\n",
       "      <td>6.398244e-05</td>\n",
       "      <td>0.835044</td>\n",
       "      <td>-0.573947</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.005574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>False</td>\n",
       "      <td>2.274395e-03</td>\n",
       "      <td>2.442280</td>\n",
       "      <td>0.533203</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.031731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR6859-1</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>False</td>\n",
       "      <td>6.175251e-05</td>\n",
       "      <td>1.295335</td>\n",
       "      <td>-0.256874</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.005634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>False</td>\n",
       "      <td>1.372886e-04</td>\n",
       "      <td>2.656352</td>\n",
       "      <td>0.680668</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.008041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND6</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>False</td>\n",
       "      <td>9.634841e-01</td>\n",
       "      <td>2.466404</td>\n",
       "      <td>0.154140</td>\n",
       "      <td>0.590065</td>\n",
       "      <td>0.741395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TE</th>\n",
       "      <td>MT-TE</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.600667e-01</td>\n",
       "      <td>1.603787</td>\n",
       "      <td>-0.044396</td>\n",
       "      <td>0.083929</td>\n",
       "      <td>0.301820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CYB</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>False</td>\n",
       "      <td>4.367693e+00</td>\n",
       "      <td>4.765751</td>\n",
       "      <td>-0.499747</td>\n",
       "      <td>3.874830</td>\n",
       "      <td>1.104192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TT</th>\n",
       "      <td>MT-TT</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>False</td>\n",
       "      <td>6.573967e-02</td>\n",
       "      <td>0.624316</td>\n",
       "      <td>-0.719108</td>\n",
       "      <td>0.040580</td>\n",
       "      <td>0.186848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TP</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>True</td>\n",
       "      <td>4.385102e-01</td>\n",
       "      <td>2.705558</td>\n",
       "      <td>1.115482</td>\n",
       "      <td>0.204806</td>\n",
       "      <td>0.501265</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58870 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_symbol     feature_type          ensemblid  highly_variable  \\\n",
       "DDX11L1          DDX11L1  Gene Expression  ENSG00000223972.5            False   \n",
       "WASH7P            WASH7P  Gene Expression  ENSG00000227232.5            False   \n",
       "MIR6859-1      MIR6859-1  Gene Expression  ENSG00000278267.1            False   \n",
       "MIR1302-2HG  MIR1302-2HG  Gene Expression  ENSG00000243485.5            False   \n",
       "MIR1302-2      MIR1302-2  Gene Expression  ENSG00000284332.1            False   \n",
       "...                  ...              ...                ...              ...   \n",
       "MT-ND6            MT-ND6  Gene Expression  ENSG00000198695.2            False   \n",
       "MT-TE              MT-TE  Gene Expression  ENSG00000210194.1            False   \n",
       "MT-CYB            MT-CYB  Gene Expression  ENSG00000198727.2            False   \n",
       "MT-TT              MT-TT  Gene Expression  ENSG00000210195.2            False   \n",
       "MT-TP              MT-TP  Gene Expression  ENSG00000210196.2             True   \n",
       "\n",
       "                    means  dispersions  dispersions_norm      mean       std  \n",
       "DDX11L1      6.398244e-05     0.835044         -0.573947  0.000039  0.005574  \n",
       "WASH7P       2.274395e-03     2.442280          0.533203  0.001080  0.031731  \n",
       "MIR6859-1    6.175251e-05     1.295335         -0.256874  0.000033  0.005634  \n",
       "MIR1302-2HG  1.372886e-04     2.656352          0.680668  0.000048  0.008041  \n",
       "MIR1302-2    1.000000e-12          NaN          0.000000  0.000000  1.000000  \n",
       "...                   ...          ...               ...       ...       ...  \n",
       "MT-ND6       9.634841e-01     2.466404          0.154140  0.590065  0.741395  \n",
       "MT-TE        1.600667e-01     1.603787         -0.044396  0.083929  0.301820  \n",
       "MT-CYB       4.367693e+00     4.765751         -0.499747  3.874830  1.104192  \n",
       "MT-TT        6.573967e-02     0.624316         -0.719108  0.040580  0.186848  \n",
       "MT-TP        4.385102e-01     2.705558          1.115482  0.204806  0.501265  \n",
       "\n",
       "[58870 rows x 9 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2ca5f4f6-3e97-440c-b03e-005bcd26b2b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ASCL1', 'POU3F2', 'MYT1L', 'LMX1A', 'FOXA2']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "ename": "OSError",
     "evalue": "[Errno 5] Unable to extend file properly, errno = 9, error message = 'bad file descriptor' (file write failed: time = Fri Sep  6 16:27:05 2024\n, filename = '/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/ASCL1_POU3F2_MYT1L_LMX1A_FOXA2.h5ad', file descriptor = 54, errno = 5, error message = 'Input/output error', buf = 0x1063e960, total write size = 272, bytes this sub-write = 272, bytes actually written = 18446744073709551615, offset = 0)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/anndata/_io/h5ad.py:104\u001b[0m, in \u001b[0;36mwrite_h5ad\u001b[0;34m(filepath, adata, as_dense, dataset_kwargs, **kwargs)\u001b[0m\n\u001b[1;32m    103\u001b[0m     write_elem(f, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mraw\u001b[39m\u001b[38;5;124m\"\u001b[39m, adata\u001b[38;5;241m.\u001b[39mraw, dataset_kwargs\u001b[38;5;241m=\u001b[39mdataset_kwargs)\n\u001b[0;32m--> 104\u001b[0m \u001b[43mwrite_elem\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mobs\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43madata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mobs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdataset_kwargs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdataset_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    105\u001b[0m write_elem(f, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvar\u001b[39m\u001b[38;5;124m\"\u001b[39m, adata\u001b[38;5;241m.\u001b[39mvar, dataset_kwargs\u001b[38;5;241m=\u001b[39mdataset_kwargs)\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/anndata/_io/specs/registry.py:359\u001b[0m, in \u001b[0;36mwrite_elem\u001b[0;34m(store, k, elem, dataset_kwargs)\u001b[0m\n\u001b[1;32m    342\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    343\u001b[0m \u001b[38;5;124;03mWrite an element to a storage group using anndata encoding.\u001b[39;00m\n\u001b[1;32m    344\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    357\u001b[0m \u001b[38;5;124;03m    E.g. for zarr this would be `chunks`, `compressor`.\u001b[39;00m\n\u001b[1;32m    358\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m--> 359\u001b[0m \u001b[43mWriter\u001b[49m\u001b[43m(\u001b[49m\u001b[43m_REGISTRY\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwrite_elem\u001b[49m\u001b[43m(\u001b[49m\u001b[43mstore\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43melem\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdataset_kwargs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdataset_kwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/anndata/_io/utils.py:243\u001b[0m, in \u001b[0;36mreport_write_key_on_error.<locals>.func_wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    242\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 243\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    244\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/anndata/_io/specs/registry.py:309\u001b[0m, in \u001b[0;36mWriter.write_elem\u001b[0;34m(self, store, k, elem, dataset_kwargs, modifiers)\u001b[0m\n\u001b[1;32m    308\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcallback \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 309\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mwrite_func\u001b[49m\u001b[43m(\u001b[49m\u001b[43mstore\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43melem\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdataset_kwargs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdataset_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    310\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcallback(\n\u001b[1;32m    311\u001b[0m     write_func,\n\u001b[1;32m    312\u001b[0m     store,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    316\u001b[0m     iospec\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mregistry\u001b[38;5;241m.\u001b[39mget_spec(elem),\n\u001b[1;32m    317\u001b[0m )\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/anndata/_io/specs/registry.py:57\u001b[0m, in \u001b[0;36mwrite_spec.<locals>.decorator.<locals>.wrapper\u001b[0;34m(g, k, *args, **kwargs)\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[38;5;129m@wraps\u001b[39m(func)\n\u001b[1;32m     56\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mwrapper\u001b[39m(g: GroupStorageType, k: \u001b[38;5;28mstr\u001b[39m, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[0;32m---> 57\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mg\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     58\u001b[0m     g[k]\u001b[38;5;241m.\u001b[39mattrs\u001b[38;5;241m.\u001b[39msetdefault(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mencoding-type\u001b[39m\u001b[38;5;124m\"\u001b[39m, spec\u001b[38;5;241m.\u001b[39mencoding_type)\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/anndata/_io/specs/methods.py:704\u001b[0m, in \u001b[0;36mwrite_dataframe\u001b[0;34m(f, key, df, _writer, dataset_kwargs)\u001b[0m\n\u001b[1;32m    701\u001b[0m \u001b[38;5;66;03m# ._values is \"the best\" array representation. It's the true array backing the\u001b[39;00m\n\u001b[1;32m    702\u001b[0m \u001b[38;5;66;03m# object, where `.values` is always a np.ndarray and .array is always a pandas\u001b[39;00m\n\u001b[1;32m    703\u001b[0m \u001b[38;5;66;03m# array.\u001b[39;00m\n\u001b[0;32m--> 704\u001b[0m \u001b[43m_writer\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwrite_elem\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    705\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgroup\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex_name\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mindex\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_values\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdataset_kwargs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdataset_kwargs\u001b[49m\n\u001b[1;32m    706\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    707\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m colname, series \u001b[38;5;129;01min\u001b[39;00m df\u001b[38;5;241m.\u001b[39mitems():\n\u001b[1;32m    708\u001b[0m     \u001b[38;5;66;03m# TODO: this should write the \"true\" representation of the series (i.e. the underlying array or ndarray depending)\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/anndata/_io/utils.py:243\u001b[0m, in \u001b[0;36mreport_write_key_on_error.<locals>.func_wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    242\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 243\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    244\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/anndata/_io/specs/registry.py:309\u001b[0m, in \u001b[0;36mWriter.write_elem\u001b[0;34m(self, store, k, elem, dataset_kwargs, modifiers)\u001b[0m\n\u001b[1;32m    308\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcallback \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 309\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mwrite_func\u001b[49m\u001b[43m(\u001b[49m\u001b[43mstore\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43melem\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdataset_kwargs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdataset_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    310\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcallback(\n\u001b[1;32m    311\u001b[0m     write_func,\n\u001b[1;32m    312\u001b[0m     store,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    316\u001b[0m     iospec\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mregistry\u001b[38;5;241m.\u001b[39mget_spec(elem),\n\u001b[1;32m    317\u001b[0m )\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/anndata/_io/specs/registry.py:57\u001b[0m, in \u001b[0;36mwrite_spec.<locals>.decorator.<locals>.wrapper\u001b[0;34m(g, k, *args, **kwargs)\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[38;5;129m@wraps\u001b[39m(func)\n\u001b[1;32m     56\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mwrapper\u001b[39m(g: GroupStorageType, k: \u001b[38;5;28mstr\u001b[39m, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[0;32m---> 57\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mg\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     58\u001b[0m     g[k]\u001b[38;5;241m.\u001b[39mattrs\u001b[38;5;241m.\u001b[39msetdefault(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mencoding-type\u001b[39m\u001b[38;5;124m\"\u001b[39m, spec\u001b[38;5;241m.\u001b[39mencoding_type)\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/anndata/_io/specs/methods.py:414\u001b[0m, in \u001b[0;36mwrite_vlen_string_array\u001b[0;34m(f, k, elem, _writer, dataset_kwargs)\u001b[0m\n\u001b[1;32m    413\u001b[0m str_dtype \u001b[38;5;241m=\u001b[39m h5py\u001b[38;5;241m.\u001b[39mspecial_dtype(vlen\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mstr\u001b[39m)\n\u001b[0;32m--> 414\u001b[0m \u001b[43mf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcreate_dataset\u001b[49m\u001b[43m(\u001b[49m\u001b[43mk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43melem\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mastype\u001b[49m\u001b[43m(\u001b[49m\u001b[43mstr_dtype\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstr_dtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mdataset_kwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/sw/pkgs/arc/python3.11-anaconda/2024.02-1/lib/python3.11/site-packages/h5py/_hl/group.py:183\u001b[0m, in \u001b[0;36mGroup.create_dataset\u001b[0;34m(self, name, shape, dtype, data, **kwds)\u001b[0m\n\u001b[1;32m    181\u001b[0m         group \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrequire_group(parent_path)\n\u001b[0;32m--> 183\u001b[0m dsid \u001b[38;5;241m=\u001b[39m \u001b[43mdataset\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmake_new_dset\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgroup\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mshape\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    184\u001b[0m dset \u001b[38;5;241m=\u001b[39m dataset\u001b[38;5;241m.\u001b[39mDataset(dsid)\n",
      "File \u001b[0;32m/sw/pkgs/arc/python3.11-anaconda/2024.02-1/lib/python3.11/site-packages/h5py/_hl/dataset.py:166\u001b[0m, in \u001b[0;36mmake_new_dset\u001b[0;34m(parent, shape, dtype, data, name, chunks, compression, shuffle, fletcher32, maxshape, compression_opts, fillvalue, scaleoffset, track_times, external, track_order, dcpl, dapl, efile_prefix, virtual_prefix, allow_unknown_filter, rdcc_nslots, rdcc_nbytes, rdcc_w0)\u001b[0m\n\u001b[1;32m    165\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (data \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m) \u001b[38;5;129;01mand\u001b[39;00m (\u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(data, Empty)):\n\u001b[0;32m--> 166\u001b[0m     \u001b[43mdset_id\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwrite\u001b[49m\u001b[43m(\u001b[49m\u001b[43mh5s\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mALL\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mh5s\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mALL\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    168\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m dset_id\n",
      "File \u001b[0;32mh5py/_objects.pyx:54\u001b[0m, in \u001b[0;36mh5py._objects.with_phil.wrapper\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mh5py/_objects.pyx:55\u001b[0m, in \u001b[0;36mh5py._objects.with_phil.wrapper\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mh5py/h5d.pyx:283\u001b[0m, in \u001b[0;36mh5py.h5d.DatasetID.write\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mh5py/_proxy.pyx:146\u001b[0m, in \u001b[0;36mh5py._proxy.dset_rw\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mOSError\u001b[0m: [Errno 5] Can't write data (file write failed: time = Fri Sep  6 16:27:04 2024\n, filename = '/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/ASCL1_POU3F2_MYT1L_LMX1A_FOXA2.h5ad', file descriptor = 54, errno = 5, error message = 'Input/output error', buf = 0x1063e960, total write size = 272, bytes this sub-write = 272, bytes actually written = 18446744073709551615, offset = 0)",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 30\u001b[0m\n\u001b[1;32m     27\u001b[0m output_path \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(output_directory, file_name)\n\u001b[1;32m     29\u001b[0m \u001b[38;5;66;03m# Save the concatenated AnnData object to the file\u001b[39;00m\n\u001b[0;32m---> 30\u001b[0m \u001b[43mconcatenated_adata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwrite_h5ad\u001b[49m\u001b[43m(\u001b[49m\u001b[43moutput_path\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     31\u001b[0m \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1929\u001b[0m, in \u001b[0;36mAnnData.write_h5ad\u001b[0;34m(self, filename, compression, compression_opts, as_dense)\u001b[0m\n\u001b[1;32m   1926\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m filename \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1927\u001b[0m     filename \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfilename\n\u001b[0;32m-> 1929\u001b[0m \u001b[43mwrite_h5ad\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1930\u001b[0m \u001b[43m    \u001b[49m\u001b[43mPath\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1931\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1932\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcompression\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1933\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression_opts\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcompression_opts\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1934\u001b[0m \u001b[43m    \u001b[49m\u001b[43mas_dense\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mas_dense\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1935\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1937\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39misbacked:\n\u001b[1;32m   1938\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfile\u001b[38;5;241m.\u001b[39mfilename \u001b[38;5;241m=\u001b[39m filename\n",
      "File \u001b[0;32m~/.local/lib/python3.11/site-packages/anndata/_io/h5ad.py:77\u001b[0m, in \u001b[0;36mwrite_h5ad\u001b[0;34m(filepath, adata, as_dense, dataset_kwargs, **kwargs)\u001b[0m\n\u001b[1;32m     74\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m adata\u001b[38;5;241m.\u001b[39misbacked:  \u001b[38;5;66;03m# close so that we can reopen below\u001b[39;00m\n\u001b[1;32m     75\u001b[0m     adata\u001b[38;5;241m.\u001b[39mfile\u001b[38;5;241m.\u001b[39mclose()\n\u001b[0;32m---> 77\u001b[0m \u001b[43m\u001b[49m\u001b[38;5;28;43;01mwith\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mh5py\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mFile\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mas\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mf\u001b[49m\u001b[43m:\u001b[49m\n\u001b[1;32m     78\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;66;43;03m# TODO: Use spec writing system for this\u001b[39;49;00m\n\u001b[1;32m     79\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;66;43;03m# Currently can't use write_dispatched here because this function is also called to do an\u001b[39;49;00m\n\u001b[1;32m     80\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;66;43;03m# inplace update of a backed object, which would delete \"/\"\u001b[39;49;00m\n\u001b[1;32m     81\u001b[0m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mf\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\n\u001b[1;32m     82\u001b[0m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mattrs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msetdefault\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding-type\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43manndata\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32mh5py/_objects.pyx:54\u001b[0m, in \u001b[0;36mh5py._objects.with_phil.wrapper\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mh5py/_objects.pyx:55\u001b[0m, in \u001b[0;36mh5py._objects.with_phil.wrapper\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m/sw/pkgs/arc/python3.11-anaconda/2024.02-1/lib/python3.11/site-packages/h5py/_hl/files.py:604\u001b[0m, in \u001b[0;36mFile.__exit__\u001b[0;34m(self, *args)\u001b[0m\n\u001b[1;32m    601\u001b[0m \u001b[38;5;129m@with_phil\u001b[39m\n\u001b[1;32m    602\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__exit__\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39margs):\n\u001b[1;32m    603\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mid:\n\u001b[0;32m--> 604\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mclose\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/sw/pkgs/arc/python3.11-anaconda/2024.02-1/lib/python3.11/site-packages/h5py/_hl/files.py:586\u001b[0m, in \u001b[0;36mFile.close\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    580\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mid\u001b[38;5;241m.\u001b[39mvalid:\n\u001b[1;32m    581\u001b[0m     \u001b[38;5;66;03m# We have to explicitly murder all open objects related to the file\u001b[39;00m\n\u001b[1;32m    582\u001b[0m \n\u001b[1;32m    583\u001b[0m     \u001b[38;5;66;03m# Close file-resident objects first, then the files.\u001b[39;00m\n\u001b[1;32m    584\u001b[0m     \u001b[38;5;66;03m# Otherwise we get errors in MPI mode.\u001b[39;00m\n\u001b[1;32m    585\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mid\u001b[38;5;241m.\u001b[39m_close_open_objects(h5f\u001b[38;5;241m.\u001b[39mOBJ_LOCAL \u001b[38;5;241m|\u001b[39m \u001b[38;5;241m~\u001b[39mh5f\u001b[38;5;241m.\u001b[39mOBJ_FILE)\n\u001b[0;32m--> 586\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mid\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_close_open_objects\u001b[49m\u001b[43m(\u001b[49m\u001b[43mh5f\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mOBJ_LOCAL\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m|\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mh5f\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mOBJ_FILE\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    588\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mid\u001b[38;5;241m.\u001b[39mclose()\n\u001b[1;32m    589\u001b[0m     _objects\u001b[38;5;241m.\u001b[39mnonlocal_close()\n",
      "File \u001b[0;32mh5py/_objects.pyx:54\u001b[0m, in \u001b[0;36mh5py._objects.with_phil.wrapper\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mh5py/_objects.pyx:55\u001b[0m, in \u001b[0;36mh5py._objects.with_phil.wrapper\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mh5py/h5f.pyx:360\u001b[0m, in \u001b[0;36mh5py.h5f.FileID._close_open_objects\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mOSError\u001b[0m: [Errno 5] Unable to extend file properly, errno = 9, error message = 'bad file descriptor' (file write failed: time = Fri Sep  6 16:27:05 2024\n, filename = '/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/ASCL1_POU3F2_MYT1L_LMX1A_FOXA2.h5ad', file descriptor = 54, errno = 5, error message = 'Input/output error', buf = 0x1063e960, total write size = 272, bytes this sub-write = 272, bytes actually written = 18446744073709551615, offset = 0)"
     ]
    }
   ],
   "source": [
    "output_directory = \"/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed\"\n",
    "scalars = [0.5, 0.75, 1.001]\n",
    "for i in range(len(df['TFs'])):\n",
    "    val = df['TFs'].iloc[i]\n",
    "    TFs = val.split()\n",
    "    if validateTFs(TFs, adata):\n",
    "        print(TFs)\n",
    "        adataDict = iterate_perturb_counts(adata.copy(), TFs, scalars)\n",
    "\n",
    "        # Concatenate all AnnData objects along the observations axis\n",
    "        concatenated_adata = ad.concat(list(adataDict.values()), axis=0)\n",
    "\n",
    "        # Save reprogramming metadata into the concatenated_adata.obs table\n",
    "        concatenated_adata.obs['Source_cells'] = df['Source cells'].iloc[i]\n",
    "        concatenated_adata.obs['Target_cells'] = df['Target cells'].iloc[i]\n",
    "        concatenated_adata.obs['Treatment'] = df['Treatment'].iloc[i]\n",
    "        concatenated_adata.obs['Species'] = df['Species'].iloc[i]\n",
    "        concatenated_adata.obs['Cell_Transplantation'] = df['Cell Transplantation'].iloc[i]\n",
    "        concatenated_adata.obs['Published_Year'] = df['Published Year'].iloc[i]\n",
    "        concatenated_adata.obs['PMID'] = df['PMID'].iloc[i]\n",
    "        \n",
    "        # Join the TFs list into a string for the filename\n",
    "        TFs_str = \"_\".join(TFs)\n",
    "        \n",
    "        # Generate the file path for saving\n",
    "        file_name = f\"{TFs_str}.h5ad\"\n",
    "        output_path = os.path.join(output_directory, file_name)\n",
    "        \n",
    "        # Save the concatenated AnnData object to the file\n",
    "        concatenated_adata.write_h5ad(output_path)\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a12afe42-e76a-4a99-bd2e-45a13a2c2097",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organ_tissue</th>\n",
       "      <th>method</th>\n",
       "      <th>donor</th>\n",
       "      <th>anatomical_information</th>\n",
       "      <th>n_counts_UMIs</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>cell_ontology_class</th>\n",
       "      <th>free_annotation</th>\n",
       "      <th>manually_annotated</th>\n",
       "      <th>compartment</th>\n",
       "      <th>gender</th>\n",
       "      <th>U</th>\n",
       "      <th>Source_cells</th>\n",
       "      <th>Target_cells</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>Species</th>\n",
       "      <th>Cell_Transplantation</th>\n",
       "      <th>Published_Year</th>\n",
       "      <th>PMID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1</th>\n",
       "      <td>Liver</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP6</td>\n",
       "      <td>nan</td>\n",
       "      <td>10423.0</td>\n",
       "      <td>2986</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>Stellate/Fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>male</td>\n",
       "      <td>0.500</td>\n",
       "      <td>Embryonic Fibroblasts and Postnatal Fibroblasts</td>\n",
       "      <td>Neuronal Cells</td>\n",
       "      <td>ASCL1, BRN2, MYT1L, LMX1A, and FOXA2</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>21646515_29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1</th>\n",
       "      <td>Liver</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP6</td>\n",
       "      <td>nan</td>\n",
       "      <td>11223.0</td>\n",
       "      <td>3248</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>Stellate/Fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>male</td>\n",
       "      <td>0.500</td>\n",
       "      <td>Embryonic Fibroblasts and Postnatal Fibroblasts</td>\n",
       "      <td>Neuronal Cells</td>\n",
       "      <td>ASCL1, BRN2, MYT1L, LMX1A, and FOXA2</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>21646515_29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1</th>\n",
       "      <td>Liver</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP6</td>\n",
       "      <td>nan</td>\n",
       "      <td>14695.0</td>\n",
       "      <td>4469</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>Stellate/Fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>male</td>\n",
       "      <td>0.500</td>\n",
       "      <td>Embryonic Fibroblasts and Postnatal Fibroblasts</td>\n",
       "      <td>Neuronal Cells</td>\n",
       "      <td>ASCL1, BRN2, MYT1L, LMX1A, and FOXA2</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>21646515_29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1</th>\n",
       "      <td>Liver</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP6</td>\n",
       "      <td>nan</td>\n",
       "      <td>8977.0</td>\n",
       "      <td>2875</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>Stellate/Fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>male</td>\n",
       "      <td>0.500</td>\n",
       "      <td>Embryonic Fibroblasts and Postnatal Fibroblasts</td>\n",
       "      <td>Neuronal Cells</td>\n",
       "      <td>ASCL1, BRN2, MYT1L, LMX1A, and FOXA2</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>21646515_29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1</th>\n",
       "      <td>Liver</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP6</td>\n",
       "      <td>nan</td>\n",
       "      <td>12109.0</td>\n",
       "      <td>3239</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>Stellate/Fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>male</td>\n",
       "      <td>0.500</td>\n",
       "      <td>Embryonic Fibroblasts and Postnatal Fibroblasts</td>\n",
       "      <td>Neuronal Cells</td>\n",
       "      <td>ASCL1, BRN2, MYT1L, LMX1A, and FOXA2</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>21646515_29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_Vasculature_aorta_SS2_B114585_B133324_Stromal_P3_S99</th>\n",
       "      <td>Vasculature</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>nan</td>\n",
       "      <td>308152.0</td>\n",
       "      <td>2090</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>female</td>\n",
       "      <td>1.001</td>\n",
       "      <td>Embryonic Fibroblasts and Postnatal Fibroblasts</td>\n",
       "      <td>Neuronal Cells</td>\n",
       "      <td>ASCL1, BRN2, MYT1L, LMX1A, and FOXA2</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>21646515_29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_Vasculature_aorta_SS2_B114585_B133324_Stromal_P7_S103</th>\n",
       "      <td>Vasculature</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>nan</td>\n",
       "      <td>387043.0</td>\n",
       "      <td>2132</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>female</td>\n",
       "      <td>1.001</td>\n",
       "      <td>Embryonic Fibroblasts and Postnatal Fibroblasts</td>\n",
       "      <td>Neuronal Cells</td>\n",
       "      <td>ASCL1, BRN2, MYT1L, LMX1A, and FOXA2</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>21646515_29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_Vasculature_aorta_SS2_B114585_B133324_Stromal_P9_S105</th>\n",
       "      <td>Vasculature</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>nan</td>\n",
       "      <td>412608.0</td>\n",
       "      <td>3140</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>female</td>\n",
       "      <td>1.001</td>\n",
       "      <td>Embryonic Fibroblasts and Postnatal Fibroblasts</td>\n",
       "      <td>Neuronal Cells</td>\n",
       "      <td>ASCL1, BRN2, MYT1L, LMX1A, and FOXA2</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>21646515_29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_Vasculature_aorta_SS2_B113343_B133091_Immune_L6_S270</th>\n",
       "      <td>Vasculature</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>aorta</td>\n",
       "      <td>341884.0</td>\n",
       "      <td>4379</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>female</td>\n",
       "      <td>1.001</td>\n",
       "      <td>Embryonic Fibroblasts and Postnatal Fibroblasts</td>\n",
       "      <td>Neuronal Cells</td>\n",
       "      <td>ASCL1, BRN2, MYT1L, LMX1A, and FOXA2</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>21646515_29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_Vasculature_aorta_SS2_B113343_B133091_Immune_N6_S318</th>\n",
       "      <td>Vasculature</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>aorta</td>\n",
       "      <td>141115.0</td>\n",
       "      <td>5842</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>female</td>\n",
       "      <td>1.001</td>\n",
       "      <td>Embryonic Fibroblasts and Postnatal Fibroblasts</td>\n",
       "      <td>Neuronal Cells</td>\n",
       "      <td>ASCL1, BRN2, MYT1L, LMX1A, and FOXA2</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>21646515_29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>114453 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   organ_tissue     method  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1                    Liver        10X   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1                    Liver        10X   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1                    Liver        10X   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1                    Liver        10X   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1                    Liver        10X   \n",
       "...                                                         ...        ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  Vasculature  smartseq2   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  Vasculature  smartseq2   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  Vasculature  smartseq2   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  Vasculature  smartseq2   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  Vasculature  smartseq2   \n",
       "\n",
       "                                                   donor  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1              TSP6   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1              TSP6   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1              TSP6   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1              TSP6   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1              TSP6   \n",
       "...                                                  ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  TSP2   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  TSP2   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  TSP2   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  TSP2   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  TSP2   \n",
       "\n",
       "                                                   anatomical_information  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1                                nan   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1                                nan   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1                                nan   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1                                nan   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1                                nan   \n",
       "...                                                                   ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...                    nan   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...                    nan   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...                    nan   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...                  aorta   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...                  aorta   \n",
       "\n",
       "                                                    n_counts_UMIs  n_genes  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1                    10423.0     2986   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1                    11223.0     3248   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1                    14695.0     4469   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1                     8977.0     2875   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1                    12109.0     3239   \n",
       "...                                                           ...      ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...       308152.0     2090   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...       387043.0     2132   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...       412608.0     3140   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...       341884.0     4379   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...       141115.0     5842   \n",
       "\n",
       "                                                   cell_ontology_class  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1                      fibroblast   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1                      fibroblast   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1                      fibroblast   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1                      fibroblast   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1                      fibroblast   \n",
       "...                                                                ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...          fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...          fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...          fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...          fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...          fibroblast   \n",
       "\n",
       "                                                        free_annotation  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1              Stellate/Fibroblast   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1              Stellate/Fibroblast   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1              Stellate/Fibroblast   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1              Stellate/Fibroblast   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1              Stellate/Fibroblast   \n",
       "...                                                                 ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...           fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...           fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...           fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...           fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...           fibroblast   \n",
       "\n",
       "                                                    manually_annotated  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1                            True   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1                            True   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1                            True   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1                            True   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1                            True   \n",
       "...                                                                ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...                True   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...                True   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...                True   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...                True   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...                True   \n",
       "\n",
       "                                                   compartment  gender      U  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1                 stromal    male  0.500   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1                 stromal    male  0.500   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1                 stromal    male  0.500   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1                 stromal    male  0.500   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1                 stromal    male  0.500   \n",
       "...                                                        ...     ...    ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...     stromal  female  1.001   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...     stromal  female  1.001   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...     stromal  female  1.001   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...     stromal  female  1.001   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...     stromal  female  1.001   \n",
       "\n",
       "                                                                                       Source_cells  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1              Embryonic Fibroblasts and Postnatal Fibroblasts   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1              Embryonic Fibroblasts and Postnatal Fibroblasts   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1              Embryonic Fibroblasts and Postnatal Fibroblasts   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1              Embryonic Fibroblasts and Postnatal Fibroblasts   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1              Embryonic Fibroblasts and Postnatal Fibroblasts   \n",
       "...                                                                                             ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  Embryonic Fibroblasts and Postnatal Fibroblasts   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  Embryonic Fibroblasts and Postnatal Fibroblasts   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  Embryonic Fibroblasts and Postnatal Fibroblasts   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  Embryonic Fibroblasts and Postnatal Fibroblasts   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  Embryonic Fibroblasts and Postnatal Fibroblasts   \n",
       "\n",
       "                                                      Target_cells  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1              Neuronal Cells   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1              Neuronal Cells   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1              Neuronal Cells   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1              Neuronal Cells   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1              Neuronal Cells   \n",
       "...                                                            ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  Neuronal Cells   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  Neuronal Cells   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  Neuronal Cells   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  Neuronal Cells   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  Neuronal Cells   \n",
       "\n",
       "                                                                               Treatment  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1              ASCL1, BRN2, MYT1L, LMX1A, and FOXA2   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1              ASCL1, BRN2, MYT1L, LMX1A, and FOXA2   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1              ASCL1, BRN2, MYT1L, LMX1A, and FOXA2   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1              ASCL1, BRN2, MYT1L, LMX1A, and FOXA2   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1              ASCL1, BRN2, MYT1L, LMX1A, and FOXA2   \n",
       "...                                                                                  ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  ASCL1, BRN2, MYT1L, LMX1A, and FOXA2   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  ASCL1, BRN2, MYT1L, LMX1A, and FOXA2   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  ASCL1, BRN2, MYT1L, LMX1A, and FOXA2   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  ASCL1, BRN2, MYT1L, LMX1A, and FOXA2   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  ASCL1, BRN2, MYT1L, LMX1A, and FOXA2   \n",
       "\n",
       "                                                   Species  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1               Human   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1               Human   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1               Human   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1               Human   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1               Human   \n",
       "...                                                    ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...   Human   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...   Human   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...   Human   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...   Human   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...   Human   \n",
       "\n",
       "                                                    Cell_Transplantation  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1                               NaN   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1                               NaN   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1                               NaN   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1                               NaN   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1                               NaN   \n",
       "...                                                                  ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...                   NaN   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...                   NaN   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...                   NaN   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...                   NaN   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...                   NaN   \n",
       "\n",
       "                                                    Published_Year  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1                      2011.0   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1                      2011.0   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1                      2011.0   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1                      2011.0   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1                      2011.0   \n",
       "...                                                            ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...          2011.0   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...          2011.0   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...          2011.0   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...          2011.0   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...          2011.0   \n",
       "\n",
       "                                                           PMID  \n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1              21646515_29  \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1              21646515_29  \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1              21646515_29  \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1              21646515_29  \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1              21646515_29  \n",
       "...                                                         ...  \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  21646515_29  \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  21646515_29  \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  21646515_29  \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  21646515_29  \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  21646515_29  \n",
       "\n",
       "[114453 rows x 19 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concatenated_adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c4134986-5695-49de-9d78-c4497fcff9d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>ensemblid</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>False</td>\n",
       "      <td>6.398244e-05</td>\n",
       "      <td>0.835044</td>\n",
       "      <td>-0.573947</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.005574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>False</td>\n",
       "      <td>2.274395e-03</td>\n",
       "      <td>2.442280</td>\n",
       "      <td>0.533203</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.031731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR6859-1</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>False</td>\n",
       "      <td>6.175251e-05</td>\n",
       "      <td>1.295335</td>\n",
       "      <td>-0.256874</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.005634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>False</td>\n",
       "      <td>1.372886e-04</td>\n",
       "      <td>2.656352</td>\n",
       "      <td>0.680668</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.008041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND6</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>False</td>\n",
       "      <td>9.634841e-01</td>\n",
       "      <td>2.466404</td>\n",
       "      <td>0.154140</td>\n",
       "      <td>0.590065</td>\n",
       "      <td>0.741395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TE</th>\n",
       "      <td>MT-TE</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.600667e-01</td>\n",
       "      <td>1.603787</td>\n",
       "      <td>-0.044396</td>\n",
       "      <td>0.083929</td>\n",
       "      <td>0.301820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CYB</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>False</td>\n",
       "      <td>4.367693e+00</td>\n",
       "      <td>4.765751</td>\n",
       "      <td>-0.499747</td>\n",
       "      <td>3.874830</td>\n",
       "      <td>1.104192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TT</th>\n",
       "      <td>MT-TT</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>False</td>\n",
       "      <td>6.573967e-02</td>\n",
       "      <td>0.624316</td>\n",
       "      <td>-0.719108</td>\n",
       "      <td>0.040580</td>\n",
       "      <td>0.186848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TP</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>True</td>\n",
       "      <td>4.385102e-01</td>\n",
       "      <td>2.705558</td>\n",
       "      <td>1.115482</td>\n",
       "      <td>0.204806</td>\n",
       "      <td>0.501265</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58870 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_symbol     feature_type          ensemblid  highly_variable  \\\n",
       "DDX11L1          DDX11L1  Gene Expression  ENSG00000223972.5            False   \n",
       "WASH7P            WASH7P  Gene Expression  ENSG00000227232.5            False   \n",
       "MIR6859-1      MIR6859-1  Gene Expression  ENSG00000278267.1            False   \n",
       "MIR1302-2HG  MIR1302-2HG  Gene Expression  ENSG00000243485.5            False   \n",
       "MIR1302-2      MIR1302-2  Gene Expression  ENSG00000284332.1            False   \n",
       "...                  ...              ...                ...              ...   \n",
       "MT-ND6            MT-ND6  Gene Expression  ENSG00000198695.2            False   \n",
       "MT-TE              MT-TE  Gene Expression  ENSG00000210194.1            False   \n",
       "MT-CYB            MT-CYB  Gene Expression  ENSG00000198727.2            False   \n",
       "MT-TT              MT-TT  Gene Expression  ENSG00000210195.2            False   \n",
       "MT-TP              MT-TP  Gene Expression  ENSG00000210196.2             True   \n",
       "\n",
       "                    means  dispersions  dispersions_norm      mean       std  \n",
       "DDX11L1      6.398244e-05     0.835044         -0.573947  0.000039  0.005574  \n",
       "WASH7P       2.274395e-03     2.442280          0.533203  0.001080  0.031731  \n",
       "MIR6859-1    6.175251e-05     1.295335         -0.256874  0.000033  0.005634  \n",
       "MIR1302-2HG  1.372886e-04     2.656352          0.680668  0.000048  0.008041  \n",
       "MIR1302-2    1.000000e-12          NaN          0.000000  0.000000  1.000000  \n",
       "...                   ...          ...               ...       ...       ...  \n",
       "MT-ND6       9.634841e-01     2.466404          0.154140  0.590065  0.741395  \n",
       "MT-TE        1.600667e-01     1.603787         -0.044396  0.083929  0.301820  \n",
       "MT-CYB       4.367693e+00     4.765751         -0.499747  3.874830  1.104192  \n",
       "MT-TT        6.573967e-02     0.624316         -0.719108  0.040580  0.186848  \n",
       "MT-TP        4.385102e-01     2.705558          1.115482  0.204806  0.501265  \n",
       "\n",
       "[58870 rows x 9 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f93f44c0-099c-4799-9156-e982f5731d8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR6859-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND6</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CYB</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TP</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58870 rows × 0 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [DDX11L1, WASH7P, MIR6859-1, MIR1302-2HG, MIR1302-2, FAM138A, OR4G4P, OR4G11P, OR4F5, AL627309.1, AL627309.3, CICP27, AL627309.6, AL627309.7, AL627309.2, AL627309.5, RNU6-1100P, AL627309.4, FO538757.1, WASH9P, MIR6859-2, AP006222.1, RPL23AP24, AL732372.1, AL732372.2, WBP1LP7, OR4F29, CICP7, AL732372.3, RF00026, AL669831.3, AC114498.1, MTND1P23, MTND2P28, MTCO1P12, AC114498.2, MTCO2P12, MTATP8P1, MTATP6P1, MTCO3P12, WBP1LP6, OR4F16, CICP3, AL669831.1, RNU6-1199P, AL669831.2, AL669831.5, AL669831.7, AL669831.4, FAM87B, LINC01128, LINC00115, FAM41C, TUBB8P11, AL669831.6, AL645608.6, AL645608.2, AL645608.4, LINC02593, SAMD11, NOC2L, KLHL17, PLEKHN1, PERM1, AL645608.7, HES4, ISG15, AL645608.3, AL645608.1, AGRN, AL645608.5, AL390719.1, AL645608.8, RNF223, C1orf159, AL390719.3, LINC01342, MIR200B, MIR200A, MIR429, AL390719.2, TTLL10-AS1, TTLL10, TNFRSF18, TNFRSF4, SDF4, B3GALT6, C1QTNF12, AL162741.1, UBE2J2, LINC01786, SCNN1D, ACAP3, MIR6726, PUSL1, INTS11, MIR6727, AL139287.1, CPTP, TAS1R3, ...]\n",
       "\n",
       "[58870 rows x 0 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concatenated_adata.var"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6bbd49e-6ef9-4f73-9f9e-5a0bf71482cd",
   "metadata": {},
   "source": [
    "# Day 6\n",
    "\n",
    "Make perturbations for all TFs in the [Human Transcription Factors Database](https://humantfs.ccbr.utoronto.ca/download.php)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44c9ace9-d19a-4c12-b0e4-a48e8f32b330",
   "metadata": {},
   "source": [
    "## Set up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7e59d9b7-a261-4825-bb15-854b8fb185d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import scanpy as sp\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8b6af9ac-8c43-400a-895c-91ded6f2bcd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def iterate_perturb_counts(adata, tf_list, scalar_list):\n",
    "    \"\"\"\n",
    "    Applies perturbations to the expression data of specified transcription factors across multiple scalars \n",
    "    and stores the resulting AnnData objects in a dictionary.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Iterates over a list of scalar values.\n",
    "    2. For each scalar, creates a copy of the AnnData object to preserve the original data.\n",
    "    3. Applies the `perturb_counts` function to scale the expression data of genes listed in `tf_list` by\n",
    "       the maximum gene expression of each cell and the current scalar.\n",
    "    4. Stores the perturbed AnnData object in a dictionary with the scalar as the key.\n",
    "\n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols (transcription factors) to be perturbed.\n",
    "    scalar_list (list): A list of scalar values for scaling the gene expression.\n",
    "    adata (AnnData): The AnnData object containing gene expression data (cells x genes).\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary where keys are scalar values and values are the corresponding perturbed AnnData objects.\n",
    "    \"\"\"\n",
    "    \n",
    "    adata_dict = {}\n",
    "    \n",
    "    for scalar in scalar_list:\n",
    "        # Create a copy of the AnnData object for each scalar value\n",
    "        adata_temp = adata.copy()\n",
    "        \n",
    "        # Apply perturb_counts to the copied AnnData object\n",
    "        perturbed_adata = perturb_counts(adata_temp, tf_list, scalar)\n",
    "        \n",
    "        # Store the perturbed AnnData object in the dictionary with scalar as the key\n",
    "        adata_dict[scalar] = perturbed_adata\n",
    "    \n",
    "    return adata_dict\n",
    "\n",
    "def perturb_counts(adata, tf_list, scalar): \n",
    "    \"\"\"\n",
    "    Applies a perturbation to the expression data of specific genes in an AnnData object.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Computes the maximum gene expression level for each cell.\n",
    "    2. Applies a scaling operation to the expression levels of genes listed in `tf_list`.\n",
    "       - Each entry of these genes in the matrix is multiplied by the maximum expression level \n",
    "         of its respective cell and a specified scalar value.\n",
    "    3. Updates the AnnData object with new columns:\n",
    "       - 'scaled': A boolean column indicating whether each gene is in the `tf_list`.\n",
    "       - 'scaled_by': Contains the scaling factor used for each gene (the product of the maximum \n",
    "         expression level of each cell and the scalar), or `1` if the gene was not in `tf_list`.\n",
    "    \n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols to be perturbed.\n",
    "    scalar (float): The scalar value used to scale the expression levels.\n",
    "    adata (AnnData): The AnnData object containing gene expression data.\n",
    "\n",
    "    Returns:\n",
    "    AnnData: The updated AnnData object with applied perturbations and new columns.\n",
    "    \"\"\"\n",
    "\n",
    "    # Create a boolean mask for genes in tf_list\n",
    "    gene_mask = adata.var['gene_symbol'].isin(tf_list)\n",
    "    \n",
    "    # Save the original state of the parameter objects, in case some tfs do not translate (failsafe)\n",
    "    original_X = adata.X.copy()\n",
    "    original_gene_mask = gene_mask.copy()\n",
    "    \n",
    "    # Compute maximum expression level of each cell\n",
    "    max_exp = np.max(adata.X, axis=1)\n",
    "\n",
    "    \"\"\"This is new today. v \"\"\"\n",
    "    # Raise an error if any of the gene names in tf_list do not match column names (we will manually update these in adata):\n",
    "    missing_genes = [gene for gene in tf_list if gene not in adata.var['gene_symbol'].values]\n",
    "    \n",
    "    if missing_genes:\n",
    "        # Restore original parameter objects\n",
    "        adata.X = original_X\n",
    "        gene_mask = original_gene_mask\n",
    "        raise ValueError(f\"Genes {missing_genes} not found in anndata object\")\n",
    "\n",
    "    else:    \n",
    "        \n",
    "        # Apply the scaling operation to the specified genes\n",
    "        adata.X[:, gene_mask] = max_exp * scalar\n",
    "        \n",
    "        # Add/Update 'scaled' column in var\n",
    "        adata.var['scaled'] = gene_mask\n",
    "        \n",
    "        # Add/Update 'scaled_by' column in var\n",
    "        adata.var['scaled_by'] = scalar  # Default value for genes not in tf_list\n",
    "\n",
    "        # Add/Update 'scaled_by' column in var\n",
    "        adata.obs['U'] = scalar  # Default value for genes not in tf_list\n",
    "    \n",
    "    return adata\n",
    "\n",
    "def validateTFs(TFs, adata):\n",
    "    adata_gene_list = adata.var['gene_symbol'].values.tolist()\n",
    "    for TF in TFs:\n",
    "        if TF not in adata_gene_list:\n",
    "            return False\n",
    "    return True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a01995a-15db-4439-a1eb-73d77721229f",
   "metadata": {},
   "source": [
    "## Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c9a87ba0-d25a-4044-bb67-0390dfe11563",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('data/HumanTFs_v_1.01.csv')\n",
    "\n",
    "DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/jpic/\"\n",
    "FILE = \"fibroblast.h5ad\"\n",
    "adata = sp.read_h5ad(os.path.join(DATAPATH, FILE))\n",
    "adata_gene_list = adata.var['gene_symbol'].values.tolist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f1dde077-ac13-4a0c-88db-b40bd3d898ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[df['Is TF?'] == 'Yes']\n",
    "all_tfs = list(df['HGNC symbol'].values)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a56e66ee-72a3-4a9f-8b63-cc1e0bc804bf",
   "metadata": {},
   "source": [
    "## Make Perturbation Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "053b3dcc-164f-445f-8bcb-b7951fb93d4a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ARID3C']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ARID5A']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ARID5B']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['KDM5B']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ARID2']\n"
     ]
    }
   ],
   "source": [
    "output_directory = \"/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed\"\n",
    "scalars = [0.5, 0.75, 1.001]\n",
    "for tf in all_tfs:\n",
    "    TFs = [tf]\n",
    "    if validateTFs(TFs, adata):\n",
    "        # Join the TFs list into a string for the filename\n",
    "        TFs_str = \"_\".join(TFs)\n",
    "       \n",
    "        # Generate the file path for saving\n",
    "        file_name = f\"{TFs_str}.h5ad\"\n",
    "        output_path = os.path.join(output_directory, file_name)\n",
    "\n",
    "        # prevents writing duplicate files\n",
    "        if os.path.exists(output_path):\n",
    "            continue\n",
    "\n",
    "        # display some information\n",
    "        print(TFs)\n",
    "\n",
    "        adataDict = iterate_perturb_counts(adata, TFs, scalars)\n",
    "\n",
    "        # Concatenate all AnnData objects along the observations axis\n",
    "        concatenated_adata = ad.concat(list(adataDict.values()), axis=0)\n",
    "\n",
    "        # Save reprogramming metadata into the concatenated_adata.obs table\n",
    "        concatenated_adata.obs['HumanTFs'] = True\n",
    "       \n",
    "        # Save the concatenated AnnData object to the file\n",
    "        concatenated_adata.write_h5ad(output_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c46c8982-c9d9-4e03-9440-e31d35f73caf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organ_tissue</th>\n",
       "      <th>method</th>\n",
       "      <th>donor</th>\n",
       "      <th>anatomical_information</th>\n",
       "      <th>n_counts_UMIs</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>cell_ontology_class</th>\n",
       "      <th>free_annotation</th>\n",
       "      <th>manually_annotated</th>\n",
       "      <th>compartment</th>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1</th>\n",
       "      <td>Liver</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP6</td>\n",
       "      <td>nan</td>\n",
       "      <td>10423.0</td>\n",
       "      <td>2986</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>Stellate/Fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1</th>\n",
       "      <td>Liver</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP6</td>\n",
       "      <td>nan</td>\n",
       "      <td>11223.0</td>\n",
       "      <td>3248</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>Stellate/Fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1</th>\n",
       "      <td>Liver</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP6</td>\n",
       "      <td>nan</td>\n",
       "      <td>14695.0</td>\n",
       "      <td>4469</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>Stellate/Fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1</th>\n",
       "      <td>Liver</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP6</td>\n",
       "      <td>nan</td>\n",
       "      <td>8977.0</td>\n",
       "      <td>2875</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>Stellate/Fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1</th>\n",
       "      <td>Liver</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP6</td>\n",
       "      <td>nan</td>\n",
       "      <td>12109.0</td>\n",
       "      <td>3239</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>Stellate/Fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_Vasculature_aorta_SS2_B114585_B133324_Stromal_P3_S99</th>\n",
       "      <td>Vasculature</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>nan</td>\n",
       "      <td>308152.0</td>\n",
       "      <td>2090</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_Vasculature_aorta_SS2_B114585_B133324_Stromal_P7_S103</th>\n",
       "      <td>Vasculature</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>nan</td>\n",
       "      <td>387043.0</td>\n",
       "      <td>2132</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_Vasculature_aorta_SS2_B114585_B133324_Stromal_P9_S105</th>\n",
       "      <td>Vasculature</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>nan</td>\n",
       "      <td>412608.0</td>\n",
       "      <td>3140</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_Vasculature_aorta_SS2_B113343_B133091_Immune_L6_S270</th>\n",
       "      <td>Vasculature</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>aorta</td>\n",
       "      <td>341884.0</td>\n",
       "      <td>4379</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_Vasculature_aorta_SS2_B113343_B133091_Immune_N6_S318</th>\n",
       "      <td>Vasculature</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>aorta</td>\n",
       "      <td>141115.0</td>\n",
       "      <td>5842</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>114453 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   organ_tissue     method  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1                    Liver        10X   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1                    Liver        10X   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1                    Liver        10X   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1                    Liver        10X   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1                    Liver        10X   \n",
       "...                                                         ...        ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  Vasculature  smartseq2   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  Vasculature  smartseq2   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  Vasculature  smartseq2   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  Vasculature  smartseq2   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  Vasculature  smartseq2   \n",
       "\n",
       "                                                   donor  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1              TSP6   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1              TSP6   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1              TSP6   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1              TSP6   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1              TSP6   \n",
       "...                                                  ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  TSP2   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  TSP2   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  TSP2   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  TSP2   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  TSP2   \n",
       "\n",
       "                                                   anatomical_information  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1                                nan   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1                                nan   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1                                nan   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1                                nan   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1                                nan   \n",
       "...                                                                   ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...                    nan   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...                    nan   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...                    nan   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...                  aorta   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...                  aorta   \n",
       "\n",
       "                                                    n_counts_UMIs  n_genes  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1                    10423.0     2986   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1                    11223.0     3248   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1                    14695.0     4469   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1                     8977.0     2875   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1                    12109.0     3239   \n",
       "...                                                           ...      ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...       308152.0     2090   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...       387043.0     2132   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...       412608.0     3140   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...       341884.0     4379   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...       141115.0     5842   \n",
       "\n",
       "                                                   cell_ontology_class  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1                      fibroblast   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1                      fibroblast   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1                      fibroblast   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1                      fibroblast   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1                      fibroblast   \n",
       "...                                                                ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...          fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...          fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...          fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...          fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...          fibroblast   \n",
       "\n",
       "                                                        free_annotation  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1              Stellate/Fibroblast   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1              Stellate/Fibroblast   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1              Stellate/Fibroblast   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1              Stellate/Fibroblast   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1              Stellate/Fibroblast   \n",
       "...                                                                 ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...           fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...           fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...           fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...           fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...           fibroblast   \n",
       "\n",
       "                                                    manually_annotated  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1                            True   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1                            True   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1                            True   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1                            True   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1                            True   \n",
       "...                                                                ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...                True   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...                True   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...                True   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...                True   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...                True   \n",
       "\n",
       "                                                   compartment  gender  \n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1                 stromal    male  \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1                 stromal    male  \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1                 stromal    male  \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1                 stromal    male  \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1                 stromal    male  \n",
       "...                                                        ...     ...  \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...     stromal  female  \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...     stromal  female  \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...     stromal  female  \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...     stromal  female  \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...     stromal  female  \n",
       "\n",
       "[114453 rows x 11 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concatenated_adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a4bb996-bd90-41e4-8890-dfcc4558cf2b",
   "metadata": {},
   "source": [
    "# Day 5\n",
    "\n",
    "Helping Nat make some changes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fd2f5f56-5dab-4687-b5dd-9712376a1a0c",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m output_directory \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      2\u001b[0m scalars \u001b[38;5;241m=\u001b[39m [\u001b[38;5;241m0.5\u001b[39m, \u001b[38;5;241m0.75\u001b[39m, \u001b[38;5;241m1.001\u001b[39m]\n\u001b[0;32m----> 3\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;28mlen\u001b[39m(\u001b[43mdf\u001b[49m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTFs\u001b[39m\u001b[38;5;124m'\u001b[39m])):\n\u001b[1;32m      4\u001b[0m     val \u001b[38;5;241m=\u001b[39m df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTFs\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39miloc[i]\n\u001b[1;32m      5\u001b[0m     val \u001b[38;5;241m=\u001b[39m val\u001b[38;5;241m.\u001b[39mreplace(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m \u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'df' is not defined"
     ]
    }
   ],
   "source": [
    "output_directory = \"/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed\"\n",
    "scalars = [0.5, 0.75, 1.001]\n",
    "for i in range(len(df['TFs'])):\n",
    "    val = df['TFs'].iloc[i]\n",
    "    val = val.replace(',',' ')\n",
    "    val = val.replace(';',' ')\n",
    "    val = val.replace(':',' ')\n",
    "    TFs = val.split(' ')\n",
    "    if not validateTFs(TFs, adata):\n",
    "        print(TFs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0a4381c-c93a-4846-b044-e9be592086cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_directory = \"/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed\"\n",
    "scalars = [0.5, 0.75, 1.001]\n",
    "for i in range(len(df['TFs'])):\n",
    "    val = df['TFs'].iloc[i]\n",
    "    val = val.replace(',',' ')\n",
    "    val = val.replace(';',' ')\n",
    "    val = val.replace(':',' ')\n",
    "    TFs = val.split(' ')\n",
    "    if validateTFs(TFs, adata):\n",
    "        # Join the TFs list into a string for the filename\n",
    "        TFs_str = \"_\".join(TFs)\n",
    "\n",
    "        # Generate the file path for saving\n",
    "        file_name = f\"{TFs_str}.h5ad\"\n",
    "        output_path = os.path.join(output_directory, file_name)\n",
    "\n",
    "        # prevents writing duplicate files\n",
    "        if os.path.exists(output_filepath):\n",
    "            continue\n",
    "\n",
    "        # display some information\n",
    "        print(TFs)\n",
    "\n",
    "        adataDict = iterate_perturb_counts(adata, TFs, scalars)\n",
    "\n",
    "        # Concatenate all AnnData objects along the observations axis\n",
    "        concatenated_adata = ad.concat(list(adataDict.values()), axis=0)\n",
    "\n",
    "        # Save reprogramming metadata into the concatenated_adata.obs table\n",
    "        concatenated_adata.obs['Source_cells'] = df['Source cells'].iloc[i]\n",
    "        concatenated_adata.obs['Target_cells'] = df['Target cells'].iloc[i]\n",
    "        concatenated_adata.obs['Treatment'] = df['Treatment'].iloc[i]\n",
    "        concatenated_adata.obs['Species'] = df['Species'].iloc[i]\n",
    "        concatenated_adata.obs['Cell_Transplantation'] = df['Cell Transplantation'].iloc[i]\n",
    "        concatenated_adata.obs['Published_Year'] = df['Published Year'].iloc[i]\n",
    "        concatenated_adata.obs['PMID'] = df['PMID'].iloc[i]\n",
    "        \n",
    "        # Save the concatenated AnnData object to the file\n",
    "        concatenated_adata.write_h5ad(output_path)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bc8fda0-befd-4615-beb3-bb4deb9126b4",
   "metadata": {},
   "source": [
    "# Day 4\n",
    "\n",
    "**Focus:** check out Nats code (debug a bit) and create a few perturbations\n",
    "- changes made to NO's code:\n",
    "    1. iterate_perturb_counts: changes the order of the arguments to `adata, tf_list, scalar_list`\n",
    "    2. perturb_counts: changes the order of the arguments to `adata, tf_list, scalar_list`\n",
    "    3. perturb_counts: there was an issue with the use of `[: np.newaxis]` with respect to `max_exp`, which is a `coo_matrix` (special type of sparse matrix). Code was modified to address an issue being thrown here.\n",
    "- new function:\n",
    "    1. validateTFs(TFs, adata): this checks if all the transcription factors are present in the adata\n",
    "- pertrubation driver (`Perform Perturbations and Create new files`):\n",
    "    1. loads Fibroblast data from Tabula Sapiens\n",
    "    2. loads `.csv` file of known reprogrmaming protocols\n",
    "    3. for each set of TFs that are validated by `validateTFs`:\n",
    "        1. use `iterate_perturb_counts` to generate perturbations with scalars `[0.5, 0.75, 1]`\n",
    "        2. concatenate the dataframes to make a single dataframe\n",
    "        3. save metadata from reprogramming protocol (i.e. PMID, source/targets, etc.)\n",
    "        4. save the new anndata as a `.h5ad` file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9481aae9-17b0-4aa6-a4f1-6e2c078b6a1c",
   "metadata": {},
   "source": [
    "## Nat's Code with some modifications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8fe9a1b0-72b1-4ca1-990f-591b1b56b4f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import scanpy as sp\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "88b2cdca-2ed9-47c8-8884-37bb6eb8ebbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "def iterate_perturb_counts(adata, tf_list, scalar_list):\n",
    "    \"\"\"\n",
    "    Applies perturbations to the expression data of specified transcription factors across multiple scalars \n",
    "    and stores the resulting AnnData objects in a dictionary.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Iterates over a list of scalar values.\n",
    "    2. For each scalar, creates a copy of the AnnData object to preserve the original data.\n",
    "    3. Applies the `perturb_counts` function to scale the expression data of genes listed in `tf_list` by\n",
    "       the maximum gene expression of each cell and the current scalar.\n",
    "    4. Stores the perturbed AnnData object in a dictionary with the scalar as the key.\n",
    "\n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols (transcription factors) to be perturbed.\n",
    "    scalar_list (list): A list of scalar values for scaling the gene expression.\n",
    "    adata (AnnData): The AnnData object containing gene expression data (cells x genes).\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary where keys are scalar values and values are the corresponding perturbed AnnData objects.\n",
    "    \"\"\"\n",
    "    \n",
    "    adata_dict = {}\n",
    "    \n",
    "    for scalar in scalar_list:\n",
    "        # Create a copy of the AnnData object for each scalar value\n",
    "        adata_temp = adata.copy()\n",
    "        \n",
    "        # Apply perturb_counts to the copied AnnData object\n",
    "        perturbed_adata = perturb_counts(adata_temp, tf_list, scalar)\n",
    "        \n",
    "        # Store the perturbed AnnData object in the dictionary with scalar as the key\n",
    "        adata_dict[scalar] = perturbed_adata\n",
    "    \n",
    "    return adata_dict\n",
    "\n",
    "def perturb_counts(adata, tf_list, scalar): \n",
    "    \"\"\"\n",
    "    Applies a perturbation to the expression data of specific genes in an AnnData object.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Computes the maximum gene expression level for each cell.\n",
    "    2. Applies a scaling operation to the expression levels of genes listed in `tf_list`.\n",
    "       - Each entry of these genes in the matrix is multiplied by the maximum expression level \n",
    "         of its respective cell and a specified scalar value.\n",
    "    3. Updates the AnnData object with new columns:\n",
    "       - 'scaled': A boolean column indicating whether each gene is in the `tf_list`.\n",
    "       - 'scaled_by': Contains the scaling factor used for each gene (the product of the maximum \n",
    "         expression level of each cell and the scalar), or `1` if the gene was not in `tf_list`.\n",
    "    \n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols to be perturbed.\n",
    "    scalar (float): The scalar value used to scale the expression levels.\n",
    "    adata (AnnData): The AnnData object containing gene expression data.\n",
    "\n",
    "    Returns:\n",
    "    AnnData: The updated AnnData object with applied perturbations and new columns.\n",
    "    \"\"\"\n",
    "\n",
    "    # Create a boolean mask for genes in tf_list\n",
    "    gene_mask = adata.var['gene_symbol'].isin(tf_list)\n",
    "    \n",
    "    # Save the original state of the parameter objects, in case some tfs do not translate (failsafe)\n",
    "    original_X = adata.X.copy()\n",
    "    original_gene_mask = gene_mask.copy()\n",
    "    \n",
    "    # Compute maximum expression level of each cell\n",
    "    max_exp = np.max(adata.X, axis=1)\n",
    "\n",
    "    \"\"\"This is new today. v \"\"\"\n",
    "    # Raise an error if any of the gene names in tf_list do not match column names (we will manually update these in adata):\n",
    "    missing_genes = [gene for gene in tf_list if gene not in adata.var['gene_symbol'].values]\n",
    "    \n",
    "    if missing_genes:\n",
    "        # Restore original parameter objects\n",
    "        adata.X = original_X\n",
    "        gene_mask = original_gene_mask\n",
    "        raise ValueError(f\"Genes {missing_genes} not found in anndata object\")\n",
    "\n",
    "    else:    \n",
    "        \n",
    "        # Apply the scaling operation to the specified genes\n",
    "        adata.X[:, gene_mask] = max_exp * scalar\n",
    "        \n",
    "        # Add/Update 'scaled' column in var\n",
    "        adata.var['scaled'] = gene_mask\n",
    "        \n",
    "        # Add/Update 'scaled_by' column in var\n",
    "        adata.var['scaled_by'] = scalar  # Default value for genes not in tf_list\n",
    "    \n",
    "    return adata\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e8bfe84-6ed9-4daf-9c0b-bc570f63af30",
   "metadata": {},
   "source": [
    "## New code to validate lists of TFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "6d52acad-9d42-45e4-aa9f-21c439b91c73",
   "metadata": {},
   "outputs": [],
   "source": [
    "def validateTFs(TFs, adata):\n",
    "    adata_gene_list = adata.var['gene_symbol'].values.tolist()\n",
    "    for TF in TFs:\n",
    "        if TF not in adata_gene_list:\n",
    "            return False\n",
    "    return True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42c85448-2758-406e-964c-b741e4f8fdae",
   "metadata": {},
   "source": [
    "## Load data and perturbations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "add5caf2-aca2-44f2-9936-7d3b3879f539",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('data/first_5_recepies_8_29_2024.csv')\n",
    "\n",
    "DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/jpic/\"\n",
    "FILE = \"fibroblast.h5ad\"\n",
    "adata = sp.read_h5ad(os.path.join(DATAPATH, FILE))\n",
    "adata_gene_list = adata.var['gene_symbol'].values.tolist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "92300a6c-c06d-4b47-9972-0e30ab525621",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>ensemblid</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>False</td>\n",
       "      <td>6.398244e-05</td>\n",
       "      <td>0.835044</td>\n",
       "      <td>-0.573947</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.005574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>False</td>\n",
       "      <td>2.274395e-03</td>\n",
       "      <td>2.442280</td>\n",
       "      <td>0.533203</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.031731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR6859-1</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>False</td>\n",
       "      <td>6.175251e-05</td>\n",
       "      <td>1.295335</td>\n",
       "      <td>-0.256874</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.005634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>False</td>\n",
       "      <td>1.372886e-04</td>\n",
       "      <td>2.656352</td>\n",
       "      <td>0.680668</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.008041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND6</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>False</td>\n",
       "      <td>9.634841e-01</td>\n",
       "      <td>2.466404</td>\n",
       "      <td>0.154140</td>\n",
       "      <td>0.590065</td>\n",
       "      <td>0.741395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TE</th>\n",
       "      <td>MT-TE</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.600667e-01</td>\n",
       "      <td>1.603787</td>\n",
       "      <td>-0.044396</td>\n",
       "      <td>0.083929</td>\n",
       "      <td>0.301820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CYB</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>False</td>\n",
       "      <td>4.367693e+00</td>\n",
       "      <td>4.765751</td>\n",
       "      <td>-0.499747</td>\n",
       "      <td>3.874830</td>\n",
       "      <td>1.104192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TT</th>\n",
       "      <td>MT-TT</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>False</td>\n",
       "      <td>6.573967e-02</td>\n",
       "      <td>0.624316</td>\n",
       "      <td>-0.719108</td>\n",
       "      <td>0.040580</td>\n",
       "      <td>0.186848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TP</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>True</td>\n",
       "      <td>4.385102e-01</td>\n",
       "      <td>2.705558</td>\n",
       "      <td>1.115482</td>\n",
       "      <td>0.204806</td>\n",
       "      <td>0.501265</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58870 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_symbol     feature_type          ensemblid  highly_variable  \\\n",
       "DDX11L1          DDX11L1  Gene Expression  ENSG00000223972.5            False   \n",
       "WASH7P            WASH7P  Gene Expression  ENSG00000227232.5            False   \n",
       "MIR6859-1      MIR6859-1  Gene Expression  ENSG00000278267.1            False   \n",
       "MIR1302-2HG  MIR1302-2HG  Gene Expression  ENSG00000243485.5            False   \n",
       "MIR1302-2      MIR1302-2  Gene Expression  ENSG00000284332.1            False   \n",
       "...                  ...              ...                ...              ...   \n",
       "MT-ND6            MT-ND6  Gene Expression  ENSG00000198695.2            False   \n",
       "MT-TE              MT-TE  Gene Expression  ENSG00000210194.1            False   \n",
       "MT-CYB            MT-CYB  Gene Expression  ENSG00000198727.2            False   \n",
       "MT-TT              MT-TT  Gene Expression  ENSG00000210195.2            False   \n",
       "MT-TP              MT-TP  Gene Expression  ENSG00000210196.2             True   \n",
       "\n",
       "                    means  dispersions  dispersions_norm      mean       std  \n",
       "DDX11L1      6.398244e-05     0.835044         -0.573947  0.000039  0.005574  \n",
       "WASH7P       2.274395e-03     2.442280          0.533203  0.001080  0.031731  \n",
       "MIR6859-1    6.175251e-05     1.295335         -0.256874  0.000033  0.005634  \n",
       "MIR1302-2HG  1.372886e-04     2.656352          0.680668  0.000048  0.008041  \n",
       "MIR1302-2    1.000000e-12          NaN          0.000000  0.000000  1.000000  \n",
       "...                   ...          ...               ...       ...       ...  \n",
       "MT-ND6       9.634841e-01     2.466404          0.154140  0.590065  0.741395  \n",
       "MT-TE        1.600667e-01     1.603787         -0.044396  0.083929  0.301820  \n",
       "MT-CYB       4.367693e+00     4.765751         -0.499747  3.874830  1.104192  \n",
       "MT-TT        6.573967e-02     0.624316         -0.719108  0.040580  0.186848  \n",
       "MT-TP        4.385102e-01     2.705558          1.115482  0.204806  0.501265  \n",
       "\n",
       "[58870 rows x 9 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c23d9def-0bff-4d13-b0ca-9f999a2d785f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4fddcbf2-e587-47ce-aef0-58caab9b60f3",
   "metadata": {},
   "source": [
    "## Perform Perturbations and Create new files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "16b9ab50-73ff-4fb0-be7b-e5cb79d1cce0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "50"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df['TFs'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36840d95-6396-4502-a81b-aba8075406ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['SOX2']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['SOX2', 'HMGA2']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ASCL1', 'PAX6']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['SOX2', 'GATA3', 'NEUROD1']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['SOX2', 'PAX6']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['SOX18']\n"
     ]
    }
   ],
   "source": [
    "output_directory = \"/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed\"\n",
    "scalars = [0.5, 0.75, 1.001]\n",
    "for i in range(len(df['TFs'])):\n",
    "    val = df['TFs'].iloc[i]\n",
    "    val = val.replace(',',' ')\n",
    "    val = val.replace(';',' ')\n",
    "    val = val.replace(':',' ')\n",
    "    TFs = val.split(' ')\n",
    "    if validateTFs(TFs, adata):\n",
    "        print(TFs)\n",
    "        adataDict = iterate_perturb_counts(adata, TFs, scalars)\n",
    "\n",
    "        # Concatenate all AnnData objects along the observations axis\n",
    "        concatenated_adata = ad.concat(list(adataDict.values()), axis=0)\n",
    "\n",
    "        # Save reprogramming metadata into the concatenated_adata.obs table\n",
    "        concatenated_adata.obs['Source_cells'] = df['Source cells'].iloc[i]\n",
    "        concatenated_adata.obs['Target_cells'] = df['Target cells'].iloc[i]\n",
    "        concatenated_adata.obs['Treatment'] = df['Treatment'].iloc[i]\n",
    "        concatenated_adata.obs['Species'] = df['Species'].iloc[i]\n",
    "        concatenated_adata.obs['Cell_Transplantation'] = df['Cell Transplantation'].iloc[i]\n",
    "        concatenated_adata.obs['Published_Year'] = df['Published Year'].iloc[i]\n",
    "        concatenated_adata.obs['PMID'] = df['PMID'].iloc[i]\n",
    "        \n",
    "        # Join the TFs list into a string for the filename\n",
    "        TFs_str = \"_\".join(TFs)\n",
    "        \n",
    "        # Generate the file path for saving\n",
    "        file_name = f\"{TFs_str}.h5ad\"\n",
    "        output_path = os.path.join(output_directory, file_name)\n",
    "        \n",
    "        # Save the concatenated AnnData object to the file\n",
    "        concatenated_adata.write_h5ad(output_path)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da6da5ad-dd98-4677-82ce-7a84cee28040",
   "metadata": {},
   "source": [
    "## Scratch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "47876647-2a68-4659-8c58-ccdd763f1747",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'adata' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[4], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43madata\u001b[49m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'adata' is not defined"
     ]
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "08cc57e2-26ef-42d3-a4d6-f398a5d1ed28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'fibroblast'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = \"fibroblast.h5ad\"\n",
    "file[:-5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "4e5e980d-0a87-4173-ae6f-e2710de5bd54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organ_tissue</th>\n",
       "      <th>method</th>\n",
       "      <th>donor</th>\n",
       "      <th>anatomical_information</th>\n",
       "      <th>n_counts_UMIs</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>cell_ontology_class</th>\n",
       "      <th>free_annotation</th>\n",
       "      <th>manually_annotated</th>\n",
       "      <th>compartment</th>\n",
       "      <th>gender</th>\n",
       "      <th>Source_cells</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1</th>\n",
       "      <td>Liver</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP6</td>\n",
       "      <td>nan</td>\n",
       "      <td>10423.0</td>\n",
       "      <td>2986</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>Stellate/Fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>male</td>\n",
       "      <td>Embryonic Fibroblasts and Adult Skin Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1</th>\n",
       "      <td>Liver</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP6</td>\n",
       "      <td>nan</td>\n",
       "      <td>11223.0</td>\n",
       "      <td>3248</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>Stellate/Fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>male</td>\n",
       "      <td>Embryonic Fibroblasts and Adult Skin Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1</th>\n",
       "      <td>Liver</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP6</td>\n",
       "      <td>nan</td>\n",
       "      <td>14695.0</td>\n",
       "      <td>4469</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>Stellate/Fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>male</td>\n",
       "      <td>Embryonic Fibroblasts and Adult Skin Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1</th>\n",
       "      <td>Liver</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP6</td>\n",
       "      <td>nan</td>\n",
       "      <td>8977.0</td>\n",
       "      <td>2875</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>Stellate/Fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>male</td>\n",
       "      <td>Embryonic Fibroblasts and Adult Skin Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1</th>\n",
       "      <td>Liver</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP6</td>\n",
       "      <td>nan</td>\n",
       "      <td>12109.0</td>\n",
       "      <td>3239</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>Stellate/Fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>male</td>\n",
       "      <td>Embryonic Fibroblasts and Adult Skin Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_Vasculature_aorta_SS2_B114585_B133324_Stromal_P3_S99</th>\n",
       "      <td>Vasculature</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>nan</td>\n",
       "      <td>308152.0</td>\n",
       "      <td>2090</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>female</td>\n",
       "      <td>Embryonic Fibroblasts and Adult Skin Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_Vasculature_aorta_SS2_B114585_B133324_Stromal_P7_S103</th>\n",
       "      <td>Vasculature</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>nan</td>\n",
       "      <td>387043.0</td>\n",
       "      <td>2132</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>female</td>\n",
       "      <td>Embryonic Fibroblasts and Adult Skin Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_Vasculature_aorta_SS2_B114585_B133324_Stromal_P9_S105</th>\n",
       "      <td>Vasculature</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>nan</td>\n",
       "      <td>412608.0</td>\n",
       "      <td>3140</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>female</td>\n",
       "      <td>Embryonic Fibroblasts and Adult Skin Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_Vasculature_aorta_SS2_B113343_B133091_Immune_L6_S270</th>\n",
       "      <td>Vasculature</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>aorta</td>\n",
       "      <td>341884.0</td>\n",
       "      <td>4379</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>female</td>\n",
       "      <td>Embryonic Fibroblasts and Adult Skin Fibroblasts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_Vasculature_aorta_SS2_B113343_B133091_Immune_N6_S318</th>\n",
       "      <td>Vasculature</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>aorta</td>\n",
       "      <td>141115.0</td>\n",
       "      <td>5842</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>fibroblast</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>female</td>\n",
       "      <td>Embryonic Fibroblasts and Adult Skin Fibroblasts</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>114453 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   organ_tissue     method  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1                    Liver        10X   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1                    Liver        10X   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1                    Liver        10X   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1                    Liver        10X   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1                    Liver        10X   \n",
       "...                                                         ...        ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  Vasculature  smartseq2   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  Vasculature  smartseq2   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  Vasculature  smartseq2   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  Vasculature  smartseq2   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  Vasculature  smartseq2   \n",
       "\n",
       "                                                   donor  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1              TSP6   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1              TSP6   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1              TSP6   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1              TSP6   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1              TSP6   \n",
       "...                                                  ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  TSP2   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  TSP2   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  TSP2   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  TSP2   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  TSP2   \n",
       "\n",
       "                                                   anatomical_information  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1                                nan   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1                                nan   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1                                nan   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1                                nan   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1                                nan   \n",
       "...                                                                   ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...                    nan   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...                    nan   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...                    nan   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...                  aorta   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...                  aorta   \n",
       "\n",
       "                                                    n_counts_UMIs  n_genes  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1                    10423.0     2986   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1                    11223.0     3248   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1                    14695.0     4469   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1                     8977.0     2875   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1                    12109.0     3239   \n",
       "...                                                           ...      ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...       308152.0     2090   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...       387043.0     2132   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...       412608.0     3140   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...       341884.0     4379   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...       141115.0     5842   \n",
       "\n",
       "                                                   cell_ontology_class  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1                      fibroblast   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1                      fibroblast   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1                      fibroblast   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1                      fibroblast   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1                      fibroblast   \n",
       "...                                                                ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...          fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...          fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...          fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...          fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...          fibroblast   \n",
       "\n",
       "                                                        free_annotation  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1              Stellate/Fibroblast   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1              Stellate/Fibroblast   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1              Stellate/Fibroblast   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1              Stellate/Fibroblast   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1              Stellate/Fibroblast   \n",
       "...                                                                 ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...           fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...           fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...           fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...           fibroblast   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...           fibroblast   \n",
       "\n",
       "                                                    manually_annotated  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1                            True   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1                            True   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1                            True   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1                            True   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1                            True   \n",
       "...                                                                ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...                True   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...                True   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...                True   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...                True   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...                True   \n",
       "\n",
       "                                                   compartment  gender  \\\n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1                 stromal    male   \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1                 stromal    male   \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1                 stromal    male   \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1                 stromal    male   \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1                 stromal    male   \n",
       "...                                                        ...     ...   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...     stromal  female   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...     stromal  female   \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...     stromal  female   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...     stromal  female   \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...     stromal  female   \n",
       "\n",
       "                                                                                        Source_cells  \n",
       "AACAAAGTCCGGCTTT_TSP6_Liver_NA_10X_1_1              Embryonic Fibroblasts and Adult Skin Fibroblasts  \n",
       "AATGACCAGTCTTCGA_TSP6_Liver_NA_10X_1_1              Embryonic Fibroblasts and Adult Skin Fibroblasts  \n",
       "ACCACAATCCGAGAAG_TSP6_Liver_NA_10X_1_1              Embryonic Fibroblasts and Adult Skin Fibroblasts  \n",
       "ACTTCGCCAACTTCTT_TSP6_Liver_NA_10X_1_1              Embryonic Fibroblasts and Adult Skin Fibroblasts  \n",
       "AGAAGTATCAGGGATG_TSP6_Liver_NA_10X_1_1              Embryonic Fibroblasts and Adult Skin Fibroblasts  \n",
       "...                                                                                              ...  \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  Embryonic Fibroblasts and Adult Skin Fibroblasts  \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  Embryonic Fibroblasts and Adult Skin Fibroblasts  \n",
       "TSP2_Vasculature_aorta_SS2_B114585_B133324_Stro...  Embryonic Fibroblasts and Adult Skin Fibroblasts  \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  Embryonic Fibroblasts and Adult Skin Fibroblasts  \n",
       "TSP2_Vasculature_aorta_SS2_B113343_B133091_Immu...  Embryonic Fibroblasts and Adult Skin Fibroblasts  \n",
       "\n",
       "[114453 rows x 12 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concatenated_adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "0283673e-43cb-44bb-a190-b1392a6d9c68",
   "metadata": {},
   "outputs": [],
   "source": [
    "concatenated_adata.obs['Source_cells'] = df['Source cells'].iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4c2cc24-c4b1-424c-812a-780deb83b707",
   "metadata": {},
   "outputs": [],
   "source": [
    "concatenated_adata.obs['Source_cells'] = df['Source cells'].iloc[i]\n",
    "concatenated_adata.obs['Target_cells'] = df['Target cells'].iloc[i]\n",
    "concatenated_adata.obs['Treatment'] = df['Treatment'].iloc[i]\n",
    "concatenated_adata.obs['Species'] = df['Species'].iloc[i]\n",
    "concatenated_adata.obs['Cell_Transplantation'] = df['Cell Transplantation'].iloc[i]\n",
    "concatenated_adata.obs['Published_Year'] = df['Published Year'].iloc[i]\n",
    "concatenated_adata.obs['PMID'] = df['PMID'].iloc[i]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "31a27add-339d-48f3-9d10-bda2b3eef4b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'SOX2_HMGA2'"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TFs_str = \"_\".join(TFs)\n",
    "TFs_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "f1f186fe-1f31-4eca-86d6-74060703f9c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "import anndata as ad\n",
    "# Concatenate all AnnData objects along the observations axis\n",
    "concatenated_adata = ad.concat(list(adataDict.values()), axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "879d6fd8-a23e-41b5-b6c5-936fd950f6dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "114453"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(list(concatenated_adata.obs.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "0a36498c-789f-4e5a-8568-cadf38f90ec8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "38151"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(list(set(list(concatenated_adata.obs.index))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "6ac32003-3df5-4028-863d-111a09deac65",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n"
     ]
    }
   ],
   "source": [
    "adata.X[:, 0] = max_exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "79486346-e33d-4f85-b658-229ee92408b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "+\n",
      "Ptf1a\n",
      "NEUROD\n",
      "2)\n",
      "c-MYC\n",
      "(ETS\n",
      "or\n",
      "OCT4\n",
      "OCT9\n",
      "BRN2\n",
      "L-MYC\n",
      "NURR1\n",
      "SV40\n",
      "OSTERIX\n",
      "P53\n",
      "PU.1\n",
      "Knockdown\n",
      "Large\n",
      "Variant\n",
      "LXH3\n",
      "Antigen\n",
      "HNF6\n",
      "Any\n",
      "and\n",
      "OCT3/4\n",
      "ER71/ETV2\n",
      "of\n",
      "AP-2A\n",
      "LEF-1\n",
      "Three\n",
      "NGN2\n",
      "NF-κB\n",
      "N-MYC\n",
      "HB9\n",
      "OCT6\n",
      "LXMLA\n",
      "PPARG2\n",
      "MASH1\n",
      "the\n",
      "Two\n",
      "Factors\n",
      "T\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('data/first_5_recepies_8_29_2024.csv')\n",
    "words = []\n",
    "for val in df['TFs']:\n",
    "    val = val.replace(',',' ')\n",
    "    val = val.replace(';',' ')\n",
    "    val = val.replace(':',' ')\n",
    "    words += val.split(' ')\n",
    "words = list(set(words))\n",
    "\n",
    "for word in words:\n",
    "#     print(word in adata_gene_list)\n",
    "    if word not in adata_gene_list:\n",
    "        print(word)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2f7cd0f-a2f8-4df6-8f50-ed706627a385",
   "metadata": {},
   "source": [
    "# Day 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d7f2e538-8981-463c-9d75-0d73c910a0f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sp\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6290a8c0-f4ff-4f67-bd0b-cfef149312a8",
   "metadata": {},
   "source": [
    "# Day 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "16016131-a65e-4860-809c-3c0d057190ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sp\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b494dee-614e-4446-8f26-a6fcf78cc0e0",
   "metadata": {},
   "source": [
    "## Build Embeddings of Target Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07929ab3-2711-4b76-9438-cf8579e35f71",
   "metadata": {},
   "source": [
    "### Neurons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d17dd340-d902-4d72-84c9-ca247cfe4a09",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sp\n",
    "import os\n",
    "\n",
    "# location to data\n",
    "DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/public_data/cellXgene\"\n",
    "FILE = \"neurons.h5ad\"\n",
    "\n",
    "# load data\n",
    "adata = sp.read_h5ad(os.path.join(DATAPATH, FILE))\n",
    "\n",
    "# filter to only neuron data\n",
    "neurons_adata = adata[adata.obs['cell_type'] == 'neuron']\n",
    "\n",
    "# this data needs to be embedded with all the models\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebb1d3e3-29af-466a-9f8f-027486c8c52d",
   "metadata": {},
   "source": [
    "## Build Embedding Driver"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e12bf22c-83f3-4e15-916b-773f1ce6ec2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main(job_number, parameter_file):\n",
    "    \"\"\"\n",
    "    Main function for performing in silico reprogramming experiments. It selects the correct model, source, target, \n",
    "    transcription factors (TFs), and other parameters based on the job number in a SLURM array job. Then, it processes \n",
    "    the data and saves the results.\n",
    "    \n",
    "    Args:\n",
    "        job_number (int): The SLURM job array index to determine which parameters to use.\n",
    "        parameter_file (str): Path to the CSV file containing model parameters.\n",
    "    \n",
    "    Returns:\n",
    "        int: Return code (0 for success).\n",
    "    \"\"\"\n",
    "\n",
    "    OUTPUTPATH = \"/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot\"\n",
    "    \n",
    "    # Load parameters from CSV\n",
    "    df_embedding_parameters = pd.read_csv(parameter_file)\n",
    "    TFs    = df_embedding_parameters['TFs'].values[job_number]\n",
    "    model  = df_embedding_parameters['model'].values[job_number]\n",
    "    source = df_embedding_parameters['source'].values[job_number]\n",
    "    target = df_embedding_parameters['target'].values[job_number]\n",
    "    PMID   = df_embedding_parameters['PMID'].values[job_number]\n",
    "    year   = df_embedding_parameters['year'].values[job_number]\n",
    "    organ  = df_embedding_parameters['organ'].values[job_number]\n",
    "\n",
    "    print(f\"Starting job {job_number} with model {model}, TFs {TFs}, source {source}, target {target}, PMID {PMID}, year {year}\")\n",
    "\n",
    "    # Load the source data (Fibroblasts)\n",
    "    source_adata = load_fibroblasts()\n",
    "    print(\"Source data loaded\")\n",
    "\n",
    "    # TODO: Load the target data\n",
    "    targets_adata = load_targets()\n",
    "    print(\"Target data loaded\")\n",
    "\n",
    "    # Perturb the source data\n",
    "    perturbed_adata = perturbation_model(source_adata, TFs)\n",
    "    print(\"Data perturbed\")\n",
    "\n",
    "    # Combine source, perturbed, and target data\n",
    "    combined_adata = combine_adata([source_adata, perturbed_adata, targets_adata])\n",
    "\n",
    "    # Add a new 'reprogrammed type' column to indicate the source, target, and reprogrammed cells\n",
    "    combined_adata.obs['reprogrammed type'] = ['source'] * source_adata.shape[0] + \\\n",
    "                                              ['reprogrammed'] * perturbed_adata.shape[0] + \\\n",
    "                                              ['target'] * targets_adata.shape[0]\n",
    "    print(\"Combined data and added 'reprogrammed type' column\")\n",
    "\n",
    "    # Add columns to the 'obs' field of combined_adata for metadata tracking\n",
    "    combined_adata.obs['model'] = model\n",
    "    combined_adata.obs['TFs'] = ','.join(TFs)\n",
    "    combined_adata.obs['PMID'] = PMID\n",
    "    combined_adata.obs['year'] = year\n",
    "    combined_adata.obs['organ'] = organ\n",
    "    print(\"Added metadata columns to combined data\")\n",
    "\n",
    "    # Generate embeddings using the specified model\n",
    "    if model == 'geneformer':\n",
    "        adata_embedded = embed_geneformer(combined_adata)\n",
    "    elif model == 'tGPT':\n",
    "        adata_embedded = embed_tGPT(combined_adata)\n",
    "    elif model == 'scGTP':\n",
    "        adata_embedded = embed_scGTP([combined_adata])\n",
    "    else:\n",
    "        raise ValueError(f\"Unknown model: {model}\")\n",
    "\n",
    "    print(f\"Embeddings generated using model: {model}\")\n",
    "\n",
    "    # Save the results to a file named according to the job number\n",
    "    result_filepath = os.path.join(OUTPUTPATH, f\"results_job_{job_number}.h5ad\")\n",
    "    adata_embedded.write(result_filepath)\n",
    "    print(f\"Results saved to {result_filepath}\")\n",
    "\n",
    "    return 0\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cad2d9cc-78c2-4d3c-a3c2-b0710a27000c",
   "metadata": {},
   "source": [
    "# Day 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "af06fdf6-e799-4be5-89d7-ec85d7278c1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sp\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bb10707-df05-4e7d-a9a4-bd301ab5cd69",
   "metadata": {},
   "source": [
    "## Perturbation Model Discussion\n",
    "\n",
    "E.V. = expression values\n",
    "\n",
    "Possible algorithm:\n",
    "```\n",
    "1. find highest E.V.  for a single cell\n",
    "2. find expression value of TFs being modified\n",
    "3. have a value k for the number of different concentrations we want to test\n",
    "4. choose k different amounts to increase the TFs from there measured E.V. to the 150% maximum E.V.\n",
    "   - make an arbitray choice and code it up\n",
    "```\n",
    "\n",
    "**A reasonable person could write this 10s of different ways**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a06f0a9c-18e0-4d3a-8efd-7993accb1048",
   "metadata": {},
   "source": [
    "## Visualize Input Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a6474b05-224f-4677-902e-5b4962a4bf23",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/\"\n",
    "FILE = \"TS_epithelial.h5ad\"\n",
    "\n",
    "adata = sp.read_h5ad(os.path.join(DATAPATH, FILE))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "047a5de2-be21-489c-8184-b90f010b06a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 104148 × 58870\n",
       "    obs: 'organ_tissue', 'method', 'donor', 'anatomical_information', 'n_counts_UMIs', 'n_genes', 'cell_ontology_class', 'free_annotation', 'manually_annotated', 'compartment', 'gender'\n",
       "    var: 'gene_symbol', 'feature_type', 'ensemblid', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
       "    uns: '_scvi', '_training_mode', 'dendrogram_cell_type_tissue', 'dendrogram_computational_compartment_assignment', 'dendrogram_consensus_prediction', 'dendrogram_tissue_cell_type', 'donor_colors', 'donor_method_colors', 'hvg', 'method_colors', 'neighbors', 'organ_tissue_colors', 'sex_colors', 'tissue_colors', 'umap'\n",
       "    obsm: 'X_pca', 'X_scvi', 'X_scvi_umap', 'X_umap'\n",
       "    layers: 'decontXcounts', 'raw_counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f010c3f2-3294-4f18-bcad-8a9b3abf9b2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>ensemblid</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>False</td>\n",
       "      <td>6.398244e-05</td>\n",
       "      <td>0.835044</td>\n",
       "      <td>-0.573947</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.005574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>False</td>\n",
       "      <td>2.274395e-03</td>\n",
       "      <td>2.442280</td>\n",
       "      <td>0.533203</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.031731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR6859-1</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>False</td>\n",
       "      <td>6.175251e-05</td>\n",
       "      <td>1.295335</td>\n",
       "      <td>-0.256874</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.005634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>False</td>\n",
       "      <td>1.372886e-04</td>\n",
       "      <td>2.656352</td>\n",
       "      <td>0.680668</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.008041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND6</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>False</td>\n",
       "      <td>9.634841e-01</td>\n",
       "      <td>2.466404</td>\n",
       "      <td>0.154140</td>\n",
       "      <td>0.590065</td>\n",
       "      <td>0.741395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TE</th>\n",
       "      <td>MT-TE</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.600667e-01</td>\n",
       "      <td>1.603787</td>\n",
       "      <td>-0.044396</td>\n",
       "      <td>0.083929</td>\n",
       "      <td>0.301820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CYB</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>False</td>\n",
       "      <td>4.367693e+00</td>\n",
       "      <td>4.765751</td>\n",
       "      <td>-0.499747</td>\n",
       "      <td>3.874830</td>\n",
       "      <td>1.104192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TT</th>\n",
       "      <td>MT-TT</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>False</td>\n",
       "      <td>6.573967e-02</td>\n",
       "      <td>0.624316</td>\n",
       "      <td>-0.719108</td>\n",
       "      <td>0.040580</td>\n",
       "      <td>0.186848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TP</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>True</td>\n",
       "      <td>4.385102e-01</td>\n",
       "      <td>2.705558</td>\n",
       "      <td>1.115482</td>\n",
       "      <td>0.204806</td>\n",
       "      <td>0.501265</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58870 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_symbol     feature_type          ensemblid  highly_variable  \\\n",
       "DDX11L1          DDX11L1  Gene Expression  ENSG00000223972.5            False   \n",
       "WASH7P            WASH7P  Gene Expression  ENSG00000227232.5            False   \n",
       "MIR6859-1      MIR6859-1  Gene Expression  ENSG00000278267.1            False   \n",
       "MIR1302-2HG  MIR1302-2HG  Gene Expression  ENSG00000243485.5            False   \n",
       "MIR1302-2      MIR1302-2  Gene Expression  ENSG00000284332.1            False   \n",
       "...                  ...              ...                ...              ...   \n",
       "MT-ND6            MT-ND6  Gene Expression  ENSG00000198695.2            False   \n",
       "MT-TE              MT-TE  Gene Expression  ENSG00000210194.1            False   \n",
       "MT-CYB            MT-CYB  Gene Expression  ENSG00000198727.2            False   \n",
       "MT-TT              MT-TT  Gene Expression  ENSG00000210195.2            False   \n",
       "MT-TP              MT-TP  Gene Expression  ENSG00000210196.2             True   \n",
       "\n",
       "                    means  dispersions  dispersions_norm      mean       std  \n",
       "DDX11L1      6.398244e-05     0.835044         -0.573947  0.000039  0.005574  \n",
       "WASH7P       2.274395e-03     2.442280          0.533203  0.001080  0.031731  \n",
       "MIR6859-1    6.175251e-05     1.295335         -0.256874  0.000033  0.005634  \n",
       "MIR1302-2HG  1.372886e-04     2.656352          0.680668  0.000048  0.008041  \n",
       "MIR1302-2    1.000000e-12          NaN          0.000000  0.000000  1.000000  \n",
       "...                   ...          ...               ...       ...       ...  \n",
       "MT-ND6       9.634841e-01     2.466404          0.154140  0.590065  0.741395  \n",
       "MT-TE        1.600667e-01     1.603787         -0.044396  0.083929  0.301820  \n",
       "MT-CYB       4.367693e+00     4.765751         -0.499747  3.874830  1.104192  \n",
       "MT-TT        6.573967e-02     0.624316         -0.719108  0.040580  0.186848  \n",
       "MT-TP        4.385102e-01     2.705558          1.115482  0.204806  0.501265  \n",
       "\n",
       "[58870 rows x 9 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "030f83ed-4955-4962-a629-f003027303e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<104148x1 sparse matrix of type '<class 'numpy.float32'>'\n",
       "\twith 104148 stored elements in COOrdinate format>"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X.max(axis=1) # what is the value of the highest expressed gene for each cell?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "c210614f-0e87-4a8a-b1c5-1cc95c7380ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0])"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TF = 'DDX11L1'\n",
    "index = np.where(adata.var['gene_symbol'] == TF)[0]\n",
    "index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "9dc3d5ad-f2f0-446d-83f3-c63179d369c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DDX11L1            DDX11L1\n",
       "WASH7P              WASH7P\n",
       "MIR6859-1        MIR6859-1\n",
       "MIR1302-2HG    MIR1302-2HG\n",
       "MIR1302-2        MIR1302-2\n",
       "                  ...     \n",
       "MT-ND6              MT-ND6\n",
       "MT-TE                MT-TE\n",
       "MT-CYB              MT-CYB\n",
       "MT-TT                MT-TT\n",
       "MT-TP                MT-TP\n",
       "Name: gene_symbol, Length: 58870, dtype: object"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var['gene_symbol']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "45904679-08f7-4ec7-b746-4017140120c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organ_tissue</th>\n",
       "      <th>method</th>\n",
       "      <th>donor</th>\n",
       "      <th>anatomical_information</th>\n",
       "      <th>n_counts_UMIs</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>cell_ontology_class</th>\n",
       "      <th>free_annotation</th>\n",
       "      <th>manually_annotated</th>\n",
       "      <th>compartment</th>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AACAGGGCATGTGCTA_TSP6_Liver_NA_10X_1_1</th>\n",
       "      <td>Liver</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP6</td>\n",
       "      <td>nan</td>\n",
       "      <td>14151.0</td>\n",
       "      <td>3404</td>\n",
       "      <td>hepatocyte</td>\n",
       "      <td>Hepatocyte</td>\n",
       "      <td>True</td>\n",
       "      <td>epithelial</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AATGAAGTCTAAGGAA_TSP6_Liver_NA_10X_1_1</th>\n",
       "      <td>Liver</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP6</td>\n",
       "      <td>nan</td>\n",
       "      <td>10303.0</td>\n",
       "      <td>2863</td>\n",
       "      <td>hepatocyte</td>\n",
       "      <td>Hepatocyte</td>\n",
       "      <td>True</td>\n",
       "      <td>epithelial</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACAGAAAAGCTGTCCG_TSP6_Liver_NA_10X_1_1</th>\n",
       "      <td>Liver</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP6</td>\n",
       "      <td>nan</td>\n",
       "      <td>9391.0</td>\n",
       "      <td>2738</td>\n",
       "      <td>hepatocyte</td>\n",
       "      <td>Hepatocyte</td>\n",
       "      <td>True</td>\n",
       "      <td>epithelial</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACGATGTTCGACTCCT_TSP6_Liver_NA_10X_1_1</th>\n",
       "      <td>Liver</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP6</td>\n",
       "      <td>nan</td>\n",
       "      <td>4634.0</td>\n",
       "      <td>1634</td>\n",
       "      <td>hepatocyte</td>\n",
       "      <td>Hepatocyte</td>\n",
       "      <td>True</td>\n",
       "      <td>epithelial</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACGTTCCAGAACCCGA_TSP6_Liver_NA_10X_1_1</th>\n",
       "      <td>Liver</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP6</td>\n",
       "      <td>nan</td>\n",
       "      <td>33370.0</td>\n",
       "      <td>6590</td>\n",
       "      <td>intrahepatic cholangiocyte</td>\n",
       "      <td>BECS</td>\n",
       "      <td>True</td>\n",
       "      <td>epithelial</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGACAGTAGATGGCGT_TSP2_Vasculature_Aorta_10X_1_2</th>\n",
       "      <td>Vasculature</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>Aorta</td>\n",
       "      <td>10354.0</td>\n",
       "      <td>2456</td>\n",
       "      <td>epithelial cell</td>\n",
       "      <td>epithelial cell</td>\n",
       "      <td>True</td>\n",
       "      <td>epithelial</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGTTGGATCGACATCA_TSP2_Vasculature_Aorta_10X_1_2</th>\n",
       "      <td>Vasculature</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>Aorta</td>\n",
       "      <td>15677.0</td>\n",
       "      <td>3234</td>\n",
       "      <td>epithelial cell</td>\n",
       "      <td>epithelial cell</td>\n",
       "      <td>True</td>\n",
       "      <td>epithelial</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGGATGGTGGCTACC_TSP2_Vasculature_Aorta_10X_1_2</th>\n",
       "      <td>Vasculature</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>Aorta</td>\n",
       "      <td>18638.0</td>\n",
       "      <td>3368</td>\n",
       "      <td>epithelial cell</td>\n",
       "      <td>epithelial cell</td>\n",
       "      <td>True</td>\n",
       "      <td>epithelial</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTTGATTTCTTGCAAG_TSP2_Vasculature_Aorta_10X_2_2</th>\n",
       "      <td>Vasculature</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>Aorta</td>\n",
       "      <td>36555.0</td>\n",
       "      <td>5102</td>\n",
       "      <td>epithelial cell</td>\n",
       "      <td>epithelial cell</td>\n",
       "      <td>True</td>\n",
       "      <td>epithelial</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGACGGCATGGAATA_TSP2_Vasculature_Aorta_10X_2_2</th>\n",
       "      <td>Vasculature</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>Aorta</td>\n",
       "      <td>6943.0</td>\n",
       "      <td>1799</td>\n",
       "      <td>epithelial cell</td>\n",
       "      <td>epithelial cell</td>\n",
       "      <td>True</td>\n",
       "      <td>epithelial</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>104148 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                organ_tissue method donor  \\\n",
       "AACAGGGCATGTGCTA_TSP6_Liver_NA_10X_1_1                 Liver    10X  TSP6   \n",
       "AATGAAGTCTAAGGAA_TSP6_Liver_NA_10X_1_1                 Liver    10X  TSP6   \n",
       "ACAGAAAAGCTGTCCG_TSP6_Liver_NA_10X_1_1                 Liver    10X  TSP6   \n",
       "ACGATGTTCGACTCCT_TSP6_Liver_NA_10X_1_1                 Liver    10X  TSP6   \n",
       "ACGTTCCAGAACCCGA_TSP6_Liver_NA_10X_1_1                 Liver    10X  TSP6   \n",
       "...                                                      ...    ...   ...   \n",
       "TGACAGTAGATGGCGT_TSP2_Vasculature_Aorta_10X_1_2  Vasculature    10X  TSP2   \n",
       "TGTTGGATCGACATCA_TSP2_Vasculature_Aorta_10X_1_2  Vasculature    10X  TSP2   \n",
       "TTGGATGGTGGCTACC_TSP2_Vasculature_Aorta_10X_1_2  Vasculature    10X  TSP2   \n",
       "CTTGATTTCTTGCAAG_TSP2_Vasculature_Aorta_10X_2_2  Vasculature    10X  TSP2   \n",
       "TCGACGGCATGGAATA_TSP2_Vasculature_Aorta_10X_2_2  Vasculature    10X  TSP2   \n",
       "\n",
       "                                                anatomical_information  \\\n",
       "AACAGGGCATGTGCTA_TSP6_Liver_NA_10X_1_1                             nan   \n",
       "AATGAAGTCTAAGGAA_TSP6_Liver_NA_10X_1_1                             nan   \n",
       "ACAGAAAAGCTGTCCG_TSP6_Liver_NA_10X_1_1                             nan   \n",
       "ACGATGTTCGACTCCT_TSP6_Liver_NA_10X_1_1                             nan   \n",
       "ACGTTCCAGAACCCGA_TSP6_Liver_NA_10X_1_1                             nan   \n",
       "...                                                                ...   \n",
       "TGACAGTAGATGGCGT_TSP2_Vasculature_Aorta_10X_1_2                  Aorta   \n",
       "TGTTGGATCGACATCA_TSP2_Vasculature_Aorta_10X_1_2                  Aorta   \n",
       "TTGGATGGTGGCTACC_TSP2_Vasculature_Aorta_10X_1_2                  Aorta   \n",
       "CTTGATTTCTTGCAAG_TSP2_Vasculature_Aorta_10X_2_2                  Aorta   \n",
       "TCGACGGCATGGAATA_TSP2_Vasculature_Aorta_10X_2_2                  Aorta   \n",
       "\n",
       "                                                 n_counts_UMIs  n_genes  \\\n",
       "AACAGGGCATGTGCTA_TSP6_Liver_NA_10X_1_1                 14151.0     3404   \n",
       "AATGAAGTCTAAGGAA_TSP6_Liver_NA_10X_1_1                 10303.0     2863   \n",
       "ACAGAAAAGCTGTCCG_TSP6_Liver_NA_10X_1_1                  9391.0     2738   \n",
       "ACGATGTTCGACTCCT_TSP6_Liver_NA_10X_1_1                  4634.0     1634   \n",
       "ACGTTCCAGAACCCGA_TSP6_Liver_NA_10X_1_1                 33370.0     6590   \n",
       "...                                                        ...      ...   \n",
       "TGACAGTAGATGGCGT_TSP2_Vasculature_Aorta_10X_1_2        10354.0     2456   \n",
       "TGTTGGATCGACATCA_TSP2_Vasculature_Aorta_10X_1_2        15677.0     3234   \n",
       "TTGGATGGTGGCTACC_TSP2_Vasculature_Aorta_10X_1_2        18638.0     3368   \n",
       "CTTGATTTCTTGCAAG_TSP2_Vasculature_Aorta_10X_2_2        36555.0     5102   \n",
       "TCGACGGCATGGAATA_TSP2_Vasculature_Aorta_10X_2_2         6943.0     1799   \n",
       "\n",
       "                                                        cell_ontology_class  \\\n",
       "AACAGGGCATGTGCTA_TSP6_Liver_NA_10X_1_1                           hepatocyte   \n",
       "AATGAAGTCTAAGGAA_TSP6_Liver_NA_10X_1_1                           hepatocyte   \n",
       "ACAGAAAAGCTGTCCG_TSP6_Liver_NA_10X_1_1                           hepatocyte   \n",
       "ACGATGTTCGACTCCT_TSP6_Liver_NA_10X_1_1                           hepatocyte   \n",
       "ACGTTCCAGAACCCGA_TSP6_Liver_NA_10X_1_1           intrahepatic cholangiocyte   \n",
       "...                                                                     ...   \n",
       "TGACAGTAGATGGCGT_TSP2_Vasculature_Aorta_10X_1_2             epithelial cell   \n",
       "TGTTGGATCGACATCA_TSP2_Vasculature_Aorta_10X_1_2             epithelial cell   \n",
       "TTGGATGGTGGCTACC_TSP2_Vasculature_Aorta_10X_1_2             epithelial cell   \n",
       "CTTGATTTCTTGCAAG_TSP2_Vasculature_Aorta_10X_2_2             epithelial cell   \n",
       "TCGACGGCATGGAATA_TSP2_Vasculature_Aorta_10X_2_2             epithelial cell   \n",
       "\n",
       "                                                 free_annotation  \\\n",
       "AACAGGGCATGTGCTA_TSP6_Liver_NA_10X_1_1                Hepatocyte   \n",
       "AATGAAGTCTAAGGAA_TSP6_Liver_NA_10X_1_1                Hepatocyte   \n",
       "ACAGAAAAGCTGTCCG_TSP6_Liver_NA_10X_1_1                Hepatocyte   \n",
       "ACGATGTTCGACTCCT_TSP6_Liver_NA_10X_1_1                Hepatocyte   \n",
       "ACGTTCCAGAACCCGA_TSP6_Liver_NA_10X_1_1                      BECS   \n",
       "...                                                          ...   \n",
       "TGACAGTAGATGGCGT_TSP2_Vasculature_Aorta_10X_1_2  epithelial cell   \n",
       "TGTTGGATCGACATCA_TSP2_Vasculature_Aorta_10X_1_2  epithelial cell   \n",
       "TTGGATGGTGGCTACC_TSP2_Vasculature_Aorta_10X_1_2  epithelial cell   \n",
       "CTTGATTTCTTGCAAG_TSP2_Vasculature_Aorta_10X_2_2  epithelial cell   \n",
       "TCGACGGCATGGAATA_TSP2_Vasculature_Aorta_10X_2_2  epithelial cell   \n",
       "\n",
       "                                                 manually_annotated  \\\n",
       "AACAGGGCATGTGCTA_TSP6_Liver_NA_10X_1_1                         True   \n",
       "AATGAAGTCTAAGGAA_TSP6_Liver_NA_10X_1_1                         True   \n",
       "ACAGAAAAGCTGTCCG_TSP6_Liver_NA_10X_1_1                         True   \n",
       "ACGATGTTCGACTCCT_TSP6_Liver_NA_10X_1_1                         True   \n",
       "ACGTTCCAGAACCCGA_TSP6_Liver_NA_10X_1_1                         True   \n",
       "...                                                             ...   \n",
       "TGACAGTAGATGGCGT_TSP2_Vasculature_Aorta_10X_1_2                True   \n",
       "TGTTGGATCGACATCA_TSP2_Vasculature_Aorta_10X_1_2                True   \n",
       "TTGGATGGTGGCTACC_TSP2_Vasculature_Aorta_10X_1_2                True   \n",
       "CTTGATTTCTTGCAAG_TSP2_Vasculature_Aorta_10X_2_2                True   \n",
       "TCGACGGCATGGAATA_TSP2_Vasculature_Aorta_10X_2_2                True   \n",
       "\n",
       "                                                compartment  gender  \n",
       "AACAGGGCATGTGCTA_TSP6_Liver_NA_10X_1_1           epithelial    male  \n",
       "AATGAAGTCTAAGGAA_TSP6_Liver_NA_10X_1_1           epithelial    male  \n",
       "ACAGAAAAGCTGTCCG_TSP6_Liver_NA_10X_1_1           epithelial    male  \n",
       "ACGATGTTCGACTCCT_TSP6_Liver_NA_10X_1_1           epithelial    male  \n",
       "ACGTTCCAGAACCCGA_TSP6_Liver_NA_10X_1_1           epithelial    male  \n",
       "...                                                     ...     ...  \n",
       "TGACAGTAGATGGCGT_TSP2_Vasculature_Aorta_10X_1_2  epithelial  female  \n",
       "TGTTGGATCGACATCA_TSP2_Vasculature_Aorta_10X_1_2  epithelial  female  \n",
       "TTGGATGGTGGCTACC_TSP2_Vasculature_Aorta_10X_1_2  epithelial  female  \n",
       "CTTGATTTCTTGCAAG_TSP2_Vasculature_Aorta_10X_2_2  epithelial  female  \n",
       "TCGACGGCATGGAATA_TSP2_Vasculature_Aorta_10X_2_2  epithelial  female  \n",
       "\n",
       "[104148 rows x 11 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "06bf7ead-d96d-4ced-8911-99e29193fc79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['hepatocyte',\n",
       " 'intrahepatic cholangiocyte',\n",
       " 'tracheal goblet cell',\n",
       " 'ciliated cell',\n",
       " 'ionocyte',\n",
       " 'secretory cell',\n",
       " 'basal cell',\n",
       " 'mucus secreting cell',\n",
       " 'serous cell of epithelium of trachea',\n",
       " 'acinar cell of salivary gland',\n",
       " 'duct epithelial cell',\n",
       " 'myoepithelial cell',\n",
       " 'epithelial cell',\n",
       " 'keratinocyte',\n",
       " 'luminal epithelial cell of mammary gland',\n",
       " 'epithelial cell of uterus',\n",
       " 'ciliated epithelial cell',\n",
       " 'conjunctival epithelial cell',\n",
       " 'eye photoreceptor cell',\n",
       " 'limbal stem cell',\n",
       " 'epithelial cell of lacrimal sac',\n",
       " 'corneal keratocyte',\n",
       " 'retinal pigment epithelial cell',\n",
       " 'corneal epithelial cell',\n",
       " 'ocular surface cell',\n",
       " 'ciliary body',\n",
       " 'pancreatic acinar cell',\n",
       " 'pancreatic ductal cell',\n",
       " 'pancreatic beta cell',\n",
       " 'basal cell of prostate epithelium',\n",
       " 'hillock-club cell of prostate epithelium',\n",
       " 'luminal cell of prostate epithelium',\n",
       " 'salivary gland cell',\n",
       " 'medullary thymic epithelial cell',\n",
       " 'bladder urothelial cell',\n",
       " 'enterocyte of epithelium of large intestine',\n",
       " 'large intestine goblet cell',\n",
       " 'paneth cell of epithelium of large intestine',\n",
       " 'transit amplifying cell of large intestine',\n",
       " 'intestinal crypt stem cell of large intestine',\n",
       " 'intestinal enteroendocrine cell',\n",
       " 'intestinal tuft cell',\n",
       " 'type ii pneumocyte',\n",
       " 'respiratory mucous cell',\n",
       " 'club cell of prostate epithelium',\n",
       " 'hillock cell of prostate epithelium',\n",
       " 'enterocyte of epithelium of small intestine',\n",
       " 'transit amplifying cell of small intestine',\n",
       " 'small intestine goblet cell',\n",
       " 'paneth cell of epithelium of small intestine',\n",
       " 'intestinal crypt stem cell of small intestine',\n",
       " 'kidney epithelial cell',\n",
       " 'immature enterocyte',\n",
       " 'intestinal crypt stem cell',\n",
       " 'mature enterocyte',\n",
       " 'goblet cell',\n",
       " 'lung ciliated cell',\n",
       " 'respiratory goblet cell',\n",
       " 'serous cell of epithelium of bronchus',\n",
       " 'type i pneumocyte',\n",
       " 'club cell',\n",
       " 'pulmonary ionocyte',\n",
       " 'pancreatic pp cell',\n",
       " 'pancreatic alpha cell',\n",
       " 'pancreatic delta cell',\n",
       " 'duodenum glandular cell',\n",
       " 'mesothelial cell']"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(adata.obs['cell_ontology_class'].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82aa3850-58e1-41b4-9b2a-60db4bea28b1",
   "metadata": {},
   "source": [
    "## Build driver"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1563c2da-5366-4738-a33b-2bfe1e881eb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def load_fibroblasts():\n",
    "    \"\"\"\n",
    "    Loads the single-cell RNA sequencing data for fibroblast cells.\n",
    "\n",
    "    The function loads a specified .h5ad file containing cellular data, filters the data\n",
    "    to only include fibroblasts, and returns the AnnData object.\n",
    "\n",
    "    Returns:\n",
    "        AnnData: The AnnData object containing data for fibroblast cells only.\n",
    "    \"\"\"\n",
    "\n",
    "    # Path to the data\n",
    "    DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/jpic/\"\n",
    "    FILE = \"fibroblast.h5ad\"\n",
    "    \n",
    "    # Load the data\n",
    "    adata = sc.read_h5ad(os.path.join(DATAPATH, FILE))\n",
    "    \n",
    "    return fibroblasts\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "\n",
    "def main(job_number, parameter_file):\n",
    "    \"\"\"\n",
    "    Main function for performing in silico reprogramming experiments. It selects the correct model, source, target, \n",
    "    transcription factors (TFs), and other parameters based on the job number in a SLURM array job. Then, it processes \n",
    "    the data and saves the results.\n",
    "    \n",
    "    Args:\n",
    "        job_number (int): The SLURM job array index to determine which parameters to use.\n",
    "        parameter_file (str): Path to the CSV file containing model parameters.\n",
    "    \n",
    "    Returns:\n",
    "        int: Return code (0 for success).\n",
    "    \"\"\"\n",
    "\n",
    "    OUTPUTPATH = \"/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot\"\n",
    "    \n",
    "    # Load parameters from CSV\n",
    "    df_embedding_parameters = pd.read_csv(parameter_file)\n",
    "    TFs    = df_embedding_parameters['TFs'].values[job_number]\n",
    "    model  = df_embedding_parameters['model'].values[job_number]\n",
    "    source = df_embedding_parameters['source'].values[job_number]\n",
    "    target = df_embedding_parameters['target'].values[job_number]\n",
    "    PMID   = df_embedding_parameters['PMID'].values[job_number]\n",
    "    year   = df_embedding_parameters['year'].values[job_number]\n",
    "    organ  = df_embedding_parameters['organ'].values[job_number]\n",
    "\n",
    "    print(f\"Starting job {job_number} with model {model}, TFs {TFs}, source {source}, target {target}, PMID {PMID}, year {year}\")\n",
    "\n",
    "    # Load the source data (Fibroblasts)\n",
    "    source_adata = load_fibroblasts()\n",
    "    print(\"Source data loaded\")\n",
    "\n",
    "    # TODO: Load the target data\n",
    "    targets_adata = load_targets()\n",
    "    print(\"Target data loaded\")\n",
    "\n",
    "    # Perturb the source data\n",
    "    perturbed_adata = perturbation_model(source_adata, TFs)\n",
    "    print(\"Data perturbed\")\n",
    "\n",
    "    # Combine source, perturbed, and target data\n",
    "    combined_adata = combine_adata([source_adata, perturbed_adata, targets_adata])\n",
    "\n",
    "    # Add a new 'reprogrammed type' column to indicate the source, target, and reprogrammed cells\n",
    "    combined_adata.obs['reprogrammed type'] = ['source'] * source_adata.shape[0] + \\\n",
    "                                              ['reprogrammed'] * perturbed_adata.shape[0] + \\\n",
    "                                              ['target'] * targets_adata.shape[0]\n",
    "    print(\"Combined data and added 'reprogrammed type' column\")\n",
    "\n",
    "    # Add columns to the 'obs' field of combined_adata for metadata tracking\n",
    "    combined_adata.obs['model'] = model\n",
    "    combined_adata.obs['TFs'] = ','.join(TFs)\n",
    "    combined_adata.obs['PMID'] = PMID\n",
    "    combined_adata.obs['year'] = year\n",
    "    combined_adata.obs['organ'] = organ\n",
    "    print(\"Added metadata columns to combined data\")\n",
    "\n",
    "    # Generate embeddings using the specified model\n",
    "    if model == 'geneformer':\n",
    "        adata_embedded = embed_geneformer(combined_adata)\n",
    "    elif model == 'tGPT':\n",
    "        adata_embedded = embed_tGPT(combined_adata)\n",
    "    elif model == 'scGTP':\n",
    "        adata_embedded = embed_scGTP([combined_adata])\n",
    "    else:\n",
    "        raise ValueError(f\"Unknown model: {model}\")\n",
    "\n",
    "    print(f\"Embeddings generated using model: {model}\")\n",
    "\n",
    "    # Save the results to a file named according to the job number\n",
    "    result_filepath = os.path.join(OUTPUTPATH, f\"results_job_{job_number}.h5ad\")\n",
    "    adata_embedded.write(result_filepath)\n",
    "    print(f\"Results saved to {result_filepath}\")\n",
    "\n",
    "    return 0\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee4f3c98-3474-4971-a559-f0a294dba782",
   "metadata": {},
   "source": [
    "## Build parameter dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "be9f04aa-ab4c-44fc-86e0-f69d7f21451d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>TFs</th>\n",
       "      <th>model</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ASCL1, LXMLA, NURR1]</td>\n",
       "      <td>geneformer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ASCL1, LXMLA, NURR1]</td>\n",
       "      <td>tGPT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ASCL1, LXMLA, NURR1]</td>\n",
       "      <td>scGTP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[BRN2, ASCL1, MYT1L, NEUROD1]</td>\n",
       "      <td>geneformer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[BRN2, ASCL1, MYT1L, NEUROD1]</td>\n",
       "      <td>tGPT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[BRN2, ASCL1, MYT1L, NEUROD1]</td>\n",
       "      <td>scGTP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ASCL1, BRN2, MYT1L, LMX1A, FOXA2]</td>\n",
       "      <td>geneformer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ASCL1, BRN2, MYT1L, LMX1A, FOXA2]</td>\n",
       "      <td>tGPT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ASCL1, BRN2, MYT1L, LMX1A, FOXA2]</td>\n",
       "      <td>scGTP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[SOX2]</td>\n",
       "      <td>geneformer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[SOX2]</td>\n",
       "      <td>tGPT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[SOX2]</td>\n",
       "      <td>scGTP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[MASH1, NGN2, SOX2, NURR1, PITX3, P53]</td>\n",
       "      <td>geneformer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[MASH1, NGN2, SOX2, NURR1, PITX3, P53]</td>\n",
       "      <td>tGPT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[MASH1, NGN2, SOX2, NURR1, PITX3, P53]</td>\n",
       "      <td>scGTP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    source  target                                     TFs       model\n",
       "0      NaN     NaN                   [ASCL1, LXMLA, NURR1]  geneformer\n",
       "1      NaN     NaN                   [ASCL1, LXMLA, NURR1]        tGPT\n",
       "2      NaN     NaN                   [ASCL1, LXMLA, NURR1]       scGTP\n",
       "3      NaN     NaN           [BRN2, ASCL1, MYT1L, NEUROD1]  geneformer\n",
       "4      NaN     NaN           [BRN2, ASCL1, MYT1L, NEUROD1]        tGPT\n",
       "5      NaN     NaN           [BRN2, ASCL1, MYT1L, NEUROD1]       scGTP\n",
       "6      NaN     NaN      [ASCL1, BRN2, MYT1L, LMX1A, FOXA2]  geneformer\n",
       "7      NaN     NaN      [ASCL1, BRN2, MYT1L, LMX1A, FOXA2]        tGPT\n",
       "8      NaN     NaN      [ASCL1, BRN2, MYT1L, LMX1A, FOXA2]       scGTP\n",
       "9      NaN     NaN                                  [SOX2]  geneformer\n",
       "10     NaN     NaN                                  [SOX2]        tGPT\n",
       "11     NaN     NaN                                  [SOX2]       scGTP\n",
       "12     NaN     NaN  [MASH1, NGN2, SOX2, NURR1, PITX3, P53]  geneformer\n",
       "13     NaN     NaN  [MASH1, NGN2, SOX2, NURR1, PITX3, P53]        tGPT\n",
       "14     NaN     NaN  [MASH1, NGN2, SOX2, NURR1, PITX3, P53]       scGTP"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "embedding_parameters = {\n",
    "    'source': [],\n",
    "    'target': [],\n",
    "    'TFs'   : [],\n",
    "    'model' : []\n",
    "}\n",
    "models = ['geneformer', 'tGPT', 'scGTP']\n",
    "\n",
    "df = pd.read_csv('data/first_5_recepies_8_29_2024.csv')\n",
    "\n",
    "for i in range(5):\n",
    "    TFs = df['TFs'].values[i].split()\n",
    "    source = df['Source'].values[i]\n",
    "    target = df['Target'].values[i]\n",
    "    for model in models:\n",
    "        embedding_parameters['TFs'].append(TFs)\n",
    "        embedding_parameters['source'].append(source)\n",
    "        embedding_parameters['target'].append(source)\n",
    "        embedding_parameters['model'].append(model)\n",
    "\n",
    "df_embedding_parameters = pd.DataFrame(embedding_parameters)\n",
    "\n",
    "df_embedding_parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2ad22cdb-f1ac-48b6-8053-b69d91309c76",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source cells</th>\n",
       "      <th>Source</th>\n",
       "      <th>Target cells</th>\n",
       "      <th>Target</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>TFs</th>\n",
       "      <th>Species</th>\n",
       "      <th>Cell Transplantation</th>\n",
       "      <th>Published Year</th>\n",
       "      <th>PMID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Embryonic Fibroblasts and Adult Skin Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Dopamine Neurons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ASCL1, LXMLA, and NURR1</td>\n",
       "      <td>ASCL1 LXMLA NURR1</td>\n",
       "      <td>Both</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2011</td>\n",
       "      <td>2172534_27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Fetal Fibroblasts and Postnatal Foreskin Fibro...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neuronal Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BRN2, ASCL1, MYT1L, and NEUROD1</td>\n",
       "      <td>BRN2 ASCL1 MYT1L NEUROD1</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2011</td>\n",
       "      <td>21617644_28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Embryonic Fibroblasts and Postnatal Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neuronal Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ASCL1, BRN2, MYT1L, LMX1A, and FOXA2</td>\n",
       "      <td>ASCL1 BRN2 MYT1L LMX1A FOXA2</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2011</td>\n",
       "      <td>21646515_29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Embryonic Fibroblasts and Fetal Foreskin Fibro...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neural Stem Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SOX2</td>\n",
       "      <td>SOX2</td>\n",
       "      <td>Both</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2012</td>\n",
       "      <td>22683203_31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Fibroblasts (IMR90 Cells)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Dopaminergic Neurons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MASH1, NGN2, SOX2, NURR1, and PITX3 + A Domina...</td>\n",
       "      <td>MASH1 NGN2 SOX2 NURR1 PITX3 P53</td>\n",
       "      <td>Human</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2014</td>\n",
       "      <td>25129808_214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Adult Skin Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Motor Neurons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NEUROG2, SOX11, ISL1, and LXH3</td>\n",
       "      <td>NEUROG2 SOX11 ISL1 LXH3</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2016</td>\n",
       "      <td>26725112_46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Umbilical Cord Blood Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neural Stem Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SOX2 and HMGA2</td>\n",
       "      <td>SOX2 HMGA2</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017</td>\n",
       "      <td>28844127_33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Fibroblast-Like Cells from Retinal Tissues</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neuronal Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ASCL1 and PAX6</td>\n",
       "      <td>ASCL1 PAX6</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017</td>\n",
       "      <td>28697461_217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Pluripotent Stem Cell-Derived Cardiomyocytes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neuronal Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BRN2, ASCL1, MYT1L, and NEUROD1</td>\n",
       "      <td>BRN2 ASCL1 MYT1L NEUROD1</td>\n",
       "      <td>Both</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017</td>\n",
       "      <td>28327614_218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Motor Neurons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ASCL1, ISL1, NEUROD1, BRN2, HB9, LHX3, MYT1L, ...</td>\n",
       "      <td>ASCL1 ISL1 NEUROD1 BRN2 HB9 LHX3 MYT1L NGN2</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017</td>\n",
       "      <td>28099929_219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neurocytes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SOX2, GATA3, and NEUROD1</td>\n",
       "      <td>SOX2 GATA3 NEUROD1</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017</td>\n",
       "      <td>28587331_62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Dermal Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neural Precursor Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SOX2 and PAX6</td>\n",
       "      <td>SOX2 PAX6</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018</td>\n",
       "      <td>30450440_34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Embryonic Fibroblasts and Newborn Foreskin Fib...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neural Stem Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ptf1a</td>\n",
       "      <td>Ptf1a</td>\n",
       "      <td>Both</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2018</td>\n",
       "      <td>30030434_35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Adult Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neural Precursor Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SOX2; SOX2 and PAX6; SOX2 and LMX1A; SOX2, LMX...</td>\n",
       "      <td>SOX2; SOX2 and PAX6; SOX2 and LMX1A; SOX2, LMX...</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018</td>\n",
       "      <td>30018471_36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Umbilical Cord Mesenchymal Stem Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neuronal Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SOX2, ASCL1, and NEUROG2</td>\n",
       "      <td>SOX2, ASCL1, and NEUROG2</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018</td>\n",
       "      <td>29937717_221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Pericytes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neuronal Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ASCL1 and SOX2</td>\n",
       "      <td>ASCL1 and SOX2</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018</td>\n",
       "      <td>29915193_181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Cord Blood CD133(+) Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neuronal Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FOXM1, SOX2, MYC, SALL4, and STAT6</td>\n",
       "      <td>FOXM1, SOX2, MYC, SALL4, and STAT6</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018</td>\n",
       "      <td>29653196_38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Peripheral CD34(+) Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neural Stem Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AR, SOX2, SMAD3, MYC, JUN, WT1, TAL1, SPI1, an...</td>\n",
       "      <td>AR, SOX2, SMAD3, MYC, JUN, WT1, TAL1, SPI1, an...</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018</td>\n",
       "      <td>29653196_38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Urine Epithelial-Like Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neural Stem Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>POU3F2, SOX2, BACH1, AR, PBX1, and NANOG</td>\n",
       "      <td>POU3F2, SOX2, BACH1, AR, PBX1, and NANOG</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018</td>\n",
       "      <td>29653196_38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Astrocytes and Foreskin Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Noradrenergic Neurons</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ASCL1, PHOX2B, AP-2A, GATA3, HAND2, NURR1, and...</td>\n",
       "      <td>ASCL1, PHOX2B, AP-2A, GATA3, HAND2, NURR1, and...</td>\n",
       "      <td>Both</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2019</td>\n",
       "      <td>31315047_222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Bone Marrow-Derived Cells, Fibroblasts, and Ke...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Neural Precursor Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MSI1, NGN2, and MBD2</td>\n",
       "      <td>MSI1, NGN2, and MBD2</td>\n",
       "      <td>Human</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2019</td>\n",
       "      <td>31196173_37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Cardiac Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cardiomyocytes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GATA4, MEF2C, TBX5, MESP1, MYOCD</td>\n",
       "      <td>GATA4, MEF2C, TBX5, MESP1, MYOCD</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2013</td>\n",
       "      <td>23861494_52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Embryonic Stem Cells-Derived Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cardiomyocytes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GATA4, MEF2C, TBX5, ESRRG, MESP1, ZFPM2, MYOCD</td>\n",
       "      <td>GATA4, MEF2C, TBX5, ESRRG, MESP1, ZFPM2, MYOCD</td>\n",
       "      <td>Human</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2013</td>\n",
       "      <td>24319660_53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Embryonic Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hepatocytes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HNF1A, HNF4A, HNF6, ATF5, PROX1, CEBPA, and MY...</td>\n",
       "      <td>HNF1A, HNF4A, HNF6, ATF5, PROX1, CEBPA, and MY...</td>\n",
       "      <td>Human</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2014</td>\n",
       "      <td>24582926_57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Fetal and Adult Fibroblasts and Adipose Tissue...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hepatocytes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FOXA3, HNF1A, and HNF4A + SV40 Large T Antigen</td>\n",
       "      <td>FOXA3, HNF1A, and HNF4A + SV40 Large T Antigen</td>\n",
       "      <td>Human</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2014</td>\n",
       "      <td>24582927_58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Fibroblasts (BJ and MRC-5 Cells)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hepatocytes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HNF1A, and Any Two of the Three Factors: FOXA1...</td>\n",
       "      <td>HNF1A, and Any Two of the Three Factors: FOXA1...</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2014</td>\n",
       "      <td>24963715_59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Fetal Lung Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hepatocytes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATF5, PROX1, FOXA2, FOXA3, and HNF4A</td>\n",
       "      <td>ATF5, PROX1, FOXA2, FOXA3, and HNF4A</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017</td>\n",
       "      <td>29192290_61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hepatocytes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OCT4, FOXA2, HNF1A, and GATA3</td>\n",
       "      <td>OCT4, FOXA2, HNF1A, and GATA3</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017</td>\n",
       "      <td>28587331_62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Liver Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>β Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDX1, PAX4, and MAFA</td>\n",
       "      <td>PDX1, PAX4, and MAFA</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2014</td>\n",
       "      <td>24504462_66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Fibroblasts (3T3 Cells) and Adult Dermal Fibro...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Megakaryocyte</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NFE2, MAFG, and MAFK</td>\n",
       "      <td>NFE2, MAFG, and MAFK</td>\n",
       "      <td>Both</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2012</td>\n",
       "      <td>22855609_75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Antigen-Presenting Dendritic Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PU.1, IRF8, and BATF3</td>\n",
       "      <td>PU.1, IRF8, and BATF3</td>\n",
       "      <td>Both</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018</td>\n",
       "      <td>30530727_78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Neonatal Foreskin Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Chondrogenic Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>c-MYC, KLF4, and SOX9</td>\n",
       "      <td>c-MYC, KLF4, and SOX9</td>\n",
       "      <td>Human</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2013</td>\n",
       "      <td>24146984_224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Dermal Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Osteoblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OCT3/4 and OCT6 or OCT9 + L-MYC, c-MYC, or N-MYC</td>\n",
       "      <td>OCT3/4 and OCT6 or OCT9 + L-MYC, c-MYC, or N-MYC</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2015</td>\n",
       "      <td>26499074_79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Osteoblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RUNX2, OCT4, OSTERIX, and L-MYC</td>\n",
       "      <td>RUNX2, OCT4, OSTERIX, and L-MYC</td>\n",
       "      <td>Human</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2015</td>\n",
       "      <td>25918395_80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Gingival Fibroblasts and Adult Dermal Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Osteoblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OCT4, OSTERIX, and L-MYC</td>\n",
       "      <td>OCT4, OSTERIX, and L-MYC</td>\n",
       "      <td>Human</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2016</td>\n",
       "      <td>26990860_81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Dermal Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Myoblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MYOD1 and MYCL</td>\n",
       "      <td>MYOD1 and MYCL</td>\n",
       "      <td>Human</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2017</td>\n",
       "      <td>28501623_83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Embryonic Fibroblasts and Newborn Foreskin Fib...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Brown Fat Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRDM16 and CEBPB</td>\n",
       "      <td>PRDM16 and CEBPB</td>\n",
       "      <td>Both</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2009</td>\n",
       "      <td>19641492_95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Iris-Derived Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Photoreceptor Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CRX, RAX, and NEUROD</td>\n",
       "      <td>CRX, RAX, and NEUROD</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2012</td>\n",
       "      <td>22558175_85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Embryonic Fibroblasts and Adult Tail-Tip Derma...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Melanocytes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MITF, SOX10, and PAX3</td>\n",
       "      <td>MITF, SOX10, and PAX3</td>\n",
       "      <td>Both</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2014</td>\n",
       "      <td>25510211_97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Adipose Tissue-Derived Stromal Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Endothelial Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SOX18</td>\n",
       "      <td>SOX18</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2014</td>\n",
       "      <td>25290189_93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Dermal Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Photoreceptor Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CRX, RAX, OTX2, and NEUROD</td>\n",
       "      <td>CRX, RAX, OTX2, and NEUROD</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2014</td>\n",
       "      <td>24456169_86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sweat Gland Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NF-κB and LEF-1</td>\n",
       "      <td>NF-κB and LEF-1</td>\n",
       "      <td>Human</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2015</td>\n",
       "      <td>26566868_91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Amniotic Fluid Stemm Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pluripotent Stem Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OCT4</td>\n",
       "      <td>OCT4</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2016</td>\n",
       "      <td>27019637_225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Embryonic Fibroblasts, Adult Tail-Tip Dermal F...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Renal Tubular Epithelial Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EMX2, HNF1B, HNF4A, and PAX8</td>\n",
       "      <td>EMX2, HNF1B, HNF4A, and PAX8</td>\n",
       "      <td>Both</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2016</td>\n",
       "      <td>27820600_99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Endothelial Progenitor Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Smooth Muscle Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MYOCD</td>\n",
       "      <td>MYOCD</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2016</td>\n",
       "      <td>26874281_89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Postnatal Dermal Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Endothelial Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ER71/ETV2 (ETS Variant 2)</td>\n",
       "      <td>ER71/ETV2 (ETS Variant 2)</td>\n",
       "      <td>Human</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2017</td>\n",
       "      <td>28003219_94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Dermal Fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Adipocytes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PPARG2</td>\n",
       "      <td>PPARG2</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017</td>\n",
       "      <td>28982679_88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Embryonic Fibroblasts and Umbilical Vein Endot...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intestine Progenitor Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HNF4A, FOXA3, GATA6, and CDX2</td>\n",
       "      <td>HNF4A, FOXA3, GATA6, and CDX2</td>\n",
       "      <td>Both</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2017</td>\n",
       "      <td>28943029_100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Embryonic Fibroblasts and Adult Dermal Fibrobl...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Smooth Muscle Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MYOCD, GATA6, and MEF2C</td>\n",
       "      <td>MYOCD, GATA6, and MEF2C</td>\n",
       "      <td>Both</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018</td>\n",
       "      <td>30026272_90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Renal Proximal Tubular Epithelial (HK2) Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nephron Progenitor Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNAI2, EYA1, and SIX1</td>\n",
       "      <td>SNAI2, EYA1, and SIX1</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019</td>\n",
       "      <td>30827514_103</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Source cells  Source  \\\n",
       "0    Embryonic Fibroblasts and Adult Skin Fibroblasts     NaN   \n",
       "1   Fetal Fibroblasts and Postnatal Foreskin Fibro...     NaN   \n",
       "2     Embryonic Fibroblasts and Postnatal Fibroblasts     NaN   \n",
       "3   Embryonic Fibroblasts and Fetal Foreskin Fibro...     NaN   \n",
       "4                           Fibroblasts (IMR90 Cells)     NaN   \n",
       "5                              Adult Skin Fibroblasts     NaN   \n",
       "6                          Umbilical Cord Blood Cells     NaN   \n",
       "7          Fibroblast-Like Cells from Retinal Tissues     NaN   \n",
       "8        Pluripotent Stem Cell-Derived Cardiomyocytes     NaN   \n",
       "9                                         Fibroblasts     NaN   \n",
       "10                                        Fibroblasts     NaN   \n",
       "11                                 Dermal Fibroblasts     NaN   \n",
       "12  Embryonic Fibroblasts and Newborn Foreskin Fib...     NaN   \n",
       "13                                  Adult Fibroblasts     NaN   \n",
       "14              Umbilical Cord Mesenchymal Stem Cells     NaN   \n",
       "15                                          Pericytes     NaN   \n",
       "16                          Cord Blood CD133(+) Cells     NaN   \n",
       "17                           Peripheral CD34(+) Cells     NaN   \n",
       "18                        Urine Epithelial-Like Cells     NaN   \n",
       "19                Astrocytes and Foreskin Fibroblasts     NaN   \n",
       "20  Bone Marrow-Derived Cells, Fibroblasts, and Ke...     NaN   \n",
       "21                                Cardiac Fibroblasts     NaN   \n",
       "22           Embryonic Stem Cells-Derived Fibroblasts     NaN   \n",
       "23                              Embryonic Fibroblasts     NaN   \n",
       "24  Fetal and Adult Fibroblasts and Adipose Tissue...     NaN   \n",
       "25                   Fibroblasts (BJ and MRC-5 Cells)     NaN   \n",
       "26                             Fetal Lung Fibroblasts     NaN   \n",
       "27                                        Fibroblasts     NaN   \n",
       "28                                        Liver Cells     NaN   \n",
       "29  Fibroblasts (3T3 Cells) and Adult Dermal Fibro...     NaN   \n",
       "30                                        Fibroblasts     NaN   \n",
       "31                      Neonatal Foreskin Fibroblasts     NaN   \n",
       "32                                 Dermal Fibroblasts     NaN   \n",
       "33                                        Fibroblasts     NaN   \n",
       "34  Gingival Fibroblasts and Adult Dermal Fibroblasts     NaN   \n",
       "35                                 Dermal Fibroblasts     NaN   \n",
       "36  Embryonic Fibroblasts and Newborn Foreskin Fib...     NaN   \n",
       "37                                 Iris-Derived Cells     NaN   \n",
       "38  Embryonic Fibroblasts and Adult Tail-Tip Derma...     NaN   \n",
       "39               Adipose Tissue-Derived Stromal Cells     NaN   \n",
       "40                                 Dermal Fibroblasts     NaN   \n",
       "41                                        Fibroblasts     NaN   \n",
       "42                         Amniotic Fluid Stemm Cells     NaN   \n",
       "43  Embryonic Fibroblasts, Adult Tail-Tip Dermal F...     NaN   \n",
       "44                       Endothelial Progenitor Cells     NaN   \n",
       "45                       Postnatal Dermal Fibroblasts     NaN   \n",
       "46                                 Dermal Fibroblasts     NaN   \n",
       "47  Embryonic Fibroblasts and Umbilical Vein Endot...     NaN   \n",
       "48  Embryonic Fibroblasts and Adult Dermal Fibrobl...     NaN   \n",
       "49      Renal Proximal Tubular Epithelial (HK2) Cells     NaN   \n",
       "\n",
       "                          Target cells  Target  \\\n",
       "0                     Dopamine Neurons     NaN   \n",
       "1                       Neuronal Cells     NaN   \n",
       "2                       Neuronal Cells     NaN   \n",
       "3                    Neural Stem Cells     NaN   \n",
       "4                 Dopaminergic Neurons     NaN   \n",
       "5                        Motor Neurons     NaN   \n",
       "6                    Neural Stem Cells     NaN   \n",
       "7                       Neuronal Cells     NaN   \n",
       "8                       Neuronal Cells     NaN   \n",
       "9                        Motor Neurons     NaN   \n",
       "10                          Neurocytes     NaN   \n",
       "11              Neural Precursor Cells     NaN   \n",
       "12                   Neural Stem Cells     NaN   \n",
       "13              Neural Precursor Cells     NaN   \n",
       "14                      Neuronal Cells     NaN   \n",
       "15                      Neuronal Cells     NaN   \n",
       "16                      Neuronal Cells     NaN   \n",
       "17                   Neural Stem Cells     NaN   \n",
       "18                   Neural Stem Cells     NaN   \n",
       "19               Noradrenergic Neurons     NaN   \n",
       "20              Neural Precursor Cells     NaN   \n",
       "21                      Cardiomyocytes     NaN   \n",
       "22                      Cardiomyocytes     NaN   \n",
       "23                         Hepatocytes     NaN   \n",
       "24                         Hepatocytes     NaN   \n",
       "25                         Hepatocytes     NaN   \n",
       "26                         Hepatocytes     NaN   \n",
       "27                         Hepatocytes     NaN   \n",
       "28                             β Cells     NaN   \n",
       "29                       Megakaryocyte     NaN   \n",
       "30  Antigen-Presenting Dendritic Cells     NaN   \n",
       "31                  Chondrogenic Cells     NaN   \n",
       "32                         Osteoblasts     NaN   \n",
       "33                         Osteoblasts     NaN   \n",
       "34                         Osteoblasts     NaN   \n",
       "35                           Myoblasts     NaN   \n",
       "36                     Brown Fat Cells     NaN   \n",
       "37                 Photoreceptor Cells     NaN   \n",
       "38                         Melanocytes     NaN   \n",
       "39                   Endothelial Cells     NaN   \n",
       "40                 Photoreceptor Cells     NaN   \n",
       "41                   Sweat Gland Cells     NaN   \n",
       "42              Pluripotent Stem Cells     NaN   \n",
       "43      Renal Tubular Epithelial Cells     NaN   \n",
       "44                 Smooth Muscle Cells     NaN   \n",
       "45                   Endothelial Cells     NaN   \n",
       "46                          Adipocytes     NaN   \n",
       "47          Intestine Progenitor Cells     NaN   \n",
       "48                 Smooth Muscle Cells     NaN   \n",
       "49            Nephron Progenitor Cells     NaN   \n",
       "\n",
       "                                            Treatment  \\\n",
       "0                             ASCL1, LXMLA, and NURR1   \n",
       "1                     BRN2, ASCL1, MYT1L, and NEUROD1   \n",
       "2                ASCL1, BRN2, MYT1L, LMX1A, and FOXA2   \n",
       "3                                                SOX2   \n",
       "4   MASH1, NGN2, SOX2, NURR1, and PITX3 + A Domina...   \n",
       "5                      NEUROG2, SOX11, ISL1, and LXH3   \n",
       "6                                      SOX2 and HMGA2   \n",
       "7                                      ASCL1 and PAX6   \n",
       "8                     BRN2, ASCL1, MYT1L, and NEUROD1   \n",
       "9   ASCL1, ISL1, NEUROD1, BRN2, HB9, LHX3, MYT1L, ...   \n",
       "10                           SOX2, GATA3, and NEUROD1   \n",
       "11                                      SOX2 and PAX6   \n",
       "12                                              Ptf1a   \n",
       "13  SOX2; SOX2 and PAX6; SOX2 and LMX1A; SOX2, LMX...   \n",
       "14                           SOX2, ASCL1, and NEUROG2   \n",
       "15                                     ASCL1 and SOX2   \n",
       "16                 FOXM1, SOX2, MYC, SALL4, and STAT6   \n",
       "17  AR, SOX2, SMAD3, MYC, JUN, WT1, TAL1, SPI1, an...   \n",
       "18           POU3F2, SOX2, BACH1, AR, PBX1, and NANOG   \n",
       "19  ASCL1, PHOX2B, AP-2A, GATA3, HAND2, NURR1, and...   \n",
       "20                               MSI1, NGN2, and MBD2   \n",
       "21                   GATA4, MEF2C, TBX5, MESP1, MYOCD   \n",
       "22     GATA4, MEF2C, TBX5, ESRRG, MESP1, ZFPM2, MYOCD   \n",
       "23  HNF1A, HNF4A, HNF6, ATF5, PROX1, CEBPA, and MY...   \n",
       "24     FOXA3, HNF1A, and HNF4A + SV40 Large T Antigen   \n",
       "25  HNF1A, and Any Two of the Three Factors: FOXA1...   \n",
       "26               ATF5, PROX1, FOXA2, FOXA3, and HNF4A   \n",
       "27                      OCT4, FOXA2, HNF1A, and GATA3   \n",
       "28                               PDX1, PAX4, and MAFA   \n",
       "29                               NFE2, MAFG, and MAFK   \n",
       "30                              PU.1, IRF8, and BATF3   \n",
       "31                              c-MYC, KLF4, and SOX9   \n",
       "32   OCT3/4 and OCT6 or OCT9 + L-MYC, c-MYC, or N-MYC   \n",
       "33                    RUNX2, OCT4, OSTERIX, and L-MYC   \n",
       "34                           OCT4, OSTERIX, and L-MYC   \n",
       "35                                     MYOD1 and MYCL   \n",
       "36                                   PRDM16 and CEBPB   \n",
       "37                               CRX, RAX, and NEUROD   \n",
       "38                              MITF, SOX10, and PAX3   \n",
       "39                                              SOX18   \n",
       "40                         CRX, RAX, OTX2, and NEUROD   \n",
       "41                                    NF-κB and LEF-1   \n",
       "42                                               OCT4   \n",
       "43                       EMX2, HNF1B, HNF4A, and PAX8   \n",
       "44                                              MYOCD   \n",
       "45                          ER71/ETV2 (ETS Variant 2)   \n",
       "46                                             PPARG2   \n",
       "47                      HNF4A, FOXA3, GATA6, and CDX2   \n",
       "48                            MYOCD, GATA6, and MEF2C   \n",
       "49                              SNAI2, EYA1, and SIX1   \n",
       "\n",
       "                                                  TFs Species  \\\n",
       "0                                   ASCL1 LXMLA NURR1    Both   \n",
       "1                            BRN2 ASCL1 MYT1L NEUROD1   Human   \n",
       "2                        ASCL1 BRN2 MYT1L LMX1A FOXA2   Human   \n",
       "3                                                SOX2    Both   \n",
       "4                     MASH1 NGN2 SOX2 NURR1 PITX3 P53   Human   \n",
       "5                             NEUROG2 SOX11 ISL1 LXH3   Human   \n",
       "6                                          SOX2 HMGA2   Human   \n",
       "7                                          ASCL1 PAX6   Human   \n",
       "8                            BRN2 ASCL1 MYT1L NEUROD1    Both   \n",
       "9         ASCL1 ISL1 NEUROD1 BRN2 HB9 LHX3 MYT1L NGN2   Human   \n",
       "10                                 SOX2 GATA3 NEUROD1   Human   \n",
       "11                                          SOX2 PAX6   Human   \n",
       "12                                              Ptf1a    Both   \n",
       "13  SOX2; SOX2 and PAX6; SOX2 and LMX1A; SOX2, LMX...   Human   \n",
       "14                           SOX2, ASCL1, and NEUROG2   Human   \n",
       "15                                     ASCL1 and SOX2   Human   \n",
       "16                 FOXM1, SOX2, MYC, SALL4, and STAT6   Human   \n",
       "17  AR, SOX2, SMAD3, MYC, JUN, WT1, TAL1, SPI1, an...   Human   \n",
       "18           POU3F2, SOX2, BACH1, AR, PBX1, and NANOG   Human   \n",
       "19  ASCL1, PHOX2B, AP-2A, GATA3, HAND2, NURR1, and...    Both   \n",
       "20                               MSI1, NGN2, and MBD2   Human   \n",
       "21                   GATA4, MEF2C, TBX5, MESP1, MYOCD   Human   \n",
       "22     GATA4, MEF2C, TBX5, ESRRG, MESP1, ZFPM2, MYOCD   Human   \n",
       "23  HNF1A, HNF4A, HNF6, ATF5, PROX1, CEBPA, and MY...   Human   \n",
       "24     FOXA3, HNF1A, and HNF4A + SV40 Large T Antigen   Human   \n",
       "25  HNF1A, and Any Two of the Three Factors: FOXA1...   Human   \n",
       "26               ATF5, PROX1, FOXA2, FOXA3, and HNF4A   Human   \n",
       "27                      OCT4, FOXA2, HNF1A, and GATA3   Human   \n",
       "28                               PDX1, PAX4, and MAFA   Human   \n",
       "29                               NFE2, MAFG, and MAFK    Both   \n",
       "30                              PU.1, IRF8, and BATF3    Both   \n",
       "31                              c-MYC, KLF4, and SOX9   Human   \n",
       "32   OCT3/4 and OCT6 or OCT9 + L-MYC, c-MYC, or N-MYC   Human   \n",
       "33                    RUNX2, OCT4, OSTERIX, and L-MYC   Human   \n",
       "34                           OCT4, OSTERIX, and L-MYC   Human   \n",
       "35                                     MYOD1 and MYCL   Human   \n",
       "36                                   PRDM16 and CEBPB    Both   \n",
       "37                               CRX, RAX, and NEUROD   Human   \n",
       "38                              MITF, SOX10, and PAX3    Both   \n",
       "39                                              SOX18   Human   \n",
       "40                         CRX, RAX, OTX2, and NEUROD   Human   \n",
       "41                                    NF-κB and LEF-1   Human   \n",
       "42                                               OCT4   Human   \n",
       "43                       EMX2, HNF1B, HNF4A, and PAX8    Both   \n",
       "44                                              MYOCD   Human   \n",
       "45                          ER71/ETV2 (ETS Variant 2)   Human   \n",
       "46                                             PPARG2   Human   \n",
       "47                      HNF4A, FOXA3, GATA6, and CDX2    Both   \n",
       "48                            MYOCD, GATA6, and MEF2C    Both   \n",
       "49                              SNAI2, EYA1, and SIX1   Human   \n",
       "\n",
       "   Cell Transplantation  Published Year          PMID  \n",
       "0                   Yes            2011    2172534_27  \n",
       "1                   NaN            2011   21617644_28  \n",
       "2                   NaN            2011   21646515_29  \n",
       "3                   Yes            2012   22683203_31  \n",
       "4                   Yes            2014  25129808_214  \n",
       "5                   NaN            2016   26725112_46  \n",
       "6                   NaN            2017   28844127_33  \n",
       "7                   NaN            2017  28697461_217  \n",
       "8                   NaN            2017  28327614_218  \n",
       "9                   NaN            2017  28099929_219  \n",
       "10                  NaN            2017   28587331_62  \n",
       "11                  NaN            2018   30450440_34  \n",
       "12                  Yes            2018   30030434_35  \n",
       "13                  NaN            2018   30018471_36  \n",
       "14                  NaN            2018  29937717_221  \n",
       "15                  NaN            2018  29915193_181  \n",
       "16                  NaN            2018   29653196_38  \n",
       "17                  NaN            2018   29653196_38  \n",
       "18                  NaN            2018   29653196_38  \n",
       "19                  Yes            2019  31315047_222  \n",
       "20                  Yes            2019   31196173_37  \n",
       "21                  NaN            2013   23861494_52  \n",
       "22                  Yes            2013   24319660_53  \n",
       "23                  Yes            2014   24582926_57  \n",
       "24                  Yes            2014   24582927_58  \n",
       "25                  NaN            2014   24963715_59  \n",
       "26                  NaN            2017   29192290_61  \n",
       "27                  NaN            2017   28587331_62  \n",
       "28                  NaN            2014   24504462_66  \n",
       "29                  Yes            2012   22855609_75  \n",
       "30                  NaN            2018   30530727_78  \n",
       "31                  Yes            2013  24146984_224  \n",
       "32                  NaN            2015   26499074_79  \n",
       "33                  Yes            2015   25918395_80  \n",
       "34                  Yes            2016   26990860_81  \n",
       "35                  Yes            2017   28501623_83  \n",
       "36                  Yes            2009   19641492_95  \n",
       "37                  NaN            2012   22558175_85  \n",
       "38                  Yes            2014   25510211_97  \n",
       "39                  NaN            2014   25290189_93  \n",
       "40                  NaN            2014   24456169_86  \n",
       "41                  Yes            2015   26566868_91  \n",
       "42                  NaN            2016  27019637_225  \n",
       "43                  Yes            2016   27820600_99  \n",
       "44                  NaN            2016   26874281_89  \n",
       "45                  Yes            2017   28003219_94  \n",
       "46                  NaN            2017   28982679_88  \n",
       "47                  Yes            2017  28943029_100  \n",
       "48                  NaN            2018   30026272_90  \n",
       "49                  NaN            2019  30827514_103  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08e02020-7129-4182-9bef-6535490e26a9",
   "metadata": {},
   "source": [
    "# Day 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6c18574c-9e46-4c1d-9b77-b4f888ae0430",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "624a5d48-99dc-4880-809d-519448cbfb82",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load known reprogramming regiems\n",
    "df = pd.read_csv('data/known-regiems-T1.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7457114f-57d5-45e9-ad8d-ecaa06a76d1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len(TFs)=126\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['PROX1',\n",
       " 'P53',\n",
       " 'N-MYC',\n",
       " 'LEF-1',\n",
       " 'PRDM16',\n",
       " 'NFE2',\n",
       " 'RUNX1',\n",
       " 'A',\n",
       " 'Ptf1a',\n",
       " 'SALL4',\n",
       " 'MESP1',\n",
       " 'PAX6',\n",
       " 'MYT1L',\n",
       " 'GATA3',\n",
       " 'SNAI2',\n",
       " 'ER71/ETV2',\n",
       " 'MAFA',\n",
       " 'CEBPB',\n",
       " 'Antigen',\n",
       " 'LXMLA',\n",
       " 'Two',\n",
       " 'SOX2;',\n",
       " 'FOXA3',\n",
       " 'HAND2',\n",
       " 'MBD2',\n",
       " 'NEUROD1',\n",
       " 'BACH1',\n",
       " 'AR',\n",
       " 'ZFPM2',\n",
       " 'RUNX2',\n",
       " 'CRX',\n",
       " 'GATA4',\n",
       " 'CEBPA',\n",
       " 'OCT4',\n",
       " 'FOXM1',\n",
       " 'MYC',\n",
       " 'PAX4',\n",
       " 'MITF',\n",
       " 'CDX2',\n",
       " 'LMX1A',\n",
       " 'OTX2',\n",
       " 'Variant',\n",
       " 'BRN2',\n",
       " 'MAFG',\n",
       " 'HB9',\n",
       " 'EYA1',\n",
       " 'PBX1',\n",
       " 'and',\n",
       " 'Factors:',\n",
       " 'SV40',\n",
       " 'NANOG',\n",
       " 'HNF1B',\n",
       " 'TP53',\n",
       " 'HNF6',\n",
       " 'LXH3',\n",
       " 'SOX10',\n",
       " 'OSTERIX',\n",
       " 'Large',\n",
       " 'TAL1',\n",
       " 'ESRRG',\n",
       " 'NEUROD',\n",
       " 'MASH1',\n",
       " 'SOX9',\n",
       " 'PHOX2A',\n",
       " 'FOXA1',\n",
       " 'STAT6',\n",
       " 'Any',\n",
       " 'SOX18',\n",
       " 'PHOX2B',\n",
       " 'the',\n",
       " '(ETS',\n",
       " 'WT1',\n",
       " 'MEF2C',\n",
       " 'IRF8',\n",
       " 'NEUROG2',\n",
       " 'Dominant-Negative',\n",
       " 'PU.1',\n",
       " 'SOX11',\n",
       " 'LHX3',\n",
       " 'RAX',\n",
       " 'NGN2',\n",
       " 'ASCL1',\n",
       " 'BATF3',\n",
       " 'NURR1',\n",
       " 'SOX2',\n",
       " 'AP-2A',\n",
       " 'MAFK',\n",
       " 'or',\n",
       " 'PAX6;',\n",
       " 'JUN',\n",
       " 'ISL1',\n",
       " 'Knockdown',\n",
       " 'EMX2',\n",
       " 'GATA6',\n",
       " '+',\n",
       " 'LMX1A;',\n",
       " 'POU3F2',\n",
       " 'c-MYC',\n",
       " 'HMGA2',\n",
       " 'HNF1A',\n",
       " 'SMAD3',\n",
       " 'OCT3/4',\n",
       " 'OCT9',\n",
       " 'MYOCD',\n",
       " 'L-MYC',\n",
       " 'T',\n",
       " 'SPI1',\n",
       " 'PPARG2',\n",
       " 'TBX5',\n",
       " 'PDX1',\n",
       " 'NF-κB',\n",
       " 'PITX3',\n",
       " 'MYOD1',\n",
       " 'PAX3',\n",
       " 'FOXA2',\n",
       " 'PAX8',\n",
       " 'of',\n",
       " '2)',\n",
       " 'HNF4A',\n",
       " 'SIX1',\n",
       " 'Three',\n",
       " 'OCT6',\n",
       " 'MSI1',\n",
       " 'MYCL',\n",
       " 'KLF4',\n",
       " 'ATF5']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get list of unique transcription factors\n",
    "TFs = []\n",
    "for regime in df['TFs'].unique():\n",
    "    TFs += regime.replace(',', '').split()\n",
    "TFs = list(set(TFs))\n",
    "print(f\"{len(TFs)=}\")\n",
    "TFs"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
