{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3be52c6f-6647-4d93-b0bf-3e05e87917e1",
   "metadata": {},
   "source": [
    "# Reprogramming Recepies\n",
    "\n",
    "Auth: Nat Oliven, Joshua Pickard\n",
    "\n",
    "Date: August 26, 2024"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "999b35ad-3c67-4023-bfcd-0da39d281c6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sp\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65d943dc-4396-42fe-b736-103d5d647760",
   "metadata": {},
   "source": [
    "# Day 10\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d209edff-4229-4a0c-a833-0638eed7342c",
   "metadata": {},
   "source": [
    "## Load Data (prev day)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5ef1294c-24df-4d55-b991-68234685928c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>ensemblid</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>False</td>\n",
       "      <td>6.398244e-05</td>\n",
       "      <td>0.835044</td>\n",
       "      <td>-0.573947</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.005574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>False</td>\n",
       "      <td>2.274395e-03</td>\n",
       "      <td>2.442280</td>\n",
       "      <td>0.533203</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.031731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR6859-1</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>False</td>\n",
       "      <td>6.175251e-05</td>\n",
       "      <td>1.295335</td>\n",
       "      <td>-0.256874</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.005634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>False</td>\n",
       "      <td>1.372886e-04</td>\n",
       "      <td>2.656352</td>\n",
       "      <td>0.680668</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.008041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_symbol     feature_type          ensemblid  highly_variable  \\\n",
       "DDX11L1          DDX11L1  Gene Expression  ENSG00000223972.5            False   \n",
       "WASH7P            WASH7P  Gene Expression  ENSG00000227232.5            False   \n",
       "MIR6859-1      MIR6859-1  Gene Expression  ENSG00000278267.1            False   \n",
       "MIR1302-2HG  MIR1302-2HG  Gene Expression  ENSG00000243485.5            False   \n",
       "MIR1302-2      MIR1302-2  Gene Expression  ENSG00000284332.1            False   \n",
       "\n",
       "                    means  dispersions  dispersions_norm      mean       std  \n",
       "DDX11L1      6.398244e-05     0.835044         -0.573947  0.000039  0.005574  \n",
       "WASH7P       2.274395e-03     2.442280          0.533203  0.001080  0.031731  \n",
       "MIR6859-1    6.175251e-05     1.295335         -0.256874  0.000033  0.005634  \n",
       "MIR1302-2HG  1.372886e-04     2.656352          0.680668  0.000048  0.008041  \n",
       "MIR1302-2    1.000000e-12          NaN          0.000000  0.000000  1.000000  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load reprogramming recipes\n",
    "df = pd.read_csv('https://raw.githubusercontent.com/Jpickard1/scFoundationModels/main/notebooks/reprogramming/data/HumanTFs_v_1.01.csv')\n",
    "df.head()\n",
    "\n",
    "# Load firboblast source cells\n",
    "DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/unperturbed\"\n",
    "FILE = \"fibroblast.h5ad\"\n",
    "adata = sp.read_h5ad(os.path.join(DATAPATH, FILE))\n",
    "\n",
    "\n",
    "adata.var.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3764a4f8-37bf-4aed-a35a-2af9c32329f0",
   "metadata": {},
   "source": [
    "## Check for invalid ensembl ids (day 9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "074d8ef0-0eb4-4523-955e-b2798e5e293e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TFs not found in counts_ensid_list:\n",
      "ZNF73_HUMAN\n",
      "ENSG00000204828\n",
      "DUX1_HUMAN\n",
      "DUX3_HUMAN\n",
      "ENSG00000262156\n",
      "ENSG00000196101\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def translate_ids(list):\n",
    "    # Extract base IDs from list_b by removing version numbers\n",
    "    new_list = [id.split('.')[0] for id in list]\n",
    "    return new_list\n",
    "\n",
    "#ensembl ids from the counts matrix\n",
    "counts_ensid_list = adata.var['ensemblid'].values.tolist()\n",
    "# there were version numbers ending with \" .'#' \" that needed to be removed\n",
    "counts_ensid_list = translate_ids(counts_ensid_list)\n",
    "\n",
    "\n",
    "# ensembl ids from the transcription factor list, which in this case is also the perturbation list (testing one at a time)\n",
    "tf_ensid_list = df['Ensembl ID'].values.tolist()\n",
    "\n",
    "# def validate_ensembl_tfs(tf_ensid_list, counts_ensid_list):\n",
    "#     for TF in tf_ensid_list:\n",
    "#         if TF not in counts_ensid_list:\n",
    "#             print(TF)\n",
    "#             return False\n",
    "#     return True\n",
    "\n",
    "\n",
    "def validate_ensembl_tfs(tf_ensid_list, counts_ensid_list):\n",
    "    missing_tfs = []\n",
    "    for TF in tf_ensid_list:\n",
    "        if TF not in counts_ensid_list:\n",
    "            missing_tfs.append(TF)\n",
    "    \n",
    "    if missing_tfs:\n",
    "        print(\"TFs not found in counts_ensid_list:\")\n",
    "        for tf in missing_tfs:\n",
    "            print(tf)\n",
    "        return False\n",
    "    \n",
    "    return True\n",
    "\n",
    "\n",
    "validate_ensembl_tfs(tf_ensid_list, counts_ensid_list)\n",
    "\n",
    "# when we perturb each tf in tf_ensid_list, we are just going to skip the one problematic one\n",
    "problem_ens_ids = ['ZNF73_HUMAN', 'ENSG00000204828', 'DUX1_HUMAN', 'DUX3_HUMAN', 'ENSG00000262156', 'ENSG00000196101']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3ec21c30-6c8b-40a8-82ac-44d7d4842030",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2765\n",
      "6\n",
      "2759\n"
     ]
    }
   ],
   "source": [
    "print(len(tf_ensid_list))\n",
    "print(len(problem_ens_ids))\n",
    "tf_ensid_list = [id for id in tf_ensid_list if id not in problem_ens_ids]\n",
    "print(len(tf_ensid_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b6a640db-48d1-409e-a836-763d9cc93eb4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>ensemblid</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>False</td>\n",
       "      <td>6.398244e-05</td>\n",
       "      <td>0.835044</td>\n",
       "      <td>-0.573947</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.005574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>False</td>\n",
       "      <td>2.274395e-03</td>\n",
       "      <td>2.442280</td>\n",
       "      <td>0.533203</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.031731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR6859-1</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>False</td>\n",
       "      <td>6.175251e-05</td>\n",
       "      <td>1.295335</td>\n",
       "      <td>-0.256874</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.005634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>False</td>\n",
       "      <td>1.372886e-04</td>\n",
       "      <td>2.656352</td>\n",
       "      <td>0.680668</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.008041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_symbol     feature_type          ensemblid  highly_variable  \\\n",
       "DDX11L1          DDX11L1  Gene Expression  ENSG00000223972.5            False   \n",
       "WASH7P            WASH7P  Gene Expression  ENSG00000227232.5            False   \n",
       "MIR6859-1      MIR6859-1  Gene Expression  ENSG00000278267.1            False   \n",
       "MIR1302-2HG  MIR1302-2HG  Gene Expression  ENSG00000243485.5            False   \n",
       "MIR1302-2      MIR1302-2  Gene Expression  ENSG00000284332.1            False   \n",
       "\n",
       "                    means  dispersions  dispersions_norm      mean       std  \n",
       "DDX11L1      6.398244e-05     0.835044         -0.573947  0.000039  0.005574  \n",
       "WASH7P       2.274395e-03     2.442280          0.533203  0.001080  0.031731  \n",
       "MIR6859-1    6.175251e-05     1.295335         -0.256874  0.000033  0.005634  \n",
       "MIR1302-2HG  1.372886e-04     2.656352          0.680668  0.000048  0.008041  \n",
       "MIR1302-2    1.000000e-12          NaN          0.000000  0.000000  1.000000  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var.head()  # Display the first 10 gene names\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0ef0f84b-55a5-4f1f-8a83-55d0ae5288bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DDX11L1        ENSG00000223972\n",
      "WASH7P         ENSG00000227232\n",
      "MIR6859-1      ENSG00000278267\n",
      "MIR1302-2HG    ENSG00000243485\n",
      "MIR1302-2      ENSG00000284332\n",
      "FAM138A        ENSG00000237613\n",
      "OR4G4P         ENSG00000268020\n",
      "OR4G11P        ENSG00000240361\n",
      "OR4F5          ENSG00000186092\n",
      "AL627309.1     ENSG00000238009\n",
      "Name: ensemblid, dtype: object\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "# Check the first few cleaned Ensembl IDs to verify the format\n",
    "print(cleaned_ensembl_ids[:10])\n",
    "\n",
    "# Check if the specific missing gene is in the cleaned list\n",
    "print('ENSG00000137203' in cleaned_ensembl_ids.values)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b191db3b-890b-493a-913f-e8b6cd6ac484",
   "metadata": {},
   "outputs": [],
   "source": [
    "def iterate_perturb_id_counts(adata, tf_list, scalar_list):\n",
    "    \"\"\"\n",
    "    Applies perturbations to the expression data of specified transcription factors across multiple scalars \n",
    "    and stores the resulting AnnData objects in a dictionary.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Iterates over a list of scalar values.\n",
    "    2. For each scalar, creates a copy of the AnnData object to preserve the original data.\n",
    "    3. Applies the `perturb_counts` function to scale the expression data of genes listed in `tf_list` by\n",
    "       the maximum gene expression of each cell and the current scalar.\n",
    "    4. Stores the perturbed AnnData object in a dictionary with the scalar as the key.\n",
    "\n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols (transcription factors) to be perturbed.\n",
    "    scalar_list (list): A list of scalar values for scaling the gene expression.\n",
    "    adata (AnnData): The AnnData object containing gene expression data (cells x genes).\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary where keys are scalar values and values are the corresponding perturbed AnnData objects.\n",
    "    \"\"\"\n",
    "    \n",
    "    adata_dict = {}\n",
    "    \n",
    "    for scalar in scalar_list:\n",
    "        # Create a copy of the AnnData object for each scalar value\n",
    "        adata_temp = adata.copy()\n",
    "        \n",
    "        # Apply perturb_counts to the copied AnnData object\n",
    "        perturbed_adata = perturb_counts(adata_temp, tf_list, scalar)\n",
    "        \n",
    "        # Store the perturbed AnnData object in the dictionary with scalar as the key\n",
    "        adata_dict[scalar] = perturbed_adata\n",
    "    \n",
    "    return adata_dict\n",
    "\n",
    "def perturb_id_counts(adata, tf_list, scalar): \n",
    "    \"\"\"\n",
    "    Applies a perturbation to the expression data of specific genes in an AnnData object.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Computes the maximum gene expression level for each cell.\n",
    "    2. Applies a scaling operation to the expression levels of genes listed in `tf_list`.\n",
    "       - Each entry of these genes in the matrix is multiplied by the maximum expression level \n",
    "         of its respective cell and a specified scalar value.\n",
    "    3. Updates the AnnData object with new columns:\n",
    "       - 'scaled': A boolean column indicating whether each gene is in the `tf_list`.\n",
    "       - 'scaled_by': Contains the scaling factor used for each gene (the product of the maximum \n",
    "         expression level of each cell and the scalar), or `1` if the gene was not in `tf_list`.\n",
    "    \n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols to be perturbed.\n",
    "    scalar (float): The scalar value used to scale the expression levels.\n",
    "    adata (AnnData): The AnnData object containing gene expression data.\n",
    "\n",
    "    Returns:\n",
    "    AnnData: The updated AnnData object with applied perturbations and new columns.\n",
    "    \"\"\"\n",
    "\n",
    "    # Create a boolean mask for genes in tf_list\n",
    "    gene_mask = adata.var['ensemblid'].isin(tf_list)\n",
    "    \n",
    "    # Save the original state of the parameter objects, in case some tfs do not translate (failsafe)\n",
    "    original_X = adata.X.copy()\n",
    "    original_gene_mask = gene_mask.copy()\n",
    "    \n",
    "    # Compute maximum expression level of each cell\n",
    "    max_exp = np.max(adata.X, axis=1)\n",
    "\n",
    "    \"\"\"This is new today. v \"\"\"\n",
    "    # Raise an error if any of the gene names in tf_list do not match column names (we will manually update these in adata):\n",
    "    missing_genes = [gene for gene in tf_list if gene not in adata.var['ensemblid'].values]\n",
    "    \n",
    "    if missing_genes:\n",
    "        # Restore original parameter objects\n",
    "        adata.X = original_X\n",
    "        gene_mask = original_gene_mask\n",
    "        raise ValueError(f\"Genes {missing_genes} not found in anndata object\")\n",
    "\n",
    "    else:    \n",
    "        \n",
    "        # Apply the scaling operation to the specified genes\n",
    "        adata.X[:, gene_mask] = max_exp * scalar\n",
    "        \n",
    "        # Add/Update 'scaled' column in var\n",
    "        adata.var['scaled'] = gene_mask\n",
    "        \n",
    "        # Add/Update 'scaled_by' column in var\n",
    "        adata.var['scaled_by'] = scalar  # Default value for genes not in tf_list\n",
    "\n",
    "        # Add/Update 'scaled_by' column in var\n",
    "        adata.obs['U'] = scalar  # Default value for genes not in tf_list\n",
    "    # adata.var = \n",
    "    return adata\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "537631a6-0137-40c7-afbd-cefdf0d23a83",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Ensembl IDs to process: 2759\n",
      "\n",
      "Processing index: 0\n",
      "Current TFs: ['ENSG00000137203']\n",
      "Perturbing TFs: ['ENSG00000137203']\n",
      "Error encountered at index 0: Genes ['ENSG00000137203'] not found in anndata object\n",
      "\n",
      "Processing index: 1\n",
      "Current TFs: ['ENSG00000008196']\n",
      "Perturbing TFs: ['ENSG00000008196']\n",
      "Error encountered at index 1: Genes ['ENSG00000008196'] not found in anndata object\n",
      "\n",
      "Processing index: 2\n",
      "Current TFs: ['ENSG00000087510']\n",
      "Perturbing TFs: ['ENSG00000087510']\n",
      "Error encountered at index 2: Genes ['ENSG00000087510'] not found in anndata object\n",
      "\n",
      "Processing index: 3\n",
      "Current TFs: ['ENSG00000008197']\n",
      "Perturbing TFs: ['ENSG00000008197']\n",
      "Error encountered at index 3: Genes ['ENSG00000008197'] not found in anndata object\n",
      "\n",
      "Processing index: 4\n",
      "Current TFs: ['ENSG00000116819']\n",
      "Perturbing TFs: ['ENSG00000116819']\n",
      "Error encountered at index 4: Genes ['ENSG00000116819'] not found in anndata object\n",
      "\n",
      "Processing index: 5\n",
      "Current TFs: ['ENSG00000117713']\n",
      "Perturbing TFs: ['ENSG00000117713']\n",
      "Error encountered at index 5: Genes ['ENSG00000117713'] not found in anndata object\n",
      "\n",
      "Processing index: 6\n",
      "Current TFs: ['ENSG00000049618']\n",
      "Perturbing TFs: ['ENSG00000049618']\n",
      "Error encountered at index 6: Genes ['ENSG00000049618'] not found in anndata object\n",
      "\n",
      "Processing index: 7\n",
      "Current TFs: ['ENSG00000116017']\n",
      "Perturbing TFs: ['ENSG00000116017']\n",
      "Error encountered at index 7: Genes ['ENSG00000116017'] not found in anndata object\n",
      "\n",
      "Processing index: 8\n",
      "Current TFs: ['ENSG00000179361']\n",
      "Perturbing TFs: ['ENSG00000179361']\n",
      "Error encountered at index 8: Genes ['ENSG00000179361'] not found in anndata object\n",
      "\n",
      "Processing index: 9\n",
      "Current TFs: ['ENSG00000205143']\n",
      "Perturbing TFs: ['ENSG00000205143']\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "output_directory = \"/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/all-tfs\"\n",
    "scalars = [0.5, 0.75, 1.001, 1.25, 1.5]\n",
    "\n",
    "# Filter tf_ensid_list to remove problematic Ensembl IDs\n",
    "tf_ensid_list = [id for id in tf_ensid_list if id not in problem_ens_ids]\n",
    "print(f\"Total Ensembl IDs to process: {len(tf_ensid_list)}\")\n",
    "\n",
    "# Clean Ensembl IDs in adata.var['ensemblid'] to remove anything after '.'\n",
    "cleaned_ensembl_ids = adata.var['ensemblid'].str.split('.').str[0]\n",
    "\n",
    "for i in range(len(tf_ensid_list)):\n",
    "    try:\n",
    "        print(f\"\\nProcessing index: {i}\")\n",
    "        val = df['Ensembl ID'].iloc[i]\n",
    "        TFs = val.split()  # Split in case there are multiple TFs\n",
    "        print(f\"Current TFs: {TFs}\")\n",
    "\n",
    "        # Check if all TFs are in the cleaned Ensembl IDs\n",
    "        missing_in_adata = [tf for tf in TFs if tf not in cleaned_ensembl_ids.values]\n",
    "        if missing_in_adata:\n",
    "            print(f\"Skipping missing genes: {missing_in_adata}\")\n",
    "            continue  # Skip this set of TFs if any are missing\n",
    "        \n",
    "        # Validate TFs with counts_ensid_list\n",
    "        if validate_ensembl_tfs(tf_ensid_list, counts_ensid_list):\n",
    "            TFs_str = \"_\".join(TFs)\n",
    "            output_path = os.path.join(output_directory, f\"{TFs_str}.h5ad\")\n",
    "\n",
    "            # Check if the file already exists\n",
    "            if os.path.exists(output_path):\n",
    "                print(f\"{output_path} already exists: continuing to next TF!\")\n",
    "                continue\n",
    "\n",
    "            print(f\"Perturbing TFs: {TFs}\")\n",
    "            adataDict = iterate_perturb_id_counts(adata.copy(), TFs, scalars)\n",
    "\n",
    "            # Concatenate all AnnData objects along the observations axis\n",
    "            concatenated_adata = ad.concat(list(adataDict.values()), axis=0)\n",
    "\n",
    "            # Copy var information from the original AnnData\n",
    "            concatenated_adata.var = adata.var.copy()\n",
    "\n",
    "            # Save reprogramming metadata into the concatenated_adata.obs table\n",
    "            concatenated_adata.obs['Source_cells'] = df['Source cells'].iloc[i]\n",
    "            concatenated_adata.obs['Target_cells'] = df['Target cells'].iloc[i]\n",
    "            concatenated_adata.obs['Treatment'] = df['Treatment'].iloc[i]\n",
    "            concatenated_adata.obs['Species'] = df['Species'].iloc[i]\n",
    "            concatenated_adata.obs['Cell_Transplantation'] = df['Cell Transplantation'].iloc[i]\n",
    "            concatenated_adata.obs['Published_Year'] = df['Published Year'].iloc[i]\n",
    "            concatenated_adata.obs['PMID'] = df['PMID'].iloc[i]\n",
    "\n",
    "            # Save the concatenated AnnData object to the file\n",
    "            concatenated_adata.write_h5ad(output_path)\n",
    "            print(\"    File created successfully\")\n",
    "        else:\n",
    "            print(f\"The file made from {val} could not be created.\")\n",
    "    except Exception as e:\n",
    "        print(f\"Error encountered at index {i}: {e}\")\n",
    "\n",
    "print('All recipes complete!')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16d95af1-336d-48d7-870e-5ee4d98272cc",
   "metadata": {},
   "source": [
    "## Perform perturbation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "156c2462-5607-47cf-99b5-5d52837b1b5d",
   "metadata": {},
   "source": [
    "### Perturbation function definitions (old)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b7ae9936-4f9c-4108-8cf4-64aca74682db",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def iterate_perturb_counts(adata, tf_list, scalar_list):\n",
    "    \"\"\"\n",
    "    Applies perturbations to the expression data of specified transcription factors across multiple scalars \n",
    "    and stores the resulting AnnData objects in a dictionary.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Iterates over a list of scalar values.\n",
    "    2. For each scalar, creates a copy of the AnnData object to preserve the original data.\n",
    "    3. Applies the `perturb_counts` function to scale the expression data of genes listed in `tf_list` by\n",
    "       the maximum gene expression of each cell and the current scalar.\n",
    "    4. Stores the perturbed AnnData object in a dictionary with the scalar as the key.\n",
    "\n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols (transcription factors) to be perturbed.\n",
    "    scalar_list (list): A list of scalar values for scaling the gene expression.\n",
    "    adata (AnnData): The AnnData object containing gene expression data (cells x genes).\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary where keys are scalar values and values are the corresponding perturbed AnnData objects.\n",
    "    \"\"\"\n",
    "    \n",
    "    adata_dict = {}\n",
    "    \n",
    "    for scalar in scalar_list:\n",
    "        # Create a copy of the AnnData object for each scalar value\n",
    "        adata_temp = adata.copy()\n",
    "        \n",
    "        # Apply perturb_counts to the copied AnnData object\n",
    "        perturbed_adata = perturb_counts(adata_temp, tf_list, scalar)\n",
    "        \n",
    "        # Store the perturbed AnnData object in the dictionary with scalar as the key\n",
    "        adata_dict[scalar] = perturbed_adata\n",
    "    \n",
    "    return adata_dict\n",
    "\n",
    "def perturb_counts(adata, tf_list, scalar): \n",
    "    \"\"\"\n",
    "    Applies a perturbation to the expression data of specific genes in an AnnData object.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Computes the maximum gene expression level for each cell.\n",
    "    2. Applies a scaling operation to the expression levels of genes listed in `tf_list`.\n",
    "       - Each entry of these genes in the matrix is multiplied by the maximum expression level \n",
    "         of its respective cell and a specified scalar value.\n",
    "    3. Updates the AnnData object with new columns:\n",
    "       - 'scaled': A boolean column indicating whether each gene is in the `tf_list`.\n",
    "       - 'scaled_by': Contains the scaling factor used for each gene (the product of the maximum \n",
    "         expression level of each cell and the scalar), or `1` if the gene was not in `tf_list`.\n",
    "    \n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols to be perturbed.\n",
    "    scalar (float): The scalar value used to scale the expression levels.\n",
    "    adata (AnnData): The AnnData object containing gene expression data.\n",
    "\n",
    "    Returns:\n",
    "    AnnData: The updated AnnData object with applied perturbations and new columns.\n",
    "    \"\"\"\n",
    "\n",
    "    # Create a boolean mask for genes in tf_list\n",
    "    gene_mask = adata.var['gene_symbol'].isin(tf_list)\n",
    "    \n",
    "    # Save the original state of the parameter objects, in case some tfs do not translate (failsafe)\n",
    "    original_X = adata.X.copy()\n",
    "    original_gene_mask = gene_mask.copy()\n",
    "    \n",
    "    # Compute maximum expression level of each cell\n",
    "    max_exp = np.max(adata.X, axis=1)\n",
    "\n",
    "    \"\"\"This is new today. v \"\"\"\n",
    "    # Raise an error if any of the gene names in tf_list do not match column names (we will manually update these in adata):\n",
    "    missing_genes = [gene for gene in tf_list if gene not in adata.var['gene_symbol'].values]\n",
    "    \n",
    "    if missing_genes:\n",
    "        # Restore original parameter objects\n",
    "        adata.X = original_X\n",
    "        gene_mask = original_gene_mask\n",
    "        raise ValueError(f\"Genes {missing_genes} not found in anndata object\")\n",
    "\n",
    "    else:    \n",
    "        \n",
    "        # Apply the scaling operation to the specified genes\n",
    "        adata.X[:, gene_mask] = max_exp * scalar\n",
    "        \n",
    "        # Add/Update 'scaled' column in var\n",
    "        adata.var['scaled'] = gene_mask\n",
    "        \n",
    "        # Add/Update 'scaled_by' column in var\n",
    "        adata.var['scaled_by'] = scalar  # Default value for genes not in tf_list\n",
    "\n",
    "        # Add/Update 'scaled_by' column in var\n",
    "        adata.obs['U'] = scalar  # Default value for genes not in tf_list\n",
    "    # adata.var = \n",
    "    return adata\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc240769-f15a-4086-9fb6-eacb0323041e",
   "metadata": {},
   "source": [
    "### Saving perturbations (new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6a2df680-d997-4f01-8bd3-0cdefa87ccf2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2765\n",
      "2759\n",
      "['ENSG00000137203']\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Genes ['ENSG00000137203'] not found in anndata object",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[5], line 28\u001b[0m\n\u001b[1;32m     25\u001b[0m     \u001b[38;5;28;01mcontinue\u001b[39;00m\n\u001b[1;32m     27\u001b[0m \u001b[38;5;28mprint\u001b[39m(TFs)\n\u001b[0;32m---> 28\u001b[0m adataDict \u001b[38;5;241m=\u001b[39m iterate_perturb_counts(adata\u001b[38;5;241m.\u001b[39mcopy(), TFs, scalars)\n\u001b[1;32m     30\u001b[0m \u001b[38;5;66;03m# Concatenate all AnnData objects along the observations axis\u001b[39;00m\n\u001b[1;32m     31\u001b[0m concatenated_adata \u001b[38;5;241m=\u001b[39m ad\u001b[38;5;241m.\u001b[39mconcat(\u001b[38;5;28mlist\u001b[39m(adataDict\u001b[38;5;241m.\u001b[39mvalues()), axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m)\n",
      "Cell \u001b[0;32mIn[4], line 29\u001b[0m, in \u001b[0;36miterate_perturb_counts\u001b[0;34m(adata, tf_list, scalar_list)\u001b[0m\n\u001b[1;32m     26\u001b[0m adata_temp \u001b[38;5;241m=\u001b[39m adata\u001b[38;5;241m.\u001b[39mcopy()\n\u001b[1;32m     28\u001b[0m \u001b[38;5;66;03m# Apply perturb_counts to the copied AnnData object\u001b[39;00m\n\u001b[0;32m---> 29\u001b[0m perturbed_adata \u001b[38;5;241m=\u001b[39m perturb_counts(adata_temp, tf_list, scalar)\n\u001b[1;32m     31\u001b[0m \u001b[38;5;66;03m# Store the perturbed AnnData object in the dictionary with scalar as the key\u001b[39;00m\n\u001b[1;32m     32\u001b[0m adata_dict[scalar] \u001b[38;5;241m=\u001b[39m perturbed_adata\n",
      "Cell \u001b[0;32mIn[4], line 77\u001b[0m, in \u001b[0;36mperturb_counts\u001b[0;34m(adata, tf_list, scalar)\u001b[0m\n\u001b[1;32m     75\u001b[0m     adata\u001b[38;5;241m.\u001b[39mX \u001b[38;5;241m=\u001b[39m original_X\n\u001b[1;32m     76\u001b[0m     gene_mask \u001b[38;5;241m=\u001b[39m original_gene_mask\n\u001b[0;32m---> 77\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mGenes \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmissing_genes\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m not found in anndata object\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     79\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:    \n\u001b[1;32m     80\u001b[0m     \n\u001b[1;32m     81\u001b[0m     \u001b[38;5;66;03m# Apply the scaling operation to the specified genes\u001b[39;00m\n\u001b[1;32m     82\u001b[0m     adata\u001b[38;5;241m.\u001b[39mX[:, gene_mask] \u001b[38;5;241m=\u001b[39m max_exp \u001b[38;5;241m*\u001b[39m scalar\n",
      "\u001b[0;31mValueError\u001b[0m: Genes ['ENSG00000137203'] not found in anndata object"
     ]
    }
   ],
   "source": [
    "output_directory = \"/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/all-tfs\"\n",
    "scalars = [0.5, 0.75, 1.001, 1.25, 1.5]\n",
    "print(len(tf_ensid_list))\n",
    "\n",
    "tf_ensid_list = [id for id in tf_ensid_list if id not in problem_ens_ids]\n",
    "print(len(tf_ensid_list))\n",
    "\n",
    "for i in range(len(tf_ensid_list)):\n",
    "    val = df['Ensembl ID'].iloc[i]\n",
    "    # this handles the case where there is more than one tf to perturb in the same file, but in this case there is just the one\n",
    "    TFs = val.split()\n",
    "    # take out all the ensembl ids where there was no match in the couns matrix\n",
    "\n",
    "    if validate_ensembl_tfs(tf_ensid_list, counts_ensid_list):\n",
    "        # Join the TFs list into a string for the filename\n",
    "        TFs_str = \"_\".join(TFs)\n",
    "        \n",
    "        # Generate the file path for saving\n",
    "        file_name = f\"{TFs_str}.h5ad\"\n",
    "        output_path = os.path.join(output_directory, file_name)\n",
    "\n",
    "        # Check if the file already exists\n",
    "        if os.path.exists(output_path):\n",
    "            print(output_path + \" already exists: continue!\")\n",
    "            continue\n",
    "        \n",
    "        print(TFs)\n",
    "        adataDict = iterate_perturb_counts(adata.copy(), TFs, scalars)\n",
    "\n",
    "        # Concatenate all AnnData objects along the observations axis\n",
    "        concatenated_adata = ad.concat(list(adataDict.values()), axis=0)\n",
    "\n",
    "        # This is the line was added today to fix the bug\n",
    "        concatenated_adata.var = adata.var.copy()\n",
    "\n",
    "        # Save reprogramming metadata into the concatenated_adata.obs table\n",
    "        concatenated_adata.obs['Source_cells'] = df['Source cells'].iloc[i]\n",
    "        concatenated_adata.obs['Target_cells'] = df['Target cells'].iloc[i]\n",
    "        concatenated_adata.obs['Treatment'] = df['Treatment'].iloc[i]\n",
    "        concatenated_adata.obs['Species'] = df['Species'].iloc[i]\n",
    "        concatenated_adata.obs['Cell_Transplantation'] = df['Cell Transplantation'].iloc[i]\n",
    "        concatenated_adata.obs['Published_Year'] = df['Published Year'].iloc[i]\n",
    "        concatenated_adata.obs['PMID'] = df['PMID'].iloc[i]\n",
    "        \n",
    "        # Save the concatenated AnnData object to the file\n",
    "        concatenated_adata.write_h5ad(output_path)\n",
    "        print(\"    file created\")\n",
    "    else:\n",
    "        print(\"The file made from \", val ,\" could not be created\")\n",
    "print('All recipes complete!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "7c0b933c-c244-4afe-bc09-1c2ee8307ce5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['DDX11L1', 'WASH7P', 'MIR6859-1', 'MIR1302-2HG', 'MIR1302-2', 'FAM138A',\n",
       "       'OR4G4P', 'OR4G11P', 'OR4F5', 'AL627309.1',\n",
       "       ...\n",
       "       'MT-ND4', 'MT-TH', 'MT-TS2', 'MT-TL2', 'MT-ND5', 'MT-ND6', 'MT-TE',\n",
       "       'MT-CYB', 'MT-TT', 'MT-TP'],\n",
       "      dtype='object', length=58870)"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if TFs are in adata\n",
    "# missing_genes = [tf for tf in TFs if tf not in adata.var.index]\n",
    "# if missing_genes:\n",
    "#     print(f\"Missing genes in AnnData object: {missing_genes}\")\n",
    "\n",
    "adata.var.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "34c80541-cc8f-4536-8e52-f660945d8b53",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(len(tf_ensid_list)):\n",
    "    print(f\"Processing index {i}\")\n",
    "    val = df['Ensembl ID'].iloc[i]\n",
    "    TFs = val.split()\n",
    "    \n",
    "    print(f\"TFs: {TFs}\")\n",
    "    \n",
    "    tf_ensid_list_filtered = [id for id in problem_ens_ids if id not in tf_ensid_list]\n",
    "    \n",
    "    if not validate_ensembl_tfs(counts_ensid_list, tf_ensid_list_filtered):\n",
    "        print(\"Validation failed\")\n",
    "        continue\n",
    "    \n",
    "    TFs_str = \"_\".join(TFs)\n",
    "    file_name = f\"{TFs_str}.h5ad\"\n",
    "    output_path = os.path.join(output_directory, file_name)\n",
    "    \n",
    "    if os.path.exists(output_path):\n",
    "        print(f\"{output_path} already exists\")\n",
    "        continue\n",
    "    \n",
    "    print(f\"Creating file: {output_path}\")\n",
    "    \n",
    "    adataDict = iterate_perturb_counts(adata.copy(), TFs, scalars)\n",
    "    concatenated_adata = ad.concat(list(adataDict.values()), axis=0)\n",
    "    concatenated_adata.var = adata.var.copy()\n",
    "    \n",
    "    concatenated_adata.obs['Source_cells'] = df['Source cells'].iloc[i]\n",
    "    concatenated_adata.obs['Target_cells'] = df['Target cells'].iloc[i]\n",
    "    concatenated_adata.obs['Treatment'] = df['Treatment'].iloc[i]\n",
    "    concatenated_adata.obs['Species'] = df['Species'].iloc[i]\n",
    "    concatenated_adata.obs['Cell_Transplantation'] = df['Cell Transplantation'].iloc[i]\n",
    "    concatenated_adata.obs['Published_Year'] = df['Published Year'].iloc[i]\n",
    "    concatenated_adata.obs['PMID'] = df['PMID'].iloc[i]\n",
    "    \n",
    "    concatenated_adata.write_h5ad(output_path)\n",
    "    print(f\"File created: {output_path}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f0fe09e-bee4-4666-9176-9282a8418120",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2570800-7f27-4621-910d-568cafd6ddbb",
   "metadata": {},
   "source": [
    "# Day 9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0065148-4b7e-4a1d-98ad-b4f429e2029b",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" Generating the perturbation files for all of the Human TFs (perturbation of 1, individually). \n",
    "Day 8 is designed for the paper spreadsheet, but some of the TFs column is still not formatted properly (not fixing this today)\n",
    "\"\"\"\n",
    "\"\"\" Important: Some of the same variable names are being used in the all tf experiment as in the paper experiment: both in the same situations, e.g. TFs is the TFs, but different ones between experiments\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abba6f88-8b01-430c-99c8-d6d00bd7f083",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6bd3a9ca-d3a0-4351-a538-e64e270cbb0e",
   "metadata": {},
   "source": [
    "## Set up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "ad760306-773a-4bed-b655-8a8d2ac9dce1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "58870\n",
      "2765\n",
      "2765\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import scanpy as sp\n",
    "import os\n",
    "\n",
    "def iterate_perturb_counts(adata, tf_list, scalar_list):\n",
    "    \"\"\"\n",
    "    Applies perturbations to the expression data of specified transcription factors across multiple scalars \n",
    "    and stores the resulting AnnData objects in a dictionary.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Iterates over a list of scalar values.\n",
    "    2. For each scalar, creates a copy of the AnnData object to preserve the original data.\n",
    "    3. Applies the `perturb_counts` function to scale the expression data of genes listed in `tf_list` by\n",
    "       the maximum gene expression of each cell and the current scalar.\n",
    "    4. Stores the perturbed AnnData object in a dictionary with the scalar as the key.\n",
    "\n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols (transcription factors) to be perturbed.\n",
    "    scalar_list (list): A list of scalar values for scaling the gene expression.\n",
    "    adata (AnnData): The AnnData object containing gene expression data (cells x genes).\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary where keys are scalar values and values are the corresponding perturbed AnnData objects.\n",
    "    \"\"\"\n",
    "    \n",
    "    adata_dict = {}\n",
    "    \n",
    "    for scalar in scalar_list:\n",
    "        # Create a copy of the AnnData object for each scalar value\n",
    "        adata_temp = adata.copy()\n",
    "        \n",
    "        # Apply perturb_counts to the copied AnnData object\n",
    "        perturbed_adata = perturb_counts(adata_temp, tf_list, scalar)\n",
    "        \n",
    "        # Store the perturbed AnnData object in the dictionary with scalar as the key\n",
    "        adata_dict[scalar] = perturbed_adata\n",
    "    \n",
    "    return adata_dict\n",
    "\n",
    "def perturb_counts(adata, tf_list, scalar): \n",
    "    \"\"\"\n",
    "    Applies a perturbation to the expression data of specific genes in an AnnData object.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Computes the maximum gene expression level for each cell.\n",
    "    2. Applies a scaling operation to the expression levels of genes listed in `tf_list`.\n",
    "       - Each entry of these genes in the matrix is multiplied by the maximum expression level \n",
    "         of its respective cell and a specified scalar value.\n",
    "    3. Updates the AnnData object with new columns:\n",
    "       - 'scaled': A boolean column indicating whether each gene is in the `tf_list`.\n",
    "       - 'scaled_by': Contains the scaling factor used for each gene (the product of the maximum \n",
    "         expression level of each cell and the scalar), or `1` if the gene was not in `tf_list`.\n",
    "    \n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols to be perturbed.\n",
    "    scalar (float): The scalar value used to scale the expression levels.\n",
    "    adata (AnnData): The AnnData object containing gene expression data.\n",
    "\n",
    "    Returns:\n",
    "    AnnData: The updated AnnData object with applied perturbations and new columns.\n",
    "    \"\"\"\n",
    "\n",
    "    # Create a boolean mask for genes in tf_list\n",
    "    gene_mask = adata.var['gene_symbol'].isin(tf_list)\n",
    "    \n",
    "    # Save the original state of the parameter objects, in case some tfs do not translate (failsafe)\n",
    "    original_X = adata.X.copy()\n",
    "    original_gene_mask = gene_mask.copy()\n",
    "    \n",
    "    # Compute maximum expression level of each cell\n",
    "    max_exp = np.max(adata.X, axis=1)\n",
    "\n",
    "    \"\"\"This is new today. v \"\"\"\n",
    "    # Raise an error if any of the gene names in tf_list do not match column names (we will manually update these in adata):\n",
    "    missing_genes = [gene for gene in tf_list if gene not in adata.var['gene_symbol'].values]\n",
    "    \n",
    "    if missing_genes:\n",
    "        # Restore original parameter objects\n",
    "        adata.X = original_X\n",
    "        gene_mask = original_gene_mask\n",
    "        raise ValueError(f\"Genes {missing_genes} not found in anndata object\")\n",
    "\n",
    "    else:    \n",
    "        \n",
    "        # Apply the scaling operation to the specified genes\n",
    "        adata.X[:, gene_mask] = max_exp * scalar\n",
    "        \n",
    "        # Add/Update 'scaled' column in var\n",
    "        adata.var['scaled'] = gene_mask\n",
    "        \n",
    "        # Add/Update 'scaled_by' column in var\n",
    "        adata.var['scaled_by'] = scalar  # Default value for genes not in tf_list\n",
    "\n",
    "        # Add/Update 'scaled_by' column in var\n",
    "        adata.obs['U'] = scalar  # Default value for genes not in tf_list\n",
    "    # adata.var = \n",
    "    return adata\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "9af4fba0-d09a-4989-b40d-e05d4a0533c7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ZNF73_HUMAN\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "def translate_ids(list):\n",
    "    # Extract base IDs from list_b by removing version numbers\n",
    "    new_list = [id.split('.')[0] for id in list]\n",
    "    return new_list\n",
    "\n",
    "#ensembl ids from the counts matrix\n",
    "counts_ensid_list = adata.var['ensemblid'].values.tolist()\n",
    "counts_ensid_list = translate_ids(adata_id_list)\n",
    "\n",
    "\n",
    "# ensembl ids from the transcription factor list, which in this case is also the perturbation list (testing one at a time)\n",
    "tf_ensid_list = df['Ensembl ID'].values.tolist()\n",
    "\n",
    "\n",
    "def validate_ensembl_tfs(tf_ensid_list, counts_ensid_list):\n",
    "    for TF in ensembl_TFs:\n",
    "        if TF not in counts_ensid_list:\n",
    "            print(TF)\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "validate_ensembl_tfs(tf_ensid_list, counts_ensid_list)\n",
    "\n",
    "# when we perturb each tf in tf_ensid_list, we are just going to skip the one problematic one\n",
    "problem_ens_ids = ['ZNF73_HUMAN']\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e0944045-cfb8-4925-b809-54c45a530223",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Ensembl ID</th>\n",
       "      <th>HGNC symbol</th>\n",
       "      <th>DBD</th>\n",
       "      <th>Is TF?</th>\n",
       "      <th>TF assessment</th>\n",
       "      <th>Binding mode</th>\n",
       "      <th>Motif status</th>\n",
       "      <th>Final Notes</th>\n",
       "      <th>Final Comments</th>\n",
       "      <th>...</th>\n",
       "      <th>CisBP considers it a TF?</th>\n",
       "      <th>TFCat classification</th>\n",
       "      <th>Is a GO TF?</th>\n",
       "      <th>Initial assessment</th>\n",
       "      <th>Curator 1</th>\n",
       "      <th>Curator 2</th>\n",
       "      <th>TFclass considers it a TF?</th>\n",
       "      <th>Go Evidence</th>\n",
       "      <th>Pfam Domains (By ENSP ID)</th>\n",
       "      <th>Is C2H2 ZF(KRAB)?</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000137203</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>AP-2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Known motif</td>\n",
       "      <td>Monomer or homomultimer</td>\n",
       "      <td>High-throughput in vitro</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Yes</td>\n",
       "      <td>TF Gene_DNA-Binding: sequence-specific_DNA Bin...</td>\n",
       "      <td>Yes</td>\n",
       "      <td>1a1, Direct HQ evidence</td>\n",
       "      <td>Sam Lambert</td>\n",
       "      <td>Yimeng Yin</td>\n",
       "      <td>Yes</td>\n",
       "      <td>$#ENSG00000137203#GO:0000981#sequence-specific...</td>\n",
       "      <td>$#ENSP00000368928#ENSG00000137203#ENST00000379...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000008196</td>\n",
       "      <td>TFAP2B</td>\n",
       "      <td>AP-2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Known motif</td>\n",
       "      <td>Monomer or homomultimer</td>\n",
       "      <td>High-throughput in vitro</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Yes</td>\n",
       "      <td>TF Gene_DNA-Binding: sequence-specific_DNA Bin...</td>\n",
       "      <td>Yes</td>\n",
       "      <td>1a1, Direct HQ evidence</td>\n",
       "      <td>Matt Weirauch</td>\n",
       "      <td>Yimeng Yin</td>\n",
       "      <td>Yes</td>\n",
       "      <td>$#ENSG00000008196#GO:0000981#sequence-specific...</td>\n",
       "      <td>$#ENSP00000377265#ENSG00000008196#ENST00000393...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000087510</td>\n",
       "      <td>TFAP2C</td>\n",
       "      <td>AP-2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Known motif</td>\n",
       "      <td>Monomer or homomultimer</td>\n",
       "      <td>High-throughput in vitro</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>1a1, Direct HQ evidence</td>\n",
       "      <td>Matt Weirauch</td>\n",
       "      <td>Yimeng Yin</td>\n",
       "      <td>Yes</td>\n",
       "      <td>$#ENSG00000087510#GO:0001077#RNA polymerase II...</td>\n",
       "      <td>$#ENSP00000201031#ENSG00000087510#ENST00000201...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000008197</td>\n",
       "      <td>TFAP2D</td>\n",
       "      <td>AP-2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Known motif</td>\n",
       "      <td>Monomer or homomultimer</td>\n",
       "      <td>In vivo/Misc source</td>\n",
       "      <td>Only known motifs are from Transfac or HocoMoc...</td>\n",
       "      <td>Binds the same GCCTGAGGC sequence as the other...</td>\n",
       "      <td>...</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2a1, Lower confidence direct evidence</td>\n",
       "      <td>Arttu Jolma</td>\n",
       "      <td>Sam Lambert</td>\n",
       "      <td>Yes</td>\n",
       "      <td>$#ENSG00000008197#GO:0000981#sequence-specific...</td>\n",
       "      <td>$#ENSP00000008391#ENSG00000008197#ENST00000008...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000116819</td>\n",
       "      <td>TFAP2E</td>\n",
       "      <td>AP-2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Known motif</td>\n",
       "      <td>Monomer or homomultimer</td>\n",
       "      <td>High-throughput in vitro</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Yes</td>\n",
       "      <td>TF Gene_DNA-Binding: sequence-specific_DNA Bin...</td>\n",
       "      <td>Yes</td>\n",
       "      <td>1a1, Direct HQ evidence</td>\n",
       "      <td>Sam Lambert</td>\n",
       "      <td>Laura Campitelli</td>\n",
       "      <td>Yes</td>\n",
       "      <td>$#ENSG00000116819#GO:0000981#sequence-specific...</td>\n",
       "      <td>$#ENSP00000362332#ENSG00000116819#ENST00000373...</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0       Ensembl ID HGNC symbol   DBD Is TF? TF assessment  \\\n",
       "0           0  ENSG00000137203      TFAP2A  AP-2    Yes   Known motif   \n",
       "1           1  ENSG00000008196      TFAP2B  AP-2    Yes   Known motif   \n",
       "2           2  ENSG00000087510      TFAP2C  AP-2    Yes   Known motif   \n",
       "3           3  ENSG00000008197      TFAP2D  AP-2    Yes   Known motif   \n",
       "4           4  ENSG00000116819      TFAP2E  AP-2    Yes   Known motif   \n",
       "\n",
       "              Binding mode              Motif status  \\\n",
       "0  Monomer or homomultimer  High-throughput in vitro   \n",
       "1  Monomer or homomultimer  High-throughput in vitro   \n",
       "2  Monomer or homomultimer  High-throughput in vitro   \n",
       "3  Monomer or homomultimer       In vivo/Misc source   \n",
       "4  Monomer or homomultimer  High-throughput in vitro   \n",
       "\n",
       "                                         Final Notes  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3  Only known motifs are from Transfac or HocoMoc...   \n",
       "4                                                NaN   \n",
       "\n",
       "                                      Final Comments  ...  \\\n",
       "0                                                NaN  ...   \n",
       "1                                                NaN  ...   \n",
       "2                                                NaN  ...   \n",
       "3  Binds the same GCCTGAGGC sequence as the other...  ...   \n",
       "4                                                NaN  ...   \n",
       "\n",
       "  CisBP considers it a TF?                               TFCat classification  \\\n",
       "0                      Yes  TF Gene_DNA-Binding: sequence-specific_DNA Bin...   \n",
       "1                      Yes  TF Gene_DNA-Binding: sequence-specific_DNA Bin...   \n",
       "2                      Yes                                                 No   \n",
       "3                      Yes                                                 No   \n",
       "4                      Yes  TF Gene_DNA-Binding: sequence-specific_DNA Bin...   \n",
       "\n",
       "  Is a GO TF?                    Initial assessment       Curator 1  \\\n",
       "0         Yes                1a1, Direct HQ evidence    Sam Lambert   \n",
       "1         Yes                1a1, Direct HQ evidence  Matt Weirauch   \n",
       "2         Yes                1a1, Direct HQ evidence  Matt Weirauch   \n",
       "3         Yes  2a1, Lower confidence direct evidence    Arttu Jolma   \n",
       "4         Yes                1a1, Direct HQ evidence    Sam Lambert   \n",
       "\n",
       "          Curator 2 TFclass considers it a TF?   \\\n",
       "0        Yimeng Yin                         Yes   \n",
       "1        Yimeng Yin                         Yes   \n",
       "2        Yimeng Yin                         Yes   \n",
       "3       Sam Lambert                         Yes   \n",
       "4  Laura Campitelli                         Yes   \n",
       "\n",
       "                                         Go Evidence  \\\n",
       "0  $#ENSG00000137203#GO:0000981#sequence-specific...   \n",
       "1  $#ENSG00000008196#GO:0000981#sequence-specific...   \n",
       "2  $#ENSG00000087510#GO:0001077#RNA polymerase II...   \n",
       "3  $#ENSG00000008197#GO:0000981#sequence-specific...   \n",
       "4  $#ENSG00000116819#GO:0000981#sequence-specific...   \n",
       "\n",
       "                           Pfam Domains (By ENSP ID) Is C2H2 ZF(KRAB)?  \n",
       "0  $#ENSP00000368928#ENSG00000137203#ENST00000379...             False  \n",
       "1  $#ENSP00000377265#ENSG00000008196#ENST00000393...             False  \n",
       "2  $#ENSP00000201031#ENSG00000087510#ENST00000201...             False  \n",
       "3  $#ENSP00000008391#ENSG00000008197#ENST00000008...             False  \n",
       "4  $#ENSP00000362332#ENSG00000116819#ENST00000373...             False  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c4d81d2-8c5e-403a-9f01-fb6cc8d32626",
   "metadata": {},
   "source": [
    "## Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8bff13f1-e4e8-42dc-8a8d-ff0eb7a97b34",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>ensemblid</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>False</td>\n",
       "      <td>6.398244e-05</td>\n",
       "      <td>0.835044</td>\n",
       "      <td>-0.573947</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.005574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>False</td>\n",
       "      <td>2.274395e-03</td>\n",
       "      <td>2.442280</td>\n",
       "      <td>0.533203</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.031731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR6859-1</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>False</td>\n",
       "      <td>6.175251e-05</td>\n",
       "      <td>1.295335</td>\n",
       "      <td>-0.256874</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.005634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>False</td>\n",
       "      <td>1.372886e-04</td>\n",
       "      <td>2.656352</td>\n",
       "      <td>0.680668</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.008041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_symbol     feature_type          ensemblid  highly_variable  \\\n",
       "DDX11L1          DDX11L1  Gene Expression  ENSG00000223972.5            False   \n",
       "WASH7P            WASH7P  Gene Expression  ENSG00000227232.5            False   \n",
       "MIR6859-1      MIR6859-1  Gene Expression  ENSG00000278267.1            False   \n",
       "MIR1302-2HG  MIR1302-2HG  Gene Expression  ENSG00000243485.5            False   \n",
       "MIR1302-2      MIR1302-2  Gene Expression  ENSG00000284332.1            False   \n",
       "\n",
       "                    means  dispersions  dispersions_norm      mean       std  \n",
       "DDX11L1      6.398244e-05     0.835044         -0.573947  0.000039  0.005574  \n",
       "WASH7P       2.274395e-03     2.442280          0.533203  0.001080  0.031731  \n",
       "MIR6859-1    6.175251e-05     1.295335         -0.256874  0.000033  0.005634  \n",
       "MIR1302-2HG  1.372886e-04     2.656352          0.680668  0.000048  0.008041  \n",
       "MIR1302-2    1.000000e-12          NaN          0.000000  0.000000  1.000000  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load reprogramming recipes\n",
    "df = pd.read_csv('https://raw.githubusercontent.com/Jpickard1/scFoundationModels/main/notebooks/reprogramming/data/HumanTFs_v_1.01.csv')\n",
    "df.head()\n",
    "\n",
    "# Load firboblast source cells\n",
    "DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/unperturbed\"\n",
    "FILE = \"fibroblast.h5ad\"\n",
    "adata = sp.read_h5ad(os.path.join(DATAPATH, FILE))\n",
    "adata_gene_list = adata.var['gene_symbol'].values.tolist()\n",
    "\n",
    "adata.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adff81e1-ddb8-47a2-bf7b-665329fbec9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_directory = \"/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed\"\n",
    "scalars = [0.5, 0.75, 1.001]\n",
    "for i in range(len(df['TFs'])):\n",
    "    val = df['TFs'].iloc[i]\n",
    "    TFs = val.split()\n",
    "    if validateTFs(TFs, adata):\n",
    "\n",
    "        # Join the TFs list into a string for the filename\n",
    "        TFs_str = \"_\".join(TFs)\n",
    "        \n",
    "        # Generate the file path for saving\n",
    "        file_name = f\"{TFs_str}.h5ad\"\n",
    "        output_path = os.path.join(output_directory, file_name)\n",
    "\n",
    "        # Check if the file already exists\n",
    "        if os.path.exists(output_path):\n",
    "            print(output_path + \" already exist: continue!\")\n",
    "            continue\n",
    "        \n",
    "        print(TFs)\n",
    "        adataDict = iterate_perturb_counts(adata.copy(), TFs, scalars)\n",
    "\n",
    "        # Concatenate all AnnData objects along the observations axis\n",
    "        concatenated_adata = ad.concat(list(adataDict.values()), axis=0)\n",
    "\n",
    "        # This is the line was added today to fix the bug\n",
    "        concatenated_adata.var = adata.var.copy()\n",
    "\n",
    "        # Save reprogramming metadata into the concatenated_adata.obs table\n",
    "        concatenated_adata.obs['Source_cells'] = df['Source cells'].iloc[i]\n",
    "        concatenated_adata.obs['Target_cells'] = df['Target cells'].iloc[i]\n",
    "        concatenated_adata.obs['Treatment'] = df['Treatment'].iloc[i]\n",
    "        concatenated_adata.obs['Species'] = df['Species'].iloc[i]\n",
    "        concatenated_adata.obs['Cell_Transplantation'] = df['Cell Transplantation'].iloc[i]\n",
    "        concatenated_adata.obs['Published_Year'] = df['Published Year'].iloc[i]\n",
    "        concatenated_adata.obs['PMID'] = df['PMID'].iloc[i]\n",
    "        \n",
    "        # Save the concatenated AnnData object to the file\n",
    "        concatenated_adata.write_h5ad(output_path)\n",
    "        print(\"    file created\")\n",
    "\n",
    "print('All recipes complete!')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4506c304-f7a3-453e-9f2f-e5efd8bd0395",
   "metadata": {},
   "source": [
    "## Trash"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "febad736-4b90-4e8b-a2e7-dc601e528a49",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# def validateTFs(TFs, adata):\n",
    "#     adata_gene_list = adata.var['gene_symbol'].values.tolist()\n",
    "#     for TF in TFs:\n",
    "#         if TF not in adata_gene_list:\n",
    "#             return False\n",
    "#     return True\n",
    "\n",
    "# # # to contrast the previous experiment, here genes are named by ensembl id rather than one of their names\n",
    "# ensembl_TFs = df['Ensembl ID']. values.tolist()\n",
    "# # def validate_ensembl_tfs(ensembl_TFs, adata):\n",
    "# #     adata_id_list = adata.var['ensemblid'].values.tolist()\n",
    "# #     for TF in ensembl_TFs:\n",
    "# #         if TF not in adata_id_list:\n",
    "# #             return False\n",
    "# #     return True\n",
    "\n",
    "# # validate_ensembl_tfs(ensembl_TFs, adata\n",
    "\n",
    "\n",
    "\n",
    "# def validate_ensembl_tfs(ensembl_TFs, new_adata_id_list):\n",
    "#     count = 0\n",
    "#     # Extract Ensembl IDs from the adata object\n",
    "\n",
    "#     # Initialize a flag to check if all IDs are found\n",
    "#     all_found = True\n",
    "#     # Iterate through the list of Ensembl IDs to validate\n",
    "#     for TF in ensembl_TFs:\n",
    "#         if TF not in adata_id_list:\n",
    "#             # Print the missing Ensembl ID\n",
    "#             #print(f\"{TF} not found in adata.var['ensemblid']\")\n",
    "#             count += 1\n",
    "#             # Set the flag to False if an ID is missing\n",
    "#             all_found = False\n",
    "#     # Return the final status of the validation\n",
    "#     print(len(adata_id_list))\n",
    "#     print(len(ensembl_TFs))\n",
    "#     print(count)\n",
    "#     return all_found\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a8f1079-66fb-423f-9f56-e46d81c04d2a",
   "metadata": {},
   "source": [
    "# Day 8\n",
    "\n",
    "Generatting the perturbation files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6ff35e5-f45f-416f-95ca-313bd921c074",
   "metadata": {},
   "source": [
    "## Set up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2d6cd062-e13d-40a5-b679-1db068d463bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import scanpy as sp\n",
    "import os\n",
    "\n",
    "def iterate_perturb_counts(adata, tf_list, scalar_list):\n",
    "    \"\"\"\n",
    "    Applies perturbations to the expression data of specified transcription factors across multiple scalars \n",
    "    and stores the resulting AnnData objects in a dictionary.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Iterates over a list of scalar values.\n",
    "    2. For each scalar, creates a copy of the AnnData object to preserve the original data.\n",
    "    3. Applies the `perturb_counts` function to scale the expression data of genes listed in `tf_list` by\n",
    "       the maximum gene expression of each cell and the current scalar.\n",
    "    4. Stores the perturbed AnnData object in a dictionary with the scalar as the key.\n",
    "\n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols (transcription factors) to be perturbed.\n",
    "    scalar_list (list): A list of scalar values for scaling the gene expression.\n",
    "    adata (AnnData): The AnnData object containing gene expression data (cells x genes).\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary where keys are scalar values and values are the corresponding perturbed AnnData objects.\n",
    "    \"\"\"\n",
    "    \n",
    "    adata_dict = {}\n",
    "    \n",
    "    for scalar in scalar_list:\n",
    "        # Create a copy of the AnnData object for each scalar value\n",
    "        adata_temp = adata.copy()\n",
    "        \n",
    "        # Apply perturb_counts to the copied AnnData object\n",
    "        perturbed_adata = perturb_counts(adata_temp, tf_list, scalar)\n",
    "        \n",
    "        # Store the perturbed AnnData object in the dictionary with scalar as the key\n",
    "        adata_dict[scalar] = perturbed_adata\n",
    "    \n",
    "    return adata_dict\n",
    "\n",
    "def perturb_counts(adata, tf_list, scalar): \n",
    "    \"\"\"\n",
    "    Applies a perturbation to the expression data of specific genes in an AnnData object.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Computes the maximum gene expression level for each cell.\n",
    "    2. Applies a scaling operation to the expression levels of genes listed in `tf_list`.\n",
    "       - Each entry of these genes in the matrix is multiplied by the maximum expression level \n",
    "         of its respective cell and a specified scalar value.\n",
    "    3. Updates the AnnData object with new columns:\n",
    "       - 'scaled': A boolean column indicating whether each gene is in the `tf_list`.\n",
    "       - 'scaled_by': Contains the scaling factor used for each gene (the product of the maximum \n",
    "         expression level of each cell and the scalar), or `1` if the gene was not in `tf_list`.\n",
    "    \n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols to be perturbed.\n",
    "    scalar (float): The scalar value used to scale the expression levels.\n",
    "    adata (AnnData): The AnnData object containing gene expression data.\n",
    "\n",
    "    Returns:\n",
    "    AnnData: The updated AnnData object with applied perturbations and new columns.\n",
    "    \"\"\"\n",
    "\n",
    "    # Create a boolean mask for genes in tf_list\n",
    "    gene_mask = adata.var['gene_symbol'].isin(tf_list)\n",
    "    \n",
    "    # Save the original state of the parameter objects, in case some tfs do not translate (failsafe)\n",
    "    original_X = adata.X.copy()\n",
    "    original_gene_mask = gene_mask.copy()\n",
    "    \n",
    "    # Compute maximum expression level of each cell\n",
    "    max_exp = np.max(adata.X, axis=1)\n",
    "\n",
    "    \"\"\"This is new today. v \"\"\"\n",
    "    # Raise an error if any of the gene names in tf_list do not match column names (we will manually update these in adata):\n",
    "    missing_genes = [gene for gene in tf_list if gene not in adata.var['gene_symbol'].values]\n",
    "    \n",
    "    if missing_genes:\n",
    "        # Restore original parameter objects\n",
    "        adata.X = original_X\n",
    "        gene_mask = original_gene_mask\n",
    "        raise ValueError(f\"Genes {missing_genes} not found in anndata object\")\n",
    "\n",
    "    else:    \n",
    "        \n",
    "        # Apply the scaling operation to the specified genes\n",
    "        adata.X[:, gene_mask] = max_exp * scalar\n",
    "        \n",
    "        # Add/Update 'scaled' column in var\n",
    "        adata.var['scaled'] = gene_mask\n",
    "        \n",
    "        # Add/Update 'scaled_by' column in var\n",
    "        adata.var['scaled_by'] = scalar  # Default value for genes not in tf_list\n",
    "\n",
    "        # Add/Update 'scaled_by' column in var\n",
    "        adata.obs['U'] = scalar  # Default value for genes not in tf_list\n",
    "    # adata.var = \n",
    "    return adata\n",
    "\n",
    "def validateTFs(TFs, adata):\n",
    "    adata_gene_list = adata.var['gene_symbol'].values.tolist()\n",
    "    for TF in TFs:\n",
    "        if TF not in adata_gene_list:\n",
    "            return False\n",
    "    return True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "459f8409-257e-4c03-b216-62337293499c",
   "metadata": {},
   "source": [
    "## Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6f639abb-5181-4d2b-89bc-e070a74af38d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load reprogramming recipes\n",
    "df = pd.read_csv('data/recipe_table_9_6_2024.csv')\n",
    "df.head()\n",
    "\n",
    "# Load firboblast source cells\n",
    "DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/unperturbed\"\n",
    "FILE = \"fibroblast.h5ad\"\n",
    "adata = sp.read_h5ad(os.path.join(DATAPATH, FILE))\n",
    "adata_gene_list = adata.var['gene_symbol'].values.tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "844415e8-df65-428b-8d18-5601f57cd263",
   "metadata": {},
   "source": [
    "## Make Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5d4c1a67-0ac0-4e94-b059-16eca9958c0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/ASCL1_POU3F2_MYT1L_LMX1A_FOXA2.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/SOX2.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/NEUROG2_SOX11_ISL1_LHX3.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/SOX2_HMGA2.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/ASCL1_PAX6.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/SOX2_PAX6.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/PTF1A.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/SOX2_ASCL1_NEUROG2.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/ASCL1_SOX2.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/FOXM1_SOX2_MYC_SALL4_STAT6.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/AR_SOX2_SMAD3_MYC_JUN_WT1_TAL1_SPI1_RUNX1.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/POU3F2_SOX2_BACH1_AR_PBX1_NANOG.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/MSI1_NEUROG2_MBD2.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/GATA4_MEF2C_TBX5_MESP1_MYOCD.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/GATA4_MEF2C_TBX5_ESRRG_MESP1_ZFPM2_MYOCD.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/ATF5_PROX1_FOXA2_FOXA3_HNF4A.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/PDX1_PAX4_MAFA.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/NFE2_MAFG_MAFK.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/SPI1_IRF8_BATF3.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/MYC_KLF4_SOX9.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/MYOD1_MYCL.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/PRDM16_CEBPB.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/CRX_RAX_NEUROD1.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/MITF_SOX10_PAX3.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/SOX18.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/CRX_RAX_OTX2_NEUROD1.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/NFKB1_LEF1.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/EMX2_HNF1B_HNF4A_PAX8.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/MYOCD.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/ETV2.h5ad already exist: continue!\n",
      "/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed/PPARG.h5ad already exist: continue!\n",
      "['HNF4A', 'FOXA3', 'GATA6', 'CDX2']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    file created\n",
      "['MYOCD', 'GATA6', 'MEF2C']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    file created\n",
      "['SNAI2', 'EYA1', 'SIX1']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    file created\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'float' object has no attribute 'split'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 5\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;28mlen\u001b[39m(df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTFs\u001b[39m\u001b[38;5;124m'\u001b[39m])):\n\u001b[1;32m      4\u001b[0m     val \u001b[38;5;241m=\u001b[39m df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTFs\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39miloc[i]\n\u001b[0;32m----> 5\u001b[0m     TFs \u001b[38;5;241m=\u001b[39m \u001b[43mval\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msplit\u001b[49m()\n\u001b[1;32m      6\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m validateTFs(TFs, adata):\n\u001b[1;32m      7\u001b[0m \n\u001b[1;32m      8\u001b[0m         \u001b[38;5;66;03m# Join the TFs list into a string for the filename\u001b[39;00m\n\u001b[1;32m      9\u001b[0m         TFs_str \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(TFs)\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'float' object has no attribute 'split'"
     ]
    }
   ],
   "source": [
    "output_directory = \"/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed\"\n",
    "scalars = [0.5, 0.75, 1.001]\n",
    "for i in range(len(df['TFs'])):\n",
    "    val = df['TFs'].iloc[i]\n",
    "    TFs = val.split()\n",
    "    if validateTFs(TFs, adata):\n",
    "\n",
    "        # Join the TFs list into a string for the filename\n",
    "        TFs_str = \"_\".join(TFs)\n",
    "        \n",
    "        # Generate the file path for saving\n",
    "        file_name = f\"{TFs_str}.h5ad\"\n",
    "        output_path = os.path.join(output_directory, file_name)\n",
    "\n",
    "        # Check if the file already exists\n",
    "        if os.path.exists(output_path):\n",
    "            print(output_path + \" already exist: continue!\")\n",
    "            continue\n",
    "        \n",
    "        print(TFs)\n",
    "        adataDict = iterate_perturb_counts(adata.copy(), TFs, scalars)\n",
    "\n",
    "        # Concatenate all AnnData objects along the observations axis\n",
    "        concatenated_adata = ad.concat(list(adataDict.values()), axis=0)\n",
    "\n",
    "        # This is the line was added today to fix the bug\n",
    "        concatenated_adata.var = adata.var.copy()\n",
    "\n",
    "        # Save reprogramming metadata into the concatenated_adata.obs table\n",
    "        concatenated_adata.obs['Source_cells'] = df['Source cells'].iloc[i]\n",
    "        concatenated_adata.obs['Target_cells'] = df['Target cells'].iloc[i]\n",
    "        concatenated_adata.obs['Treatment'] = df['Treatment'].iloc[i]\n",
    "        concatenated_adata.obs['Species'] = df['Species'].iloc[i]\n",
    "        concatenated_adata.obs['Cell_Transplantation'] = df['Cell Transplantation'].iloc[i]\n",
    "        concatenated_adata.obs['Published_Year'] = df['Published Year'].iloc[i]\n",
    "        concatenated_adata.obs['PMID'] = df['PMID'].iloc[i]\n",
    "        \n",
    "        # Save the concatenated AnnData object to the file\n",
    "        concatenated_adata.write_h5ad(output_path)\n",
    "        print(\"    file created\")\n",
    "\n",
    "print('All recipes complete!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b4bf4e3f-0f83-40d8-9a98-5c8e33a02d80",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                                     ASCL1 LXMLA NR4A2\n",
       "1                           POU3F2 ASCL1 MYT1L NEUROD11\n",
       "2                        ASCL1 POU3F2 MYT1L LMX1A FOXA2\n",
       "3                                                  SOX2\n",
       "4                    ASCL1 NEUROG2 SOX2 NR4A2 PITX3 P53\n",
       "5                               NEUROG2 SOX11 ISL1 LHX3\n",
       "6                                            SOX2 HMGA2\n",
       "7                                            ASCL1 PAX6\n",
       "8                           POU3F2 ASCL1 MYT1L NEUROD11\n",
       "9     ASCL1 ISL1 NEUROD11 POU3F2 MNX1 LHX3 MYT1L NEU...\n",
       "10                                  SOX2 GATA3 NEUROD11\n",
       "11                                            SOX2 PAX6\n",
       "12                                                PTF1A\n",
       "13        SOX2; SOX2 PAX6; SOX2 LMX1A; SOX2 LMX1A FOXA2\n",
       "14                                   SOX2 ASCL1 NEUROG2\n",
       "15                                           ASCL1 SOX2\n",
       "16                           FOXM1 SOX2 MYC SALL4 STAT6\n",
       "17            AR SOX2 SMAD3 MYC JUN WT1 TAL1 SPI1 RUNX1\n",
       "18                      POU3F2 SOX2 BACH1 AR PBX1 NANOG\n",
       "19     ASCL1  PHOX2B  AP-2A  GATA3  HAND2  NR4A2 PHOX2A\n",
       "20                                    MSI1 NEUROG2 MBD2\n",
       "21                         GATA4 MEF2C TBX5 MESP1 MYOCD\n",
       "22             GATA4 MEF2C TBX5 ESRRG MESP1 ZFPM2 MYOCD\n",
       "23     HNF1A HNF4A ONECUT1 ATF5 PROX1 CEBPA; MYC + TP53\n",
       "24            FOXA3 HNF1A  HNF4A + SV40 Large T Antigen\n",
       "25    HNF1A FOXA1 FOXA3; HNF1A FOXA1 HNF4A; HNF1A FO...\n",
       "26                         ATF5 PROX1 FOXA2 FOXA3 HNF4A\n",
       "27                               OCT4 FOXA2 HNF1A GATA3\n",
       "28                                       PDX1 PAX4 MAFA\n",
       "29                                       NFE2 MAFG MAFK\n",
       "30                                      SPI1 IRF8 BATF3\n",
       "31                                        MYC KLF4 SOX9\n",
       "32    POU5F1 OCT6; POU3F4 MYCL; POU3F4 MYC; POU3F4 MYCN\n",
       "33                            RUNX2  OCT4  OSTERIX MYCL\n",
       "34                                   OCT4  OSTERIX MYCL\n",
       "35                                           MYOD1 MYCL\n",
       "36                                         PRDM16 CEBPB\n",
       "37                                     CRX  RAX NEUROD1\n",
       "38                                     MITF  SOX10 PAX3\n",
       "39                                                SOX18\n",
       "40                               CRX  RAX  OTX2 NEUROD1\n",
       "41                                           NFKB1 LEF1\n",
       "42                                                 OCT4\n",
       "43                              EMX2  HNF1B  HNF4A PAX8\n",
       "44                                                MYOCD\n",
       "45                                                 ETV2\n",
       "46                                                PPARG\n",
       "47                             HNF4A  FOXA3  GATA6 CDX2\n",
       "48                                   MYOCD  GATA6 MEF2C\n",
       "49                                     SNAI2  EYA1 SIX1\n",
       "50                                                  NaN\n",
       "51                                                  NaN\n",
       "52                                                  NaN\n",
       "53                                                  NaN\n",
       "Name: TFs, dtype: object"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['TFs']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61aa5087-b0ea-44c2-95c3-6bc2bec5ac89",
   "metadata": {},
   "source": [
    "## Debug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d5c74dbc-6b9f-443d-89fb-be4b752976f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs  n_vars = 38151  58870\n",
       "    obs: 'organ_tissue', 'method', 'donor', 'anatomical_information', 'n_counts_UMIs', 'n_genes', 'cell_ontology_class', 'free_annotation', 'manually_annotated', 'compartment', 'gender'\n",
       "    var: 'gene_symbol', 'feature_type', 'ensemblid', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
       "    uns: '_scvi', '_training_mode', 'dendrogram_cell_type_tissue', 'dendrogram_computational_compartment_assignment', 'dendrogram_consensus_prediction', 'dendrogram_tissue_cell_type', 'donor_colors', 'donor_method_colors', 'hvg', 'method_colors', 'neighbors', 'organ_tissue_colors', 'sex_colors', 'tissue_colors', 'umap'\n",
       "    obsm: 'X_pca', 'X_scvi', 'X_scvi_umap', 'X_umap'\n",
       "    layers: 'decontXcounts', 'raw_counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "42651dce-39f6-415a-a9a5-6d35695e1d65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>ensemblid</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>False</td>\n",
       "      <td>6.398244e-05</td>\n",
       "      <td>0.835044</td>\n",
       "      <td>-0.573947</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.005574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>False</td>\n",
       "      <td>2.274395e-03</td>\n",
       "      <td>2.442280</td>\n",
       "      <td>0.533203</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.031731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR6859-1</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>False</td>\n",
       "      <td>6.175251e-05</td>\n",
       "      <td>1.295335</td>\n",
       "      <td>-0.256874</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.005634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>False</td>\n",
       "      <td>1.372886e-04</td>\n",
       "      <td>2.656352</td>\n",
       "      <td>0.680668</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.008041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND6</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>False</td>\n",
       "      <td>9.634841e-01</td>\n",
       "      <td>2.466404</td>\n",
       "      <td>0.154140</td>\n",
       "      <td>0.590065</td>\n",
       "      <td>0.741395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TE</th>\n",
       "      <td>MT-TE</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.600667e-01</td>\n",
       "      <td>1.603787</td>\n",
       "      <td>-0.044396</td>\n",
       "      <td>0.083929</td>\n",
       "      <td>0.301820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CYB</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>False</td>\n",
       "      <td>4.367693e+00</td>\n",
       "      <td>4.765751</td>\n",
       "      <td>-0.499747</td>\n",
       "      <td>3.874830</td>\n",
       "      <td>1.104192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TT</th>\n",
       "      <td>MT-TT</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>False</td>\n",
       "      <td>6.573967e-02</td>\n",
       "      <td>0.624316</td>\n",
       "      <td>-0.719108</td>\n",
       "      <td>0.040580</td>\n",
       "      <td>0.186848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TP</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>True</td>\n",
       "      <td>4.385102e-01</td>\n",
       "      <td>2.705558</td>\n",
       "      <td>1.115482</td>\n",
       "      <td>0.204806</td>\n",
       "      <td>0.501265</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58870 rows  9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_symbol     feature_type          ensemblid  highly_variable  \\\n",
       "DDX11L1          DDX11L1  Gene Expression  ENSG00000223972.5            False   \n",
       "WASH7P            WASH7P  Gene Expression  ENSG00000227232.5            False   \n",
       "MIR6859-1      MIR6859-1  Gene Expression  ENSG00000278267.1            False   \n",
       "MIR1302-2HG  MIR1302-2HG  Gene Expression  ENSG00000243485.5            False   \n",
       "MIR1302-2      MIR1302-2  Gene Expression  ENSG00000284332.1            False   \n",
       "...                  ...              ...                ...              ...   \n",
       "MT-ND6            MT-ND6  Gene Expression  ENSG00000198695.2            False   \n",
       "MT-TE              MT-TE  Gene Expression  ENSG00000210194.1            False   \n",
       "MT-CYB            MT-CYB  Gene Expression  ENSG00000198727.2            False   \n",
       "MT-TT              MT-TT  Gene Expression  ENSG00000210195.2            False   \n",
       "MT-TP              MT-TP  Gene Expression  ENSG00000210196.2             True   \n",
       "\n",
       "                    means  dispersions  dispersions_norm      mean       std  \n",
       "DDX11L1      6.398244e-05     0.835044         -0.573947  0.000039  0.005574  \n",
       "WASH7P       2.274395e-03     2.442280          0.533203  0.001080  0.031731  \n",
       "MIR6859-1    6.175251e-05     1.295335         -0.256874  0.000033  0.005634  \n",
       "MIR1302-2HG  1.372886e-04     2.656352          0.680668  0.000048  0.008041  \n",
       "MIR1302-2    1.000000e-12          NaN          0.000000  0.000000  1.000000  \n",
       "...                   ...          ...               ...       ...       ...  \n",
       "MT-ND6       9.634841e-01     2.466404          0.154140  0.590065  0.741395  \n",
       "MT-TE        1.600667e-01     1.603787         -0.044396  0.083929  0.301820  \n",
       "MT-CYB       4.367693e+00     4.765751         -0.499747  3.874830  1.104192  \n",
       "MT-TT        6.573967e-02     0.624316         -0.719108  0.040580  0.186848  \n",
       "MT-TP        4.385102e-01     2.705558          1.115482  0.204806  0.501265  \n",
       "\n",
       "[58870 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "05498f82-b879-4452-8d69-273a28b05d1b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/scipy/sparse/_index.py:134: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.\n",
      "  self._set_arrayXarray_sparse(i, j, x)\n"
     ]
    }
   ],
   "source": [
    "padata = perturb_counts(adata, ['DDX11L1'], 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e2cdffdb-d65c-4162-b41c-15f9edfc526a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>ensemblid</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>scaled</th>\n",
       "      <th>scaled_by</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>False</td>\n",
       "      <td>6.398244e-05</td>\n",
       "      <td>0.835044</td>\n",
       "      <td>-0.573947</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.005574</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>False</td>\n",
       "      <td>2.274395e-03</td>\n",
       "      <td>2.442280</td>\n",
       "      <td>0.533203</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.031731</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR6859-1</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>False</td>\n",
       "      <td>6.175251e-05</td>\n",
       "      <td>1.295335</td>\n",
       "      <td>-0.256874</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.005634</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>False</td>\n",
       "      <td>1.372886e-04</td>\n",
       "      <td>2.656352</td>\n",
       "      <td>0.680668</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.008041</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND6</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>False</td>\n",
       "      <td>9.634841e-01</td>\n",
       "      <td>2.466404</td>\n",
       "      <td>0.154140</td>\n",
       "      <td>0.590065</td>\n",
       "      <td>0.741395</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TE</th>\n",
       "      <td>MT-TE</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.600667e-01</td>\n",
       "      <td>1.603787</td>\n",
       "      <td>-0.044396</td>\n",
       "      <td>0.083929</td>\n",
       "      <td>0.301820</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CYB</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>False</td>\n",
       "      <td>4.367693e+00</td>\n",
       "      <td>4.765751</td>\n",
       "      <td>-0.499747</td>\n",
       "      <td>3.874830</td>\n",
       "      <td>1.104192</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TT</th>\n",
       "      <td>MT-TT</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>False</td>\n",
       "      <td>6.573967e-02</td>\n",
       "      <td>0.624316</td>\n",
       "      <td>-0.719108</td>\n",
       "      <td>0.040580</td>\n",
       "      <td>0.186848</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TP</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>True</td>\n",
       "      <td>4.385102e-01</td>\n",
       "      <td>2.705558</td>\n",
       "      <td>1.115482</td>\n",
       "      <td>0.204806</td>\n",
       "      <td>0.501265</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58870 rows  11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_symbol     feature_type          ensemblid  highly_variable  \\\n",
       "DDX11L1          DDX11L1  Gene Expression  ENSG00000223972.5            False   \n",
       "WASH7P            WASH7P  Gene Expression  ENSG00000227232.5            False   \n",
       "MIR6859-1      MIR6859-1  Gene Expression  ENSG00000278267.1            False   \n",
       "MIR1302-2HG  MIR1302-2HG  Gene Expression  ENSG00000243485.5            False   \n",
       "MIR1302-2      MIR1302-2  Gene Expression  ENSG00000284332.1            False   \n",
       "...                  ...              ...                ...              ...   \n",
       "MT-ND6            MT-ND6  Gene Expression  ENSG00000198695.2            False   \n",
       "MT-TE              MT-TE  Gene Expression  ENSG00000210194.1            False   \n",
       "MT-CYB            MT-CYB  Gene Expression  ENSG00000198727.2            False   \n",
       "MT-TT              MT-TT  Gene Expression  ENSG00000210195.2            False   \n",
       "MT-TP              MT-TP  Gene Expression  ENSG00000210196.2             True   \n",
       "\n",
       "                    means  dispersions  dispersions_norm      mean       std  \\\n",
       "DDX11L1      6.398244e-05     0.835044         -0.573947  0.000039  0.005574   \n",
       "WASH7P       2.274395e-03     2.442280          0.533203  0.001080  0.031731   \n",
       "MIR6859-1    6.175251e-05     1.295335         -0.256874  0.000033  0.005634   \n",
       "MIR1302-2HG  1.372886e-04     2.656352          0.680668  0.000048  0.008041   \n",
       "MIR1302-2    1.000000e-12          NaN          0.000000  0.000000  1.000000   \n",
       "...                   ...          ...               ...       ...       ...   \n",
       "MT-ND6       9.634841e-01     2.466404          0.154140  0.590065  0.741395   \n",
       "MT-TE        1.600667e-01     1.603787         -0.044396  0.083929  0.301820   \n",
       "MT-CYB       4.367693e+00     4.765751         -0.499747  3.874830  1.104192   \n",
       "MT-TT        6.573967e-02     0.624316         -0.719108  0.040580  0.186848   \n",
       "MT-TP        4.385102e-01     2.705558          1.115482  0.204806  0.501265   \n",
       "\n",
       "             scaled  scaled_by  \n",
       "DDX11L1        True          1  \n",
       "WASH7P        False          1  \n",
       "MIR6859-1     False          1  \n",
       "MIR1302-2HG   False          1  \n",
       "MIR1302-2     False          1  \n",
       "...             ...        ...  \n",
       "MT-ND6        False          1  \n",
       "MT-TE         False          1  \n",
       "MT-CYB        False          1  \n",
       "MT-TT         False          1  \n",
       "MT-TP         False          1  \n",
       "\n",
       "[58870 rows x 11 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concatenated_adata = ad.concat([adata, padata], axis=0)\n",
    "concatenated_adata.var = adata.var.copy()\n",
    "concatenated_adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2311423c-164d-4797-abcc-553f4016b813",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>ensemblid</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>scaled</th>\n",
       "      <th>scaled_by</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>False</td>\n",
       "      <td>6.398244e-05</td>\n",
       "      <td>0.835044</td>\n",
       "      <td>-0.573947</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.005574</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>False</td>\n",
       "      <td>2.274395e-03</td>\n",
       "      <td>2.442280</td>\n",
       "      <td>0.533203</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.031731</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR6859-1</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>False</td>\n",
       "      <td>6.175251e-05</td>\n",
       "      <td>1.295335</td>\n",
       "      <td>-0.256874</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.005634</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>False</td>\n",
       "      <td>1.372886e-04</td>\n",
       "      <td>2.656352</td>\n",
       "      <td>0.680668</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.008041</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND6</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>False</td>\n",
       "      <td>9.634841e-01</td>\n",
       "      <td>2.466404</td>\n",
       "      <td>0.154140</td>\n",
       "      <td>0.590065</td>\n",
       "      <td>0.741395</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TE</th>\n",
       "      <td>MT-TE</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.600667e-01</td>\n",
       "      <td>1.603787</td>\n",
       "      <td>-0.044396</td>\n",
       "      <td>0.083929</td>\n",
       "      <td>0.301820</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CYB</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>False</td>\n",
       "      <td>4.367693e+00</td>\n",
       "      <td>4.765751</td>\n",
       "      <td>-0.499747</td>\n",
       "      <td>3.874830</td>\n",
       "      <td>1.104192</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TT</th>\n",
       "      <td>MT-TT</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>False</td>\n",
       "      <td>6.573967e-02</td>\n",
       "      <td>0.624316</td>\n",
       "      <td>-0.719108</td>\n",
       "      <td>0.040580</td>\n",
       "      <td>0.186848</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TP</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>True</td>\n",
       "      <td>4.385102e-01</td>\n",
       "      <td>2.705558</td>\n",
       "      <td>1.115482</td>\n",
       "      <td>0.204806</td>\n",
       "      <td>0.501265</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58870 rows  11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_symbol     feature_type          ensemblid  highly_variable  \\\n",
       "DDX11L1          DDX11L1  Gene Expression  ENSG00000223972.5            False   \n",
       "WASH7P            WASH7P  Gene Expression  ENSG00000227232.5            False   \n",
       "MIR6859-1      MIR6859-1  Gene Expression  ENSG00000278267.1            False   \n",
       "MIR1302-2HG  MIR1302-2HG  Gene Expression  ENSG00000243485.5            False   \n",
       "MIR1302-2      MIR1302-2  Gene Expression  ENSG00000284332.1            False   \n",
       "...                  ...              ...                ...              ...   \n",
       "MT-ND6            MT-ND6  Gene Expression  ENSG00000198695.2            False   \n",
       "MT-TE              MT-TE  Gene Expression  ENSG00000210194.1            False   \n",
       "MT-CYB            MT-CYB  Gene Expression  ENSG00000198727.2            False   \n",
       "MT-TT              MT-TT  Gene Expression  ENSG00000210195.2            False   \n",
       "MT-TP              MT-TP  Gene Expression  ENSG00000210196.2             True   \n",
       "\n",
       "                    means  dispersions  dispersions_norm      mean       std  \\\n",
       "DDX11L1      6.398244e-05     0.835044         -0.573947  0.000039  0.005574   \n",
       "WASH7P       2.274395e-03     2.442280          0.533203  0.001080  0.031731   \n",
       "MIR6859-1    6.175251e-05     1.295335         -0.256874  0.000033  0.005634   \n",
       "MIR1302-2HG  1.372886e-04     2.656352          0.680668  0.000048  0.008041   \n",
       "MIR1302-2    1.000000e-12          NaN          0.000000  0.000000  1.000000   \n",
       "...                   ...          ...               ...       ...       ...   \n",
       "MT-ND6       9.634841e-01     2.466404          0.154140  0.590065  0.741395   \n",
       "MT-TE        1.600667e-01     1.603787         -0.044396  0.083929  0.301820   \n",
       "MT-CYB       4.367693e+00     4.765751         -0.499747  3.874830  1.104192   \n",
       "MT-TT        6.573967e-02     0.624316         -0.719108  0.040580  0.186848   \n",
       "MT-TP        4.385102e-01     2.705558          1.115482  0.204806  0.501265   \n",
       "\n",
       "             scaled  scaled_by  \n",
       "DDX11L1        True          1  \n",
       "WASH7P        False          1  \n",
       "MIR6859-1     False          1  \n",
       "MIR1302-2HG   False          1  \n",
       "MIR1302-2     False          1  \n",
       "...             ...        ...  \n",
       "MT-ND6        False          1  \n",
       "MT-TE         False          1  \n",
       "MT-CYB        False          1  \n",
       "MT-TT         False          1  \n",
       "MT-TP         False          1  \n",
       "\n",
       "[58870 rows x 11 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "84f591c5-ea99-4235-a2a6-9100fca8b04e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR6859-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND6</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CYB</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TP</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58870 rows  0 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [DDX11L1, WASH7P, MIR6859-1, MIR1302-2HG, MIR1302-2, FAM138A, OR4G4P, OR4G11P, OR4F5, AL627309.1, AL627309.3, CICP27, AL627309.6, AL627309.7, AL627309.2, AL627309.5, RNU6-1100P, AL627309.4, FO538757.1, WASH9P, MIR6859-2, AP006222.1, RPL23AP24, AL732372.1, AL732372.2, WBP1LP7, OR4F29, CICP7, AL732372.3, RF00026, AL669831.3, AC114498.1, MTND1P23, MTND2P28, MTCO1P12, AC114498.2, MTCO2P12, MTATP8P1, MTATP6P1, MTCO3P12, WBP1LP6, OR4F16, CICP3, AL669831.1, RNU6-1199P, AL669831.2, AL669831.5, AL669831.7, AL669831.4, FAM87B, LINC01128, LINC00115, FAM41C, TUBB8P11, AL669831.6, AL645608.6, AL645608.2, AL645608.4, LINC02593, SAMD11, NOC2L, KLHL17, PLEKHN1, PERM1, AL645608.7, HES4, ISG15, AL645608.3, AL645608.1, AGRN, AL645608.5, AL390719.1, AL645608.8, RNF223, C1orf159, AL390719.3, LINC01342, MIR200B, MIR200A, MIR429, AL390719.2, TTLL10-AS1, TTLL10, TNFRSF18, TNFRSF4, SDF4, B3GALT6, C1QTNF12, AL162741.1, UBE2J2, LINC01786, SCNN1D, ACAP3, MIR6726, PUSL1, INTS11, MIR6727, AL139287.1, CPTP, TAS1R3, ...]\n",
       "\n",
       "[58870 rows x 0 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concatenated_adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4d79b7ad-c657-40c6-aaf0-9b80a1dc7b5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>ensemblid</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>scaled</th>\n",
       "      <th>scaled_by</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>False</td>\n",
       "      <td>6.398244e-05</td>\n",
       "      <td>0.835044</td>\n",
       "      <td>-0.573947</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.005574</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>False</td>\n",
       "      <td>2.274395e-03</td>\n",
       "      <td>2.442280</td>\n",
       "      <td>0.533203</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.031731</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR6859-1</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>False</td>\n",
       "      <td>6.175251e-05</td>\n",
       "      <td>1.295335</td>\n",
       "      <td>-0.256874</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.005634</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>False</td>\n",
       "      <td>1.372886e-04</td>\n",
       "      <td>2.656352</td>\n",
       "      <td>0.680668</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.008041</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND6</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>False</td>\n",
       "      <td>9.634841e-01</td>\n",
       "      <td>2.466404</td>\n",
       "      <td>0.154140</td>\n",
       "      <td>0.590065</td>\n",
       "      <td>0.741395</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TE</th>\n",
       "      <td>MT-TE</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.600667e-01</td>\n",
       "      <td>1.603787</td>\n",
       "      <td>-0.044396</td>\n",
       "      <td>0.083929</td>\n",
       "      <td>0.301820</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CYB</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>False</td>\n",
       "      <td>4.367693e+00</td>\n",
       "      <td>4.765751</td>\n",
       "      <td>-0.499747</td>\n",
       "      <td>3.874830</td>\n",
       "      <td>1.104192</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TT</th>\n",
       "      <td>MT-TT</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>False</td>\n",
       "      <td>6.573967e-02</td>\n",
       "      <td>0.624316</td>\n",
       "      <td>-0.719108</td>\n",
       "      <td>0.040580</td>\n",
       "      <td>0.186848</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TP</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>True</td>\n",
       "      <td>4.385102e-01</td>\n",
       "      <td>2.705558</td>\n",
       "      <td>1.115482</td>\n",
       "      <td>0.204806</td>\n",
       "      <td>0.501265</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58870 rows  11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_symbol     feature_type          ensemblid  highly_variable  \\\n",
       "DDX11L1          DDX11L1  Gene Expression  ENSG00000223972.5            False   \n",
       "WASH7P            WASH7P  Gene Expression  ENSG00000227232.5            False   \n",
       "MIR6859-1      MIR6859-1  Gene Expression  ENSG00000278267.1            False   \n",
       "MIR1302-2HG  MIR1302-2HG  Gene Expression  ENSG00000243485.5            False   \n",
       "MIR1302-2      MIR1302-2  Gene Expression  ENSG00000284332.1            False   \n",
       "...                  ...              ...                ...              ...   \n",
       "MT-ND6            MT-ND6  Gene Expression  ENSG00000198695.2            False   \n",
       "MT-TE              MT-TE  Gene Expression  ENSG00000210194.1            False   \n",
       "MT-CYB            MT-CYB  Gene Expression  ENSG00000198727.2            False   \n",
       "MT-TT              MT-TT  Gene Expression  ENSG00000210195.2            False   \n",
       "MT-TP              MT-TP  Gene Expression  ENSG00000210196.2             True   \n",
       "\n",
       "                    means  dispersions  dispersions_norm      mean       std  \\\n",
       "DDX11L1      6.398244e-05     0.835044         -0.573947  0.000039  0.005574   \n",
       "WASH7P       2.274395e-03     2.442280          0.533203  0.001080  0.031731   \n",
       "MIR6859-1    6.175251e-05     1.295335         -0.256874  0.000033  0.005634   \n",
       "MIR1302-2HG  1.372886e-04     2.656352          0.680668  0.000048  0.008041   \n",
       "MIR1302-2    1.000000e-12          NaN          0.000000  0.000000  1.000000   \n",
       "...                   ...          ...               ...       ...       ...   \n",
       "MT-ND6       9.634841e-01     2.466404          0.154140  0.590065  0.741395   \n",
       "MT-TE        1.600667e-01     1.603787         -0.044396  0.083929  0.301820   \n",
       "MT-CYB       4.367693e+00     4.765751         -0.499747  3.874830  1.104192   \n",
       "MT-TT        6.573967e-02     0.624316         -0.719108  0.040580  0.186848   \n",
       "MT-TP        4.385102e-01     2.705558          1.115482  0.204806  0.501265   \n",
       "\n",
       "             scaled  scaled_by  \n",
       "DDX11L1        True          1  \n",
       "WASH7P        False          1  \n",
       "MIR6859-1     False          1  \n",
       "MIR1302-2HG   False          1  \n",
       "MIR1302-2     False          1  \n",
       "...             ...        ...  \n",
       "MT-ND6        False          1  \n",
       "MT-TE         False          1  \n",
       "MT-CYB        False          1  \n",
       "MT-TT         False          1  \n",
       "MT-TP         False          1  \n",
       "\n",
       "[58870 rows x 11 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "padata.var"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcb287d2-59ac-4fbc-8887-8500217a591f",
   "metadata": {},
   "source": [
    "# Day 7"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "982889a0-25b9-4fb4-8195-d3593c22ed67",
   "metadata": {},
   "source": [
    "# Day 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60afe83b-fd13-436f-8dd6-1fd95fd883d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" Changes to gene names and formatting are reflected in spreadsheet\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a1edb3a-5d7c-4230-a2cb-5ec790935dd3",
   "metadata": {},
   "source": [
    "# Day 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "aa696c58-66b2-4eba-a1b1-b578eff3ff28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copied from a prev day\n",
    "\n",
    "\"\"\"\n",
    "Josh, please read: the adata.obs['scalar'] = scalar copies the scalar down for that call, associated with every cell in X. Same with ['scaled'] and ['scaled_by']\n",
    "in var. This is good in case the data is later appended into one anndata object.\n",
    "But my return from the perturb_counts loop (cell below this) is a dictionary of all of the perturb_counts, since appending along any axis will probably either overwrite\n",
    "obs or var.\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "def perturb_counts(tf_list, scalar, adata): \n",
    "    \"\"\"\n",
    "    Applies a perturbation to the expression data of specific genes in an AnnData object.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Computes the maximum gene expression level for each cell.\n",
    "    2. Applies a scaling operation to the expression levels of genes listed in `tf_list`.\n",
    "       - Each entry of these genes in the matrix is multiplied by the maximum expression level \n",
    "         of its respective cell and a specified scalar value.\n",
    "    3. Updates the AnnData object with new columns:\n",
    "       - 'scaled': A boolean column indicating whether each gene is in the `tf_list`.\n",
    "       - 'scaled_by': Contains the scaling factor used for each gene (the product of the maximum \n",
    "         expression level of each cell and the scalar), or `1` if the gene was not in `tf_list`.\n",
    "    \n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols to be perturbed.\n",
    "    scalar (float): The scalar value used to scale the expression levels.\n",
    "    adata (AnnData): The AnnData object containing gene expression data.\n",
    "\n",
    "    Returns:\n",
    "    AnnData: The updated AnnData object with applied perturbations and new columns.\n",
    "    \"\"\"\n",
    "    # Save the original state of the parameter objects, in case some tfs do not translate (failsafe)\n",
    "    original_X = adata.X.copy()\n",
    "    original_gene_mask = gene_mask.copy()\n",
    "    \n",
    "    # Compute maximum expression level of each cell\n",
    "    max_exp = np.max(adata.X, axis=1)\n",
    "    \n",
    "    # Create a boolean mask for genes in tf_list\n",
    "    gene_mask = adata.var['gene_symbol'].isin(tf_list)\n",
    "\n",
    "    \"\"\"This is new today. v \"\"\"\n",
    "    # Raise an error if any of the gene names in tf_list do not match column names (we will manually update these in adata):\n",
    "    missing_genes = [gene for gene in tf_list if gene not in adata.var['gene_symbol'].values]\n",
    "    \n",
    "    if missing_genes:\n",
    "        # Restore original parameter objects\n",
    "        adata.X = original_X\n",
    "        gene_mask = original_gene_mask\n",
    "        raise ValueError(f\"Genes {missing_genes} not found in anndata object\")\n",
    "\n",
    "    else:\n",
    "        \"\"\" This is new today. ^ \"\"\"\n",
    "        # Apply the scaling operation to the specified genes\n",
    "        adata.X[:, gene_mask] *= max_exp[:, np.newaxis] * scalar\n",
    "        \n",
    "        # Add/Update 'scaled' column in var\n",
    "        adata.var['scaled'] = gene_mask\n",
    "        \n",
    "        # Add/Update 'scaled_by' column in var\n",
    "        adata.var['scaled_by'] = 1  # Default value for genes not in tf_list\n",
    "        adata.var.loc[gene_mask, 'scaled_by'] = max_exp[:, np.newaxis] * scalar  # Correct scaling factor assignment\n",
    "    \n",
    "    return adata\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7a489b88-60e1-4387-8d03-20c129931da5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copied from a prev day\n",
    "\n",
    "import anndata\n",
    "\n",
    "def iterate_perturb_counts(tf_list, scalar_list, adata):\n",
    "    \"\"\"\n",
    "    Applies perturbations to the expression data of specified transcription factors across multiple scalars \n",
    "    and stores the resulting AnnData objects in a dictionary.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Iterates over a list of scalar values.\n",
    "    2. For each scalar, creates a copy of the AnnData object to preserve the original data.\n",
    "    3. Applies the `perturb_counts` function to scale the expression data of genes listed in `tf_list` by\n",
    "       the maximum gene expression of each cell and the current scalar.\n",
    "    4. Stores the perturbed AnnData object in a dictionary with the scalar as the key.\n",
    "\n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols (transcription factors) to be perturbed.\n",
    "    scalar_list (list): A list of scalar values for scaling the gene expression.\n",
    "    adata (AnnData): The AnnData object containing gene expression data (cells x genes).\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary where keys are scalar values and values are the corresponding perturbed AnnData objects.\n",
    "    \"\"\"\n",
    "    \n",
    "    adata_dict = {}\n",
    "    \n",
    "    for scalar in scalar_list:\n",
    "        # Create a copy of the AnnData object for each scalar value\n",
    "        adata_temp = adata.copy()\n",
    "        \n",
    "        # Apply perturb_counts to the copied AnnData object\n",
    "        perturbed_adata = perturb_counts(tf_list, scalar, adata_temp)\n",
    "        \n",
    "        # Store the perturbed AnnData object in the dictionary with scalar as the key\n",
    "        adata_dict[scalar] = perturbed_adata\n",
    "    \n",
    "    return adata_dict\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fa93426a-3ff0-44e1-8b41-5d5e89d56666",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "First 5 entries:\n",
      "['DDX11L1', 'WASH7P', 'MIR6859-1', 'MIR1302-2HG', 'MIR1302-2']\n",
      "Last 5 entries:\n",
      "['MT-ND6', 'MT-TE', 'MT-CYB', 'MT-TT', 'MT-TP']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Copied from a prev day\n",
    "\n",
    "DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/\"\n",
    "FILE = \"TS_epithelial.h5ad\"\n",
    "adata = sp.read_h5ad(os.path.join(DATAPATH, FILE))\n",
    "adata_gene_list = adata.var['gene_symbol'].values.tolist()\n",
    "\n",
    "# Print the first 5 entries\n",
    "print(\"First 5 entries:\")\n",
    "print(adata_gene_list[:5])\n",
    "\n",
    "# Print the last 5 entries\n",
    "print(\"Last 5 entries:\")\n",
    "print(adata_gene_list[-5:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "270230bf-4742-48c8-90ab-0ba4ded0513c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Entries in the table that are not genes in the counts matrix: \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['A',\n",
       " 'AP-2A',\n",
       " 'MASH1',\n",
       " 'ANY',\n",
       " 'THE',\n",
       " 'LEF-1',\n",
       " 'NEUROD',\n",
       " 'OCT6',\n",
       " 'SV40',\n",
       " 'T',\n",
       " 'LXMLA',\n",
       " 'HNF6',\n",
       " 'HB9',\n",
       " 'FACTORS:',\n",
       " '+',\n",
       " 'OCT9',\n",
       " 'LXH3',\n",
       " 'AND',\n",
       " 'ANTIGEN',\n",
       " 'THREE',\n",
       " 'KNOCKDOWN',\n",
       " '(ETS',\n",
       " 'DOMINANT-NEGATIVE',\n",
       " 'OR',\n",
       " 'OSTERIX',\n",
       " 'OCT4',\n",
       " 'BRN2',\n",
       " 'OCT3/4',\n",
       " 'C-MYC',\n",
       " 'LARGE',\n",
       " 'OF',\n",
       " 'SOX2;',\n",
       " 'P53',\n",
       " 'NGN2',\n",
       " 'TWO',\n",
       " 'L-MYC',\n",
       " 'PPARG2',\n",
       " 'PAX6;',\n",
       " 'N-MYC',\n",
       " '2)',\n",
       " 'ER71/ETV2',\n",
       " 'NURR1',\n",
       " 'PU.1',\n",
       " 'NF-B',\n",
       " 'LMX1A;',\n",
       " 'VARIANT']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# outside the function so I can manipulate these directly.\n",
    "# I also checked whether it matters if I do case insensitive (capitalize everything then compare) or case sensitive.\n",
    "# Unsurprisingly, case sensitive has more discrepancies (45 vs. 44), with the one extra that was picked up as \"Ptf1a\".\n",
    "# I left the case insensitive version. \n",
    "\n",
    "# get a list of words (potential genes, also includes and, + , etc. ) from the table from the review paper\n",
    "table_1_df = pd.read_csv(\"/home/oliven/scFoundationModels/notebooks/reprogramming/data/table_1_data_from_paper_9_1.csv\")\n",
    "combined_string = ' '.join(table_1_df['Treatment'].astype(str)).replace(',', '')\n",
    "word_list = combined_string.split()\n",
    "\n",
    "# get a list of genes that appear in the data matrix\n",
    "DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/\"\n",
    "FILE = \"TS_epithelial.h5ad\"\n",
    "adata = sp.read_h5ad(os.path.join(DATAPATH, FILE))\n",
    "adata_gene_list = adata.var['gene_symbol'].values.tolist()\n",
    "\n",
    "# New today\n",
    "def check_valid_tfs(word_list, adata_gene_list):\n",
    "\n",
    "    word_list_upper = [word.upper() for word in word_list]\n",
    "    adata_gene_list_upper = [gene.upper() for gene in adata_gene_list]\n",
    "\n",
    "    # print what does not overlap\n",
    "    not_valid_gene = set(word_list_upper) - set(adata_gene_list_upper)\n",
    "\n",
    "    print(\"Entries in the table that are not genes in the counts matrix: \")\n",
    "    \n",
    "    return list(not_valid_gene)\n",
    "    \n",
    "check_valid_tfs(word_list, adata_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "930e2060-2c16-4f3b-9f64-cf1807778708",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for those genes with multiple aliases, checking which are valid\n",
    "# we have to worry about making this case insensitive just in case\n",
    "multiple_alias_dict = {}\n",
    "multiple_alias_dict.update({\n",
    "    'P53': ['BCC7','BMFS5', 'LFS1', 'TRP53'],\n",
    "    'OCT3/4': ['POU5F1', 'OCT3', 'OCT4', 'OTF4', 'MGC22487'],\n",
    "    'MASH1': ['HASH1', 'BHLHa46', 'ASH1', 'ASH-1', 'ASCL1'],\n",
    "    'HNF6': ['HNF6', 'HNF6A', 'ONECUT1'],\n",
    "    'HB9': ['MNX1', 'HOXHB9', 'SCRA1', 'HLXB9', 'GC07M156491', 'GC07M156786', 'GC07M150530'],\n",
    "    'PPARG2': ['PPARG', 'NR1C3', 'PPARG1', 'PPARgamma', 'PPAR-Gamma', 'PPARG5', 'CIMT1', 'GLM1'], \n",
    "    'PU.1': ['SPI1', 'SPI-A', 'SFPI1', 'SPI-1', 'OF', 'AGM10'],\n",
    "    'N-MYC': ['MYCN', 'BHLHe37', 'N-Myc', 'NMYC', 'MYCNOT', 'MYCNsORF', 'MYCNsPEP', 'BHLHE37', 'FGLDS1', 'MODED', 'MPAPA', 'ODED'],\n",
    "    # oct9 had a strange genecards lookup\n",
    "    'OCT9': ['POU3F4', 'SLC22A16'],\n",
    "    'LEF-1': ['TCF1ALPHA', 'TCF7L3', 'TCF10', 'LEF1'],\n",
    "    # these next two were listed as ER71/ETV2\n",
    "    'ER71/ETV2': ['ER71', 'ETV2', 'ETSRP71'],\n",
    "    # sv40 had a strange genecards lookup\n",
    "    'SV40': [''],\n",
    "    # this one didn't show up\n",
    "    'LXH3': ['M2-LHX3', 'M2LHX3', 'CPHD3', 'LIM3'],\n",
    "    'NGN2': ['NEUROG2', 'BHLHA8', 'MATH4A', 'Math4a', 'ATOH4', 'Ngn-2', 'BHLHA8', 'Atoh4', 'NGN-2'],\n",
    "    'LMX1A': ['LMX1.1', 'LMX1', 'LMX-1.', 'DFNA7'],\n",
    "    # NF-Kb had  a strange genecards lookup\n",
    "    'NF-B': ['NFkb1', 'NFKB1'], \n",
    "    'L-MYC': ['MYCL', 'LMYC', 'BHLHe38', 'MYCL1', 'BHLHE38'],\n",
    "    'BRN2': ['POU3F2', 'BRN2', 'OCT7', 'POUF3', 'OTF7', 'Brain-2', 'OTF-7', 'Brn-2', 'Oct-7', 'N-Oct3'], \n",
    "    'NURR1': ['NR4A2', 'TINUR', 'NOT', 'HZF3', 'NURR1', 'RNR1', 'IDLDP'],\n",
    "    'SOX2': ['SRY-Box 2', 'MCOPS3', 'ANOP3'],\n",
    "    'NEUROD': ['NEUROD1', 'BHLHa3', 'BETA2', 'BHF-1', 'MODY6', 'NeuroD1', 'BHLHA3', 'T2D'],\n",
    "    'C-MYC': ['MYC', 'C-MYC', 'MYCC', 'MRTL', 'BHLHE39'],\n",
    "    # AP-2A had  a strange genecards lookup\n",
    "    'AP-2A': [' '],\n",
    "    'PAX6': ['D11S812E', 'WAGR', 'AN2', 'AN', 'AN1', 'ASGD5', 'FVH1', 'MGDA'],\n",
    "    'OSTERIX': ['SP7', 'OSX', 'OI11', 'OI12'],\n",
    "    # OCT6 had a strange genecards lookup\n",
    "    'OCT6': ['POU3F1', 'SCIP', 'OTF6', 'OTF-6'],\n",
    "   \n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eed97378-5da6-46b7-9b51-d1b8fca15442",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4a2881f1-9966-4946-a2c4-d27fb6de066c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P53 was not found in gene list.\n",
      "OCT3/4 was found in gene list as ['POU5F1']\n",
      "MASH1 was found in gene list as ['ASCL1']\n",
      "HNF6 was found in gene list as ['ONECUT1']\n",
      "HB9 was found in gene list as ['MNX1']\n",
      "PPARG2 was found in gene list as ['PPARG']\n",
      "PU.1 was found in gene list as ['SPI1']\n",
      "N-MYC was found in gene list as ['MYCN']\n",
      "OCT9 was found in gene list as ['POU3F4', 'SLC22A16']\n",
      "LEF-1 was found in gene list as ['LEF1']\n",
      "ER71/ETV2 was found in gene list as ['ETV2']\n",
      "SV40 was not found in gene list.\n",
      "LXH3 was not found in gene list.\n",
      "NGN2 was found in gene list as ['NEUROG2']\n",
      "LMX1A was not found in gene list.\n",
      "NF-B was found in gene list as ['NFKB1', 'NFKB1']\n",
      "L-MYC was found in gene list as ['MYCL']\n",
      "BRN2 was found in gene list as ['POU3F2']\n",
      "NURR1 was found in gene list as ['NR4A2']\n",
      "SOX2 was not found in gene list.\n",
      "NEUROD was found in gene list as ['NEUROD1', 'NEUROD1']\n",
      "C-MYC was found in gene list as ['MYC']\n",
      "AP-2A was not found in gene list.\n",
      "PAX6 was not found in gene list.\n",
      "OSTERIX was not found in gene list.\n",
      "OCT6 was found in gene list as ['POU3F1']\n"
     ]
    }
   ],
   "source": [
    "# day 4, technically (migrate this)\n",
    "adata_gene_list = adata.var['gene_symbol'].values.tolist()\n",
    "def identify_gene_name_translations(multiple_alias_dict, adata_gene_list):\n",
    "    \"\"\"\n",
    "    Identifies whether any values from multiple_alias_dict appear in adata_gene_list, case-insensitive.\n",
    "    \n",
    "    Parameters:\n",
    "    multiple_alias_dict (dict): A dictionary where keys are gene names and values are lists of aliases.\n",
    "    adata_gene_list (list): A list of gene names to check against, case-insensitive.\n",
    "\n",
    "    Prints:\n",
    "    For each key, whether it was found in the gene list along with the matching values.\n",
    "    \"\"\"\n",
    "    # Convert the gene list to uppercase for case-insensitive comparison\n",
    "    adata_gene_list_upper = [gene.upper() for gene in adata_gene_list]\n",
    "\n",
    "    # Loop through each key and values in the dictionary\n",
    "    for key, values in multiple_alias_dict.items():\n",
    "        # Convert each alias to uppercase\n",
    "        values_upper = [value.upper() for value in values]\n",
    "        \n",
    "        # Check if any alias is present in the gene list\n",
    "        found_values = [value for value in values_upper if value in adata_gene_list_upper]\n",
    "        \n",
    "        # Print appropriate message based on whether any values were found\n",
    "        if found_values:\n",
    "            print(f\"{key} was found in gene list as {found_values}\")\n",
    "        else:\n",
    "            print(f\"{key} was not found in gene list.\")\n",
    "\n",
    "identify_gene_name_translations(multiple_alias_dict, adata_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fe5db023-1ae7-46cf-a4d8-7a9eaf5b193d",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "'break' outside loop (668683560.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[8], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    break\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m 'break' outside loop\n"
     ]
    }
   ],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79ba1127-ee3c-478f-8298-4a738d540bc6",
   "metadata": {},
   "source": [
    "# Day 4: Mine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f02c0cbf-311c-4de6-9e83-8dae191d7ccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Copied from a prev day\n",
    "\n",
    "DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/\"\n",
    "FILE = \"TS_epithelial.h5ad\"\n",
    "adata = sp.read_h5ad(os.path.join(DATAPATH, FILE))\n",
    "adata_gene_list = adata.var['gene_symbol'].values.tolist()\n",
    "\n",
    "# Print the first 5 entries\n",
    "print(\"First 5 entries:\")\n",
    "print(adata_gene_list[:5])\n",
    "\n",
    "# Print the last 5 entries\n",
    "print(\"Last 5 entries:\")\n",
    "print(adata_gene_list[-5:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee05aa9b-e46f-4a68-a047-704b9fb4effb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# outside the function so I can manipulate these directly.\n",
    "# I also checked whether it matters if I do case insensitive (capitalize everything then compare) or case sensitive.\n",
    "# Unsurprisingly, case sensitive has more discrepancies (45 vs. 44), with the one extra that was picked up as \"Ptf1a\".\n",
    "# I left the case insensitive version. \n",
    "\n",
    "# get a list of words (potential genes, also includes and, + , etc. ) from the table from the review paper\n",
    "table_1_df = pd.read_csv(\"/home/oliven/scFoundationModels/notebooks/reprogramming/data/table_1_data_from_paper_9_1.csv\")\n",
    "combined_string = ' '.join(table_1_df['TFs'].astype(str)).replace(',', '')\n",
    "word_list = combined_string.split()\n",
    "\n",
    "# get a list of genes that appear in the data matrix\n",
    "DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/\"\n",
    "FILE = \"TS_epithelial.h5ad\"\n",
    "adata = sp.read_h5ad(os.path.join(DATAPATH, FILE))\n",
    "adata_gene_list = adata.var['gene_symbol'].values.tolist()\n",
    "\n",
    "# New today\n",
    "def check_valid_tfs(word_list, adata_gene_list):\n",
    "\n",
    "    word_list_upper = [word.upper() for word in word_list]\n",
    "    adata_gene_list_upper = [gene.upper() for gene in adata_gene_list]\n",
    "\n",
    "    # print what does not overlap\n",
    "    not_valid_gene = set(word_list_upper) - set(adata_gene_list_upper)\n",
    "\n",
    "    print(\"Entries in the table that are not genes in the counts matrix: \")\n",
    "    \n",
    "    return list(not_valid_gene)\n",
    "    \n",
    "check_valid_tfs(word_list, adata_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25d64c66-8ea7-4ea6-928e-b907f456735b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for those genes with multiple aliases, checking which are valid\n",
    "# we have to worry about making this case insensitive just in case\n",
    "multiple_alias_dict = {}\n",
    "multiple_alias_dict.update({\n",
    "    'p53': ['BCC7','BMFS5', 'LFS1', 'TRP53'],\n",
    "    'OCT3/4': ['POU5F1', 'OCT3', 'OCT4', 'OTF4', 'MGC22487'],\n",
    "    'MASH1': ['HASH1', 'BHLHa46', 'ASH1', 'ASH-1', 'ASCL1'],\n",
    "    'HNF6': ['HNF6', 'HNF6A', 'ONECUT1'],\n",
    "    'HB9': ['MNX1', 'HOXHB9', 'SCRA1', 'HLXB9', 'GC07M156491', 'GC07M156786', 'GC07M150530'],\n",
    "    'PPARG2': ['PPARG', 'NR1C3', 'PPARG1', 'PPARgamma', 'PPAR-Gamma', 'PPARG5', 'CIMT1', 'GLM1'], \n",
    "    'PU.1': ['SPI1', 'SPI-A', 'SFPI1', 'SPI-1', 'OF', 'AGM10'],\n",
    "    'N-MYC': ['MYCN', 'BHLHe37', 'N-Myc', 'NMYC', 'MYCNOT', 'MYCNsORF', 'MYCNsPEP', 'BHLHE37', 'FGLDS1', 'MODED', 'MPAPA', 'ODED'],\n",
    "    # oct9 had a strange genecards lookup\n",
    "    'OCT9': ['POU3F4', 'SLC22A16']\n",
    "    'LEF-1': ['TCF1ALPHA', 'TCF7L3', 'TCF10', 'LEF1'],\n",
    "    # these next two were listed as ER71/ETV2\n",
    "    'ER71/ETV2': ['ER71', 'ETV2', 'ETSRP71'],\n",
    "    # sv40 had a strange genecards lookup\n",
    "    'SV40': [''],\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8932fa31-967f-4f7f-ab6b-845791e191c7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0183d92-2ee0-4b6f-a711-9cdc55a23260",
   "metadata": {},
   "outputs": [],
   "source": [
    "# day 4, technically (migrate this)\n",
    "adata_gene_list = adata.var['gene_symbol'].values.tolist()\n",
    "def identify_gene_name_translations(multiple_alias_dict, adata_gene_list):\n",
    "    \"\"\"\n",
    "    Identifies whether any values from multiple_alias_dict appear in adata_gene_list, case-insensitive.\n",
    "    \n",
    "    Parameters:\n",
    "    multiple_alias_dict (dict): A dictionary where keys are gene names and values are lists of aliases.\n",
    "    adata_gene_list (list): A list of gene names to check against, case-insensitive.\n",
    "\n",
    "    Prints:\n",
    "    For each key, whether it was found in the gene list along with the matching values.\n",
    "    \"\"\"\n",
    "    # Convert the gene list to uppercase for case-insensitive comparison\n",
    "    adata_gene_list_upper = [gene.upper() for gene in adata_gene_list]\n",
    "\n",
    "    # Loop through each key and values in the dictionary\n",
    "    for key, values in multiple_alias_dict.items():\n",
    "        # Convert each alias to uppercase\n",
    "        values_upper = [value.upper() for value in values]\n",
    "        \n",
    "        # Check if any alias is present in the gene list\n",
    "        found_values = [value for value in values_upper if value in adata_gene_list_upper]\n",
    "        \n",
    "        # Print appropriate message based on whether any values were found\n",
    "        if found_values:\n",
    "            print(f\"{key} was found in gene list as {found_values}\")\n",
    "        else:\n",
    "            print(f\"{key} was not found in gene list.\")\n",
    "\n",
    "identify_gene_name_translations(multiple_alias_dict, adata_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8e04100-5836-4bb5-985b-252c739d7fce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The above list is small enough that I can manually check it.\n",
    "\n",
    "# Things to remove from word_list:\n",
    "not_genes = ['Variant', 'Large',  '(ETS', '2)', 'Knockdown', ]\n",
    "\n",
    "# Valid genes to replace/rename in word_list:\n",
    "genes_to_translate = ['LMX1A;', #'P53']\n",
    "translated_names = ['LMX1A',]\n",
    "\n",
    "# These ones might have appeared as multiple entries, etc. b/c of spacing. easiest way was to delete and add back\n",
    "genes_to_add = ['ETS2',]\n",
    "\n",
    "# replace then subtract and add. [--------------]\n",
    "new_word_list = set(word_list) - set(not_genes)\n",
    "\n",
    "# Running the function one more time to check:\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "502b55c4-7e15-47c1-8173-4859a1a32303",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" Renamed med_nonz to max_exp to be more accurate. \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1607b7d-0394-4092-b3d7-1f5337304ae9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copied from a prev day\n",
    "\n",
    "\"\"\"\n",
    "Josh, please read: the adata.obs['scalar'] = scalar copies the scalar down for that call, associated with every cell in X. Same with ['scaled'] and ['scaled_by']\n",
    "in var. This is good in case the data is later appended into one anndata object.\n",
    "But my return from the perturb_counts loop (cell below this) is a dictionary of all of the perturb_counts, since appending along any axis will probably either overwrite\n",
    "obs or var.\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "def perturb_counts(tf_list, scalar, adata): \n",
    "    \"\"\"\n",
    "    Applies a perturbation to the expression data of specific genes in an AnnData object.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Computes the maximum gene expression level for each cell.\n",
    "    2. Applies a scaling operation to the expression levels of genes listed in `tf_list`.\n",
    "       - Each entry of these genes in the matrix is multiplied by the maximum expression level \n",
    "         of its respective cell and a specified scalar value.\n",
    "    3. Updates the AnnData object with new columns:\n",
    "       - 'scaled': A boolean column indicating whether each gene is in the `tf_list`.\n",
    "       - 'scaled_by': Contains the scaling factor used for each gene (the product of the maximum \n",
    "         expression level of each cell and the scalar), or `1` if the gene was not in `tf_list`.\n",
    "    \n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols to be perturbed.\n",
    "    scalar (float): The scalar value used to scale the expression levels.\n",
    "    adata (AnnData): The AnnData object containing gene expression data.\n",
    "\n",
    "    Returns:\n",
    "    AnnData: The updated AnnData object with applied perturbations and new columns.\n",
    "    \"\"\"\n",
    "    # Save the original state of the parameter objects, in case some tfs do not translate (failsafe)\n",
    "    original_X = adata.X.copy()\n",
    "    original_gene_mask = gene_mask.copy()\n",
    "    \n",
    "    # Compute maximum expression level of each cell\n",
    "    max_exp = np.max(adata.X, axis=1)\n",
    "    \n",
    "    # Create a boolean mask for genes in tf_list\n",
    "    gene_mask = adata.var['gene_symbol'].isin(tf_list)\n",
    "\n",
    "    \"\"\"This is new today. v \"\"\"\n",
    "    # Raise an error if any of the gene names in tf_list do not match column names (we will manually update these in adata):\n",
    "    missing_genes = [gene for gene in tf_list if gene not in adata.var['gene_symbol'].values]\n",
    "    \n",
    "    if missing_genes:\n",
    "        # Restore original parameter objects\n",
    "        adata.X = original_X\n",
    "        gene_mask = original_gene_mask\n",
    "        raise ValueError(f\"Genes {missing_genes} not found in anndata object\")\n",
    "\n",
    "    else:\n",
    "        \"\"\" This is new today. ^ \"\"\"\n",
    "        # Apply the scaling operation to the specified genes\n",
    "        adata.X[:, gene_mask] *= max_exp[:, np.newaxis] * scalar\n",
    "        \n",
    "        # Add/Update 'scaled' column in var\n",
    "        adata.var['scaled'] = gene_mask\n",
    "        \n",
    "        # Add/Update 'scaled_by' column in var\n",
    "        adata.var['scaled_by'] = 1  # Default value for genes not in tf_list\n",
    "        adata.var.loc[gene_mask, 'scaled_by'] = max_exp[:, np.newaxis] * scalar  # Correct scaling factor assignment\n",
    "    \n",
    "    return adata\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15a083ac-8ec8-4b5e-8084-3299bf915df4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copied from a prev day\n",
    "\n",
    "import anndata\n",
    "\n",
    "def iterate_perturb_counts(tf_list, scalar_list, adata):\n",
    "    \"\"\"\n",
    "    Applies perturbations to the expression data of specified transcription factors across multiple scalars \n",
    "    and stores the resulting AnnData objects in a dictionary.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Iterates over a list of scalar values.\n",
    "    2. For each scalar, creates a copy of the AnnData object to preserve the original data.\n",
    "    3. Applies the `perturb_counts` function to scale the expression data of genes listed in `tf_list` by\n",
    "       the maximum gene expression of each cell and the current scalar.\n",
    "    4. Stores the perturbed AnnData object in a dictionary with the scalar as the key.\n",
    "\n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols (transcription factors) to be perturbed.\n",
    "    scalar_list (list): A list of scalar values for scaling the gene expression.\n",
    "    adata (AnnData): The AnnData object containing gene expression data (cells x genes).\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary where keys are scalar values and values are the corresponding perturbed AnnData objects.\n",
    "    \"\"\"\n",
    "    \n",
    "    adata_dict = {}\n",
    "    \n",
    "    for scalar in scalar_list:\n",
    "        # Create a copy of the AnnData object for each scalar value\n",
    "        adata_temp = adata.copy()\n",
    "        \n",
    "        # Apply perturb_counts to the copied AnnData object\n",
    "        perturbed_adata = perturb_counts(tf_list, scalar, adata_temp)\n",
    "        \n",
    "        # Store the perturbed AnnData object in the dictionary with scalar as the key\n",
    "        adata_dict[scalar] = perturbed_adata\n",
    "    \n",
    "    return adata_dict\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d543ea4-8e81-4a18-8a6a-9d58cca3794f",
   "metadata": {},
   "source": [
    "# Day 4: Copied From Josh's Notebook\n",
    "\n",
    "**Focus:** check out Nats code (debug a bit) and create a few perturbations\n",
    "- changes made to NO's code:\n",
    "    1. iterate_perturb_counts: changes the order of the arguments to `adata, tf_list, scalar_list`\n",
    "    2. perturb_counts: changes the order of the arguments to `adata, tf_list, scalar_list`\n",
    "    3. perturb_counts: there was an issue with the use of `[: np.newaxis]` with respect to `max_exp`, which is a `coo_matrix` (special type of sparse matrix). Code was modified to address an issue being thrown here.\n",
    "- new function:\n",
    "    1. validateTFs(TFs, adata): this checks if all the transcription factors are present in the adata\n",
    "- pertrubation driver (`Perform Perturbations and Create new files`):\n",
    "    1. loads Fibroblast data from Tabula Sapiens\n",
    "    2. loads `.csv` file of known reprogrmaming protocols\n",
    "    3. for each set of TFs that are validated by `validateTFs`:\n",
    "        1. use `iterate_perturb_counts` to generate perturbations with scalars `[0.5, 0.75, 1]`\n",
    "        2. concatenate the dataframes to make a single dataframe\n",
    "        3. save metadata from reprogramming protocol (i.e. PMID, source/targets, etc.)\n",
    "        4. save the new anndata as a `.h5ad` file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbd7570a-143e-42ce-9da0-092a30f2a5de",
   "metadata": {},
   "source": [
    "## Nat's Code with some modifications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec0b62fc-2de4-4c72-989b-dd5b2c2447ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import scanpy as sp\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d74e22ec-6676-4879-94a7-7ecdf360fd4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def iterate_perturb_counts(adata, tf_list, scalar_list):\n",
    "    \"\"\"\n",
    "    Applies perturbations to the expression data of specified transcription factors across multiple scalars \n",
    "    and stores the resulting AnnData objects in a dictionary.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Iterates over a list of scalar values.\n",
    "    2. For each scalar, creates a copy of the AnnData object to preserve the original data.\n",
    "    3. Applies the `perturb_counts` function to scale the expression data of genes listed in `tf_list` by\n",
    "       the maximum gene expression of each cell and the current scalar.\n",
    "    4. Stores the perturbed AnnData object in a dictionary with the scalar as the key.\n",
    "\n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols (transcription factors) to be perturbed.\n",
    "    scalar_list (list): A list of scalar values for scaling the gene expression.\n",
    "    adata (AnnData): The AnnData object containing gene expression data (cells x genes).\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary where keys are scalar values and values are the corresponding perturbed AnnData objects.\n",
    "    \"\"\"\n",
    "    \n",
    "    adata_dict = {}\n",
    "    \n",
    "    for scalar in scalar_list:\n",
    "        # Create a copy of the AnnData object for each scalar value\n",
    "        adata_temp = adata.copy()\n",
    "        \n",
    "        # Apply perturb_counts to the copied AnnData object\n",
    "        perturbed_adata = perturb_counts(adata_temp, tf_list, scalar)\n",
    "        \n",
    "        # Store the perturbed AnnData object in the dictionary with scalar as the key\n",
    "        adata_dict[scalar] = perturbed_adata\n",
    "    \n",
    "    return adata_dict\n",
    "\n",
    "def perturb_counts(adata, tf_list, scalar): \n",
    "    \"\"\"\n",
    "    Applies a perturbation to the expression data of specific genes in an AnnData object.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Computes the maximum gene expression level for each cell.\n",
    "    2. Applies a scaling operation to the expression levels of genes listed in `tf_list`.\n",
    "       - Each entry of these genes in the matrix is multiplied by the maximum expression level \n",
    "         of its respective cell and a specified scalar value.\n",
    "    3. Updates the AnnData object with new columns:\n",
    "       - 'scaled': A boolean column indicating whether each gene is in the `tf_list`.\n",
    "       - 'scaled_by': Contains the scaling factor used for each gene (the product of the maximum \n",
    "         expression level of each cell and the scalar), or `1` if the gene was not in `tf_list`.\n",
    "    \n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols to be perturbed.\n",
    "    scalar (float): The scalar value used to scale the expression levels.\n",
    "    adata (AnnData): The AnnData object containing gene expression data.\n",
    "\n",
    "    Returns:\n",
    "    AnnData: The updated AnnData object with applied perturbations and new columns.\n",
    "    \"\"\"\n",
    "\n",
    "    # Create a boolean mask for genes in tf_list\n",
    "    gene_mask = adata.var['gene_symbol'].isin(tf_list)\n",
    "    \n",
    "    # Save the original state of the parameter objects, in case some tfs do not translate (failsafe)\n",
    "    original_X = adata.X.copy()\n",
    "    original_gene_mask = gene_mask.copy()\n",
    "    \n",
    "    # Compute maximum expression level of each cell\n",
    "    max_exp = np.max(adata.X, axis=1)\n",
    "\n",
    "    \"\"\"This is new today. v \"\"\"\n",
    "    # Raise an error if any of the gene names in tf_list do not match column names (we will manually update these in adata):\n",
    "    missing_genes = [gene for gene in tf_list if gene not in adata.var['gene_symbol'].values]\n",
    "    \n",
    "    if missing_genes:\n",
    "        # Restore original parameter objects\n",
    "        adata.X = original_X\n",
    "        gene_mask = original_gene_mask\n",
    "        raise ValueError(f\"Genes {missing_genes} not found in anndata object\")\n",
    "\n",
    "    else:    \n",
    "        \n",
    "        # Apply the scaling operation to the specified genes\n",
    "        adata.X[:, gene_mask] = max_exp * scalar\n",
    "        \n",
    "        # Add/Update 'scaled' column in var\n",
    "        adata.var['scaled'] = gene_mask\n",
    "        \n",
    "        # Add/Update 'scaled_by' column in var\n",
    "        adata.var['scaled_by'] = scalar  # Default value for genes not in tf_list\n",
    "    \n",
    "    return adata\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f5645f9-2865-4256-9a67-16714be224ba",
   "metadata": {},
   "source": [
    "## New code to validate lists of TFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8458b5b2-728b-4868-91ac-dd439523a620",
   "metadata": {},
   "outputs": [],
   "source": [
    "def validateTFs(TFs, adata):\n",
    "    adata_gene_list = adata.var['gene_symbol'].values.tolist()\n",
    "    for TF in TFs:\n",
    "        if TF not in adata_gene_list:\n",
    "            return False\n",
    "    return True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7cf8a3e-cbf6-4a03-810d-52d07b426076",
   "metadata": {},
   "source": [
    "## Load data and perturbations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a2da050-9ac3-4a40-82fe-ff3c9f125689",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('data/first_5_recepies_8_29_2024.csv')\n",
    "\n",
    "DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/jpic/\"\n",
    "FILE = \"fibroblast.h5ad\"\n",
    "adata = sp.read_h5ad(os.path.join(DATAPATH, FILE))\n",
    "adata_gene_list = adata.var['gene_symbol'].values.tolist()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7754f880-759d-4ceb-ad5e-c7aec8535f99",
   "metadata": {},
   "source": [
    "## Perform Perturbations and Create new files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aacf265d-6a26-476b-826a-08997f1c40eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df['TFs'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b5c916f-8cca-4155-a81b-318d253ca6a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_directory = \"/nfs/turbo/umms-indikar/shared/projects/DARPA_AI/in-silico-reprogramming/one-shot/perturbed\"\n",
    "scalars = [0.5, 0.75, 1.001]\n",
    "for i in range(len(df['TFs'])):\n",
    "    val = df['TFs'].iloc[i]\n",
    "    val = val.replace(',',' ')\n",
    "    val = val.replace(';',' ')\n",
    "    val = val.replace(':',' ')\n",
    "    TFs = val.split(' ')\n",
    "    if validateTFs(TFs, adata):\n",
    "        print(TFs)\n",
    "        adataDict = iterate_perturb_counts(adata, TFs, scalars)\n",
    "\n",
    "        # Concatenate all AnnData objects along the observations axis\n",
    "        concatenated_adata = ad.concat(list(adataDict.values()), axis=0)\n",
    "\n",
    "        # Save reprogramming metadata into the concatenated_adata.obs table\n",
    "        concatenated_adata.obs['Source_cells'] = df['Source cells'].iloc[i]\n",
    "        concatenated_adata.obs['Target_cells'] = df['Target cells'].iloc[i]\n",
    "        concatenated_adata.obs['Treatment'] = df['Treatment'].iloc[i]\n",
    "        concatenated_adata.obs['Species'] = df['Species'].iloc[i]\n",
    "        concatenated_adata.obs['Cell_Transplantation'] = df['Cell Transplantation'].iloc[i]\n",
    "        concatenated_adata.obs['Published_Year'] = df['Published Year'].iloc[i]\n",
    "        concatenated_adata.obs['PMID'] = df['PMID'].iloc[i]\n",
    "        \n",
    "        # Join the TFs list into a string for the filename\n",
    "        TFs_str = \"_\".join(TFs)\n",
    "        \n",
    "        # Generate the file path for saving\n",
    "        file_name = f\"{TFs_str}.h5ad\"\n",
    "        output_path = os.path.join(output_directory, file_name)\n",
    "        \n",
    "        # Save the concatenated AnnData object to the file\n",
    "        concatenated_adata.write_h5ad(output_path)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52a3a04e-5927-424b-9a17-fb698e0d5d4b",
   "metadata": {},
   "source": [
    "# Day 3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dae743d0-4445-45da-aa67-60dbfd13aa6e",
   "metadata": {},
   "source": [
    "# Day 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ca734bb-76c1-453b-b0bb-e2caaff6be0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Copied from a prev day\n",
    "\n",
    "DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/\"\n",
    "FILE = \"TS_epithelial.h5ad\"\n",
    "adata = sp.read_h5ad(os.path.join(DATAPATH, FILE))\n",
    "adata_gene_list = adata.var['gene_symbol'].values.tolist()\n",
    "\n",
    "# Print the first 5 entries\n",
    "print(\"First 5 entries:\")\n",
    "print(adata_gene_list[:5])\n",
    "\n",
    "# Print the last 5 entries\n",
    "print(\"Last 5 entries:\")\n",
    "print(adata_gene_list[-5:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c44c1301-ecaa-40f0-bcc4-aba1e08e04e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# outside the function so I can manipulate these directly.\n",
    "# I also checked whether it matters if I do case insensitive (capitalize everything then compare) or case sensitive.\n",
    "# Unsurprisingly, case sensitive has more discrepancies (45 vs. 44), with the one extra that was picked up as \"Ptf1a\".\n",
    "# I left the case insensitive version. \n",
    "\n",
    "# get a list of words (potential genes, also includes and, + , etc. ) from the table from the review paper\n",
    "table_1_df = pd.read_csv(\"/home/oliven/scFoundationModels/notebooks/reprogramming/data/table_1_data_from_paper_9_1.csv\")\n",
    "combined_string = ' '.join(table_1_df['TFs'].astype(str)).replace(',', '')\n",
    "word_list = combined_string.split()\n",
    "\n",
    "# get a list of genes that appear in the data matrix\n",
    "DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/\"\n",
    "FILE = \"TS_epithelial.h5ad\"\n",
    "adata = sp.read_h5ad(os.path.join(DATAPATH, FILE))\n",
    "adata_gene_list = adata.var['gene_symbol'].values.tolist()\n",
    "\n",
    "# New today\n",
    "def check_valid_tfs(word_list, adata_gene_list):\n",
    "\n",
    "    word_list_upper = [word.upper() for word in word_list]\n",
    "    adata_gene_list_upper = [gene.upper() for gene in adata_gene_list]\n",
    "\n",
    "    # print what does not overlap\n",
    "    not_valid_gene = set(word_list_upper) - set(adata_gene_list_upper)\n",
    "\n",
    "    print(\"Entries in the table that are not genes in the counts matrix: \")\n",
    "    \n",
    "    return list(not_valid_gene)\n",
    "    \n",
    "check_valid_tfs(word_list, adata_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62011aea-3548-4942-847a-d17616055966",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for those genes with multiple aliases, checking which are valid\n",
    "# we have to worry about making this case insensitive just in case\n",
    "multiple_alias_dict = {}\n",
    "multiple_alias_dict.update({\n",
    "    'p53': ['BCC7','BMFS5', 'LFS1', 'TRP53'],\n",
    "    'OCT3/4': ['POU5F1', 'OCT3', 'OCT4', 'OTF4', 'MGC22487'],\n",
    "    'MASH1': ['HASH1', 'BHLHa46', 'ASH1', 'ASH-1', 'ASCL1'],\n",
    "    'HNF6': ['HNF6', 'HNF6A', 'ONECUT1'],\n",
    "    'HB9': ['MNX1', 'HOXHB9', 'SCRA1', 'HLXB9', 'GC07M156491', 'GC07M156786', 'GC07M150530'],\n",
    "    'PPARG2': ['PPARG', 'NR1C3', 'PPARG1', 'PPARgamma', 'PPAR-Gamma', 'PPARG5', 'CIMT1', 'GLM1'], \n",
    "    'PU.1': ['SPI1', 'SPI-A', 'SFPI1', 'SPI-1', 'OF', 'AGM10'],\n",
    "    'N-MYC': ['MYCN', 'BHLHe37', 'N-Myc', 'NMYC', 'MYCNOT', 'MYCNsORF', 'MYCNsPEP', 'BHLHE37', 'FGLDS1', 'MODED', 'MPAPA', 'ODED'],\n",
    "    # oct9 had a strange genecards lookup\n",
    "    'OCT9': ['POU3F4', 'SLC22A16']\n",
    "    'LEF-1': ['TCF1ALPHA', 'TCF7L3', 'TCF10', 'LEF1'],\n",
    "    # these next two were listed as ER71/ETV2\n",
    "    'ER71/ETV2': ['ER71', 'ETV2', 'ETSRP71'],\n",
    "    # sv40 had a strange genecards lookup\n",
    "    'SV40': [''],\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81f439e6-738c-4cba-8067-fc8a28af8a5f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b1df6ef-73f6-4d86-8dfe-335d4a067b3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# day 4, technically (migrate this)\n",
    "adata_gene_list = adata.var['gene_symbol'].values.tolist()\n",
    "def identify_gene_name_translations(multiple_alias_dict, adata_gene_list):\n",
    "    \"\"\"\n",
    "    Identifies whether any values from multiple_alias_dict appear in adata_gene_list, case-insensitive.\n",
    "    \n",
    "    Parameters:\n",
    "    multiple_alias_dict (dict): A dictionary where keys are gene names and values are lists of aliases.\n",
    "    adata_gene_list (list): A list of gene names to check against, case-insensitive.\n",
    "\n",
    "    Prints:\n",
    "    For each key, whether it was found in the gene list along with the matching values.\n",
    "    \"\"\"\n",
    "    # Convert the gene list to uppercase for case-insensitive comparison\n",
    "    adata_gene_list_upper = [gene.upper() for gene in adata_gene_list]\n",
    "\n",
    "    # Loop through each key and values in the dictionary\n",
    "    for key, values in multiple_alias_dict.items():\n",
    "        # Convert each alias to uppercase\n",
    "        values_upper = [value.upper() for value in values]\n",
    "        \n",
    "        # Check if any alias is present in the gene list\n",
    "        found_values = [value for value in values_upper if value in adata_gene_list_upper]\n",
    "        \n",
    "        # Print appropriate message based on whether any values were found\n",
    "        if found_values:\n",
    "            print(f\"{key} was found in gene list as {found_values}\")\n",
    "        else:\n",
    "            print(f\"{key} was not found in gene list.\")\n",
    "\n",
    "identify_gene_name_translations(multiple_alias_dict, adata_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac00a2c8-771e-463a-bdbb-3668d31b6868",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The above list is small enough that I can manually check it.\n",
    "\n",
    "# Things to remove from word_list:\n",
    "not_genes = ['Variant', 'Large',  '(ETS', '2)', 'Knockdown', ]\n",
    "\n",
    "# Valid genes to replace/rename in word_list:\n",
    "genes_to_translate = ['LMX1A;', #'P53']\n",
    "translated_names = ['LMX1A',]\n",
    "\n",
    "# These ones might have appeared as multiple entries, etc. b/c of spacing. easiest way was to delete and add back\n",
    "genes_to_add = ['ETS2',]\n",
    "\n",
    "# replace then subtract and add. [--------------]\n",
    "new_word_list = set(word_list) - set(not_genes)\n",
    "\n",
    "# Running the function one more time to check:\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1bed245-f122-47cd-bc76-814e682c23a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" Renamed med_nonz to max_exp to be more accurate. \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dee7615-c3ae-4b05-a473-e3274728c49b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copied from a prev day\n",
    "\n",
    "\"\"\"\n",
    "Josh, please read: the adata.obs['scalar'] = scalar copies the scalar down for that call, associated with every cell in X. Same with ['scaled'] and ['scaled_by']\n",
    "in var. This is good in case the data is later appended into one anndata object.\n",
    "But my return from the perturb_counts loop (cell below this) is a dictionary of all of the perturb_counts, since appending along any axis will probably either overwrite\n",
    "obs or var.\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "def perturb_counts(tf_list, scalar, adata): \n",
    "    \"\"\"\n",
    "    Applies a perturbation to the expression data of specific genes in an AnnData object.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Computes the maximum gene expression level for each cell.\n",
    "    2. Applies a scaling operation to the expression levels of genes listed in `tf_list`.\n",
    "       - Each entry of these genes in the matrix is multiplied by the maximum expression level \n",
    "         of its respective cell and a specified scalar value.\n",
    "    3. Updates the AnnData object with new columns:\n",
    "       - 'scaled': A boolean column indicating whether each gene is in the `tf_list`.\n",
    "       - 'scaled_by': Contains the scaling factor used for each gene (the product of the maximum \n",
    "         expression level of each cell and the scalar), or `1` if the gene was not in `tf_list`.\n",
    "    \n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols to be perturbed.\n",
    "    scalar (float): The scalar value used to scale the expression levels.\n",
    "    adata (AnnData): The AnnData object containing gene expression data.\n",
    "\n",
    "    Returns:\n",
    "    AnnData: The updated AnnData object with applied perturbations and new columns.\n",
    "    \"\"\"\n",
    "    # Save the original state of the parameter objects, in case some tfs do not translate (failsafe)\n",
    "    original_X = adata.X.copy()\n",
    "    original_gene_mask = gene_mask.copy()\n",
    "    \n",
    "    # Compute maximum expression level of each cell\n",
    "    max_exp = np.max(adata.X, axis=1)\n",
    "    \n",
    "    # Create a boolean mask for genes in tf_list\n",
    "    gene_mask = adata.var['gene_symbol'].isin(tf_list)\n",
    "\n",
    "    \"\"\"This is new today. v \"\"\"\n",
    "    # Raise an error if any of the gene names in tf_list do not match column names (we will manually update these in adata):\n",
    "    missing_genes = [gene for gene in tf_list if gene not in adata.var['gene_symbol'].values]\n",
    "    \n",
    "    if missing_genes:\n",
    "        # Restore original parameter objects\n",
    "        adata.X = original_X\n",
    "        gene_mask = original_gene_mask\n",
    "        raise ValueError(f\"Genes {missing_genes} not found in anndata object\")\n",
    "\n",
    "    else:\n",
    "        \"\"\" This is new today. ^ \"\"\"\n",
    "        # Apply the scaling operation to the specified genes\n",
    "        adata.X[:, gene_mask] *= max_exp[:, np.newaxis] * scalar\n",
    "        \n",
    "        # Add/Update 'scaled' column in var\n",
    "        adata.var['scaled'] = gene_mask\n",
    "        \n",
    "        # Add/Update 'scaled_by' column in var\n",
    "        adata.var['scaled_by'] = 1  # Default value for genes not in tf_list\n",
    "        adata.var.loc[gene_mask, 'scaled_by'] = max_exp[:, np.newaxis] * scalar  # Correct scaling factor assignment\n",
    "    \n",
    "    return adata\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6eeb06cf-f71b-4c2c-b8a2-9d7b81346258",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copied from a prev day\n",
    "\n",
    "import anndata\n",
    "\n",
    "def iterate_perturb_counts(tf_list, scalar_list, adata):\n",
    "    \"\"\"\n",
    "    Applies perturbations to the expression data of specified transcription factors across multiple scalars \n",
    "    and stores the resulting AnnData objects in a dictionary.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Iterates over a list of scalar values.\n",
    "    2. For each scalar, creates a copy of the AnnData object to preserve the original data.\n",
    "    3. Applies the `perturb_counts` function to scale the expression data of genes listed in `tf_list` by\n",
    "       the maximum gene expression of each cell and the current scalar.\n",
    "    4. Stores the perturbed AnnData object in a dictionary with the scalar as the key.\n",
    "\n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols (transcription factors) to be perturbed.\n",
    "    scalar_list (list): A list of scalar values for scaling the gene expression.\n",
    "    adata (AnnData): The AnnData object containing gene expression data (cells x genes).\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary where keys are scalar values and values are the corresponding perturbed AnnData objects.\n",
    "    \"\"\"\n",
    "    \n",
    "    adata_dict = {}\n",
    "    \n",
    "    for scalar in scalar_list:\n",
    "        # Create a copy of the AnnData object for each scalar value\n",
    "        adata_temp = adata.copy()\n",
    "        \n",
    "        # Apply perturb_counts to the copied AnnData object\n",
    "        perturbed_adata = perturb_counts(tf_list, scalar, adata_temp)\n",
    "        \n",
    "        # Store the perturbed AnnData object in the dictionary with scalar as the key\n",
    "        adata_dict[scalar] = perturbed_adata\n",
    "    \n",
    "    return adata_dict\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddd82877-b10f-48f4-894a-ebdf8117d831",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We want each recipe (returned as a dictionary of anndata objects, one adata object for each scalar)\n",
    "def save_perturb_to_turbo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56e0a638-b15e-4589-94d6-0275ada43a5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Testing on one of the tf lists from the file.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cad2d9cc-78c2-4d3c-a3c2-b0710a27000c",
   "metadata": {},
   "source": [
    "# Day 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bb10707-df05-4e7d-a9a4-bd301ab5cd69",
   "metadata": {},
   "source": [
    "## Perturbation Model Discussion\n",
    "\n",
    "E.V. = expression values\n",
    "\n",
    "Possible algorithm:\n",
    "```\n",
    "1. find highest E.V.  for a single cell\n",
    "2. find expression value of TFs being modified\n",
    "3. have a value k for the number of different concentrations we want to test\n",
    "4. choose k different amounts to increase the TFs from there measured E.V. to the 150% maximum E.V.\n",
    "   - make an arbitray choice and code it up\n",
    "```\n",
    "\n",
    "**A reasonable person could write this 10s of different ways**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "206882c6-cec6-4a8b-9605-bfdc8b058823",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "# median nonzero value of each row\n",
    "# the w stands for working. I just dont want to screw up the original.\n",
    "DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/\"\n",
    "FILE = \"TS_epithelial.h5ad\"\n",
    "adata_w = sp.read_h5ad(os.path.join(DATAPATH, FILE))\n",
    "adata_w\n",
    "# def median_nonzero(col):\n",
    "#     nonzero_vals = col[col != 0]  # Extract nonzero values\n",
    "#     if len(nonzero_vals) == 0:    # If no nonzero values, return NaN\n",
    "#         return np.nan\n",
    "#     return np.median(nonzero_vals)\n",
    "\n",
    "# # Apply the function to each column and store the results\n",
    "# med_nonz = np.apply_along_axis(median_nonzero, axis=0, arr=adata_w.X)\n",
    "# adata_w.var['med_nonz'] = med_nonz\n",
    "# adata_w.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e8d3635-dfd9-4412-9d95-cfcbfdf0fcc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "\n",
    "X = adata_w.X.toarray() if not isinstance(adata_w.X, np.ndarray) else adata_w.X\n",
    "\n",
    "def median_nonzero(col):\n",
    "    nonzero_vals = col[col != 0] \n",
    "    return np.median(nonzero_vals) if len(nonzero_vals) > 0 else 0\n",
    "\n",
    "#perform and save results of fn\n",
    "med_nonz = np.apply_along_axis(median_nonzero, axis=0, arr=X)\n",
    "adata_w.var['med_nonz'] = med_nonz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06553f13-9b94-4a5f-9039-bcf183b14ac6",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_w.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92f2c287-8651-4280-b369-60bcae264beb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# seeing what it looks like before tf_list changes the first 3 rows\n",
    "# Convert to dense if it's sparse and display the first five rows\n",
    "import numpy as np\n",
    "\n",
    "# Convert to a dense array if necessary\n",
    "dense_X = adata_w.X.toarray() if not isinstance(adata_w.X, np.ndarray) else adata_w.X\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "621349c1-2e81-4842-b701-ef86075bebc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# problem, scaling by a factor of the max expressed gene in that cell means that you could be scaling by different genes for each cell,\n",
    "# when the cells are all of the same type. for each get the median nonzero expression\n",
    "\n",
    "# for testing purposes: v\n",
    "tf_list = ['DDX11L1', 'WASH7P', 'MIR6859-1']\n",
    "tf = 'DDX11L1'\n",
    "# for testing purposes: ^\n",
    "\n",
    "#mask = adata_w.obs_names.isin(tf_list)\n",
    "mask = np.where(adata_w.var['gene_symbol'] == tf)[0]\n",
    "adata_w.X[mask, :] = adata_w.X[mask, :] * adata_w.var['med_nonz'].values \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dff775d-3988-4580-b7b4-c348ba958f5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/\"\n",
    "FILE = \"TS_epithelial.h5ad\"\n",
    "adata = sp.read_h5ad(os.path.join(DATAPATH, FILE))\n",
    "(adata_w.X[mask, :] - adata.X[mask, :]).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78ea8504-7206-427a-b4ad-5d0a58f8fff4",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_w.X[mask, :]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef84ecb9-b34c-4db6-82e2-b5d99fa1a35e",
   "metadata": {},
   "source": [
    "### Scaling by median nonzero entry of each gene across all cells (old)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "073a102f-4053-420a-9d04-b21410d9853d",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "# import numpy as np\n",
    "# DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/\"\n",
    "# FILE = \"TS_epithelial.h5ad\"\n",
    "# adata_w = sp.read_h5ad(os.path.join(DATAPATH, FILE))\n",
    "# #tf = 'DDX11L1'\n",
    "# tf_list =['DDX11L1', 'WASH7P', 'MIR6859-1']\n",
    "\n",
    "# def median_nonzero(col):\n",
    "#     nonzero_vals = col[col != 0] \n",
    "#     return np.median(nonzero_vals) if len(nonzero_vals) > 0 else 0\n",
    "\n",
    "# # requires scalar is a scalar\n",
    "# def perturb_counts(tf_list, scalar, adata): \n",
    "#     # compute nonzero median expression of each gene across cells, save to var\n",
    "#     med_nonz = np.apply_along_axis(median_nonzero, axis=0, arr=adata.X)\n",
    "#     adata.var['med_nonz'] = med_nonz\n",
    "\n",
    "#     # filter by desired tf(s), and apply the nonzero_median scaling operation to only these  \n",
    "#     mask = np.where(adata.var['gene_symbol'].isin(tf_list))[0]\n",
    "#     adata.X[mask, :] = adata.X[mask, :] * adata.var['med_nonz'].values * scalar\n",
    "#     return adata\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fde1fc75-0e49-4800-9c33-d147e4bac06d",
   "metadata": {},
   "source": [
    "### Scaling by max gene expression within each cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7ebbee6-3805-417a-9425-e99f2b8b35db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# old version with (I believe) improper mask that was being applied to rows and not columns\n",
    "\n",
    "# import numpy as np\n",
    "# DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/\"\n",
    "# FILE = \"TS_epithelial.h5ad\"\n",
    "# adata_w = sp.read_h5ad(os.path.join(DATAPATH, FILE))\n",
    "# #tf = 'DDX11L1'\n",
    "# tf_list =['DDX11L1', 'WASH7P', 'MIR6859-1']\n",
    "\n",
    "\n",
    "# # requires scalar is a scalar\n",
    "# def perturb_counts(tf_list, scalar, adata): \n",
    "#     # compute nonzero median expression of each gene across cells, save to var\n",
    "#     med_nonz = med_nonz = np.max(adata.X, axis=1)\n",
    "#     adata.obs['med_nonz'] = med_nonz\n",
    "\n",
    "#     # filter by desired tf(s), and apply the nonzero_median scaling operation to only these  \n",
    "#     mask = np.where(adata.var['gene_symbol'].isin(tf_list))[0]\n",
    "#     adata.X[mask, :] = adata.X[mask, :] * adata.obs['med_nonz'].values * scalar\n",
    "#     return adata\n",
    "\n",
    "# old version without extra obs and var rows telling what was scaled and by how much\n",
    "\n",
    "\n",
    "# def perturb_counts(tf_list, scalar, adata): \n",
    "#     # Compute maximum expression level of each cell and save it to obs\n",
    "#     med_nonz = np.max(adata.X, axis=1)\n",
    "#     adata.obs['med_nonz'] = med_nonz\n",
    "    \n",
    "#     # apply operation only to genes in tf_list\n",
    "#     mask = np.where(adata.var['gene_symbol'].isin(tf_list))[0]\n",
    "#     adata.X[:, mask] = adata.X[:, mask] * adata.obs['med_nonz'].values[:, np.newaxis] * scalar\n",
    "    \n",
    "#     return adata\n",
    "\n",
    "# version before gpt optimized\n",
    "\n",
    "# def perturb_counts(tf_list, scalar, adata): \n",
    "#     # Compute maximum expression level of each cell and save it to obs\n",
    "#     med_nonz = np.max(adata.X, axis=1)\n",
    "#     adata.obs['med_nonz'] = med_nonz\n",
    "    \n",
    "#     # Add a new obs column called 'scalar' containing the scalar value for each row\n",
    "#     adata.obs['scalar'] = scalar\n",
    "    \n",
    "#     # Create a mask for genes in tf_list\n",
    "#     mask = np.where(adata.var['gene_symbol'].isin(tf_list))[0]\n",
    "#     # Apply the scaling operation to the specified genes\n",
    "#     adata.X[:, mask] = adata.X[:, mask] * adata.obs['med_nonz'].values[:, np.newaxis] * scalar\n",
    "    \n",
    "#     # Add a new var column called 'scaled' with True for genes in tf_list and False otherwise\n",
    "#     adata.var['scaled'] = adata.var['gene_symbol'].isin(tf_list)\n",
    "    \n",
    "#     # Add a new var column 'scaled_by'\n",
    "#     adata.var['scaled_by'] = 1\n",
    "#     # Set scaling factor for genes in tf_list\n",
    "\n",
    "#     ############I asked gpt to do this line and am unsure if it is correct. checking now.\n",
    "#     adata.var.loc[adata.var['scaled'], 'scaled_by'] = adata.obs['med_nonz'].values[:, np.newaxis] * scalar\n",
    "#     ###############\n",
    "    \n",
    "#     return adata\n",
    "\n",
    "# loop version before gpt optimized\n",
    "\n",
    "# # requires that within each perturbation, all of the transcription factors in tf_list are scaled by the same amount, that is, (scalar * \"max gene expression in that cell\")\n",
    "# # requires adata is cells x genes\n",
    "\n",
    "# import anndata\n",
    "\n",
    "# def iterate_perturb_counts(tf_list, scalar_list, adata):\n",
    "#     adata_dict = {}\n",
    "    \n",
    "#     for scalar in scalar_list:\n",
    "#         adata_temp = adata.copy()\n",
    "#         perturbed_adata = perturb_counts(tf_list, scalar, adata_temp)\n",
    "#         adata_dict[scalar] = perturbed_adata\n",
    "    \n",
    "#     return adata_dict\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99114ac8-38af-488b-8856-acc26632b46a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\"\"\"\n",
    "Josh, please read: the adata.obs['scalar'] = scalar copies the scalar down for that call, associated with every cell in X. Same with ['scaled'] and ['scaled_by']\n",
    "in var. This is good in case the data is later appended into one anndata object.\n",
    "But my return from the perturb_counts loop (cell below this) is a dictionary of all of the perturb_counts, since appending along any axis will probably either overwrite\n",
    "obs or var.\n",
    "\"\"\"\n",
    "import numpy as np\n",
    "\n",
    "def perturb_counts(tf_list, scalar, adata): \n",
    "    \"\"\"\n",
    "    Applies a perturbation to the expression data of specific genes in an AnnData object.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Computes the maximum gene expression level for each cell.\n",
    "    2. Applies a scaling operation to the expression levels of genes listed in `tf_list`.\n",
    "       - Each entry of these genes in the matrix is multiplied by the maximum expression level \n",
    "         of its respective cell and a specified scalar value.\n",
    "    3. Updates the AnnData object with new columns:\n",
    "       - 'scaled': A boolean column indicating whether each gene is in the `tf_list`.\n",
    "       - 'scaled_by': Contains the scaling factor used for each gene (the product of the maximum \n",
    "         expression level of each cell and the scalar), or `1` if the gene was not in `tf_list`.\n",
    "    \n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols to be perturbed.\n",
    "    scalar (float): The scalar value used to scale the expression levels.\n",
    "    adata (AnnData): The AnnData object containing gene expression data.\n",
    "\n",
    "    Returns:\n",
    "    AnnData: The updated AnnData object with applied perturbations and new columns.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Compute maximum expression level of each cell\n",
    "    med_nonz = np.max(adata.X, axis=1)\n",
    "    \n",
    "    # Create a boolean mask for genes in tf_list\n",
    "    gene_mask = adata.var['gene_symbol'].isin(tf_list)\n",
    "    \n",
    "    # Apply the scaling operation to the specified genes\n",
    "    adata.X[:, gene_mask] *= med_nonz[:, np.newaxis] * scalar\n",
    "    \n",
    "    # Add/Update 'scaled' column in var\n",
    "    adata.var['scaled'] = gene_mask\n",
    "    \n",
    "    # Add/Update 'scaled_by' column in var\n",
    "    adata.var['scaled_by'] = 1  # Default value for genes not in tf_list\n",
    "    adata.var.loc[gene_mask, 'scaled_by'] = med_nonz[:, np.newaxis] * scalar  # Correct scaling factor assignment\n",
    "    \n",
    "    return adata\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4a79c50-3dbb-4d6f-8723-e6f967f41893",
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata\n",
    "\n",
    "def iterate_perturb_counts(tf_list, scalar_list, adata):\n",
    "    \"\"\"\n",
    "    Applies perturbations to the expression data of specified transcription factors across multiple scalars \n",
    "    and stores the resulting AnnData objects in a dictionary.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Iterates over a list of scalar values.\n",
    "    2. For each scalar, creates a copy of the AnnData object to preserve the original data.\n",
    "    3. Applies the `perturb_counts` function to scale the expression data of genes listed in `tf_list` by\n",
    "       the maximum gene expression of each cell and the current scalar.\n",
    "    4. Stores the perturbed AnnData object in a dictionary with the scalar as the key.\n",
    "\n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols (transcription factors) to be perturbed.\n",
    "    scalar_list (list): A list of scalar values for scaling the gene expression.\n",
    "    adata (AnnData): The AnnData object containing gene expression data (cells x genes).\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary where keys are scalar values and values are the corresponding perturbed AnnData objects.\n",
    "    \"\"\"\n",
    "    \n",
    "    adata_dict = {}\n",
    "    \n",
    "    for scalar in scalar_list:\n",
    "        # Create a copy of the AnnData object for each scalar value\n",
    "        adata_temp = adata.copy()\n",
    "        \n",
    "        # Apply perturb_counts to the copied AnnData object\n",
    "        perturbed_adata = perturb_counts(tf_list, scalar, adata_temp)\n",
    "        \n",
    "        # Store the perturbed AnnData object in the dictionary with scalar as the key\n",
    "        adata_dict[scalar] = perturbed_adata\n",
    "    \n",
    "    return adata_dict\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a06f0a9c-18e0-4d3a-8efd-7993accb1048",
   "metadata": {},
   "source": [
    "## Visualize Input Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8542f13-3c5c-4130-9766-0d0a9ce55e61",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6474b05-224f-4677-902e-5b4962a4bf23",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/\"\n",
    "FILE = \"TS_epithelial.h5ad\"\n",
    "\n",
    "adata = sp.read_h5ad(os.path.join(DATAPATH, FILE))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "047a5de2-be21-489c-8184-b90f010b06a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f010c3f2-3294-4f18-bcad-8a9b3abf9b2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "030f83ed-4955-4962-a629-f003027303e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X.max(axis=1) # what is the value of the highest expressed gene for each cell?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c210614f-0e87-4a8a-b1c5-1cc95c7380ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "TF = 'DDX11L1'\n",
    "index = np.where(adata.var['gene_symbol'] == TF)[0]\n",
    "index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dc3d5ad-f2f0-446d-83f3-c63179d369c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var['gene_symbol']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45904679-08f7-4ec7-b746-4017140120c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82aa3850-58e1-41b4-9b2a-60db4bea28b1",
   "metadata": {},
   "source": [
    "## Build driver"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1563c2da-5366-4738-a33b-2bfe1e881eb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def main(job_number, parameter_file):\n",
    "    \"\"\"\n",
    "    This is the main function for the array job to perform the reprogramming experiment. job_number is a single parameter\n",
    "    that will be used to look up in a parameter table which model, reprogramming recipe, and other information relevant\n",
    "    to the test.\n",
    "    \"\"\"\n",
    "\n",
    "    # Determine embedding parameters and recipie\n",
    "    df_embedding_parameters = pd.read_csv(parameter_file)\n",
    "    TFs    = df_embedding_parameters['TFs'].values[job_number]\n",
    "    model  = df_embedding_parameters['model'].values[job_number]\n",
    "    source = df_embedding_parameters['source'].values[job_number]\n",
    "    target = df_embedding_parameters['target'].values[job_number]\n",
    "\n",
    "    # Load the source data\n",
    "    adata = \n",
    "\n",
    "    # Perturb the data\n",
    "    perturbed_adata = perturbation_model(adata, TFs)\n",
    "\n",
    "    # Generate embeddings\n",
    "    if model == 'geneformer':\n",
    "        adata_embedded = embed_geneformer([source_adata, perturbed_adata, target_adata])\n",
    "    elif model == 'tGPT':\n",
    "        adata_embedded = embed_tGPT([source_adata, perturbed_adata, target_adata])\n",
    "    elif model == 'scGTP':\n",
    "        adata_embedded = embed_scGTP([source_adata, perturbed_adata, target_adata])\n",
    "\n",
    "    # Save the results to a file\n",
    "\n",
    "    return 0\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee4f3c98-3474-4971-a559-f0a294dba782",
   "metadata": {},
   "source": [
    "## Build parameter dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be9f04aa-ab4c-44fc-86e0-f69d7f21451d",
   "metadata": {},
   "outputs": [],
   "source": [
    "embedding_parameters = {\n",
    "    'source': [],\n",
    "    'target': [],\n",
    "    'TFs'   : [],\n",
    "    'model' : []\n",
    "}\n",
    "models = ['geneformer', 'tGPT', 'scGTP']\n",
    "\n",
    "df = pd.read_csv('data/first_5_recepies_8_29_2024.csv')\n",
    "\n",
    "for i in range(5):\n",
    "    TFs = df['TFs'].values[i].split()\n",
    "    source = df['Source'].values[i]\n",
    "    target = df['Target'].values[i]\n",
    "    for model in models:\n",
    "        embedding_parameters['TFs'].append(TFs)\n",
    "        embedding_parameters['source'].append(source)\n",
    "        embedding_parameters['target'].append(source)\n",
    "        embedding_parameters['model'].append(model)\n",
    "\n",
    "df_embedding_parameters = pd.DataFrame(embedding_parameters)\n",
    "\n",
    "df_embedding_parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ad22cdb-f1ac-48b6-8053-b69d91309c76",
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08e02020-7129-4182-9bef-6535490e26a9",
   "metadata": {},
   "source": [
    "# Day 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c18574c-9e46-4c1d-9b77-b4f888ae0430",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "624a5d48-99dc-4880-809d-519448cbfb82",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load known reprogramming regiems\n",
    "df = pd.read_csv('data/known-regiems-T1.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7457114f-57d5-45e9-ad8d-ecaa06a76d1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get list of unique transcription factors\n",
    "TFs = []\n",
    "for regime in df['TFs'].unique():\n",
    "    TFs += regime.replace(',', '').split()\n",
    "TFs = list(set(TFs))\n",
    "print(f\"{len(TFs)=}\")\n",
    "TFs"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
