{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c53d654e-3d58-4854-8684-52ad2b603313",
   "metadata": {},
   "source": [
    "# Tables to Describe Tabula Sapiens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "0865a76d-09c0-41b4-9af2-ba7becf415fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import numpy\n",
    "import pickle\n",
    "from scipy.sparse import issparse, csr_matrix\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pickle\n",
    "\n",
    "def getFiles(category=None, model=None):\n",
    "    h5ad_files = [] # list of all files\n",
    "    targets    = [] # list of target\n",
    "    for root, dirs, files in os.walk(DATAPATH):\n",
    "        for file in files:\n",
    "            if file.endswith('.h5ad'):\n",
    "                h5ad_files.append(os.path.join(root, file))\n",
    "    for root, dirs, files in os.walk(EMBPATH):\n",
    "        for file in files:\n",
    "            if file.endswith('.h5ad'):\n",
    "                h5ad_files.append(os.path.join(root, file))\n",
    "    for file in h5ad_files:\n",
    "        if category is None and model in file:\n",
    "            targets.append(file)\n",
    "        elif model is None and category in file:\n",
    "            targets.append(file)\n",
    "        elif model == 'raw' and category in file:\n",
    "            if 'geneformer' not in file and 'scgpt' not in file and 'Tgpt' not in file:\n",
    "                targets.append(file)\n",
    "        elif model is not None and category is not None and model in file and category in file:\n",
    "            targets.append(file)\n",
    "    return targets\n",
    "\n",
    "def format_number(x):\n",
    "    if isinstance(x, (float, np.float64)):\n",
    "        return f\"{x:.2g}\"\n",
    "    return x\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f958bca2-f3f9-40f5-9488-167e87699ac5",
   "metadata": {},
   "source": [
    "## Assemble Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a520a601-b170-4fee-8e09-d8c83d340589",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data has been loaded from: /nfs/turbo/umms-indikar/shared/projects/foundation_models/experiments/tabulaSapiens/data_summary.pkl\n",
      "You should not need to run the cells below\n"
     ]
    }
   ],
   "source": [
    "data_summary_file = '/nfs/turbo/umms-indikar/shared/projects/foundation_models/experiments/tabulaSapiens/data_summary.pkl'\n",
    "with open(data_summary_file, 'rb') as file:\n",
    "    data = pickle.load(file)\n",
    "\n",
    "print(\"Data has been loaded from: \" + data_summary_file)\n",
    "print(\"You should not need to run the cells below\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f96bb820-d141-4c6b-855f-a231961a8fa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATAPATH = \"/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/\"\n",
    "EMBPATH  = \"/nfs/turbo/umms-indikar/shared/projects/foundation_models/experiments/tabulaSapiens\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "febae2e8-855d-45a9-990a-b857c2aea3ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TS_germ line', 'TS_stromal', 'TS_Large_Intestine', 'TS_Lymph_Node', 'TS_Skin', 'TS_epithelial', 'TS_Spleen', 'TS_Vasculature', 'TS_Kidney', 'TS_Salivary_Gland', 'TS_Thymus', 'TS_Pancreas', 'TS_Fat', 'TS_Bladder', 'TS_endothelial', 'TS_Blood', 'TS_Trachea', 'TS_Bone_Marrow', 'TS_Tongue', 'TS_Lung', 'TS_Liver', 'TS_Small_Intestine', 'TS_immune', 'TS_Uterus', 'TS_Eye', 'TS_Mammary', 'TS_Prostate', 'TS_Heart', 'TS_Muscle']\n",
      "['raw', 'geneformer', 'Tgpt', 'scgpt']\n"
     ]
    }
   ],
   "source": [
    "h5ad_files = []\n",
    "for root, dirs, files in os.walk(DATAPATH):\n",
    "    for file in files:\n",
    "        if file.endswith('.h5ad'):\n",
    "            h5ad_files.append(os.path.join(root, file))\n",
    "\n",
    "tsCategories = []\n",
    "for file in h5ad_files:\n",
    "    tsCategories.append(file.split('.')[0].split('/')[-1])\n",
    "tsCategories.remove('TabulaSapiens')\n",
    "print(tsCategories)\n",
    "models = ['raw', 'geneformer', 'Tgpt', 'scgpt']\n",
    "print(models)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b75b96db-2db6-4c54-91d5-e2b623200017",
   "metadata": {},
   "source": [
    "### Rank and Effective Ranks of the Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a4e3a8a1-7788-4d53-980b-ee24a2037a3c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 10888843\n",
      "Submitted job 1/116\n",
      "Submitted batch job 10888844\n",
      "Submitted job 2/116\n",
      "Submitted batch job 10888845\n",
      "Submitted job 3/116\n",
      "Submitted batch job 10888846\n",
      "Submitted job 4/116\n",
      "Submitted batch job 10888847\n",
      "Submitted job 5/116\n",
      "Submitted batch job 10888848\n",
      "Submitted job 6/116\n",
      "Submitted batch job 10888849\n",
      "Submitted job 7/116\n",
      "Submitted batch job 10888850\n",
      "Submitted job 8/116\n",
      "Submitted batch job 10888851\n",
      "Submitted job 9/116\n",
      "Submitted batch job 10888852\n",
      "Submitted job 10/116\n",
      "Submitted batch job 10888853\n",
      "Submitted job 11/116\n",
      "Submitted batch job 10888854\n",
      "Submitted job 12/116\n",
      "Submitted batch job 10888855\n",
      "Submitted job 13/116\n",
      "Submitted batch job 10888856\n",
      "Submitted job 14/116\n",
      "Submitted batch job 10888857\n",
      "Submitted job 15/116\n",
      "Submitted batch job 10888858\n",
      "Submitted job 16/116\n",
      "Submitted batch job 10888859\n",
      "Submitted job 17/116\n",
      "Submitted batch job 10888860\n",
      "Submitted job 18/116\n",
      "Submitted batch job 10888861\n",
      "Submitted job 19/116\n",
      "Submitted batch job 10888862\n",
      "Submitted job 20/116\n",
      "Submitted batch job 10888863\n",
      "Submitted job 21/116\n",
      "Submitted batch job 10888864\n",
      "Submitted job 22/116\n",
      "Submitted batch job 10888865\n",
      "Submitted job 23/116\n",
      "Submitted batch job 10888866\n",
      "Submitted job 24/116\n",
      "Submitted batch job 10888867\n",
      "Submitted job 25/116\n",
      "Submitted batch job 10888868\n",
      "Submitted job 26/116\n",
      "Submitted batch job 10888869\n",
      "Submitted job 27/116\n",
      "Submitted batch job 10888870\n",
      "Submitted job 28/116\n",
      "Submitted batch job 10888871\n",
      "Submitted job 29/116\n"
     ]
    }
   ],
   "source": [
    "## Comput the SVD of each of dataset in each embedding space\n",
    "import subprocess\n",
    "\n",
    "# Define the number of jobs (assuming one job per file)\n",
    "num_jobs = len(tsCategories) * len(models)\n",
    "\n",
    "# SLURM parameters\n",
    "job_name       = \"TS-SVDs\"\n",
    "partition      = \"standard\"  # Specify your SLURM partition\n",
    "nodes          = 1\n",
    "tasks_per_node = 1\n",
    "cpus_per_task  = 1\n",
    "memory         = \"100G\"      # Adjust as per your memory requirements\n",
    "account        = 'indikar0'\n",
    "time           = \"0:30:00\"\n",
    "\n",
    "# Loop over each category and model to submit jobs\n",
    "job_id = 1\n",
    "for category in tsCategories:\n",
    "    for model in ['raw']: # models:\n",
    "        file = getFiles(category=category, model=model)\n",
    "        if len(file) == 0:\n",
    "            continue\n",
    "        file = file[0]\n",
    "        if model == 'raw':\n",
    "            time = \"5:00:00\"\n",
    "        else:\n",
    "            time = \"1:00:00\"\n",
    "\n",
    "        # Construct command to run the Python script with arguments\n",
    "        command = f\"python svd_job.py {file} {category} {model}\"\n",
    "\n",
    "        # SLURM job submission command\n",
    "        sbatch_cmd = f\"sbatch --job-name {job_name} \" \\\n",
    "                     f\"--partition {partition} \" \\\n",
    "                     f\"--nodes={nodes} \" \\\n",
    "                     f\"--time={time} \"\\\n",
    "                     f\"--account={account} \" \\\n",
    "                     f\"--ntasks-per-node={tasks_per_node} \" \\\n",
    "                     f\"--cpus-per-task={cpus_per_task} \" \\\n",
    "                     f\"--mem={memory} \" \\\n",
    "                     f\"--output=/home/jpic/scFoundationProject/scFoundationModels/notebooks/datasets/logs/%x-%j.log \" \\\n",
    "                     f\"--wrap=\\\"{command}\\\"\"\n",
    "        # print(sbatch_cmd)\n",
    "        # Submit the job using subprocess\n",
    "        subprocess.run(sbatch_cmd, shell=True)\n",
    "        print(f\"Submitted job {job_id}/{num_jobs}\")\n",
    "        job_id += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "82703b31-5206-4328-83d0-238d07cce7f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>data</th>\n",
       "      <th>model</th>\n",
       "      <th>numCells</th>\n",
       "      <th>dimension</th>\n",
       "      <th>sigmas</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>TS_germ line</td>\n",
       "      <td>raw</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>None</td>\n",
       "      <td>TS_germ line</td>\n",
       "      <td>geneformer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>None</td>\n",
       "      <td>TS_germ line</td>\n",
       "      <td>Tgpt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>None</td>\n",
       "      <td>TS_germ line</td>\n",
       "      <td>scgpt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/founda...</td>\n",
       "      <td>TS_stromal</td>\n",
       "      <td>raw</td>\n",
       "      <td>82478.0</td>\n",
       "      <td>58870.0</td>\n",
       "      <td>[527.04584, 527.44977, 527.73, 527.86005, 528....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/founda...</td>\n",
       "      <td>TS_Heart</td>\n",
       "      <td>scgpt</td>\n",
       "      <td>11505.0</td>\n",
       "      <td>512.0</td>\n",
       "      <td>[102.1326, 24.3429, 10.272291, 9.001001, 6.298...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/founda...</td>\n",
       "      <td>TS_Muscle</td>\n",
       "      <td>raw</td>\n",
       "      <td>30746.0</td>\n",
       "      <td>58870.0</td>\n",
       "      <td>[334.51364, 334.55432, 334.70996, 334.8416, 33...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/founda...</td>\n",
       "      <td>TS_Muscle</td>\n",
       "      <td>geneformer</td>\n",
       "      <td>30746.0</td>\n",
       "      <td>512.0</td>\n",
       "      <td>[2352.0562, 660.7629, 193.3488, 154.38156, 150...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/founda...</td>\n",
       "      <td>TS_Muscle</td>\n",
       "      <td>Tgpt</td>\n",
       "      <td>30746.0</td>\n",
       "      <td>1024.0</td>\n",
       "      <td>[874.7479394410863, 2.6449773579307213e-05, 1....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>/nfs/turbo/umms-indikar/shared/projects/founda...</td>\n",
       "      <td>TS_Muscle</td>\n",
       "      <td>scgpt</td>\n",
       "      <td>30746.0</td>\n",
       "      <td>512.0</td>\n",
       "      <td>[168.26132, 26.562035, 19.433113, 14.5399885, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>116 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  file          data  \\\n",
       "0                                                 None  TS_germ line   \n",
       "1                                                 None  TS_germ line   \n",
       "2                                                 None  TS_germ line   \n",
       "3                                                 None  TS_germ line   \n",
       "4    /nfs/turbo/umms-indikar/shared/projects/founda...    TS_stromal   \n",
       "..                                                 ...           ...   \n",
       "111  /nfs/turbo/umms-indikar/shared/projects/founda...      TS_Heart   \n",
       "112  /nfs/turbo/umms-indikar/shared/projects/founda...     TS_Muscle   \n",
       "113  /nfs/turbo/umms-indikar/shared/projects/founda...     TS_Muscle   \n",
       "114  /nfs/turbo/umms-indikar/shared/projects/founda...     TS_Muscle   \n",
       "115  /nfs/turbo/umms-indikar/shared/projects/founda...     TS_Muscle   \n",
       "\n",
       "          model  numCells  dimension  \\\n",
       "0           raw       NaN        NaN   \n",
       "1    geneformer       NaN        NaN   \n",
       "2          Tgpt       NaN        NaN   \n",
       "3         scgpt       NaN        NaN   \n",
       "4           raw   82478.0    58870.0   \n",
       "..          ...       ...        ...   \n",
       "111       scgpt   11505.0      512.0   \n",
       "112         raw   30746.0    58870.0   \n",
       "113  geneformer   30746.0      512.0   \n",
       "114        Tgpt   30746.0     1024.0   \n",
       "115       scgpt   30746.0      512.0   \n",
       "\n",
       "                                                sigmas  \n",
       "0                                                 None  \n",
       "1                                                 None  \n",
       "2                                                 None  \n",
       "3                                                 None  \n",
       "4    [527.04584, 527.44977, 527.73, 527.86005, 528....  \n",
       "..                                                 ...  \n",
       "111  [102.1326, 24.3429, 10.272291, 9.001001, 6.298...  \n",
       "112  [334.51364, 334.55432, 334.70996, 334.8416, 33...  \n",
       "113  [2352.0562, 660.7629, 193.3488, 154.38156, 150...  \n",
       "114  [874.7479394410863, 2.6449773579307213e-05, 1....  \n",
       "115  [168.26132, 26.562035, 19.433113, 14.5399885, ...  \n",
       "\n",
       "[116 rows x 6 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result = {\n",
    "    'file': [],\n",
    "    'data': [],\n",
    "    'model': [],\n",
    "    'numCells': [],\n",
    "    'dimension': [],\n",
    "    'sigmas': []\n",
    "}\n",
    "for category in tsCategories:\n",
    "    for model in models:\n",
    "        output_file = f'/nfs/turbo/umms-indikar/shared/projects/foundation_models/experiments/tabulaSapiens/results/result_{category}_{model}.pkl'        \n",
    "        if os.path.exists(output_file):\n",
    "            with open(output_file, 'rb') as file:\n",
    "                single_results = pickle.load(file)\n",
    "                result['file'].append(output_file)\n",
    "                result['data'].append(single_results['data'])\n",
    "                result['model'].append(single_results['model'])\n",
    "                result['numCells'].append(single_results['numCells'])\n",
    "                result['dimension'].append(single_results['dimension'])\n",
    "                result['sigmas'].append(single_results['sigmas'])\n",
    "        else:\n",
    "            # File does not exist, append None values\n",
    "            result['file'].append(None)\n",
    "            result['data'].append(category)\n",
    "            result['model'].append(model)\n",
    "            result['numCells'].append(None)\n",
    "            result['dimension'].append(None)\n",
    "            result['sigmas'].append(None)\n",
    "\n",
    "df = pd.DataFrame(result)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "e4328ef8-c58a-484c-afcf-43ba405f68d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def determine_rank_fixed_threshold(singular_values, epsilon=1e-10):\n",
    "    return np.sum(singular_values > epsilon)\n",
    "\n",
    "def determine_rank_energy_threshold(singular_values, energy_threshold=0.99):\n",
    "    total_energy = np.sum(singular_values**2)\n",
    "    cumulative_energy = np.cumsum(singular_values**2)\n",
    "    rank = np.searchsorted(cumulative_energy, energy_threshold * total_energy) + 1\n",
    "    return rank\n",
    "\n",
    "def determine_rank_relative_threshold(singular_values, relative_threshold=1e-3):\n",
    "    max_singular_value = np.max(singular_values)\n",
    "    threshold = relative_threshold * max_singular_value\n",
    "    return np.sum(singular_values > threshold)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "5a676933-6fc6-48b0-a3f8-e60af3193da4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "807dc2b3-6147-44b8-b482-8693629e932a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{tabular}{l|rrrrrrrrrrrrrrrr}\n",
      "\n",
      "\\multicolumn{1}{c}{} & \\multicolumn{4}{c}{Largest Singular Value $\\sigma_1$} & \\multicolumn{4}{c}{$\\sigma_r> \\varepsilon$ (\\varepsilon=10)} & \\multicolumn{4}{c}{$\\sum_i^r\\sigma_i\\approx\\varepsilon\\sum\\sigma_i$ (\\varepsilon=0.9)} & \\multicolumn{4}{c}{$\\sigma_r\\leq \\varepsilon\\sigma_1$ (\\varepsilon=0.05)} \\\\\n",
      "\\cmidrule(lr){2-5} \\cmidrule(lr){6-9}\\cmidrule(lr){10-13}\\cmidrule(lr){14-17} & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt \\\\\n",
      "\\hline\\hline\n",
      "Bladder & 647 & 2123 & 10113 & 149 & 127 & 189 & 250 & 6 & 12 & 2 & 250 & 2 & 21 & 6 & 36 & 10 \\\\\n",
      "Blood & 932 & 2893 & 13721 & 211 & 242 & 279 & 250 & 8 & 13 & 2 & 250 & 2 & 20 & 7 & 23 & 8 \\\\\n",
      "Bone Marrow & 477 & 1462 & 7180 & 105 & 69 & 150 & 250 & 4 & 11 & 2 & 250 & 2 & 18 & 7 & 45 & 11 \\\\\n",
      "Eye & 418 & 1412 & 7877 & 97 & 59 & 126 & 250 & 4 & 11 & 2 & 250 & 2 & 22 & 6 & 51 & 9 \\\\\n",
      "Fat & 575 & 1979 & 10502 & 137 & 98 & 170 & 250 & 5 & 11 & 2 & 250 & 1 & 21 & 5 & 28 & 10 \\\\\n",
      "Heart & 535 & 1504 & 10480 & 102 & 1 & 120 & 250 & 3 & 1 & 2 & 250 & 1 & 1 & 5 & 19 & 7 \\\\\n",
      "Kidney & 419 & 1364 & 9466 & 95 & 46 & 108 & 250 & 2 & 7 & 2 & 250 & 1 & 16 & 5 & 18 & 5 \\\\\n",
      "Large Intestine & 472 & 1557 & 6713 & 112 & 73 & 125 & 250 & 3 & 10 & 2 & 250 & 1 & 20 & 6 & 40 & 7 \\\\\n",
      "Liver & 282 & 945 & 5450 & 68 & 30 & 70 & 250 & 2 & 9 & 2 & 250 & 1 & 20 & 6 & 73 & 7 \\\\\n",
      "Lung & 778 & 2661 & 13596 & 179 & 169 & 216 & 250 & 8 & 12 & 2 & 250 & 2 & 20 & 5 & 26 & 8 \\\\\n",
      "Lymph Node & 1019 & 3033 & 14709 & 224 & 218 & 255 & 250 & 8 & 9 & 1 & 250 & 1 & 16 & 5 & 20 & 6 \\\\\n",
      "Mammary & 439 & 1439 & 7325 & 102 & 67 & 127 & 250 & 2 & 12 & 2 & 250 & 1 & 19 & 6 & 31 & 9 \\\\\n",
      "Muscle & 875 & 2352 & 12043 & 168 & 1 & 217 & 250 & 7 & 1 & 2 & 250 & 1 & 1 & 7 & 29 & 9 \\\\\n",
      "Pancreas & 471 & 1573 & 7682 & 110 & 88 & 145 & 250 & 4 & 16 & 2 & 250 & 2 & 22 & 6 & 38 & 9 \\\\\n",
      "Prostate & 505 & 1716 & 8769 & 123 & 77 & 154 & 250 & 4 & 10 & 2 & 250 & 1 & 21 & 6 & 28 & 8 \\\\\n",
      "Salivary Gland & 667 & 2180 & 9947 & 158 & 149 & 206 & 250 & 6 & 14 & 2 & 250 & 1 & 22 & 7 & 29 & 8 \\\\\n",
      "Skin & 392 & 1292 & 6373 & 93 & 56 & 111 & 250 & 2 & 13 & 2 & 250 & 1 & 19 & 6 & 41 & 9 \\\\\n",
      "Small Intestine & 471 & 1485 & 6774 & 108 & 70 & 121 & 250 & 3 & 10 & 2 & 250 & 1 & 19 & 5 & 39 & 6 \\\\\n",
      "Spleen & 794 & 2443 & 11999 & 176 & 164 & 226 & 250 & 7 & 11 & 2 & 250 & 1 & 19 & 5 & 26 & 9 \\\\\n",
      "Thymus & 762 & 2427 & 12131 & 177 & 158 & 214 & 250 & 7 & 11 & 2 & 250 & 1 & 19 & 5 & 32 & 9 \\\\\n",
      "Tongue & 510 & 1673 & 8927 & 115 & 94 & 152 & 250 & 5 & 13 & 2 & 250 & 2 & 23 & 6 & 32 & 9 \\\\\n",
      "Trachea & 383 & 1321 & 6915 & 93 & 56 & 117 & 250 & 3 & 14 & 2 & 250 & 1 & 24 & 7 & 35 & 9 \\\\\n",
      "Uterus & 341 & 1154 & 6309 & 80 & 40 & 91 & 250 & 3 & 11 & 2 & 250 & 1 & 20 & 5 & 36 & 9 \\\\\n",
      "Vasculature & 510 & 1796 & 9856 & 120 & 79 & 150 & 250 & 5 & 11 & 2 & 250 & 1 & 21 & 5 & 32 & 8 \\\\\n",
      "endothelial & 690 & 2397 & 14168 & 171 & 124 & 209 & 250 & 7 & 11 & 2 & 250 & 1 & 21 & 5 & 29 & 7 \\\\\n",
      "epithelial & 1315 & 4432 & 21387 & 302 & 392 & 339 & 250 & 21 & 13 & 2 & 250 & 3 & 22 & 7 & 47 & 13 \\\\\n",
      "germ line &  &  &  &  &  &  &  &  &  &  &  &  &  &  &  &  \\\\\n",
      "immune & 2196 & 6744 & 31711 & 491 & 653 & 440 & 250 & 26 & 12 & 2 & 250 & 1 & 19 & 6 & 32 & 10 \\\\\n",
      "stromal & 1154 & 3981 & 21517 & 275 & 275 & 301 & 250 & 14 & 10 & 2 & 250 & 1 & 20 & 5 & 34 & 9 \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "v1 = 10\n",
    "v2 = 0.90\n",
    "v3 = 0.05\n",
    "maxSv = []\n",
    "eng90p = []\n",
    "relS1 = []\n",
    "fixed = []\n",
    "for i in range(df.shape[0]):\n",
    "    sigmas = df['sigmas'].iloc[i]\n",
    "    if sigmas is None:\n",
    "        maxSv.append(None)\n",
    "        eng90p.append(None)\n",
    "        relS1.append(None)\n",
    "        fixed.append(None)\n",
    "    else:\n",
    "        maxSv.append(max(sigmas))\n",
    "        fixed.append(determine_rank_fixed_threshold(sigmas, epsilon=v1))\n",
    "        eng90p.append(determine_rank_energy_threshold(sigmas, energy_threshold=v2))\n",
    "        relS1.append(determine_rank_relative_threshold(sigmas, relative_threshold=v3))\n",
    "\n",
    "df['maxSv'] = maxSv\n",
    "df['fixed'] = fixed\n",
    "df['eng90p'] = eng90p\n",
    "df['relS1'] = relS1\n",
    "\n",
    "pivoted_df = df.pivot(index='data', columns='model', values=['maxSv', 'fixed', 'eng90p', 'relS1'])\n",
    "\n",
    "# Flatten the multi-index columns\n",
    "# pivoted_df.columns = [f'{col[1]} {col[0]}' for col in pivoted_df.columns]\n",
    "\n",
    "# Reset the index to turn the 'data' index back into a column\n",
    "pivoted_df.reset_index(inplace=True)\n",
    "\n",
    "latex_table = pivoted_df.to_latex(index=False, escape=False, float_format=\"%.f\")\n",
    "latex_table = latex_table.replace(\n",
    "    'TS_', ''\n",
    ").replace(\n",
    "    '\\midrule', '\\hline\\hline'\n",
    ").replace(\n",
    "    'NaN', ''\n",
    ").replace(\n",
    "    r'\\begin{tabular}{lrrrrrrrrrrrrrrrr}', r'\\begin{tabular}{l|rrrrrrrrrrrrrrrr}'\n",
    ").replace(\n",
    "    '{4}{r}', '{4}{c}'\n",
    ").replace(\n",
    "    r'\\toprule', ''\n",
    ").replace(\n",
    "    'data', '\\multicolumn{1}{c}{}'\n",
    ").replace('_', ' ').replace(\n",
    "    'maxSv', 'Largest Singular Value $\\sigma_1$'\n",
    ").replace(\n",
    "    'fixed', (r'$\\sigma_r> \\varepsilon$ ' + f'(\\\\varepsilon={v1})')\n",
    ").replace(\n",
    "    'eng90p', (r'$\\sum_i^r\\sigma_i\\approx\\varepsilon\\sum\\sigma_i$ ' + f'(\\\\varepsilon={v2})')\n",
    ").replace(\n",
    "    'relS1', (r'$\\sigma_r\\leq \\varepsilon\\sigma_1$ ' + f'(\\\\varepsilon={v3})')\n",
    ").replace(\n",
    "    r' & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt',\n",
    "    r'\\cmidrule(lr){2-5} \\cmidrule(lr){6-9}\\cmidrule(lr){10-13}\\cmidrule(lr){14-17} & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt'\n",
    ")\n",
    "print(latex_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "a8524a3b-97f7-401d-b963-64e379febd0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{tabular}{l|rrrrrrrrrrrrrrrr}\n",
      "\n",
      "\\multicolumn{1}{c}{} & \\multicolumn{4}{c}{Largest Singular Value $\\sigma_1$} & \\multicolumn{4}{c}{$\\sigma_r> \\varepsilon$ (\\varepsilon=50)} & \\multicolumn{4}{c}{$\\sum_i^r\\sigma_i\\approx\\varepsilon\\sum\\sigma_i$ (\\varepsilon=0.99)} & \\multicolumn{4}{c}{$\\sigma_r\\leq \\varepsilon\\sigma_1$ (\\varepsilon=0.001)} \\\\\n",
      "\\cmidrule(lr){2-5} \\cmidrule(lr){6-9}\\cmidrule(lr){10-13}\\cmidrule(lr){14-17} & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt \\\\\n",
      "\\hline\\hline\n",
      "Bladder & 647 & 2123 & 10113 & 149 & 10 & 20 & 250 & 1 & 377 & 48 & 250 & 25 & 1022 & 459 & 250 & 376 \\\\\n",
      "Blood & 932 & 2893 & 13721 & 211 & 18 & 35 & 250 & 2 & 396 & 67 & 250 & 14 & 1022 & 471 & 250 & 324 \\\\\n",
      "Bone Marrow & 477 & 1462 & 7180 & 105 & 8 & 12 & 250 & 1 & 344 & 66 & 250 & 22 & 1022 & 467 & 250 & 369 \\\\\n",
      "Eye & 418 & 1412 & 7877 & 97 & 6 & 10 & 250 & 1 & 341 & 50 & 250 & 26 & 1022 & 462 & 250 & 379 \\\\\n",
      "Fat & 575 & 1979 & 10502 & 137 & 9 & 14 & 250 & 1 & 345 & 43 & 250 & 17 & 1022 & 455 & 250 & 332 \\\\\n",
      "Heart & 535 & 1504 & 10480 & 102 & 1 & 9 & 250 & 1 & 1 & 37 & 250 & 13 & 1 & 453 & 250 & 317 \\\\\n",
      "Kidney & 419 & 1364 & 9466 & 95 & 5 & 7 & 250 & 1 & 261 & 39 & 250 & 9 & 1020 & 453 & 250 & 278 \\\\\n",
      "Large Intestine & 472 & 1557 & 6713 & 112 & 8 & 8 & 250 & 1 & 348 & 39 & 250 & 13 & 1022 & 457 & 250 & 307 \\\\\n",
      "Liver & 282 & 945 & 5450 & 68 & 3 & 6 & 250 & 1 & 294 & 48 & 250 & 16 & 1021 & 458 & 250 & 314 \\\\\n",
      "Lung & 778 & 2661 & 13596 & 179 & 13 & 25 & 250 & 1 & 366 & 42 & 250 & 22 & 1022 & 454 & 250 & 354 \\\\\n",
      "Lymph Node & 1019 & 3033 & 14709 & 224 & 17 & 28 & 250 & 1 & 330 & 47 & 250 & 11 & 1022 & 460 & 250 & 295 \\\\\n",
      "Mammary & 439 & 1439 & 7325 & 102 & 6 & 9 & 250 & 1 & 361 & 48 & 250 & 18 & 1022 & 459 & 250 & 346 \\\\\n",
      "Muscle & 875 & 2352 & 12043 & 168 & 1 & 21 & 250 & 1 & 1 & 52 & 250 & 16 & 1 & 463 & 250 & 345 \\\\\n",
      "Pancreas & 471 & 1573 & 7682 & 110 & 9 & 13 & 250 & 1 & 393 & 51 & 250 & 19 & 1022 & 462 & 250 & 348 \\\\\n",
      "Prostate & 505 & 1716 & 8769 & 123 & 8 & 13 & 250 & 1 & 328 & 48 & 250 & 15 & 1022 & 461 & 250 & 324 \\\\\n",
      "Salivary Gland & 667 & 2180 & 9947 & 158 & 12 & 19 & 250 & 1 & 388 & 55 & 250 & 17 & 1022 & 465 & 250 & 336 \\\\\n",
      "Skin & 392 & 1292 & 6373 & 93 & 6 & 7 & 250 & 1 & 374 & 49 & 250 & 17 & 1022 & 460 & 250 & 339 \\\\\n",
      "Small Intestine & 471 & 1485 & 6774 & 108 & 7 & 8 & 250 & 1 & 336 & 42 & 250 & 13 & 1022 & 456 & 250 & 307 \\\\\n",
      "Spleen & 794 & 2443 & 11999 & 176 & 14 & 23 & 250 & 1 & 355 & 56 & 250 & 17 & 1022 & 464 & 250 & 337 \\\\\n",
      "Thymus & 762 & 2427 & 12131 & 177 & 13 & 21 & 250 & 1 & 358 & 49 & 250 & 17 & 1022 & 461 & 250 & 329 \\\\\n",
      "Tongue & 510 & 1673 & 8927 & 115 & 8 & 13 & 250 & 1 & 378 & 48 & 250 & 18 & 1022 & 460 & 250 & 353 \\\\\n",
      "Trachea & 383 & 1321 & 6915 & 93 & 7 & 9 & 250 & 1 & 360 & 50 & 250 & 23 & 1022 & 458 & 250 & 353 \\\\\n",
      "Uterus & 341 & 1154 & 6309 & 80 & 4 & 7 & 250 & 1 & 339 & 44 & 250 & 22 & 1022 & 455 & 250 & 354 \\\\\n",
      "Vasculature & 510 & 1796 & 9856 & 120 & 8 & 14 & 250 & 1 & 346 & 42 & 250 & 20 & 1022 & 452 & 250 & 356 \\\\\n",
      "endothelial & 690 & 2397 & 14168 & 171 & 13 & 19 & 250 & 1 & 338 & 45 & 250 & 14 & 1022 & 460 & 250 & 332 \\\\\n",
      "epithelial & 1315 & 4432 & 21387 & 302 & 31 & 70 & 250 & 2 & 387 & 53 & 250 & 35 & 1022 & 465 & 250 & 415 \\\\\n",
      "germ line &  &  &  &  &  &  &  &  &  &  &  &  &  &  &  &  \\\\\n",
      "immune & 2196 & 6744 & 31711 & 491 & 63 & 132 & 250 & 3 & 382 & 61 & 250 & 24 & 1022 & 468 & 250 & 375 \\\\\n",
      "stromal & 1154 & 3981 & 21517 & 275 & 23 & 50 & 250 & 1 & 339 & 45 & 250 & 19 & 1022 & 458 & 250 & 362 \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "v1 = 50\n",
    "v2 = 0.99\n",
    "v3 = 0.001\n",
    "maxSv = []\n",
    "eng90p = []\n",
    "relS1 = []\n",
    "fixed = []\n",
    "for i in range(df.shape[0]):\n",
    "    sigmas = df['sigmas'].iloc[i]\n",
    "    if sigmas is None:\n",
    "        maxSv.append(None)\n",
    "        eng90p.append(None)\n",
    "        relS1.append(None)\n",
    "        fixed.append(None)\n",
    "    else:\n",
    "        maxSv.append(max(sigmas))\n",
    "        fixed.append(determine_rank_fixed_threshold(sigmas, epsilon=v1))\n",
    "        eng90p.append(determine_rank_energy_threshold(sigmas, energy_threshold=v2))\n",
    "        relS1.append(determine_rank_relative_threshold(sigmas, relative_threshold=v3))\n",
    "\n",
    "df['maxSv'] = maxSv\n",
    "df['fixed'] = fixed\n",
    "df['eng90p'] = eng90p\n",
    "df['relS1'] = relS1\n",
    "\n",
    "pivoted_df = df.pivot(index='data', columns='model', values=['maxSv', 'fixed', 'eng90p', 'relS1'])\n",
    "\n",
    "# Flatten the multi-index columns\n",
    "# pivoted_df.columns = [f'{col[1]} {col[0]}' for col in pivoted_df.columns]\n",
    "\n",
    "# Reset the index to turn the 'data' index back into a column\n",
    "pivoted_df.reset_index(inplace=True)\n",
    "\n",
    "latex_table = pivoted_df.to_latex(index=False, escape=False, float_format=\"%.f\")\n",
    "latex_table = latex_table.replace(\n",
    "    'TS_', ''\n",
    ").replace(\n",
    "    '\\midrule', '\\hline\\hline'\n",
    ").replace(\n",
    "    'NaN', ''\n",
    ").replace(\n",
    "    r'\\begin{tabular}{lrrrrrrrrrrrrrrrr}', r'\\begin{tabular}{l|rrrrrrrrrrrrrrrr}'\n",
    ").replace(\n",
    "    '{4}{r}', '{4}{c}'\n",
    ").replace(\n",
    "    r'\\toprule', ''\n",
    ").replace(\n",
    "    'data', '\\multicolumn{1}{c}{}'\n",
    ").replace('_', ' ').replace(\n",
    "    'maxSv', 'Largest Singular Value $\\sigma_1$'\n",
    ").replace(\n",
    "    'fixed', (r'$\\sigma_r> \\varepsilon$ ' + f'(\\\\varepsilon={v1})')\n",
    ").replace(\n",
    "    'eng90p', (r'$\\sum_i^r\\sigma_i\\approx\\varepsilon\\sum\\sigma_i$ ' + f'(\\\\varepsilon={v2})')\n",
    ").replace(\n",
    "    'relS1', (r'$\\sigma_r\\leq \\varepsilon\\sigma_1$ ' + f'(\\\\varepsilon={v3})')\n",
    ").replace(\n",
    "    r' & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt',\n",
    "    r'\\cmidrule(lr){2-5} \\cmidrule(lr){6-9}\\cmidrule(lr){10-13}\\cmidrule(lr){14-17} & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt'\n",
    ")\n",
    "print(latex_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "2d025c04-f1b5-457c-bc72-8e8e9b4d176f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{tabular}{l|rrrrrrrrrrrrrrrr}\n",
      "\n",
      "\\multicolumn{1}{c}{} & \\multicolumn{4}{c}{Largest Singular Value $\\sigma_1$} & \\multicolumn{4}{c}{$\\sigma_r> \\varepsilon$ (\\varepsilon=50)} & \\multicolumn{4}{c}{$\\sum_i^r\\sigma_i\\approx\\varepsilon\\sum\\sigma_i$ (\\varepsilon=0.8)} & \\multicolumn{4}{c}{$\\sigma_r\\leq \\varepsilon\\sigma_1$ (\\varepsilon=0.5)} \\\\\n",
      "\\cmidrule(lr){2-5} \\cmidrule(lr){6-9}\\cmidrule(lr){10-13}\\cmidrule(lr){14-17} & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt \\\\\n",
      "\\hline\\hline\n",
      "Bladder & 647 & 2123 & 10113 & 149 & 10 & 20 & 250 & 1 & 3 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "Blood & 932 & 2893 & 13721 & 211 & 18 & 35 & 250 & 2 & 4 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "Bone Marrow & 477 & 1462 & 7180 & 105 & 8 & 12 & 250 & 1 & 3 & 1 & 250 & 1 & 1 & 1 & 2 & 1 \\\\\n",
      "Eye & 418 & 1412 & 7877 & 97 & 6 & 10 & 250 & 1 & 3 & 1 & 250 & 1 & 1 & 1 & 2 & 1 \\\\\n",
      "Fat & 575 & 1979 & 10502 & 137 & 9 & 14 & 250 & 1 & 3 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "Heart & 535 & 1504 & 10480 & 102 & 1 & 9 & 250 & 1 & 1 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "Kidney & 419 & 1364 & 9466 & 95 & 5 & 7 & 250 & 1 & 2 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "Large Intestine & 472 & 1557 & 6713 & 112 & 8 & 8 & 250 & 1 & 3 & 1 & 250 & 1 & 1 & 1 & 2 & 1 \\\\\n",
      "Liver & 282 & 945 & 5450 & 68 & 3 & 6 & 250 & 1 & 2 & 1 & 250 & 1 & 1 & 1 & 2 & 1 \\\\\n",
      "Lung & 778 & 2661 & 13596 & 179 & 13 & 25 & 250 & 1 & 3 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "Lymph Node & 1019 & 3033 & 14709 & 224 & 17 & 28 & 250 & 1 & 2 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "Mammary & 439 & 1439 & 7325 & 102 & 6 & 9 & 250 & 1 & 3 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "Muscle & 875 & 2352 & 12043 & 168 & 1 & 21 & 250 & 1 & 1 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "Pancreas & 471 & 1573 & 7682 & 110 & 9 & 13 & 250 & 1 & 4 & 1 & 250 & 1 & 1 & 1 & 2 & 1 \\\\\n",
      "Prostate & 505 & 1716 & 8769 & 123 & 8 & 13 & 250 & 1 & 3 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "Salivary Gland & 667 & 2180 & 9947 & 158 & 12 & 19 & 250 & 1 & 4 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "Skin & 392 & 1292 & 6373 & 93 & 6 & 7 & 250 & 1 & 3 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "Small Intestine & 471 & 1485 & 6774 & 108 & 7 & 8 & 250 & 1 & 3 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "Spleen & 794 & 2443 & 11999 & 176 & 14 & 23 & 250 & 1 & 3 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "Thymus & 762 & 2427 & 12131 & 177 & 13 & 21 & 250 & 1 & 3 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "Tongue & 510 & 1673 & 8927 & 115 & 8 & 13 & 250 & 1 & 4 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "Trachea & 383 & 1321 & 6915 & 93 & 7 & 9 & 250 & 1 & 4 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "Uterus & 341 & 1154 & 6309 & 80 & 4 & 7 & 250 & 1 & 3 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "Vasculature & 510 & 1796 & 9856 & 120 & 8 & 14 & 250 & 1 & 3 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "endothelial & 690 & 2397 & 14168 & 171 & 13 & 19 & 250 & 1 & 2 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "epithelial & 1315 & 4432 & 21387 & 302 & 31 & 70 & 250 & 2 & 3 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "germ line &  &  &  &  &  &  &  &  &  &  &  &  &  &  &  &  \\\\\n",
      "immune & 2196 & 6744 & 31711 & 491 & 63 & 132 & 250 & 3 & 3 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "stromal & 1154 & 3981 & 21517 & 275 & 23 & 50 & 250 & 1 & 2 & 1 & 250 & 1 & 1 & 1 & 1 & 1 \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "v1 = 50\n",
    "v2 = 0.8\n",
    "v3 = 0.5\n",
    "maxSv = []\n",
    "eng90p = []\n",
    "relS1 = []\n",
    "fixed = []\n",
    "for i in range(df.shape[0]):\n",
    "    sigmas = df['sigmas'].iloc[i]\n",
    "    if sigmas is None:\n",
    "        maxSv.append(None)\n",
    "        eng90p.append(None)\n",
    "        relS1.append(None)\n",
    "        fixed.append(None)\n",
    "    else:\n",
    "        maxSv.append(max(sigmas))\n",
    "        fixed.append(determine_rank_fixed_threshold(sigmas, epsilon=v1))\n",
    "        eng90p.append(determine_rank_energy_threshold(sigmas, energy_threshold=v2))\n",
    "        relS1.append(determine_rank_relative_threshold(sigmas, relative_threshold=v3))\n",
    "\n",
    "df['maxSv'] = maxSv\n",
    "df['fixed'] = fixed\n",
    "df['eng90p'] = eng90p\n",
    "df['relS1'] = relS1\n",
    "\n",
    "pivoted_df = df.pivot(index='data', columns='model', values=['maxSv', 'fixed', 'eng90p', 'relS1'])\n",
    "\n",
    "# Flatten the multi-index columns\n",
    "# pivoted_df.columns = [f'{col[1]} {col[0]}' for col in pivoted_df.columns]\n",
    "\n",
    "# Reset the index to turn the 'data' index back into a column\n",
    "pivoted_df.reset_index(inplace=True)\n",
    "\n",
    "latex_table = pivoted_df.to_latex(index=False, escape=False, float_format=\"%.f\")\n",
    "latex_table = latex_table.replace(\n",
    "    'TS_', ''\n",
    ").replace(\n",
    "    '\\midrule', '\\hline\\hline'\n",
    ").replace(\n",
    "    'NaN', ''\n",
    ").replace(\n",
    "    r'\\begin{tabular}{lrrrrrrrrrrrrrrrr}', r'\\begin{tabular}{l|rrrrrrrrrrrrrrrr}'\n",
    ").replace(\n",
    "    '{4}{r}', '{4}{c}'\n",
    ").replace(\n",
    "    r'\\toprule', ''\n",
    ").replace(\n",
    "    'data', '\\multicolumn{1}{c}{}'\n",
    ").replace('_', ' ').replace(\n",
    "    'maxSv', 'Largest Singular Value $\\sigma_1$'\n",
    ").replace(\n",
    "    'fixed', (r'$\\sigma_r> \\varepsilon$ ' + f'(\\\\varepsilon={v1})')\n",
    ").replace(\n",
    "    'eng90p', (r'$\\sum_i^r\\sigma_i\\approx\\varepsilon\\sum\\sigma_i$ ' + f'(\\\\varepsilon={v2})')\n",
    ").replace(\n",
    "    'relS1', (r'$\\sigma_r\\leq \\varepsilon\\sigma_1$ ' + f'(\\\\varepsilon={v3})')\n",
    ").replace(\n",
    "    r' & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt',\n",
    "    r'\\cmidrule(lr){2-5} \\cmidrule(lr){6-9}\\cmidrule(lr){10-13}\\cmidrule(lr){14-17} & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt & Tgpt & geneformer & raw & scgpt'\n",
    ")\n",
    "print(latex_table)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f5be163-9333-4635-b3ed-49c3d8dc2acb",
   "metadata": {},
   "source": [
    "### Summary Stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f56ac24c-3542-4d8d-9ec5-f71ade4e173c",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfData = pd.DataFrame(index = tsCategories)\n",
    "dfData.index = dfData.index.map(lambda x: x[3:])\n",
    "dfRep = pd.DataFrame(index = models)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b5700cd9-e3d2-4a48-a128-6fecd0bf0aae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_germ line.h5ad']\n",
      "{'dataset': 'germ line', 'num_cells': 11, 'female_to_male_ratio': 'M', 'umi_mean': 13773.818, 'umi_median': 11515.0, 'umi_std': 7367.68798828125, 'n_genes_mean': 2836.3636363636365, 'n_genes_median': 2775.0, 'n_genes_std': 1062.8599411707332, 'anatomy': 'nan', 'compartment': 'germ line', 'methods': '10X', 'donor': 'TSP14'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_stromal.h5ad']\n",
      "{'dataset': 'stromal', 'num_cells': 82478, 'female_to_male_ratio': 0.8408213369043633, 'umi_mean': 69257.82, 'umi_median': 10464.0, 'umi_std': 351686.34375, 'n_genes_mean': 3095.35902907442, 'n_genes_median': 2803.0, 'n_genes_std': 1462.1518160320982, 'anatomy': 'nan, Supradiaphagmatic, Parotid, Posterior, Endometrium, Myometrium, MAT, SCAT, Atria, Ventricle, exocrine, abdomen, chest, atria, ventricle, noCornea, lacrimalgland, diaphragm, rectusabdominus, whole, anterior, posterior, Sublingual, Cornea-etc, Neuroretina-etc, Sclera-etc, Distal, Proximal, Abdomen, Chest, Anterior, Diaphragm, AortaVeneCava, CoronaryArteries, distal, proximal, MedialDistal , proxmedialdistal, Neuron, Endocrine, Exocrine, Aorta, aorta', 'compartment': 'stromal', 'methods': '10X, smartseq2', 'donor': 'TSP6, TSP7, TSP4, TSP5, TSP3, TSP10, TSP12, TSP9, TSP8, TSP14, TSP15, TSP1, TSP2'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Large_Intestine.h5ad']\n",
      "{'dataset': 'Large_Intestine', 'num_cells': 13680, 'female_to_male_ratio': 0.1959087332808812, 'umi_mean': 23675.748, 'umi_median': 6036.5, 'umi_std': 112187.65625, 'n_genes_mean': 1832.6504385964913, 'n_genes_median': 1525.5, 'n_genes_std': 980.5872224616902, 'anatomy': 'Distal, Proximal, distal, proximal, nan', 'compartment': 'immune, epithelial, stromal, endothelial', 'methods': '10X, smartseq2', 'donor': 'TSP14, TSP2'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Lymph_Node.h5ad']\n",
      "{'dataset': 'Lymph_Node', 'num_cells': 53275, 'female_to_male_ratio': 0.7509120189305551, 'umi_mean': 25103.4, 'umi_median': 5949.0, 'umi_std': 119044.3203125, 'n_genes_mean': 2004.631290473956, 'n_genes_median': 1810.0, 'n_genes_std': 807.0289588636018, 'anatomy': 'Inguinal, Supradiaphagmatic, inguinal, supradiaphragmatic, nan', 'compartment': 'immune, endothelial, stromal', 'methods': '10X, smartseq2', 'donor': 'TSP7, TSP14, TSP2'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Skin.h5ad']\n",
      "{'dataset': 'Skin', 'num_cells': 9424, 'female_to_male_ratio': 'M', 'umi_mean': 115167.19, 'umi_median': 7401.5, 'umi_std': 531242.8125, 'n_genes_mean': 2161.478034804754, 'n_genes_median': 1960.0, 'n_genes_std': 974.9489827632551, 'anatomy': 'nan, abdomen, chest, Abdomen, Chest', 'compartment': 'immune, stromal, endothelial, epithelial', 'methods': '10X, smartseq2', 'donor': 'TSP10, TSP14'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_epithelial.h5ad']\n",
      "{'dataset': 'epithelial', 'num_cells': 104148, 'female_to_male_ratio': 1.1368952357503386, 'umi_mean': 70796.49, 'umi_median': 11906.5, 'umi_std': 414406.875, 'n_genes_mean': 3028.2610899873257, 'n_genes_median': 2675.0, 'n_genes_std': 1752.915742969831, 'anatomy': 'nan, Parotid, Anterior, Posterior, Endometrium, Myometrium, Conjunctiva, Atria, Ventricle, exocrine, abdomen, chest, atria, ventricle, noCornea, lacrimalgland, whole, anterior, posterior, Sublingual, Cornea-etc, Neuroretina-etc, Sclera-etc, Distal, Proximal, Abdomen, Chest, distal, proximal, MedialDistal , proxmedialdistal, Neuron, Endocrine, Exocrine, Aorta', 'compartment': 'epithelial', 'methods': '10X, smartseq2', 'donor': 'TSP6, TSP7, TSP4, TSP5, TSP3, TSP10, TSP12, TSP9, TSP8, TSP14, TSP15, TSP1, TSP2'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Spleen.h5ad']\n",
      "{'dataset': 'Spleen', 'num_cells': 34004, 'female_to_male_ratio': 0.7930816283484496, 'umi_mean': 33204.973, 'umi_median': 6628.0, 'umi_std': 146508.28125, 'n_genes_mean': 2349.5154099517704, 'n_genes_median': 2110.0, 'n_genes_std': 1167.4199364284573, 'anatomy': 'nan', 'compartment': 'immune, endothelial', 'methods': '10X, smartseq2', 'donor': 'TSP7, TSP14, TSP2'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Vasculature.h5ad']\n",
      "{'dataset': 'Vasculature', 'num_cells': 16037, 'female_to_male_ratio': 1.1703884152118014, 'umi_mean': 24278.81, 'umi_median': 13304.0, 'umi_std': 52047.7578125, 'n_genes_mean': 3616.7917939764293, 'n_genes_median': 3426.0, 'n_genes_std': 1715.1142509251742, 'anatomy': 'AortaVeneCava, CoronaryArteries, Aorta, nan, aorta', 'compartment': 'stromal, immune, endothelial, epithelial', 'methods': '10X, smartseq2', 'donor': 'TSP14, TSP2'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Kidney.h5ad']\n",
      "{'dataset': 'Kidney', 'num_cells': 9641, 'female_to_male_ratio': 'F', 'umi_mean': 24345.217, 'umi_median': 8520.0, 'umi_std': 98628.0546875, 'n_genes_mean': 2683.97323929053, 'n_genes_median': 2481.0, 'n_genes_std': 1069.4323423017286, 'anatomy': 'nan', 'compartment': 'epithelial, immune, endothelial', 'methods': '10X, smartseq2', 'donor': 'TSP2'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Salivary_Gland.h5ad']\n",
      "{'dataset': 'Salivary_Gland', 'num_cells': 27199, 'female_to_male_ratio': 0.5313027812183313, 'umi_mean': 39793.414, 'umi_median': 7092.0, 'umi_std': 214614.59375, 'n_genes_mean': 2446.1270267289237, 'n_genes_median': 2083.0, 'n_genes_std': 1698.5102734630711, 'anatomy': 'Parotid, Sublingual', 'compartment': 'epithelial, immune, stromal, endothelial', 'methods': '10X, smartseq2', 'donor': 'TSP7, TSP14'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Thymus.h5ad']\n",
      "{'dataset': 'Thymus', 'num_cells': 33664, 'female_to_male_ratio': 0.5601075169153767, 'umi_mean': 15215.986, 'umi_median': 6423.5, 'umi_std': 69065.6171875, 'n_genes_mean': 2401.8281547053234, 'n_genes_median': 2084.0, 'n_genes_std': 1066.8157077193953, 'anatomy': 'nan', 'compartment': 'immune, epithelial, stromal, endothelial', 'methods': '10X, smartseq2', 'donor': 'TSP14, TSP2'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Pancreas.h5ad']\n",
      "{'dataset': 'Pancreas', 'num_cells': 13497, 'female_to_male_ratio': 1.0147783251231528, 'umi_mean': 107647.79, 'umi_median': 12816.0, 'umi_std': 521587.8125, 'n_genes_mean': 2954.093724531377, 'n_genes_median': 2614.0, 'n_genes_std': 1713.781155263306, 'anatomy': 'exocrine, Endocrine, Exocrine', 'compartment': 'epithelial, immune, endothelial, stromal', 'methods': '10X, smartseq2', 'donor': 'TSP9, TSP1'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Fat.h5ad']\n",
      "{'dataset': 'Fat', 'num_cells': 20263, 'female_to_male_ratio': 'M', 'umi_mean': 56812.285, 'umi_median': 11511.0, 'umi_std': 410517.34375, 'n_genes_mean': 3197.3036569116125, 'n_genes_median': 3132.0, 'n_genes_std': 1367.3564784455853, 'anatomy': 'MAT, SCAT', 'compartment': 'stromal, endothelial, immune', 'methods': '10X, smartseq2', 'donor': 'TSP10, TSP14'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Bladder.h5ad']\n",
      "{'dataset': 'Bladder', 'num_cells': 24583, 'female_to_male_ratio': 7.017938682322244, 'umi_mean': 43915.56, 'umi_median': 9327.0, 'umi_std': 236138.6875, 'n_genes_mean': 2720.9129886506935, 'n_genes_median': 2399.0, 'n_genes_std': 1380.5378991879945, 'anatomy': 'nan', 'compartment': 'immune, stromal, epithelial, endothelial', 'methods': '10X, smartseq2', 'donor': 'TSP14, TSP1, TSP2'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_endothelial.h5ad']\n",
      "{'dataset': 'endothelial', 'num_cells': 31691, 'female_to_male_ratio': 0.9756249610373419, 'umi_mean': 77655.41, 'umi_median': 7436.0, 'umi_std': 388806.40625, 'n_genes_mean': 2702.6087217191002, 'n_genes_median': 2421.0, 'n_genes_std': 1262.8111113961622, 'anatomy': 'nan, Inguinal, Supradiaphagmatic, Parotid, Posterior, Endometrium, Myometrium, Orbital, MAT, SCAT, Atria, Ventricle, exocrine, abdomen, chest, atria, ventricle, lacrimalgland, noCornea, diaphragm, rectusabdominus, inguinal, supradiaphragmatic, anterior, posterior, Sublingual, Cornea-etc, Sclera-etc, Distal, Proximal, Abdomen, Chest, Anterior, Diaphragm, AortaVeneCava, CoronaryArteries, distal, proximal, MedialDistal , proxmedialdistal, Endocrine, Exocrine, Aorta, aorta', 'compartment': 'endothelial', 'methods': '10X, smartseq2', 'donor': 'TSP6, TSP7, TSP4, TSP5, TSP3, TSP10, TSP12, TSP9, TSP8, TSP14, TSP15, TSP1, TSP2'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Blood.h5ad']\n",
      "{'dataset': 'Blood', 'num_cells': 50115, 'female_to_male_ratio': 1.9578587027090835, 'umi_mean': 25557.389, 'umi_median': 7814.0, 'umi_std': 162959.703125, 'n_genes_mean': 1847.9537264292128, 'n_genes_median': 1708.0, 'n_genes_std': 1199.3767996508157, 'anatomy': 'nan', 'compartment': 'immune', 'methods': '10X, smartseq2', 'donor': 'TSP7, TSP10, TSP8, TSP14, TSP1, TSP2'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Trachea.h5ad']\n",
      "{'dataset': 'Trachea', 'num_cells': 9522, 'female_to_male_ratio': 0.8604923798358733, 'umi_mean': 57639.977, 'umi_median': 7810.0, 'umi_std': 277252.28125, 'n_genes_mean': 2834.2302037387103, 'n_genes_median': 2469.0, 'n_genes_std': 1494.6364652163534, 'anatomy': 'nan', 'compartment': 'immune, epithelial, stromal, endothelial', 'methods': '10X, smartseq2', 'donor': 'TSP6, TSP2'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Bone_Marrow.h5ad']\n",
      "{'dataset': 'Bone_Marrow', 'num_cells': 12297, 'female_to_male_ratio': 0.3849532605023088, 'umi_mean': 109504.63, 'umi_median': 7741.0, 'umi_std': 405771.40625, 'n_genes_mean': 2400.6517036675614, 'n_genes_median': 2001.0, 'n_genes_std': 1548.5289985362385, 'anatomy': 'nan, vertebralbody', 'compartment': 'immune', 'methods': '10X, smartseq2', 'donor': 'TSP11, TSP13, TSP14, TSP2'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Tongue.h5ad']\n",
      "{'dataset': 'Tongue', 'num_cells': 15020, 'female_to_male_ratio': 1.7727524460033228, 'umi_mean': 80515.84, 'umi_median': 10800.0, 'umi_std': 391156.59375, 'n_genes_mean': 2695.688548601864, 'n_genes_median': 2363.0, 'n_genes_std': 1512.0205688904314, 'anatomy': 'Anterior, Posterior, nan, anterior, posterior', 'compartment': 'epithelial, immune, stromal, endothelial', 'methods': '10X, smartseq2', 'donor': 'TSP7, TSP4, TSP14'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Lung.h5ad']\n",
      "{'dataset': 'Lung', 'num_cells': 35682, 'female_to_male_ratio': 7.2027586206896554, 'umi_mean': 68051.26, 'umi_median': 13465.5, 'umi_std': 318570.21875, 'n_genes_mean': 3275.1932066588197, 'n_genes_median': 3191.0, 'n_genes_std': 1481.6763930491543, 'anatomy': 'Distal, Proximal, nan, MedialDistal , proxmedialdistal, Neuron', 'compartment': 'epithelial, immune, endothelial, stromal', 'methods': '10X, smartseq2', 'donor': 'TSP14, TSP1, TSP2'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Liver.h5ad']\n",
      "{'dataset': 'Liver', 'num_cells': 5007, 'female_to_male_ratio': 'M', 'umi_mean': 28392.615, 'umi_median': 6329.0, 'umi_std': 217181.59375, 'n_genes_mean': 2289.529458757739, 'n_genes_median': 2072.0, 'n_genes_std': 946.8976698906137, 'anatomy': 'nan', 'compartment': 'immune, endothelial, stromal, epithelial', 'methods': '10X, smartseq2', 'donor': 'TSP6, TSP14'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Small_Intestine.h5ad']\n",
      "{'dataset': 'Small_Intestine', 'num_cells': 12467, 'female_to_male_ratio': 0.19210174029451138, 'umi_mean': 32625.623, 'umi_median': 6369.0, 'umi_std': 175116.296875, 'n_genes_mean': 1967.3262212240315, 'n_genes_median': 1741.0, 'n_genes_std': 889.9268264741113, 'anatomy': 'Distal, Proximal, distal, proximal, nan', 'compartment': 'immune, epithelial, stromal, endothelial', 'methods': '10X, smartseq2', 'donor': 'TSP14, TSP2'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_immune.h5ad']\n",
      "{'dataset': 'immune', 'num_cells': 264824, 'female_to_male_ratio': 0.9548101835790157, 'umi_mean': 36243.04, 'umi_median': 6467.0, 'umi_std': 211887.203125, 'n_genes_mean': 2130.1319026976407, 'n_genes_median': 1880.0, 'n_genes_std': 1107.4070386050994, 'anatomy': 'nan, Inguinal, Supradiaphagmatic, Parotid, Posterior, Endometrium, Myometrium, Orbital, MAT, SCAT, Atria, Ventricle, exocrine, abdomen, chest, atria, ventricle, vertebralbody, noCornea, lacrimalgland, diaphragm, rectusabdominus, whole, inguinal, supradiaphragmatic, anterior, posterior, Sublingual, Cornea-etc, Neuroretina-etc, Sclera-etc, Distal, Proximal, Abdomen, Chest, Anterior, Diaphragm, AortaVeneCava, CoronaryArteries, distal, proximal, MedialDistal , proxmedialdistal, Neuron, Endocrine, Exocrine, Aorta, aorta', 'compartment': 'immune', 'methods': '10X, smartseq2', 'donor': 'TSP6, TSP7, TSP4, TSP5, TSP10, TSP8, TSP3, TSP11, TSP12, TSP9, TSP13, TSP14, TSP15, TSP1, TSP2'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Uterus.h5ad']\n",
      "{'dataset': 'Uterus', 'num_cells': 7124, 'female_to_male_ratio': 'F', 'umi_mean': 93465.16, 'umi_median': 10123.0, 'umi_std': 593572.0625, 'n_genes_mean': 3062.347838293094, 'n_genes_median': 2722.0, 'n_genes_std': 1520.6161696220765, 'anatomy': 'Endometrium, Myometrium, nan', 'compartment': 'immune, stromal, epithelial, endothelial', 'methods': '10X, smartseq2', 'donor': 'TSP4'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Eye.h5ad']\n",
      "{'dataset': 'Eye', 'num_cells': 10650, 'female_to_male_ratio': 0.8980573872749955, 'umi_mean': 47718.64, 'umi_median': 10517.0, 'umi_std': 433921.4375, 'n_genes_mean': 3268.511549295775, 'n_genes_median': 2792.0, 'n_genes_std': 2096.543272530692, 'anatomy': 'nan, Conjunctiva, Orbital, noCornea, lacrimalgland, whole, Cornea-etc, Neuroretina-etc, Sclera-etc', 'compartment': 'epithelial, immune, stromal, endothelial', 'methods': '10X, smartseq2', 'donor': 'TSP5, TSP3, TSP15'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Mammary.h5ad']\n",
      "{'dataset': 'Mammary', 'num_cells': 11375, 'female_to_male_ratio': 'F', 'umi_mean': 86726.71, 'umi_median': 8506.0, 'umi_std': 603710.4375, 'n_genes_mean': 2504.7185934065933, 'n_genes_median': 2318.0, 'n_genes_std': 1152.741654483558, 'anatomy': 'nan', 'compartment': 'stromal, immune, epithelial, endothelial', 'methods': '10X, smartseq2', 'donor': 'TSP4'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Prostate.h5ad']\n",
      "{'dataset': 'Prostate', 'num_cells': 16375, 'female_to_male_ratio': 'M', 'umi_mean': 91032.62, 'umi_median': 7488.0, 'umi_std': 537543.625, 'n_genes_mean': 2474.9666564885497, 'n_genes_median': 2126.0, 'n_genes_std': 1373.4602589976716, 'anatomy': 'nan', 'compartment': 'epithelial, stromal, immune, endothelial, germ line', 'methods': '10X, smartseq2', 'donor': 'TSP8, TSP14'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Heart.h5ad']\n",
      "{'dataset': 'Heart', 'num_cells': 11505, 'female_to_male_ratio': 'M', 'umi_mean': 47384.58, 'umi_median': 11373.0, 'umi_std': 271071.21875, 'n_genes_mean': 3716.8061712299, 'n_genes_median': 3208.0, 'n_genes_std': 2092.2421811017066, 'anatomy': 'Atria, Ventricle, atria, ventricle', 'compartment': 'endothelial, stromal, epithelial, immune', 'methods': '10X, smartseq2', 'donor': 'TSP12'}\n",
      "['/nfs/turbo/umms-indikar/shared/projects/DGC/data/tabula_sapiens/extract/TS_Muscle.h5ad']\n",
      "{'dataset': 'Muscle', 'num_cells': 30746, 'female_to_male_ratio': 2.6628544198236836, 'umi_mean': 123840.6, 'umi_median': 9269.0, 'umi_std': 410966.4375, 'n_genes_mean': 2428.3909776881546, 'n_genes_median': 2394.0, 'n_genes_std': 917.4659647224217, 'anatomy': 'diaphragm, rectusabdominus, Abdomen, Diaphragm, nan', 'compartment': 'endothelial, stromal, immune', 'methods': 'smartseq2, 10X', 'donor': 'TSP4, TSP14, TSP1, TSP2'}\n"
     ]
    }
   ],
   "source": [
    "data = []\n",
    "\n",
    "for d in dfData.index:\n",
    "    f = getFiles(category = 'TS_' + d, model='raw') #, model='geneformer') #_geneformer.h5ad\n",
    "    print(f)\n",
    "    adata = sc.read_h5ad(f[0])\n",
    "\n",
    "    # Number of cells\n",
    "    num_cells  = adata.shape[0]\n",
    "\n",
    "    # female to mail ratio\n",
    "    gender_distribution = adata.obs['gender'].value_counts()\n",
    "    genders = list(gender_distribution.keys())\n",
    "    if len(genders) == 2:\n",
    "        fmr = gender_distribution.female / gender_distribution.male\n",
    "    elif 'male' in genders:\n",
    "        fmr = 'M'\n",
    "    else:\n",
    "        fmr = 'F'\n",
    "    \n",
    "    # unique molecular identifiers\n",
    "    umi_mean = adata.obs['n_counts_UMIs'].mean()\n",
    "    umi_median = adata.obs['n_counts_UMIs'].median()\n",
    "    umi_std = adata.obs['n_counts_UMIs'].std()\n",
    "\n",
    "    # Number of genes\n",
    "    n_genes_mean = adata.obs['n_genes'].mean()\n",
    "    n_genes_median = adata.obs['n_genes'].median()\n",
    "    n_genes_std = adata.obs['n_genes'].std()\n",
    "    \n",
    "    # Anatomical information\n",
    "    anatomy = ', '.join(adata.obs['anatomical_information'].unique())\n",
    "\n",
    "    # Compartment\n",
    "    compartment = ', '.join(adata.obs['compartment'].unique())\n",
    "\n",
    "    # Methods\n",
    "    methods = ', '.join(adata.obs['method'].unique())\n",
    "\n",
    "    # Dataset\n",
    "    ds = ', '.join(adata.obs['donor'].unique())\n",
    "\n",
    "    # Append the collected data to the list\n",
    "    data.append({\n",
    "        'dataset': d,\n",
    "        'num_cells': num_cells,\n",
    "        'female_to_male_ratio': fmr,\n",
    "        'umi_mean': umi_mean,\n",
    "        'umi_median': umi_median,\n",
    "        'umi_std': umi_std,\n",
    "        'n_genes_mean': n_genes_mean,\n",
    "        'n_genes_median': n_genes_median,\n",
    "        'n_genes_std': n_genes_std,\n",
    "        'anatomy': anatomy,\n",
    "        'compartment': compartment,\n",
    "        'methods': methods,\n",
    "        'donor': ds\n",
    "    })\n",
    "    print(data[-1])\n",
    "\n",
    "# Create a DataFrame from the collected data\n",
    "df_results = pd.DataFrame(data)\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "efd49b7b-ea2d-447c-86e6-3fe02dcdaa8f",
   "metadata": {},
   "source": [
    "# You should not run this cell again so it has been set to raw\n",
    "data_summary_file = '/nfs/turbo/umms-indikar/shared/projects/foundation_models/experiments/tabulaSapiens/data_summary.pkl'\n",
    "with open(data_summary_file, 'wb') as file:\n",
    "    pickle.dump(data, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34c258b8-c1db-4a95-bf60-2474893e9bb1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a16a9456-2b79-4ec7-a380-e25f882db3fd",
   "metadata": {},
   "source": [
    "## Dataset summary table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c7b8b8f6-6c57-4b55-952f-9f31da2227c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{tabular}{l|rrrrrrrrrrrrrr}\n",
      "\\multicolumn{2}{c}{}& & \\multicolumn{3}{c}{UMI} & \\multicolumn{3}{c}{Num. Genes} &\\multicolumn{3}{c}{Categories}&\\multicolumn{2}{c}{Assays} \\\\\n",
      "\\cmidrule(lr){4-6} \\cmidrule(lr){7-9}\\cmidrule(lr){10-12}\\cmidrule(lr){13-14}\n",
      "Dataset & \\# Cells & F/M &avg.&med.&std.&avg.&med.&std.&\\# Anatomy &\\# Donors&\\# Compartments&10X&smartseq2\\\\\n",
      "    \\hline\\hline\n",
      "germ line & 11 & M & 1.4e+04 & 1.2e+04 & 7.4e+03 & 2.8e+03 & 2.8e+03 & 1.1e+03 & 0 & 1 & 1 & \\checkmark &  \\\\\n",
      "stromal & 82478 & 0.84 & 6.9e+04 & 1e+04 & 3.5e+05 & 3.1e+03 & 2.8e+03 & 1.5e+03 & 0 & 13 & 1 & \\checkmark & \\checkmark \\\\\n",
      "Large Intestine & 13680 & 0.2 & 2.4e+04 & 6e+03 & 1.1e+05 & 1.8e+03 & 1.5e+03 & 9.8e+02 & 5 & 2 & 4 & \\checkmark & \\checkmark \\\\\n",
      "Lymph Node & 53275 & 0.75 & 2.5e+04 & 5.9e+03 & 1.2e+05 & 2e+03 & 1.8e+03 & 8.1e+02 & 5 & 3 & 3 & \\checkmark & \\checkmark \\\\\n",
      "Skin & 9424 & M & 1.2e+05 & 7.4e+03 & 5.3e+05 & 2.2e+03 & 2e+03 & 9.7e+02 & 0 & 2 & 4 & \\checkmark & \\checkmark \\\\\n",
      "epithelial & 104148 & 1.1 & 7.1e+04 & 1.2e+04 & 4.1e+05 & 3e+03 & 2.7e+03 & 1.8e+03 & 0 & 13 & 1 & \\checkmark & \\checkmark \\\\\n",
      "Spleen & 34004 & 0.79 & 3.3e+04 & 6.6e+03 & 1.5e+05 & 2.3e+03 & 2.1e+03 & 1.2e+03 & 0 & 3 & 2 & \\checkmark & \\checkmark \\\\\n",
      "Vasculature & 16037 & 1.2 & 2.4e+04 & 1.3e+04 & 5.2e+04 & 3.6e+03 & 3.4e+03 & 1.7e+03 & 5 & 2 & 4 & \\checkmark & \\checkmark \\\\\n",
      "Kidney & 9641 & F & 2.4e+04 & 8.5e+03 & 9.9e+04 & 2.7e+03 & 2.5e+03 & 1.1e+03 & 0 & 1 & 3 & \\checkmark & \\checkmark \\\\\n",
      "Salivary Gland & 27199 & 0.53 & 4e+04 & 7.1e+03 & 2.1e+05 & 2.4e+03 & 2.1e+03 & 1.7e+03 & 2 & 2 & 4 & \\checkmark & \\checkmark \\\\\n",
      "Thymus & 33664 & 0.56 & 1.5e+04 & 6.4e+03 & 6.9e+04 & 2.4e+03 & 2.1e+03 & 1.1e+03 & 0 & 2 & 4 & \\checkmark & \\checkmark \\\\\n",
      "Pancreas & 13497 & 1 & 1.1e+05 & 1.3e+04 & 5.2e+05 & 3e+03 & 2.6e+03 & 1.7e+03 & 3 & 2 & 4 & \\checkmark & \\checkmark \\\\\n",
      "Fat & 20263 & M & 5.7e+04 & 1.2e+04 & 4.1e+05 & 3.2e+03 & 3.1e+03 & 1.4e+03 & 2 & 2 & 3 & \\checkmark & \\checkmark \\\\\n",
      "Bladder & 24583 & 7 & 4.4e+04 & 9.3e+03 & 2.4e+05 & 2.7e+03 & 2.4e+03 & 1.4e+03 & 0 & 3 & 4 & \\checkmark & \\checkmark \\\\\n",
      "endothelial & 31691 & 0.98 & 7.8e+04 & 7.4e+03 & 3.9e+05 & 2.7e+03 & 2.4e+03 & 1.3e+03 & 0 & 13 & 1 & \\checkmark & \\checkmark \\\\\n",
      "Blood & 50115 & 2 & 2.6e+04 & 7.8e+03 & 1.6e+05 & 1.8e+03 & 1.7e+03 & 1.2e+03 & 0 & 6 & 1 & \\checkmark & \\checkmark \\\\\n",
      "Trachea & 9522 & 0.86 & 5.8e+04 & 7.8e+03 & 2.8e+05 & 2.8e+03 & 2.5e+03 & 1.5e+03 & 0 & 2 & 4 & \\checkmark & \\checkmark \\\\\n",
      "Bone Marrow & 12297 & 0.38 & 1.1e+05 & 7.7e+03 & 4.1e+05 & 2.4e+03 & 2e+03 & 1.5e+03 & 0 & 4 & 1 & \\checkmark & \\checkmark \\\\\n",
      "Tongue & 15020 & 1.8 & 8.1e+04 & 1.1e+04 & 3.9e+05 & 2.7e+03 & 2.4e+03 & 1.5e+03 & 5 & 3 & 4 & \\checkmark & \\checkmark \\\\\n",
      "Lung & 35682 & 7.2 & 6.8e+04 & 1.3e+04 & 3.2e+05 & 3.3e+03 & 3.2e+03 & 1.5e+03 & 6 & 3 & 4 & \\checkmark & \\checkmark \\\\\n",
      "Liver & 5007 & M & 2.8e+04 & 6.3e+03 & 2.2e+05 & 2.3e+03 & 2.1e+03 & 9.5e+02 & 0 & 2 & 4 & \\checkmark & \\checkmark \\\\\n",
      "Small Intestine & 12467 & 0.19 & 3.3e+04 & 6.4e+03 & 1.8e+05 & 2e+03 & 1.7e+03 & 8.9e+02 & 5 & 2 & 4 & \\checkmark & \\checkmark \\\\\n",
      "immune & 264824 & 0.95 & 3.6e+04 & 6.5e+03 & 2.1e+05 & 2.1e+03 & 1.9e+03 & 1.1e+03 & 0 & 15 & 1 & \\checkmark & \\checkmark \\\\\n",
      "Uterus & 7124 & F & 9.3e+04 & 1e+04 & 5.9e+05 & 3.1e+03 & 2.7e+03 & 1.5e+03 & 3 & 1 & 4 & \\checkmark & \\checkmark \\\\\n",
      "Eye & 10650 & 0.9 & 4.8e+04 & 1.1e+04 & 4.3e+05 & 3.3e+03 & 2.8e+03 & 2.1e+03 & 0 & 3 & 4 & \\checkmark & \\checkmark \\\\\n",
      "Mammary & 11375 & F & 8.7e+04 & 8.5e+03 & 6e+05 & 2.5e+03 & 2.3e+03 & 1.2e+03 & 0 & 1 & 4 & \\checkmark & \\checkmark \\\\\n",
      "Prostate & 16375 & M & 9.1e+04 & 7.5e+03 & 5.4e+05 & 2.5e+03 & 2.1e+03 & 1.4e+03 & 0 & 2 & 5 & \\checkmark & \\checkmark \\\\\n",
      "Heart & 11505 & M & 4.7e+04 & 1.1e+04 & 2.7e+05 & 3.7e+03 & 3.2e+03 & 2.1e+03 & 4 & 1 & 4 & \\checkmark & \\checkmark \\\\\n",
      "Muscle & 30746 & 2.7 & 1.2e+05 & 9.3e+03 & 4.1e+05 & 2.4e+03 & 2.4e+03 & 9.2e+02 & 5 & 4 & 3 & \\checkmark & \\checkmark \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3074957/2707408305.py:15: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  df_results = df_results.applymap(format_number)\n"
     ]
    }
   ],
   "source": [
    "df_results = pd.DataFrame(data)\n",
    "df_results['anatomy list'] = df_results['anatomy'].apply(lambda x: x.split(','))\n",
    "df_results['num. anatomy'] = df_results['anatomy list'].apply(lambda x: 0 if x[0] == 'nan' else len(x))\n",
    "df_results = df_results.drop(['anatomy list', 'anatomy'], axis=1)\n",
    "df_results['donor list'] = df_results['donor'].apply(lambda x: x.split(','))\n",
    "df_results['num. donor'] = df_results['donor list'].apply(lambda x: 0 if x[0] == 'nan' else len(x))\n",
    "df_results = df_results.drop(['donor list', 'donor'], axis=1)\n",
    "df_results['compartment list'] = df_results['compartment'].apply(lambda x: x.split(','))\n",
    "df_results['num. compartment'] = df_results['compartment list'].apply(lambda x: 0 if x[0] == 'nan' else len(x))\n",
    "df_results = df_results.drop(['compartment list', 'compartment'], axis=1)\n",
    "df_results['10X'] = df_results['methods'].apply(lambda x: '\\checkmark' if '10X' in x else '')\n",
    "df_results['smartseq2'] = df_results['methods'].apply(lambda x: '\\checkmark' if 'smartseq2' in x else '')\n",
    "df_results = df_results.drop(['methods'], axis=1)\n",
    "df_results.columns = ['Dataset', '\\# Cells', 'F/M', 'avg. UMI', 'med. UMI', 'std. UMI', 'avg. \\# genes', 'med. \\# genes', 'std. \\# genes', '\\# anatomy', '\\# donors', '\\# compartments', '10X', 'smartseq2']\n",
    "df_results = df_results.applymap(format_number)\n",
    "\n",
    "\n",
    "latex_table = df_results.to_latex(index=False, escape=False, float_format=\"%.3f\")\n",
    "\n",
    "custom_table = latex_table.replace(\n",
    "    r'\\begin{tabular}{lrlllllllllll}',\n",
    "    r'\\begin{tabular}{l|rrrrrrrrrrrr}'\n",
    ").replace(\n",
    "    r'\\toprule',\n",
    "    r'\\hline'\n",
    ").replace(\n",
    "    r'\\midrule',\n",
    "    r'\\hline\\hline'\n",
    ").replace(\n",
    "    '_', ' '\n",
    ").replace(\n",
    "    r\"\"\"\\begin{tabular}{lrlllllllrrrll}\n",
    "\\hline\n",
    "Dataset & \\# Cells & F/M & avg. UMI & med. UMI & std. UMI & avg. \\# genes & med. \\# genes & std. \\# genes & \\# anatomy & \\# donors & \\# compartments & 10X & smartseq2 \\\\\n",
    "\"\"\",\n",
    "    r\"\"\"\\begin{tabular}{l|rrrrrrrrrrrrrr}\n",
    "\\multicolumn{2}{c}{}& & \\multicolumn{3}{c}{UMI} & \\multicolumn{3}{c}{Num. Genes} &\\multicolumn{3}{c}{Categories}&\\multicolumn{2}{c}{Assays} \\\\\n",
    "\\cmidrule(lr){4-6} \\cmidrule(lr){7-9}\\cmidrule(lr){10-12}\\cmidrule(lr){13-14}\n",
    "Dataset & \\# Cells & F/M &avg.&med.&std.&avg.&med.&std.&\\# Anatomy &\\# Donors&\\# Compartments&10X&smartseq2\\\\\n",
    "    \"\"\"\n",
    ")\n",
    "\n",
    "# Print the LaTeX table\n",
    "print(custom_table)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d10a7b2c-9705-46a8-844c-c989023808fe",
   "metadata": {},
   "source": [
    "## Categories Tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "b2eca47f-a998-4cd9-ac84-f7de041e60a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{tabular}{l|p{0.8\\linewidth}}\n",
      "\\toprule\n",
      "Dataset & Anatomy \\\\\n",
      "\\hline\\hline\n",
      "germ line &  \\\\\n",
      "stromal &  Supradiaphagmatic,  Parotid,  Posterior,  Endometrium,  Myometrium,  MAT,  SCAT,  Atria,  Ventricle,  exocrine,  abdomen,  chest,  atria,  ventricle,  noCornea,  lacrimalgland,  diaphragm,  rectusabdominus,  whole,  anterior,  posterior,  Sublingual,  Cornea-etc,  Neuroretina-etc,  Sclera-etc,  Distal,  Proximal,  Abdomen,  Chest,  Anterior,  Diaphragm,  AortaVeneCava,  CoronaryArteries,  distal,  proximal,  MedialDistal ,  proxmedialdistal,  Neuron,  Endocrine,  Exocrine,  Aorta,  aorta \\\\\n",
      "Large Intestine & Distal,  Proximal,  distal,  proximal,   \\\\\n",
      "Lymph Node & Inguinal,  Supradiaphagmatic,  inguinal,  supradiaphragmatic,   \\\\\n",
      "Skin &  abdomen,  chest,  Abdomen,  Chest \\\\\n",
      "epithelial &  Parotid,  Anterior,  Posterior,  Endometrium,  Myometrium,  Conjunctiva,  Atria,  Ventricle,  exocrine,  abdomen,  chest,  atria,  ventricle,  noCornea,  lacrimalgland,  whole,  anterior,  posterior,  Sublingual,  Cornea-etc,  Neuroretina-etc,  Sclera-etc,  Distal,  Proximal,  Abdomen,  Chest,  distal,  proximal,  MedialDistal ,  proxmedialdistal,  Neuron,  Endocrine,  Exocrine,  Aorta \\\\\n",
      "Spleen &  \\\\\n",
      "Vasculature & AortaVeneCava,  CoronaryArteries,  Aorta,   aorta \\\\\n",
      "Kidney &  \\\\\n",
      "Salivary Gland & Parotid,  Sublingual \\\\\n",
      "Thymus &  \\\\\n",
      "Pancreas & exocrine,  Endocrine,  Exocrine \\\\\n",
      "Fat & MAT,  SCAT \\\\\n",
      "Bladder &  \\\\\n",
      "endothelial &  Inguinal,  Supradiaphagmatic,  Parotid,  Posterior,  Endometrium,  Myometrium,  Orbital,  MAT,  SCAT,  Atria,  Ventricle,  exocrine,  abdomen,  chest,  atria,  ventricle,  lacrimalgland,  noCornea,  diaphragm,  rectusabdominus,  inguinal,  supradiaphragmatic,  anterior,  posterior,  Sublingual,  Cornea-etc,  Sclera-etc,  Distal,  Proximal,  Abdomen,  Chest,  Anterior,  Diaphragm,  AortaVeneCava,  CoronaryArteries,  distal,  proximal,  MedialDistal ,  proxmedialdistal,  Endocrine,  Exocrine,  Aorta,  aorta \\\\\n",
      "Blood &  \\\\\n",
      "Trachea &  \\\\\n",
      "Bone Marrow &  vertebralbody \\\\\n",
      "Tongue & Anterior,  Posterior,   anterior,  posterior \\\\\n",
      "Lung & Distal,  Proximal,   MedialDistal ,  proxmedialdistal,  Neuron \\\\\n",
      "Liver &  \\\\\n",
      "Small Intestine & Distal,  Proximal,  distal,  proximal,   \\\\\n",
      "immune &  Inguinal,  Supradiaphagmatic,  Parotid,  Posterior,  Endometrium,  Myometrium,  Orbital,  MAT,  SCAT,  Atria,  Ventricle,  exocrine,  abdomen,  chest,  atria,  ventricle,  vertebralbody,  noCornea,  lacrimalgland,  diaphragm,  rectusabdominus,  whole,  inguinal,  supradiaphragmatic,  anterior,  posterior,  Sublingual,  Cornea-etc,  Neuroretina-etc,  Sclera-etc,  Distal,  Proximal,  Abdomen,  Chest,  Anterior,  Diaphragm,  AortaVeneCava,  CoronaryArteries,  distal,  proximal,  MedialDistal ,  proxmedialdistal,  Neuron,  Endocrine,  Exocrine,  Aorta,  aorta \\\\\n",
      "Uterus & Endometrium,  Myometrium,   \\\\\n",
      "Eye &  Conjunctiva,  Orbital,  noCornea,  lacrimalgland,  whole,  Cornea-etc,  Neuroretina-etc,  Sclera-etc \\\\\n",
      "Mammary &  \\\\\n",
      "Prostate &  \\\\\n",
      "Heart & Atria,  Ventricle,  atria,  ventricle \\\\\n",
      "Muscle & diaphragm,  rectusabdominus,  Abdomen,  Diaphragm,   \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Anatomy table\n",
    "df_results = pd.DataFrame(data)\n",
    "df_results = df_results.drop(['num_cells', 'female_to_male_ratio', 'num_cells', 'female_to_male_ratio', 'umi_mean',\n",
    "       'umi_median', 'umi_std', 'n_genes_mean', 'n_genes_median',\n",
    "       'n_genes_std', 'compartment', 'methods', 'donor'], axis=1)\n",
    "df_results.columns = ['Dataset', 'Anatomy']\n",
    "df_results['anatomy list'] = df_results['Anatomy'].apply(lambda x: x.split(','))\n",
    "df_results['Anatomy'] = df_results['anatomy list'].apply(lambda x: ', '.join(x))\n",
    "df_results = df_results.drop(['anatomy list'], axis=1)\n",
    "latex = df_results.to_latex(index=False)\n",
    "latex = latex.replace('nan, ', '').replace('nan,', '').replace('nan', '').replace('_', ' ').replace('\\midrule', '\\hline\\hline').replace(r'\\begin{tabular}{ll}', r'\\begin{tabular}{l|p{0.8\\linewidth}}')\n",
    "\n",
    "print(latex)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "6fb3e336-5221-46f2-9adf-111df09edf3a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{tabular}{l|l}\n",
      "\\toprule\n",
      "Dataset & Compartment \\\\\n",
      "\\hline\\hline\n",
      "germ line & germ line \\\\\n",
      "stromal & stromal \\\\\n",
      "Large Intestine & immune,  epithelial,  stromal,  endothelial \\\\\n",
      "Lymph Node & immune,  endothelial,  stromal \\\\\n",
      "Skin & immune,  stromal,  endothelial,  epithelial \\\\\n",
      "epithelial & epithelial \\\\\n",
      "Spleen & immune,  endothelial \\\\\n",
      "Vasculature & stromal,  immune,  endothelial,  epithelial \\\\\n",
      "Kidney & epithelial,  immune,  endothelial \\\\\n",
      "Salivary Gland & epithelial,  immune,  stromal,  endothelial \\\\\n",
      "Thymus & immune,  epithelial,  stromal,  endothelial \\\\\n",
      "Pancreas & epithelial,  immune,  endothelial,  stromal \\\\\n",
      "Fat & stromal,  endothelial,  immune \\\\\n",
      "Bladder & immune,  stromal,  epithelial,  endothelial \\\\\n",
      "endothelial & endothelial \\\\\n",
      "Blood & immune \\\\\n",
      "Trachea & immune,  epithelial,  stromal,  endothelial \\\\\n",
      "Bone Marrow & immune \\\\\n",
      "Tongue & epithelial,  immune,  stromal,  endothelial \\\\\n",
      "Lung & epithelial,  immune,  endothelial,  stromal \\\\\n",
      "Liver & immune,  endothelial,  stromal,  epithelial \\\\\n",
      "Small Intestine & immune,  epithelial,  stromal,  endothelial \\\\\n",
      "immune & immune \\\\\n",
      "Uterus & immune,  stromal,  epithelial,  endothelial \\\\\n",
      "Eye & epithelial,  immune,  stromal,  endothelial \\\\\n",
      "Mammary & stromal,  immune,  epithelial,  endothelial \\\\\n",
      "Prostate & epithelial,  stromal,  immune,  endothelial,  germ line \\\\\n",
      "Heart & endothelial,  stromal,  epithelial,  immune \\\\\n",
      "Muscle & endothelial,  stromal,  immune \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Compartment table\n",
    "df_results = pd.DataFrame(data)\n",
    "df_results = df_results.drop(['num_cells', 'female_to_male_ratio', 'num_cells', 'female_to_male_ratio', 'umi_mean',\n",
    "       'umi_median', 'umi_std', 'n_genes_mean', 'n_genes_median',\n",
    "       'n_genes_std', 'anatomy', 'methods', 'donor'], axis=1)\n",
    "df_results.columns = ['Dataset', 'Compartment']\n",
    "df_results['Compartment list'] = df_results['Compartment'].apply(lambda x: x.split(','))\n",
    "df_results['Compartment'] = df_results['Compartment list'].apply(lambda x: ', '.join(x))\n",
    "df_results = df_results.drop(['Compartment list'], axis=1)\n",
    "latex = df_results.to_latex(index=False)\n",
    "latex = latex.replace('nan, ', '').replace('nan,', '').replace('nan', '').replace('_', ' ').replace('\\midrule', '\\hline\\hline').replace(r'\\begin{tabular}{ll}', r'\\begin{tabular}{l|l}')\n",
    "\n",
    "print(latex)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "0865e2aa-685e-46d5-a4c3-e72ba739e9e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{tabular}{l|l}\n",
      "\\toprule\n",
      "Dataset & Donor \\\\\n",
      "\\hline\\hline\n",
      "germ line & TSP14 \\\\\n",
      "stromal & TSP6,  TSP7,  TSP4,  TSP5,  TSP3,  TSP10,  TSP12,  TSP9,  TSP8,  TSP14,  TSP15,  TSP1,  TSP2 \\\\\n",
      "Large Intestine & TSP14,  TSP2 \\\\\n",
      "Lymph Node & TSP7,  TSP14,  TSP2 \\\\\n",
      "Skin & TSP10,  TSP14 \\\\\n",
      "epithelial & TSP6,  TSP7,  TSP4,  TSP5,  TSP3,  TSP10,  TSP12,  TSP9,  TSP8,  TSP14,  TSP15,  TSP1,  TSP2 \\\\\n",
      "Spleen & TSP7,  TSP14,  TSP2 \\\\\n",
      "Vasculature & TSP14,  TSP2 \\\\\n",
      "Kidney & TSP2 \\\\\n",
      "Salivary Gland & TSP7,  TSP14 \\\\\n",
      "Thymus & TSP14,  TSP2 \\\\\n",
      "Pancreas & TSP9,  TSP1 \\\\\n",
      "Fat & TSP10,  TSP14 \\\\\n",
      "Bladder & TSP14,  TSP1,  TSP2 \\\\\n",
      "endothelial & TSP6,  TSP7,  TSP4,  TSP5,  TSP3,  TSP10,  TSP12,  TSP9,  TSP8,  TSP14,  TSP15,  TSP1,  TSP2 \\\\\n",
      "Blood & TSP7,  TSP10,  TSP8,  TSP14,  TSP1,  TSP2 \\\\\n",
      "Trachea & TSP6,  TSP2 \\\\\n",
      "Bone Marrow & TSP11,  TSP13,  TSP14,  TSP2 \\\\\n",
      "Tongue & TSP7,  TSP4,  TSP14 \\\\\n",
      "Lung & TSP14,  TSP1,  TSP2 \\\\\n",
      "Liver & TSP6,  TSP14 \\\\\n",
      "Small Intestine & TSP14,  TSP2 \\\\\n",
      "immune & TSP6,  TSP7,  TSP4,  TSP5,  TSP10,  TSP8,  TSP3,  TSP11,  TSP12,  TSP9,  TSP13,  TSP14,  TSP15,  TSP1,  TSP2 \\\\\n",
      "Uterus & TSP4 \\\\\n",
      "Eye & TSP5,  TSP3,  TSP15 \\\\\n",
      "Mammary & TSP4 \\\\\n",
      "Prostate & TSP8,  TSP14 \\\\\n",
      "Heart & TSP12 \\\\\n",
      "Muscle & TSP4,  TSP14,  TSP1,  TSP2 \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Donor table\n",
    "df_results = pd.DataFrame(data)\n",
    "df_results = df_results.drop(['num_cells', 'female_to_male_ratio', 'num_cells', 'female_to_male_ratio', 'umi_mean',\n",
    "       'umi_median', 'umi_std', 'n_genes_mean', 'n_genes_median',\n",
    "       'n_genes_std', 'anatomy', 'methods', 'compartment'], axis=1)\n",
    "df_results.columns = ['Dataset', 'Donor']\n",
    "df_results['Donor list'] = df_results['Donor'].apply(lambda x: x.split(','))\n",
    "df_results['Donor'] = df_results['Donor list'].apply(lambda x: ', '.join(x))\n",
    "df_results = df_results.drop(['Donor list'], axis=1)\n",
    "latex = df_results.to_latex(index=False)\n",
    "latex = latex.replace('nan, ', '').replace('nan,', '').replace('nan', '').replace('_', ' ').replace('\\midrule', '\\hline\\hline').replace(r'\\begin{tabular}{ll}', r'\\begin{tabular}{l|l}')\n",
    "\n",
    "print(latex)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcadfda4-3241-46be-b0f4-2783fee00f1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_results = pd.DataFrame(data)\n",
    "df_results['anatomy list'] = df_results['anatomy'].apply(lambda x: x.split(','))\n",
    "df_results['num. anatomy'] = df_results['anatomy list'].apply(lambda x: 0 if x[0] == 'nan' else len(x))\n",
    "df_results = df_results.drop(['anatomy list', 'anatomy'], axis=1)\n",
    "df_results['donor list'] = df_results['donor'].apply(lambda x: x.split(','))\n",
    "df_results['num. donor'] = df_results['donor list'].apply(lambda x: 0 if x[0] == 'nan' else len(x))\n",
    "df_results = df_results.drop(['donor list', 'donor'], axis=1)\n",
    "df_results['compartment list'] = df_results['compartment'].apply(lambda x: x.split(','))\n",
    "df_results['num. compartment'] = df_results['compartment list'].apply(lambda x: 0 if x[0] == 'nan' else len(x))\n",
    "df_results = df_results.drop(['compartment list', 'compartment'], axis=1)\n",
    "df_results['10X'] = df_results['methods'].apply(lambda x: '\\checkmark' if '10X' in x else '')\n",
    "df_results['smartseq2'] = df_results['methods'].apply(lambda x: '\\checkmark' if 'smartseq2' in x else '')\n",
    "df_results = df_results.drop(['methods'], axis=1)\n",
    "df_results.columns = ['Dataset', '\\# Cells', 'F/M', 'avg. UMI', 'med. UMI', 'std. UMI', 'avg. \\# genes', 'med. \\# genes', 'std. \\# genes', '\\# anatomy', '\\# donors', '\\# compartments', '10X', 'smartseq2']\n",
    "df_results = df_results.applymap(format_number)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf42cb96-a4f9-4f2a-864d-e40a58de3465",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "62f5c10d-9940-45ee-b982-f539881cd28c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{tabular}{l|rrrrrrrrrrrr}\n",
      "\\hline\n",
      "Dataset & Number of Cells & F/M & avg. UMI & med. UMI & std. UMI & avg. num. genes & med. num. genes & std. num. genes & anatomy & compartment & assay & donor \\\\\n",
      "\\hline\\hline\n",
      "germ line & 11 & M & 1.4e+04 & 1.2e+04 & 7.4e+03 & 2.8e+03 & 2.8e+03 & 1.1e+03 & nan & germ line & 10X & TSP14 \\\\\n",
      "stromal & 82478 & 0.84 & 6.9e+04 & 1e+04 & 3.5e+05 & 3.1e+03 & 2.8e+03 & 1.5e+03 & nan, Supradiaphagmatic, Parotid, Posterior, Endometrium, Myometrium, MAT, SCAT, Atria, Ventricle, exocrine, abdomen, chest, atria, ventricle, noCornea, lacrimalgland, diaphragm, rectusabdominus, whole, anterior, posterior, Sublingual, Cornea-etc, Neuroretina-etc, Sclera-etc, Distal, Proximal, Abdomen, Chest, Anterior, Diaphragm, AortaVeneCava, CoronaryArteries, distal, proximal, MedialDistal , proxmedialdistal, Neuron, Endocrine, Exocrine, Aorta, aorta & stromal & 10X, smartseq2 & TSP6, TSP7, TSP4, TSP5, TSP3, TSP10, TSP12, TSP9, TSP8, TSP14, TSP15, TSP1, TSP2 \\\\\n",
      "Large Intestine & 13680 & 0.2 & 2.4e+04 & 6e+03 & 1.1e+05 & 1.8e+03 & 1.5e+03 & 9.8e+02 & Distal, Proximal, distal, proximal, nan & immune, epithelial, stromal, endothelial & 10X, smartseq2 & TSP14, TSP2 \\\\\n",
      "Lymph Node & 53275 & 0.75 & 2.5e+04 & 5.9e+03 & 1.2e+05 & 2e+03 & 1.8e+03 & 8.1e+02 & Inguinal, Supradiaphagmatic, inguinal, supradiaphragmatic, nan & immune, endothelial, stromal & 10X, smartseq2 & TSP7, TSP14, TSP2 \\\\\n",
      "Skin & 9424 & M & 1.2e+05 & 7.4e+03 & 5.3e+05 & 2.2e+03 & 2e+03 & 9.7e+02 & nan, abdomen, chest, Abdomen, Chest & immune, stromal, endothelial, epithelial & 10X, smartseq2 & TSP10, TSP14 \\\\\n",
      "epithelial & 104148 & 1.1 & 7.1e+04 & 1.2e+04 & 4.1e+05 & 3e+03 & 2.7e+03 & 1.8e+03 & nan, Parotid, Anterior, Posterior, Endometrium, Myometrium, Conjunctiva, Atria, Ventricle, exocrine, abdomen, chest, atria, ventricle, noCornea, lacrimalgland, whole, anterior, posterior, Sublingual, Cornea-etc, Neuroretina-etc, Sclera-etc, Distal, Proximal, Abdomen, Chest, distal, proximal, MedialDistal , proxmedialdistal, Neuron, Endocrine, Exocrine, Aorta & epithelial & 10X, smartseq2 & TSP6, TSP7, TSP4, TSP5, TSP3, TSP10, TSP12, TSP9, TSP8, TSP14, TSP15, TSP1, TSP2 \\\\\n",
      "Spleen & 34004 & 0.79 & 3.3e+04 & 6.6e+03 & 1.5e+05 & 2.3e+03 & 2.1e+03 & 1.2e+03 & nan & immune, endothelial & 10X, smartseq2 & TSP7, TSP14, TSP2 \\\\\n",
      "Vasculature & 16037 & 1.2 & 2.4e+04 & 1.3e+04 & 5.2e+04 & 3.6e+03 & 3.4e+03 & 1.7e+03 & AortaVeneCava, CoronaryArteries, Aorta, nan, aorta & stromal, immune, endothelial, epithelial & 10X, smartseq2 & TSP14, TSP2 \\\\\n",
      "Kidney & 9641 & F & 2.4e+04 & 8.5e+03 & 9.9e+04 & 2.7e+03 & 2.5e+03 & 1.1e+03 & nan & epithelial, immune, endothelial & 10X, smartseq2 & TSP2 \\\\\n",
      "Salivary Gland & 27199 & 0.53 & 4e+04 & 7.1e+03 & 2.1e+05 & 2.4e+03 & 2.1e+03 & 1.7e+03 & Parotid, Sublingual & epithelial, immune, stromal, endothelial & 10X, smartseq2 & TSP7, TSP14 \\\\\n",
      "Thymus & 33664 & 0.56 & 1.5e+04 & 6.4e+03 & 6.9e+04 & 2.4e+03 & 2.1e+03 & 1.1e+03 & nan & immune, epithelial, stromal, endothelial & 10X, smartseq2 & TSP14, TSP2 \\\\\n",
      "Pancreas & 13497 & 1 & 1.1e+05 & 1.3e+04 & 5.2e+05 & 3e+03 & 2.6e+03 & 1.7e+03 & exocrine, Endocrine, Exocrine & epithelial, immune, endothelial, stromal & 10X, smartseq2 & TSP9, TSP1 \\\\\n",
      "Fat & 20263 & M & 5.7e+04 & 1.2e+04 & 4.1e+05 & 3.2e+03 & 3.1e+03 & 1.4e+03 & MAT, SCAT & stromal, endothelial, immune & 10X, smartseq2 & TSP10, TSP14 \\\\\n",
      "Bladder & 24583 & 7 & 4.4e+04 & 9.3e+03 & 2.4e+05 & 2.7e+03 & 2.4e+03 & 1.4e+03 & nan & immune, stromal, epithelial, endothelial & 10X, smartseq2 & TSP14, TSP1, TSP2 \\\\\n",
      "endothelial & 31691 & 0.98 & 7.8e+04 & 7.4e+03 & 3.9e+05 & 2.7e+03 & 2.4e+03 & 1.3e+03 & nan, Inguinal, Supradiaphagmatic, Parotid, Posterior, Endometrium, Myometrium, Orbital, MAT, SCAT, Atria, Ventricle, exocrine, abdomen, chest, atria, ventricle, lacrimalgland, noCornea, diaphragm, rectusabdominus, inguinal, supradiaphragmatic, anterior, posterior, Sublingual, Cornea-etc, Sclera-etc, Distal, Proximal, Abdomen, Chest, Anterior, Diaphragm, AortaVeneCava, CoronaryArteries, distal, proximal, MedialDistal , proxmedialdistal, Endocrine, Exocrine, Aorta, aorta & endothelial & 10X, smartseq2 & TSP6, TSP7, TSP4, TSP5, TSP3, TSP10, TSP12, TSP9, TSP8, TSP14, TSP15, TSP1, TSP2 \\\\\n",
      "Blood & 50115 & 2 & 2.6e+04 & 7.8e+03 & 1.6e+05 & 1.8e+03 & 1.7e+03 & 1.2e+03 & nan & immune & 10X, smartseq2 & TSP7, TSP10, TSP8, TSP14, TSP1, TSP2 \\\\\n",
      "Trachea & 9522 & 0.86 & 5.8e+04 & 7.8e+03 & 2.8e+05 & 2.8e+03 & 2.5e+03 & 1.5e+03 & nan & immune, epithelial, stromal, endothelial & 10X, smartseq2 & TSP6, TSP2 \\\\\n",
      "Bone Marrow & 12297 & 0.38 & 1.1e+05 & 7.7e+03 & 4.1e+05 & 2.4e+03 & 2e+03 & 1.5e+03 & nan, vertebralbody & immune & 10X, smartseq2 & TSP11, TSP13, TSP14, TSP2 \\\\\n",
      "Tongue & 15020 & 1.8 & 8.1e+04 & 1.1e+04 & 3.9e+05 & 2.7e+03 & 2.4e+03 & 1.5e+03 & Anterior, Posterior, nan, anterior, posterior & epithelial, immune, stromal, endothelial & 10X, smartseq2 & TSP7, TSP4, TSP14 \\\\\n",
      "Lung & 35682 & 7.2 & 6.8e+04 & 1.3e+04 & 3.2e+05 & 3.3e+03 & 3.2e+03 & 1.5e+03 & Distal, Proximal, nan, MedialDistal , proxmedialdistal, Neuron & epithelial, immune, endothelial, stromal & 10X, smartseq2 & TSP14, TSP1, TSP2 \\\\\n",
      "Liver & 5007 & M & 2.8e+04 & 6.3e+03 & 2.2e+05 & 2.3e+03 & 2.1e+03 & 9.5e+02 & nan & immune, endothelial, stromal, epithelial & 10X, smartseq2 & TSP6, TSP14 \\\\\n",
      "Small Intestine & 12467 & 0.19 & 3.3e+04 & 6.4e+03 & 1.8e+05 & 2e+03 & 1.7e+03 & 8.9e+02 & Distal, Proximal, distal, proximal, nan & immune, epithelial, stromal, endothelial & 10X, smartseq2 & TSP14, TSP2 \\\\\n",
      "immune & 264824 & 0.95 & 3.6e+04 & 6.5e+03 & 2.1e+05 & 2.1e+03 & 1.9e+03 & 1.1e+03 & nan, Inguinal, Supradiaphagmatic, Parotid, Posterior, Endometrium, Myometrium, Orbital, MAT, SCAT, Atria, Ventricle, exocrine, abdomen, chest, atria, ventricle, vertebralbody, noCornea, lacrimalgland, diaphragm, rectusabdominus, whole, inguinal, supradiaphragmatic, anterior, posterior, Sublingual, Cornea-etc, Neuroretina-etc, Sclera-etc, Distal, Proximal, Abdomen, Chest, Anterior, Diaphragm, AortaVeneCava, CoronaryArteries, distal, proximal, MedialDistal , proxmedialdistal, Neuron, Endocrine, Exocrine, Aorta, aorta & immune & 10X, smartseq2 & TSP6, TSP7, TSP4, TSP5, TSP10, TSP8, TSP3, TSP11, TSP12, TSP9, TSP13, TSP14, TSP15, TSP1, TSP2 \\\\\n",
      "Uterus & 7124 & F & 9.3e+04 & 1e+04 & 5.9e+05 & 3.1e+03 & 2.7e+03 & 1.5e+03 & Endometrium, Myometrium, nan & immune, stromal, epithelial, endothelial & 10X, smartseq2 & TSP4 \\\\\n",
      "Eye & 10650 & 0.9 & 4.8e+04 & 1.1e+04 & 4.3e+05 & 3.3e+03 & 2.8e+03 & 2.1e+03 & nan, Conjunctiva, Orbital, noCornea, lacrimalgland, whole, Cornea-etc, Neuroretina-etc, Sclera-etc & epithelial, immune, stromal, endothelial & 10X, smartseq2 & TSP5, TSP3, TSP15 \\\\\n",
      "Mammary & 11375 & F & 8.7e+04 & 8.5e+03 & 6e+05 & 2.5e+03 & 2.3e+03 & 1.2e+03 & nan & stromal, immune, epithelial, endothelial & 10X, smartseq2 & TSP4 \\\\\n",
      "Prostate & 16375 & M & 9.1e+04 & 7.5e+03 & 5.4e+05 & 2.5e+03 & 2.1e+03 & 1.4e+03 & nan & epithelial, stromal, immune, endothelial, germ line & 10X, smartseq2 & TSP8, TSP14 \\\\\n",
      "Heart & 11505 & M & 4.7e+04 & 1.1e+04 & 2.7e+05 & 3.7e+03 & 3.2e+03 & 2.1e+03 & Atria, Ventricle, atria, ventricle & endothelial, stromal, epithelial, immune & 10X, smartseq2 & TSP12 \\\\\n",
      "Muscle & 30746 & 2.7 & 1.2e+05 & 9.3e+03 & 4.1e+05 & 2.4e+03 & 2.4e+03 & 9.2e+02 & diaphragm, rectusabdominus, Abdomen, Diaphragm, nan & endothelial, stromal, immune & smartseq2, 10X & TSP4, TSP14, TSP1, TSP2 \\\\\n",
      "\\bottomrule\n",
      "\\end{tabular}\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2359138/2969974440.py:11: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  df_results = df_results.applymap(format_number)\n"
     ]
    }
   ],
   "source": [
    "def format_number(x):\n",
    "    if isinstance(x, (float, np.float64)):\n",
    "        return f\"{x:.2g}\"\n",
    "    return x\n",
    "\n",
    "df_results = pd.DataFrame(data)\n",
    "df_results.columns = ['Dataset', 'Number of Cells', 'F/M', 'avg. UMI', 'med. UMI', 'std. UMI', 'avg. num. genes', 'med. num. genes', 'std. num. genes', 'anatomy', 'compartment', 'assay', 'donor']\n",
    "# df_results.columns = df_results.columns.str.replace('_', ' ')\n",
    "\n",
    "# Apply number formatting\n",
    "df_results = df_results.applymap(format_number)\n",
    "\n",
    "# Convert DataFrame to LaTeX\n",
    "latex_table = df_results.to_latex(index=False, escape=False, float_format=\"%.3f\")\n",
    "\n",
    "custom_table = latex_table.replace(\n",
    "    r'\\begin{tabular}{lrlllllllllll}',\n",
    "    r'\\begin{tabular}{l|rrrrrrrrrrrr}'\n",
    ").replace(\n",
    "    r'\\toprule',\n",
    "    r'\\hline'\n",
    ").replace(\n",
    "    r'\\midrule',\n",
    "    r'\\hline\\hline'\n",
    ").replace(\n",
    "    '_', ' '\n",
    ")\n",
    "\n",
    "# Print the LaTeX table\n",
    "print(custom_table)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "eda386ab-9c4d-43cd-882f-7d63635e826b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on method to_latex in module pandas.core.generic:\n",
      "\n",
      "to_latex(buf: 'FilePath | WriteBuffer[str] | None' = None, columns: 'Sequence[Hashable] | None' = None, header: 'bool_t | list[str]' = True, index: 'bool_t' = True, na_rep: 'str' = 'NaN', formatters: 'FormattersType | None' = None, float_format: 'FloatFormatType | None' = None, sparsify: 'bool_t | None' = None, index_names: 'bool_t' = True, bold_rows: 'bool_t' = False, column_format: 'str | None' = None, longtable: 'bool_t | None' = None, escape: 'bool_t | None' = None, encoding: 'str | None' = None, decimal: 'str' = '.', multicolumn: 'bool_t | None' = None, multicolumn_format: 'str | None' = None, multirow: 'bool_t | None' = None, caption: 'str | tuple[str, str] | None' = None, label: 'str | None' = None, position: 'str | None' = None) -> 'str | None' method of pandas.core.frame.DataFrame instance\n",
      "    Render object to a LaTeX tabular, longtable, or nested table.\n",
      "    \n",
      "    Requires ``\\usepackage{{booktabs}}``.  The output can be copy/pasted\n",
      "    into a main LaTeX document or read from an external file\n",
      "    with ``\\input{{table.tex}}``.\n",
      "    \n",
      "    .. versionchanged:: 1.2.0\n",
      "       Added position argument, changed meaning of caption argument.\n",
      "    \n",
      "    .. versionchanged:: 2.0.0\n",
      "       Refactored to use the Styler implementation via jinja2 templating.\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    buf : str, Path or StringIO-like, optional, default None\n",
      "        Buffer to write to. If None, the output is returned as a string.\n",
      "    columns : list of label, optional\n",
      "        The subset of columns to write. Writes all columns by default.\n",
      "    header : bool or list of str, default True\n",
      "        Write out the column names. If a list of strings is given,\n",
      "        it is assumed to be aliases for the column names.\n",
      "    index : bool, default True\n",
      "        Write row names (index).\n",
      "    na_rep : str, default 'NaN'\n",
      "        Missing data representation.\n",
      "    formatters : list of functions or dict of {{str: function}}, optional\n",
      "        Formatter functions to apply to columns' elements by position or\n",
      "        name. The result of each function must be a unicode string.\n",
      "        List must be of length equal to the number of columns.\n",
      "    float_format : one-parameter function or str, optional, default None\n",
      "        Formatter for floating point numbers. For example\n",
      "        ``float_format=\"%.2f\"`` and ``float_format=\"{{:0.2f}}\".format`` will\n",
      "        both result in 0.1234 being formatted as 0.12.\n",
      "    sparsify : bool, optional\n",
      "        Set to False for a DataFrame with a hierarchical index to print\n",
      "        every multiindex key at each row. By default, the value will be\n",
      "        read from the config module.\n",
      "    index_names : bool, default True\n",
      "        Prints the names of the indexes.\n",
      "    bold_rows : bool, default False\n",
      "        Make the row labels bold in the output.\n",
      "    column_format : str, optional\n",
      "        The columns format as specified in `LaTeX table format\n",
      "        <https://en.wikibooks.org/wiki/LaTeX/Tables>`__ e.g. 'rcl' for 3\n",
      "        columns. By default, 'l' will be used for all columns except\n",
      "        columns of numbers, which default to 'r'.\n",
      "    longtable : bool, optional\n",
      "        Use a longtable environment instead of tabular. Requires\n",
      "        adding a \\usepackage{{longtable}} to your LaTeX preamble.\n",
      "        By default, the value will be read from the pandas config\n",
      "        module, and set to `True` if the option ``styler.latex.environment`` is\n",
      "        `\"longtable\"`.\n",
      "    \n",
      "        .. versionchanged:: 2.0.0\n",
      "           The pandas option affecting this argument has changed.\n",
      "    escape : bool, optional\n",
      "        By default, the value will be read from the pandas config\n",
      "        module and set to `True` if the option ``styler.format.escape`` is\n",
      "        `\"latex\"`. When set to False prevents from escaping latex special\n",
      "        characters in column names.\n",
      "    \n",
      "        .. versionchanged:: 2.0.0\n",
      "           The pandas option affecting this argument has changed, as has the\n",
      "           default value to `False`.\n",
      "    encoding : str, optional\n",
      "        A string representing the encoding to use in the output file,\n",
      "        defaults to 'utf-8'.\n",
      "    decimal : str, default '.'\n",
      "        Character recognized as decimal separator, e.g. ',' in Europe.\n",
      "    multicolumn : bool, default True\n",
      "        Use \\multicolumn to enhance MultiIndex columns.\n",
      "        The default will be read from the config module, and is set\n",
      "        as the option ``styler.sparse.columns``.\n",
      "    \n",
      "        .. versionchanged:: 2.0.0\n",
      "           The pandas option affecting this argument has changed.\n",
      "    multicolumn_format : str, default 'r'\n",
      "        The alignment for multicolumns, similar to `column_format`\n",
      "        The default will be read from the config module, and is set as the option\n",
      "        ``styler.latex.multicol_align``.\n",
      "    \n",
      "        .. versionchanged:: 2.0.0\n",
      "           The pandas option affecting this argument has changed, as has the\n",
      "           default value to \"r\".\n",
      "    multirow : bool, default True\n",
      "        Use \\multirow to enhance MultiIndex rows. Requires adding a\n",
      "        \\usepackage{{multirow}} to your LaTeX preamble. Will print\n",
      "        centered labels (instead of top-aligned) across the contained\n",
      "        rows, separating groups via clines. The default will be read\n",
      "        from the pandas config module, and is set as the option\n",
      "        ``styler.sparse.index``.\n",
      "    \n",
      "        .. versionchanged:: 2.0.0\n",
      "           The pandas option affecting this argument has changed, as has the\n",
      "           default value to `True`.\n",
      "    caption : str or tuple, optional\n",
      "        Tuple (full_caption, short_caption),\n",
      "        which results in ``\\caption[short_caption]{{full_caption}}``;\n",
      "        if a single string is passed, no short caption will be set.\n",
      "    \n",
      "        .. versionchanged:: 1.2.0\n",
      "           Optionally allow caption to be a tuple ``(full_caption, short_caption)``.\n",
      "    \n",
      "    label : str, optional\n",
      "        The LaTeX label to be placed inside ``\\label{{}}`` in the output.\n",
      "        This is used with ``\\ref{{}}`` in the main ``.tex`` file.\n",
      "    \n",
      "    position : str, optional\n",
      "        The LaTeX positional argument for tables, to be placed after\n",
      "        ``\\begin{{}}`` in the output.\n",
      "    \n",
      "        .. versionadded:: 1.2.0\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    str or None\n",
      "        If buf is None, returns the result as a string. Otherwise returns None.\n",
      "    \n",
      "    See Also\n",
      "    --------\n",
      "    io.formats.style.Styler.to_latex : Render a DataFrame to LaTeX\n",
      "        with conditional formatting.\n",
      "    DataFrame.to_string : Render a DataFrame to a console-friendly\n",
      "        tabular output.\n",
      "    DataFrame.to_html : Render a DataFrame as an HTML table.\n",
      "    \n",
      "    Notes\n",
      "    -----\n",
      "    As of v2.0.0 this method has changed to use the Styler implementation as\n",
      "    part of :meth:`.Styler.to_latex` via ``jinja2`` templating. This means\n",
      "    that ``jinja2`` is a requirement, and needs to be installed, for this method\n",
      "    to function. It is advised that users switch to using Styler, since that\n",
      "    implementation is more frequently updated and contains much more\n",
      "    flexibility with the output.\n",
      "    \n",
      "    Examples\n",
      "    --------\n",
      "    Convert a general DataFrame to LaTeX with formatting:\n",
      "    \n",
      "    >>> df = pd.DataFrame(dict(name=['Raphael', 'Donatello'],\n",
      "    ...                        age=[26, 45],\n",
      "    ...                        height=[181.23, 177.65]))\n",
      "    >>> print(df.to_latex(index=False,\n",
      "    ...                   formatters={\"name\": str.upper},\n",
      "    ...                   float_format=\"{:.1f}\".format,\n",
      "    ... ))  # doctest: +SKIP\n",
      "    \\begin{tabular}{lrr}\n",
      "    \\toprule\n",
      "    name & age & height \\\\\n",
      "    \\midrule\n",
      "    RAPHAEL & 26 & 181.2 \\\\\n",
      "    DONATELLO & 45 & 177.7 \\\\\n",
      "    \\bottomrule\n",
      "    \\end{tabular}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(df_results.to_latex)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "4f2e93af-6b84-45df-87af-718bd6e09be2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['male']"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(gender_distribution.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "d7a616e0-3347-4252-a271-2f0298cc90de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organ_tissue</th>\n",
       "      <th>method</th>\n",
       "      <th>donor</th>\n",
       "      <th>anatomical_information</th>\n",
       "      <th>n_counts_UMIs</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>cell_ontology_class</th>\n",
       "      <th>free_annotation</th>\n",
       "      <th>manually_annotated</th>\n",
       "      <th>compartment</th>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACTATTCAGGTCATCT_TSP14_Prostate_NA_10X_1_1</th>\n",
       "      <td>Prostate</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP14</td>\n",
       "      <td>nan</td>\n",
       "      <td>5104.0</td>\n",
       "      <td>1220</td>\n",
       "      <td>sperm</td>\n",
       "      <td>Sperm Cell</td>\n",
       "      <td>True</td>\n",
       "      <td>germ line</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGGTTGTCAACCCTAA_TSP14_Prostate_NA_10X_1_1</th>\n",
       "      <td>Prostate</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP14</td>\n",
       "      <td>nan</td>\n",
       "      <td>8547.0</td>\n",
       "      <td>2775</td>\n",
       "      <td>sperm</td>\n",
       "      <td>Sperm Cell</td>\n",
       "      <td>True</td>\n",
       "      <td>germ line</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATCCGTCGTGCTATTG_TSP14_Prostate_NA_10X_1_1</th>\n",
       "      <td>Prostate</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP14</td>\n",
       "      <td>nan</td>\n",
       "      <td>10919.0</td>\n",
       "      <td>2578</td>\n",
       "      <td>sperm</td>\n",
       "      <td>Sperm Cell</td>\n",
       "      <td>True</td>\n",
       "      <td>germ line</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAGCCTGAGCATGCAG_TSP14_Prostate_NA_10X_1_1</th>\n",
       "      <td>Prostate</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP14</td>\n",
       "      <td>nan</td>\n",
       "      <td>25324.0</td>\n",
       "      <td>4061</td>\n",
       "      <td>sperm</td>\n",
       "      <td>Sperm Cell</td>\n",
       "      <td>True</td>\n",
       "      <td>germ line</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGGACTCCAATGAGCG_TSP14_Prostate_NA_10X_1_1</th>\n",
       "      <td>Prostate</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP14</td>\n",
       "      <td>nan</td>\n",
       "      <td>19955.0</td>\n",
       "      <td>4062</td>\n",
       "      <td>sperm</td>\n",
       "      <td>Sperm Cell</td>\n",
       "      <td>True</td>\n",
       "      <td>germ line</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGACCCTTCATCGCCT_TSP14_Prostate_NA_10X_1_1</th>\n",
       "      <td>Prostate</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP14</td>\n",
       "      <td>nan</td>\n",
       "      <td>12453.0</td>\n",
       "      <td>2837</td>\n",
       "      <td>sperm</td>\n",
       "      <td>Sperm Cell</td>\n",
       "      <td>True</td>\n",
       "      <td>germ line</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACACACAGCGAACTG_TSP14_Prostate_NA_10X_1_2</th>\n",
       "      <td>Prostate</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP14</td>\n",
       "      <td>nan</td>\n",
       "      <td>7446.0</td>\n",
       "      <td>1819</td>\n",
       "      <td>sperm</td>\n",
       "      <td>Sperm Cell</td>\n",
       "      <td>True</td>\n",
       "      <td>germ line</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAGGAATGTATTCTCT_TSP14_Prostate_NA_10X_1_2</th>\n",
       "      <td>Prostate</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP14</td>\n",
       "      <td>nan</td>\n",
       "      <td>19495.0</td>\n",
       "      <td>3442</td>\n",
       "      <td>sperm</td>\n",
       "      <td>Sperm Cell</td>\n",
       "      <td>True</td>\n",
       "      <td>germ line</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTATCTCATCCGAAT_TSP14_Prostate_NA_10X_1_2</th>\n",
       "      <td>Prostate</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP14</td>\n",
       "      <td>nan</td>\n",
       "      <td>11515.0</td>\n",
       "      <td>2667</td>\n",
       "      <td>sperm</td>\n",
       "      <td>Sperm Cell</td>\n",
       "      <td>True</td>\n",
       "      <td>germ line</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGGGTCAAGTATTAGG_TSP14_Prostate_NA_10X_1_2</th>\n",
       "      <td>Prostate</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP14</td>\n",
       "      <td>nan</td>\n",
       "      <td>6001.0</td>\n",
       "      <td>1424</td>\n",
       "      <td>sperm</td>\n",
       "      <td>Sperm Cell</td>\n",
       "      <td>True</td>\n",
       "      <td>germ line</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGGGCAAGGTAGACC_TSP14_Prostate_NA_10X_1_2</th>\n",
       "      <td>Prostate</td>\n",
       "      <td>10X</td>\n",
       "      <td>TSP14</td>\n",
       "      <td>nan</td>\n",
       "      <td>24753.0</td>\n",
       "      <td>4315</td>\n",
       "      <td>sperm</td>\n",
       "      <td>Sperm Cell</td>\n",
       "      <td>True</td>\n",
       "      <td>germ line</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           organ_tissue method  donor  \\\n",
       "cell_id                                                                 \n",
       "ACTATTCAGGTCATCT_TSP14_Prostate_NA_10X_1_1     Prostate    10X  TSP14   \n",
       "AGGTTGTCAACCCTAA_TSP14_Prostate_NA_10X_1_1     Prostate    10X  TSP14   \n",
       "ATCCGTCGTGCTATTG_TSP14_Prostate_NA_10X_1_1     Prostate    10X  TSP14   \n",
       "GAGCCTGAGCATGCAG_TSP14_Prostate_NA_10X_1_1     Prostate    10X  TSP14   \n",
       "GGGACTCCAATGAGCG_TSP14_Prostate_NA_10X_1_1     Prostate    10X  TSP14   \n",
       "TGACCCTTCATCGCCT_TSP14_Prostate_NA_10X_1_1     Prostate    10X  TSP14   \n",
       "AACACACAGCGAACTG_TSP14_Prostate_NA_10X_1_2     Prostate    10X  TSP14   \n",
       "AAGGAATGTATTCTCT_TSP14_Prostate_NA_10X_1_2     Prostate    10X  TSP14   \n",
       "ACTATCTCATCCGAAT_TSP14_Prostate_NA_10X_1_2     Prostate    10X  TSP14   \n",
       "CGGGTCAAGTATTAGG_TSP14_Prostate_NA_10X_1_2     Prostate    10X  TSP14   \n",
       "TCGGGCAAGGTAGACC_TSP14_Prostate_NA_10X_1_2     Prostate    10X  TSP14   \n",
       "\n",
       "                                           anatomical_information  \\\n",
       "cell_id                                                             \n",
       "ACTATTCAGGTCATCT_TSP14_Prostate_NA_10X_1_1                    nan   \n",
       "AGGTTGTCAACCCTAA_TSP14_Prostate_NA_10X_1_1                    nan   \n",
       "ATCCGTCGTGCTATTG_TSP14_Prostate_NA_10X_1_1                    nan   \n",
       "GAGCCTGAGCATGCAG_TSP14_Prostate_NA_10X_1_1                    nan   \n",
       "GGGACTCCAATGAGCG_TSP14_Prostate_NA_10X_1_1                    nan   \n",
       "TGACCCTTCATCGCCT_TSP14_Prostate_NA_10X_1_1                    nan   \n",
       "AACACACAGCGAACTG_TSP14_Prostate_NA_10X_1_2                    nan   \n",
       "AAGGAATGTATTCTCT_TSP14_Prostate_NA_10X_1_2                    nan   \n",
       "ACTATCTCATCCGAAT_TSP14_Prostate_NA_10X_1_2                    nan   \n",
       "CGGGTCAAGTATTAGG_TSP14_Prostate_NA_10X_1_2                    nan   \n",
       "TCGGGCAAGGTAGACC_TSP14_Prostate_NA_10X_1_2                    nan   \n",
       "\n",
       "                                            n_counts_UMIs  n_genes  \\\n",
       "cell_id                                                              \n",
       "ACTATTCAGGTCATCT_TSP14_Prostate_NA_10X_1_1         5104.0     1220   \n",
       "AGGTTGTCAACCCTAA_TSP14_Prostate_NA_10X_1_1         8547.0     2775   \n",
       "ATCCGTCGTGCTATTG_TSP14_Prostate_NA_10X_1_1        10919.0     2578   \n",
       "GAGCCTGAGCATGCAG_TSP14_Prostate_NA_10X_1_1        25324.0     4061   \n",
       "GGGACTCCAATGAGCG_TSP14_Prostate_NA_10X_1_1        19955.0     4062   \n",
       "TGACCCTTCATCGCCT_TSP14_Prostate_NA_10X_1_1        12453.0     2837   \n",
       "AACACACAGCGAACTG_TSP14_Prostate_NA_10X_1_2         7446.0     1819   \n",
       "AAGGAATGTATTCTCT_TSP14_Prostate_NA_10X_1_2        19495.0     3442   \n",
       "ACTATCTCATCCGAAT_TSP14_Prostate_NA_10X_1_2        11515.0     2667   \n",
       "CGGGTCAAGTATTAGG_TSP14_Prostate_NA_10X_1_2         6001.0     1424   \n",
       "TCGGGCAAGGTAGACC_TSP14_Prostate_NA_10X_1_2        24753.0     4315   \n",
       "\n",
       "                                           cell_ontology_class  \\\n",
       "cell_id                                                          \n",
       "ACTATTCAGGTCATCT_TSP14_Prostate_NA_10X_1_1               sperm   \n",
       "AGGTTGTCAACCCTAA_TSP14_Prostate_NA_10X_1_1               sperm   \n",
       "ATCCGTCGTGCTATTG_TSP14_Prostate_NA_10X_1_1               sperm   \n",
       "GAGCCTGAGCATGCAG_TSP14_Prostate_NA_10X_1_1               sperm   \n",
       "GGGACTCCAATGAGCG_TSP14_Prostate_NA_10X_1_1               sperm   \n",
       "TGACCCTTCATCGCCT_TSP14_Prostate_NA_10X_1_1               sperm   \n",
       "AACACACAGCGAACTG_TSP14_Prostate_NA_10X_1_2               sperm   \n",
       "AAGGAATGTATTCTCT_TSP14_Prostate_NA_10X_1_2               sperm   \n",
       "ACTATCTCATCCGAAT_TSP14_Prostate_NA_10X_1_2               sperm   \n",
       "CGGGTCAAGTATTAGG_TSP14_Prostate_NA_10X_1_2               sperm   \n",
       "TCGGGCAAGGTAGACC_TSP14_Prostate_NA_10X_1_2               sperm   \n",
       "\n",
       "                                           free_annotation  \\\n",
       "cell_id                                                      \n",
       "ACTATTCAGGTCATCT_TSP14_Prostate_NA_10X_1_1      Sperm Cell   \n",
       "AGGTTGTCAACCCTAA_TSP14_Prostate_NA_10X_1_1      Sperm Cell   \n",
       "ATCCGTCGTGCTATTG_TSP14_Prostate_NA_10X_1_1      Sperm Cell   \n",
       "GAGCCTGAGCATGCAG_TSP14_Prostate_NA_10X_1_1      Sperm Cell   \n",
       "GGGACTCCAATGAGCG_TSP14_Prostate_NA_10X_1_1      Sperm Cell   \n",
       "TGACCCTTCATCGCCT_TSP14_Prostate_NA_10X_1_1      Sperm Cell   \n",
       "AACACACAGCGAACTG_TSP14_Prostate_NA_10X_1_2      Sperm Cell   \n",
       "AAGGAATGTATTCTCT_TSP14_Prostate_NA_10X_1_2      Sperm Cell   \n",
       "ACTATCTCATCCGAAT_TSP14_Prostate_NA_10X_1_2      Sperm Cell   \n",
       "CGGGTCAAGTATTAGG_TSP14_Prostate_NA_10X_1_2      Sperm Cell   \n",
       "TCGGGCAAGGTAGACC_TSP14_Prostate_NA_10X_1_2      Sperm Cell   \n",
       "\n",
       "                                            manually_annotated compartment  \\\n",
       "cell_id                                                                      \n",
       "ACTATTCAGGTCATCT_TSP14_Prostate_NA_10X_1_1                True   germ line   \n",
       "AGGTTGTCAACCCTAA_TSP14_Prostate_NA_10X_1_1                True   germ line   \n",
       "ATCCGTCGTGCTATTG_TSP14_Prostate_NA_10X_1_1                True   germ line   \n",
       "GAGCCTGAGCATGCAG_TSP14_Prostate_NA_10X_1_1                True   germ line   \n",
       "GGGACTCCAATGAGCG_TSP14_Prostate_NA_10X_1_1                True   germ line   \n",
       "TGACCCTTCATCGCCT_TSP14_Prostate_NA_10X_1_1                True   germ line   \n",
       "AACACACAGCGAACTG_TSP14_Prostate_NA_10X_1_2                True   germ line   \n",
       "AAGGAATGTATTCTCT_TSP14_Prostate_NA_10X_1_2                True   germ line   \n",
       "ACTATCTCATCCGAAT_TSP14_Prostate_NA_10X_1_2                True   germ line   \n",
       "CGGGTCAAGTATTAGG_TSP14_Prostate_NA_10X_1_2                True   germ line   \n",
       "TCGGGCAAGGTAGACC_TSP14_Prostate_NA_10X_1_2                True   germ line   \n",
       "\n",
       "                                           gender  \n",
       "cell_id                                            \n",
       "ACTATTCAGGTCATCT_TSP14_Prostate_NA_10X_1_1   male  \n",
       "AGGTTGTCAACCCTAA_TSP14_Prostate_NA_10X_1_1   male  \n",
       "ATCCGTCGTGCTATTG_TSP14_Prostate_NA_10X_1_1   male  \n",
       "GAGCCTGAGCATGCAG_TSP14_Prostate_NA_10X_1_1   male  \n",
       "GGGACTCCAATGAGCG_TSP14_Prostate_NA_10X_1_1   male  \n",
       "TGACCCTTCATCGCCT_TSP14_Prostate_NA_10X_1_1   male  \n",
       "AACACACAGCGAACTG_TSP14_Prostate_NA_10X_1_2   male  \n",
       "AAGGAATGTATTCTCT_TSP14_Prostate_NA_10X_1_2   male  \n",
       "ACTATCTCATCCGAAT_TSP14_Prostate_NA_10X_1_2   male  \n",
       "CGGGTCAAGTATTAGG_TSP14_Prostate_NA_10X_1_2   male  \n",
       "TCGGGCAAGGTAGACC_TSP14_Prostate_NA_10X_1_2   male  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "152281bf-2b74-44d4-b613-c98ffff6cb04",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Number of Cells</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>germ line</th>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stromal</th>\n",
       "      <td>82478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Large_Intestine</th>\n",
       "      <td>13680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lymph_Node</th>\n",
       "      <td>53275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Skin</th>\n",
       "      <td>9424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>epithelial</th>\n",
       "      <td>104148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Spleen</th>\n",
       "      <td>34004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Vasculature</th>\n",
       "      <td>16037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Kidney</th>\n",
       "      <td>9641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Salivary_Gland</th>\n",
       "      <td>27199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Thymus</th>\n",
       "      <td>33664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pancreas</th>\n",
       "      <td>13497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Fat</th>\n",
       "      <td>20263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bladder</th>\n",
       "      <td>24583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>endothelial</th>\n",
       "      <td>31691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Blood</th>\n",
       "      <td>50115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Trachea</th>\n",
       "      <td>9522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bone_Marrow</th>\n",
       "      <td>12297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tongue</th>\n",
       "      <td>15020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lung</th>\n",
       "      <td>35682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Liver</th>\n",
       "      <td>5007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Small_Intestine</th>\n",
       "      <td>12467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>immune</th>\n",
       "      <td>264824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uterus</th>\n",
       "      <td>7124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Eye</th>\n",
       "      <td>10650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mammary</th>\n",
       "      <td>11375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prostate</th>\n",
       "      <td>16375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Heart</th>\n",
       "      <td>11505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Muscle</th>\n",
       "      <td>30746</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Number of Cells\n",
       "germ line                     11\n",
       "stromal                    82478\n",
       "Large_Intestine            13680\n",
       "Lymph_Node                 53275\n",
       "Skin                        9424\n",
       "epithelial                104148\n",
       "Spleen                     34004\n",
       "Vasculature                16037\n",
       "Kidney                      9641\n",
       "Salivary_Gland             27199\n",
       "Thymus                     33664\n",
       "Pancreas                   13497\n",
       "Fat                        20263\n",
       "Bladder                    24583\n",
       "endothelial                31691\n",
       "Blood                      50115\n",
       "Trachea                     9522\n",
       "Bone_Marrow                12297\n",
       "Tongue                     15020\n",
       "Lung                       35682\n",
       "Liver                       5007\n",
       "Small_Intestine            12467\n",
       "immune                    264824\n",
       "Uterus                      7124\n",
       "Eye                        10650\n",
       "Mammary                    11375\n",
       "Prostate                   16375\n",
       "Heart                      11505\n",
       "Muscle                     30746"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfData['Number of Cells'] = numCells\n",
    "dfData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "da9d6273-a312-4d54-8c5a-f5c65c199561",
   "metadata": {},
   "outputs": [],
   "source": [
    "# female to mail ratio\n",
    "gender_distribution = adata.obs['gender'].value_counts()\n",
    "fmr = gender_distribution.female / gender_distribution.male\n",
    "\n",
    "# unique molecular identifiers\n",
    "umiMean = adata.obs['n_counts_UMIs'].mean()\n",
    "umiMedian = adata.obs['n_counts_UMIs'].median()\n",
    "umiStd = adata.obs['n_counts_UMIs'].std()\n",
    "\n",
    "# number of genes\n",
    "nGenesMean = adata.obs['n_genes'].mean()\n",
    "nGenesMedian = adata.obs['n_genes'].median()\n",
    "nGenesStd = adata.obs['n_genes'].std()\n",
    "\n",
    "# anatomical information\n",
    "anatomy = adata.obs['anatomical_information'].unique()\n",
    "\n",
    "# compartment\n",
    "compartment = adata.obs['compartment'].unique()\n",
    "\n",
    "# methods\n",
    "methods = adata.obs['method'].unique()\n",
    "\n",
    "# dataset\n",
    "ds = adata.obs['donor'].unique()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "98c7d4aa-6abd-4d3e-8014-cdd146551fda",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['endothelial', 'stromal', 'immune']\n",
       "Categories (3, object): ['endothelial', 'immune', 'stromal']"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['compartment'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "8ccf224d-ae98-419c-a651-103f9592770e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organ_tissue</th>\n",
       "      <th>method</th>\n",
       "      <th>donor</th>\n",
       "      <th>anatomical_information</th>\n",
       "      <th>n_counts_UMIs</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>cell_ontology_class</th>\n",
       "      <th>free_annotation</th>\n",
       "      <th>manually_annotated</th>\n",
       "      <th>compartment</th>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSP4_Muscle_diaphragm_SS2_B112813_B134043_Empty_J24_L003</th>\n",
       "      <td>Muscle</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP4</td>\n",
       "      <td>diaphragm</td>\n",
       "      <td>82831.0</td>\n",
       "      <td>2253</td>\n",
       "      <td>endothelial cell of vascular tree</td>\n",
       "      <td>endothelial cell of vascular tree</td>\n",
       "      <td>True</td>\n",
       "      <td>endothelial</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endothelial_B12_L003</th>\n",
       "      <td>Muscle</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP4</td>\n",
       "      <td>diaphragm</td>\n",
       "      <td>3531175.0</td>\n",
       "      <td>3896</td>\n",
       "      <td>endothelial cell of vascular tree</td>\n",
       "      <td>endothelial cell of vascular tree</td>\n",
       "      <td>True</td>\n",
       "      <td>endothelial</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endothelial_B13_L003</th>\n",
       "      <td>Muscle</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP4</td>\n",
       "      <td>diaphragm</td>\n",
       "      <td>1332680.0</td>\n",
       "      <td>1487</td>\n",
       "      <td>capillary endothelial cell</td>\n",
       "      <td>capillary endothelial cell</td>\n",
       "      <td>True</td>\n",
       "      <td>endothelial</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endothelial_B14_L003</th>\n",
       "      <td>Muscle</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP4</td>\n",
       "      <td>diaphragm</td>\n",
       "      <td>37120.0</td>\n",
       "      <td>502</td>\n",
       "      <td>endothelial cell of lymphatic vessel</td>\n",
       "      <td>endothelial cell of lymphatic vessel</td>\n",
       "      <td>True</td>\n",
       "      <td>endothelial</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endothelial_B15_L003</th>\n",
       "      <td>Muscle</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP4</td>\n",
       "      <td>diaphragm</td>\n",
       "      <td>2391365.0</td>\n",
       "      <td>3262</td>\n",
       "      <td>endothelial cell of lymphatic vessel</td>\n",
       "      <td>endothelial cell of lymphatic vessel</td>\n",
       "      <td>True</td>\n",
       "      <td>endothelial</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_Muscle_rectusabdominus_SS2_B114661_B133087_MuscleStemCell_O6_S342</th>\n",
       "      <td>Muscle</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>rectusabdominus</td>\n",
       "      <td>410010.0</td>\n",
       "      <td>1669</td>\n",
       "      <td>mesenchymal stem cell</td>\n",
       "      <td>mesenchymal stem cell</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_Muscle_rectusabdominus_SS2_B114661_B133087_MuscleStemCell_P18_S378</th>\n",
       "      <td>Muscle</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>rectusabdominus</td>\n",
       "      <td>1547776.0</td>\n",
       "      <td>2212</td>\n",
       "      <td>mesenchymal stem cell</td>\n",
       "      <td>mesenchymal stem cell</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_Muscle_rectusabdominus_SS2_B114661_B133087_MuscleStemCell_P21_S381</th>\n",
       "      <td>Muscle</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>rectusabdominus</td>\n",
       "      <td>99507.0</td>\n",
       "      <td>899</td>\n",
       "      <td>skeletal muscle satellite stem cell</td>\n",
       "      <td>skeletal muscle satellite stem cell</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_Muscle_rectusabdominus_SS2_B114661_B133087_MuscleStemCell_P7_S367</th>\n",
       "      <td>Muscle</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>rectusabdominus</td>\n",
       "      <td>200210.0</td>\n",
       "      <td>726</td>\n",
       "      <td>skeletal muscle satellite stem cell</td>\n",
       "      <td>skeletal muscle satellite stem cell</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSP2_Muscle_rectusabdominus_SS2_B114661_B133087_MuscleStemCell_P8_S368</th>\n",
       "      <td>Muscle</td>\n",
       "      <td>smartseq2</td>\n",
       "      <td>TSP2</td>\n",
       "      <td>rectusabdominus</td>\n",
       "      <td>57224.0</td>\n",
       "      <td>390</td>\n",
       "      <td>skeletal muscle satellite stem cell</td>\n",
       "      <td>skeletal muscle satellite stem cell</td>\n",
       "      <td>True</td>\n",
       "      <td>stromal</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30746 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   organ_tissue     method  \\\n",
       "cell_id                                                                      \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Empty...       Muscle  smartseq2   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...       Muscle  smartseq2   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...       Muscle  smartseq2   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...       Muscle  smartseq2   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...       Muscle  smartseq2   \n",
       "...                                                         ...        ...   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...       Muscle  smartseq2   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...       Muscle  smartseq2   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...       Muscle  smartseq2   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...       Muscle  smartseq2   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...       Muscle  smartseq2   \n",
       "\n",
       "                                                   donor  \\\n",
       "cell_id                                                    \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Empty...  TSP4   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...  TSP4   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...  TSP4   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...  TSP4   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...  TSP4   \n",
       "...                                                  ...   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...  TSP2   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...  TSP2   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...  TSP2   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...  TSP2   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...  TSP2   \n",
       "\n",
       "                                                   anatomical_information  \\\n",
       "cell_id                                                                     \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Empty...              diaphragm   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...              diaphragm   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...              diaphragm   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...              diaphragm   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...              diaphragm   \n",
       "...                                                                   ...   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...        rectusabdominus   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...        rectusabdominus   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...        rectusabdominus   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...        rectusabdominus   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...        rectusabdominus   \n",
       "\n",
       "                                                    n_counts_UMIs  n_genes  \\\n",
       "cell_id                                                                      \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Empty...        82831.0     2253   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...      3531175.0     3896   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...      1332680.0     1487   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...        37120.0      502   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...      2391365.0     3262   \n",
       "...                                                           ...      ...   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...       410010.0     1669   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...      1547776.0     2212   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...        99507.0      899   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...       200210.0      726   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...        57224.0      390   \n",
       "\n",
       "                                                                     cell_ontology_class  \\\n",
       "cell_id                                                                                    \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Empty...     endothelial cell of vascular tree   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...     endothelial cell of vascular tree   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...            capillary endothelial cell   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...  endothelial cell of lymphatic vessel   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...  endothelial cell of lymphatic vessel   \n",
       "...                                                                                  ...   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...                 mesenchymal stem cell   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...                 mesenchymal stem cell   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...   skeletal muscle satellite stem cell   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...   skeletal muscle satellite stem cell   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...   skeletal muscle satellite stem cell   \n",
       "\n",
       "                                                                         free_annotation  \\\n",
       "cell_id                                                                                    \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Empty...     endothelial cell of vascular tree   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...     endothelial cell of vascular tree   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...            capillary endothelial cell   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...  endothelial cell of lymphatic vessel   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...  endothelial cell of lymphatic vessel   \n",
       "...                                                                                  ...   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...                 mesenchymal stem cell   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...                 mesenchymal stem cell   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...   skeletal muscle satellite stem cell   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...   skeletal muscle satellite stem cell   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...   skeletal muscle satellite stem cell   \n",
       "\n",
       "                                                    manually_annotated  \\\n",
       "cell_id                                                                  \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Empty...                True   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...                True   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...                True   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...                True   \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...                True   \n",
       "...                                                                ...   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...                True   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...                True   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...                True   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...                True   \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...                True   \n",
       "\n",
       "                                                    compartment  gender  \n",
       "cell_id                                                                  \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Empty...  endothelial  female  \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...  endothelial  female  \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...  endothelial  female  \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...  endothelial  female  \n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endot...  endothelial  female  \n",
       "...                                                         ...     ...  \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...      stromal  female  \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...      stromal  female  \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...      stromal  female  \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...      stromal  female  \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087...      stromal  female  \n",
       "\n",
       "[30746 rows x 11 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "0a17a4b7-0e30-4304-98d6-e4d18d8c03f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>ensemblid</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>False</td>\n",
       "      <td>6.398244e-05</td>\n",
       "      <td>0.835044</td>\n",
       "      <td>-0.573947</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.005574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>False</td>\n",
       "      <td>2.274395e-03</td>\n",
       "      <td>2.442280</td>\n",
       "      <td>0.533203</td>\n",
       "      <td>0.001080</td>\n",
       "      <td>0.031731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR6859-1</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>False</td>\n",
       "      <td>6.175251e-05</td>\n",
       "      <td>1.295335</td>\n",
       "      <td>-0.256874</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.005634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>False</td>\n",
       "      <td>1.372886e-04</td>\n",
       "      <td>2.656352</td>\n",
       "      <td>0.680668</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>0.008041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND6</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>False</td>\n",
       "      <td>9.634841e-01</td>\n",
       "      <td>2.466404</td>\n",
       "      <td>0.154140</td>\n",
       "      <td>0.590065</td>\n",
       "      <td>0.741395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TE</th>\n",
       "      <td>MT-TE</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>False</td>\n",
       "      <td>1.600667e-01</td>\n",
       "      <td>1.603787</td>\n",
       "      <td>-0.044396</td>\n",
       "      <td>0.083929</td>\n",
       "      <td>0.301820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CYB</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>False</td>\n",
       "      <td>4.367693e+00</td>\n",
       "      <td>4.765751</td>\n",
       "      <td>-0.499747</td>\n",
       "      <td>3.874830</td>\n",
       "      <td>1.104192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TT</th>\n",
       "      <td>MT-TT</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>False</td>\n",
       "      <td>6.573967e-02</td>\n",
       "      <td>0.624316</td>\n",
       "      <td>-0.719108</td>\n",
       "      <td>0.040580</td>\n",
       "      <td>0.186848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-TP</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>True</td>\n",
       "      <td>4.385102e-01</td>\n",
       "      <td>2.705558</td>\n",
       "      <td>1.115482</td>\n",
       "      <td>0.204806</td>\n",
       "      <td>0.501265</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58870 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_symbol     feature_type          ensemblid  highly_variable  \\\n",
       "DDX11L1          DDX11L1  Gene Expression  ENSG00000223972.5            False   \n",
       "WASH7P            WASH7P  Gene Expression  ENSG00000227232.5            False   \n",
       "MIR6859-1      MIR6859-1  Gene Expression  ENSG00000278267.1            False   \n",
       "MIR1302-2HG  MIR1302-2HG  Gene Expression  ENSG00000243485.5            False   \n",
       "MIR1302-2      MIR1302-2  Gene Expression  ENSG00000284332.1            False   \n",
       "...                  ...              ...                ...              ...   \n",
       "MT-ND6            MT-ND6  Gene Expression  ENSG00000198695.2            False   \n",
       "MT-TE              MT-TE  Gene Expression  ENSG00000210194.1            False   \n",
       "MT-CYB            MT-CYB  Gene Expression  ENSG00000198727.2            False   \n",
       "MT-TT              MT-TT  Gene Expression  ENSG00000210195.2            False   \n",
       "MT-TP              MT-TP  Gene Expression  ENSG00000210196.2             True   \n",
       "\n",
       "                    means  dispersions  dispersions_norm      mean       std  \n",
       "DDX11L1      6.398244e-05     0.835044         -0.573947  0.000039  0.005574  \n",
       "WASH7P       2.274395e-03     2.442280          0.533203  0.001080  0.031731  \n",
       "MIR6859-1    6.175251e-05     1.295335         -0.256874  0.000033  0.005634  \n",
       "MIR1302-2HG  1.372886e-04     2.656352          0.680668  0.000048  0.008041  \n",
       "MIR1302-2    1.000000e-12          NaN          0.000000  0.000000  1.000000  \n",
       "...                   ...          ...               ...       ...       ...  \n",
       "MT-ND6       9.634841e-01     2.466404          0.154140  0.590065  0.741395  \n",
       "MT-TE        1.600667e-01     1.603787         -0.044396  0.083929  0.301820  \n",
       "MT-CYB       4.367693e+00     4.765751         -0.499747  3.874830  1.104192  \n",
       "MT-TT        6.573967e-02     0.624316         -0.719108  0.040580  0.186848  \n",
       "MT-TP        4.385102e-01     2.705558          1.115482  0.204806  0.501265  \n",
       "\n",
       "[58870 rows x 9 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "5418efae-19b7-41e0-aef3-43b720c307c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cell_id\n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Empty_J24_L003                   female\n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endothelial_B12_L003             female\n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endothelial_B13_L003             female\n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endothelial_B14_L003             female\n",
       "TSP4_Muscle_diaphragm_SS2_B112813_B134043_Endothelial_B15_L003             female\n",
       "                                                                            ...  \n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087_MuscleStemCell_O6_S342     female\n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087_MuscleStemCell_P18_S378    female\n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087_MuscleStemCell_P21_S381    female\n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087_MuscleStemCell_P7_S367     female\n",
       "TSP2_Muscle_rectusabdominus_SS2_B114661_B133087_MuscleStemCell_P8_S368     female\n",
       "Name: gender, Length: 30746, dtype: category\n",
       "Categories (2, object): ['female', 'male']"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['gender']  # ['compartment']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "926af20c-37cb-4841-8e43-cfa1f5e9fbcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfRep['Reference'] = ['','','','']\n",
    "dfRep['Dimension'] = ['', 512, '', '']\n",
    "dfRep['Input']     = ['NA', 'Rank Order', 'Rank Order', 'Real Value']\n",
    "dfRep"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
