{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e9a68e81-1d0e-4531-921d-6c857e165390",
   "metadata": {},
   "source": [
    "# Geneformer Embeddings for Cooper\n",
    "\n",
    "```\n",
    "Author: Joshua Pickard\n",
    "        jpic@umich.edu\n",
    "Date: September 12, 2024\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01eae1f0-0a3c-4054-9ef4-b866b071bf54",
   "metadata": {},
   "source": [
    "# Set up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d7f00801-562d-4195-bdc8-a1a5dcc1d8de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "general imports\n",
      "bioinf imports\n",
      "scgpt wont work\n",
      "tGPT will work\n",
      "geneformer will work\n",
      "general imports\n",
      "bioinf imports\n",
      "scgpt wont work\n",
      "tGPT will work\n",
      "geneformer will work\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "\n",
    "import importlib\n",
    "\n",
    "sys.path.append('../anndata2embedding')\n",
    "import embed\n",
    "importlib.reload(embed)\n",
    "from embed import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19eefc8c-c867-47ab-98d8-8762b521fb4f",
   "metadata": {},
   "source": [
    "# Load & Chekcout Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3cc825d2-055d-48b1-9a82-b34113f77d2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "inputdata = [\n",
    "    '/scratch/indikar_root/indikar1/shared_data/single_cell_fibroblast/scanpy/raw.anndata.h5ad',\n",
    "    '/scratch/indikar_root/indikar1/shared_data/HSC/hsc_landscape.anndata.h5ad'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "803f1802-f5e3-4b3a-b747-598042acaa1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000160072</th>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>1</td>\n",
       "      <td>1471764</td>\n",
       "      <td>1497848</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142611</th>\n",
       "      <td>PRDM16</td>\n",
       "      <td>1</td>\n",
       "      <td>3069167</td>\n",
       "      <td>3438621</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000157933</th>\n",
       "      <td>SKI</td>\n",
       "      <td>1</td>\n",
       "      <td>2228318</td>\n",
       "      <td>2310213</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142655</th>\n",
       "      <td>PEX14</td>\n",
       "      <td>1</td>\n",
       "      <td>10472287</td>\n",
       "      <td>10630758</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000149527</th>\n",
       "      <td>PLCH2</td>\n",
       "      <td>1</td>\n",
       "      <td>2425979</td>\n",
       "      <td>2505532</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000187191</th>\n",
       "      <td>DAZ3</td>\n",
       "      <td>Y</td>\n",
       "      <td>24763068</td>\n",
       "      <td>24813492</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000188120</th>\n",
       "      <td>DAZ1</td>\n",
       "      <td>Y</td>\n",
       "      <td>23129354</td>\n",
       "      <td>23199010</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000099721</th>\n",
       "      <td>AMELY</td>\n",
       "      <td>Y</td>\n",
       "      <td>6865917</td>\n",
       "      <td>6911752</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000242389</th>\n",
       "      <td>RBMY1E</td>\n",
       "      <td>Y</td>\n",
       "      <td>21903617</td>\n",
       "      <td>21918042</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000172352</th>\n",
       "      <td>CDY1B</td>\n",
       "      <td>Y</td>\n",
       "      <td>24045228</td>\n",
       "      <td>24048019</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19393 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_name Chromosome     Start       End Strand\n",
       "gene_id                                                        \n",
       "ENSG00000160072    ATAD3B          1   1471764   1497848      +\n",
       "ENSG00000142611    PRDM16          1   3069167   3438621      +\n",
       "ENSG00000157933       SKI          1   2228318   2310213      +\n",
       "ENSG00000142655     PEX14          1  10472287  10630758      +\n",
       "ENSG00000149527     PLCH2          1   2425979   2505532      +\n",
       "...                   ...        ...       ...       ...    ...\n",
       "ENSG00000187191      DAZ3          Y  24763068  24813492      -\n",
       "ENSG00000188120      DAZ1          Y  23129354  23199010      -\n",
       "ENSG00000099721     AMELY          Y   6865917   6911752      -\n",
       "ENSG00000242389    RBMY1E          Y  21903617  21918042      -\n",
       "ENSG00000172352     CDY1B          Y  24045228  24048019      -\n",
       "\n",
       "[19393 rows x 5 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata1 = sc.read_h5ad(inputdata[0])\n",
    "adata1.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "aa767dd8-08fc-4272-b463-a4f5a349bec7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ensemble_id</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>B_mem</th>\n",
       "      <th>B_naive</th>\n",
       "      <th>...</th>\n",
       "      <th>Mega</th>\n",
       "      <th>Mono</th>\n",
       "      <th>MDSC</th>\n",
       "      <th>Neut</th>\n",
       "      <th>Osteo</th>\n",
       "      <th>Plasma</th>\n",
       "      <th>pDC</th>\n",
       "      <th>Retic</th>\n",
       "      <th>Stromal</th>\n",
       "      <th>cluster_genes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>1653.0</td>\n",
       "      <td>39930</td>\n",
       "      <td>False</td>\n",
       "      <td>0.041392</td>\n",
       "      <td>2.130549</td>\n",
       "      <td>-0.184673</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>53.0</td>\n",
       "      <td>42149</td>\n",
       "      <td>False</td>\n",
       "      <td>0.001226</td>\n",
       "      <td>0.618202</td>\n",
       "      <td>-0.620396</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>471452.0</td>\n",
       "      <td>36875</td>\n",
       "      <td>False</td>\n",
       "      <td>10.314141</td>\n",
       "      <td>20.391302</td>\n",
       "      <td>0.641015</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>A2ML1</td>\n",
       "      <td>ENSG00000166535</td>\n",
       "      <td>6200.0</td>\n",
       "      <td>49012</td>\n",
       "      <td>True</td>\n",
       "      <td>0.392610</td>\n",
       "      <td>8.034155</td>\n",
       "      <td>1.516218</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A3GALT2</th>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>ENSG00000184389</td>\n",
       "      <td>1324.0</td>\n",
       "      <td>46926</td>\n",
       "      <td>True</td>\n",
       "      <td>0.784540</td>\n",
       "      <td>10.645084</td>\n",
       "      <td>2.268454</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>ENSG00000203995</td>\n",
       "      <td>129.0</td>\n",
       "      <td>26942</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000869</td>\n",
       "      <td>-0.580755</td>\n",
       "      <td>-0.965828</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>28036.0</td>\n",
       "      <td>36992</td>\n",
       "      <td>False</td>\n",
       "      <td>0.225987</td>\n",
       "      <td>0.316922</td>\n",
       "      <td>-0.707198</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>90723.0</td>\n",
       "      <td>41568</td>\n",
       "      <td>False</td>\n",
       "      <td>0.383848</td>\n",
       "      <td>1.318699</td>\n",
       "      <td>-0.418576</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>24856.0</td>\n",
       "      <td>28154</td>\n",
       "      <td>False</td>\n",
       "      <td>3.753854</td>\n",
       "      <td>14.622520</td>\n",
       "      <td>0.324226</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ENSG00000036549</td>\n",
       "      <td>1238.0</td>\n",
       "      <td>43157</td>\n",
       "      <td>True</td>\n",
       "      <td>0.044078</td>\n",
       "      <td>4.729982</td>\n",
       "      <td>0.564251</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19070 rows Ã— 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene_name      ensemble_id  n_counts  n_cells  highly_variable  \\\n",
       "gene_name                                                                  \n",
       "A1BG           A1BG  ENSG00000121410    1653.0    39930            False   \n",
       "A1CF           A1CF  ENSG00000148584      53.0    42149            False   \n",
       "A2M             A2M  ENSG00000175899  471452.0    36875            False   \n",
       "A2ML1         A2ML1  ENSG00000166535    6200.0    49012             True   \n",
       "A3GALT2     A3GALT2  ENSG00000184389    1324.0    46926             True   \n",
       "...             ...              ...       ...      ...              ...   \n",
       "ZYG11A       ZYG11A  ENSG00000203995     129.0    26942            False   \n",
       "ZYG11B       ZYG11B  ENSG00000162378   28036.0    36992            False   \n",
       "ZYX             ZYX  ENSG00000159840   90723.0    41568            False   \n",
       "ZZEF1         ZZEF1  ENSG00000074755   24856.0    28154            False   \n",
       "ZZZ3           ZZZ3  ENSG00000036549    1238.0    43157             True   \n",
       "\n",
       "               means  dispersions  dispersions_norm  B_mem  B_naive  ...  \\\n",
       "gene_name                                                            ...   \n",
       "A1BG        0.041392     2.130549         -0.184673  False    False  ...   \n",
       "A1CF        0.001226     0.618202         -0.620396  False    False  ...   \n",
       "A2M        10.314141    20.391302          0.641015  False    False  ...   \n",
       "A2ML1       0.392610     8.034155          1.516218  False    False  ...   \n",
       "A3GALT2     0.784540    10.645084          2.268454  False    False  ...   \n",
       "...              ...          ...               ...    ...      ...  ...   \n",
       "ZYG11A      0.000869    -0.580755         -0.965828  False    False  ...   \n",
       "ZYG11B      0.225987     0.316922         -0.707198  False    False  ...   \n",
       "ZYX         0.383848     1.318699         -0.418576  False    False  ...   \n",
       "ZZEF1       3.753854    14.622520          0.324226  False    False  ...   \n",
       "ZZZ3        0.044078     4.729982          0.564251  False    False  ...   \n",
       "\n",
       "            Mega   Mono   MDSC   Neut  Osteo  Plasma    pDC  Retic  Stromal  \\\n",
       "gene_name                                                                     \n",
       "A1BG       False  False  False  False  False   False  False  False    False   \n",
       "A1CF       False  False  False  False  False   False  False  False    False   \n",
       "A2M        False  False  False  False  False   False  False  False    False   \n",
       "A2ML1      False  False  False  False  False   False  False  False    False   \n",
       "A3GALT2    False  False  False  False  False   False  False  False    False   \n",
       "...          ...    ...    ...    ...    ...     ...    ...    ...      ...   \n",
       "ZYG11A     False  False  False  False  False   False  False  False    False   \n",
       "ZYG11B     False  False  False  False  False   False  False  False    False   \n",
       "ZYX        False  False  False  False  False   False  False  False    False   \n",
       "ZZEF1      False  False  False  False  False   False  False  False    False   \n",
       "ZZZ3       False  False  False  False  False   False  False  False    False   \n",
       "\n",
       "           cluster_genes  \n",
       "gene_name                 \n",
       "A1BG               False  \n",
       "A1CF               False  \n",
       "A2M                False  \n",
       "A2ML1              False  \n",
       "A3GALT2            False  \n",
       "...                  ...  \n",
       "ZYG11A             False  \n",
       "ZYG11B             False  \n",
       "ZYX                False  \n",
       "ZZEF1              False  \n",
       "ZZZ3               False  \n",
       "\n",
       "[19070 rows x 28 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata2 = sc.read_h5ad(inputdata[1])\n",
    "adata2.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "92d85950-556b-4d8f-97f0-0e2509389689",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10883, 19393)\n",
      "(54347, 19070)\n"
     ]
    }
   ],
   "source": [
    "print(adata1.shape)\n",
    "print(adata2.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcd99e84-c03d-4326-89f0-0fd18f9ebd16",
   "metadata": {},
   "source": [
    "# Generate embeddings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe4eaad9-bd25-49da-95e1-edbd838ae560",
   "metadata": {},
   "source": [
    "## sc Fibroblasts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ab080e4a-31ab-4f15-a77f-e1bb57658ce3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>ensemble_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000160072</th>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>1</td>\n",
       "      <td>1471764</td>\n",
       "      <td>1497848</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000160072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142611</th>\n",
       "      <td>PRDM16</td>\n",
       "      <td>1</td>\n",
       "      <td>3069167</td>\n",
       "      <td>3438621</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000157933</th>\n",
       "      <td>SKI</td>\n",
       "      <td>1</td>\n",
       "      <td>2228318</td>\n",
       "      <td>2310213</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000157933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142655</th>\n",
       "      <td>PEX14</td>\n",
       "      <td>1</td>\n",
       "      <td>10472287</td>\n",
       "      <td>10630758</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000149527</th>\n",
       "      <td>PLCH2</td>\n",
       "      <td>1</td>\n",
       "      <td>2425979</td>\n",
       "      <td>2505532</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000149527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000187191</th>\n",
       "      <td>DAZ3</td>\n",
       "      <td>Y</td>\n",
       "      <td>24763068</td>\n",
       "      <td>24813492</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000187191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000188120</th>\n",
       "      <td>DAZ1</td>\n",
       "      <td>Y</td>\n",
       "      <td>23129354</td>\n",
       "      <td>23199010</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000188120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000099721</th>\n",
       "      <td>AMELY</td>\n",
       "      <td>Y</td>\n",
       "      <td>6865917</td>\n",
       "      <td>6911752</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000099721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000242389</th>\n",
       "      <td>RBMY1E</td>\n",
       "      <td>Y</td>\n",
       "      <td>21903617</td>\n",
       "      <td>21918042</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000242389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000172352</th>\n",
       "      <td>CDY1B</td>\n",
       "      <td>Y</td>\n",
       "      <td>24045228</td>\n",
       "      <td>24048019</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000172352</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19393 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_name Chromosome     Start       End Strand  \\\n",
       "gene_id                                                           \n",
       "ENSG00000160072    ATAD3B          1   1471764   1497848      +   \n",
       "ENSG00000142611    PRDM16          1   3069167   3438621      +   \n",
       "ENSG00000157933       SKI          1   2228318   2310213      +   \n",
       "ENSG00000142655     PEX14          1  10472287  10630758      +   \n",
       "ENSG00000149527     PLCH2          1   2425979   2505532      +   \n",
       "...                   ...        ...       ...       ...    ...   \n",
       "ENSG00000187191      DAZ3          Y  24763068  24813492      -   \n",
       "ENSG00000188120      DAZ1          Y  23129354  23199010      -   \n",
       "ENSG00000099721     AMELY          Y   6865917   6911752      -   \n",
       "ENSG00000242389    RBMY1E          Y  21903617  21918042      -   \n",
       "ENSG00000172352     CDY1B          Y  24045228  24048019      -   \n",
       "\n",
       "                     ensemble_id  \n",
       "gene_id                           \n",
       "ENSG00000160072  ENSG00000160072  \n",
       "ENSG00000142611  ENSG00000142611  \n",
       "ENSG00000157933  ENSG00000157933  \n",
       "ENSG00000142655  ENSG00000142655  \n",
       "ENSG00000149527  ENSG00000149527  \n",
       "...                          ...  \n",
       "ENSG00000187191  ENSG00000187191  \n",
       "ENSG00000188120  ENSG00000188120  \n",
       "ENSG00000099721  ENSG00000099721  \n",
       "ENSG00000242389  ENSG00000242389  \n",
       "ENSG00000172352  ENSG00000172352  \n",
       "\n",
       "[19393 rows x 6 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata1.var['ensemble_id'] = adata1.var.index\n",
    "adata1.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a0f89647-1dcc-404b-be87-f7a2595bbc18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_counts_UMIs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGGGTAGCA</th>\n",
       "      <td>16001</td>\n",
       "      <td>16001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGTAAGGGT</th>\n",
       "      <td>7864</td>\n",
       "      <td>7864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCATTCAGGTAGG</th>\n",
       "      <td>15610</td>\n",
       "      <td>15610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCATTCCAGCCCT</th>\n",
       "      <td>10505</td>\n",
       "      <td>10505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCATTCGTGACCG</th>\n",
       "      <td>8775</td>\n",
       "      <td>8775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGTGGTTGTAGCCCTA</th>\n",
       "      <td>10412</td>\n",
       "      <td>10412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGTGGTTGTTGAAGGC</th>\n",
       "      <td>11811</td>\n",
       "      <td>11811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGTGTTAGTTGGTAAG</th>\n",
       "      <td>6249</td>\n",
       "      <td>6249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGTGTTAGTTTGCGAT</th>\n",
       "      <td>2804</td>\n",
       "      <td>2804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGTGTTGAGCCTATCT</th>\n",
       "      <td>13163</td>\n",
       "      <td>13163</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10883 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  n_genes  n_counts_UMIs\n",
       "cell_id                                 \n",
       "AAACCAAAGGGTAGCA    16001          16001\n",
       "AAACCAAAGTAAGGGT     7864           7864\n",
       "AAACCATTCAGGTAGG    15610          15610\n",
       "AAACCATTCCAGCCCT    10505          10505\n",
       "AAACCATTCGTGACCG     8775           8775\n",
       "...                   ...            ...\n",
       "TGTGGTTGTAGCCCTA    10412          10412\n",
       "TGTGGTTGTTGAAGGC    11811          11811\n",
       "TGTGTTAGTTGGTAAG     6249           6249\n",
       "TGTGTTAGTTTGCGAT     2804           2804\n",
       "TGTGTTGAGCCTATCT    13163          13163\n",
       "\n",
       "[10883 rows x 2 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "adata1.obs['n_counts_UMIs'] = np.sum(adata1.X, axis=1)\n",
    "adata1.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2c8ee3ba-f31f-416d-8d65-c29a68452774",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output_directory='/scratch/indikar_root/indikar1/shared_data/single_cell_fibroblast/scanpy/'\n",
      "l212: output_path='/scratch/indikar_root/indikar1/shared_data/single_cell_fibroblast/scanpy/'\n",
      "l213: type(output_path)=<class 'str'>\n",
      "l214: (output_path==None)=False\n",
      "Tokenizing the dataset\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/views.py:144: RuntimeWarning: invalid value encountered in divide\n",
      "  results = super().__array_ufunc__(\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6eef750cf476442a96f445a9360fa71d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Map (num_proc=16):   0%|          | 0/10883 [00:00<?, ? examples/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "58598f7c9d5f4212b1e4b42c21e1416a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Saving the dataset (0/1 shards):   0%|          | 0/10883 [00:00<?, ? examples/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/nfs/turbo/umms-indikar/shared/projects/geneformer/geneformer-12L-30M/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/torch/_utils.py:831: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()\n",
      "  return self.fget.__get__(instance, owner)()\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e3c694c74c4c4883af8881c82d2f8023",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1089 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "output_path = '/scratch/indikar_root/indikar1/shared_data/single_cell_fibroblast/scanpy/'\n",
    "output_file = 'geneformer_no_perturbed_jpic_9_12_2024.h5ad'\n",
    "# Embed the AnnData object using Geneformer\n",
    "embedAdGF = embed(adata1, 'geneformer',\n",
    "                  output_directory=output_path,\n",
    "                  output_file=output_file,\n",
    "                  genenameloc='ensemble_id',\n",
    "                  n_counts_column='n_counts_UMIs'\n",
    "                 )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d778dcc1-1821-4260-a3e7-6f2d1c6db140",
   "metadata": {},
   "outputs": [],
   "source": [
    "embedAdGF.obs = adata1.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0767a607-1d71-41dd-8e8f-fb1f910389c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_counts_UMIs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGGGTAGCA</th>\n",
       "      <td>16001</td>\n",
       "      <td>16001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGTAAGGGT</th>\n",
       "      <td>7864</td>\n",
       "      <td>7864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCATTCAGGTAGG</th>\n",
       "      <td>15610</td>\n",
       "      <td>15610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCATTCCAGCCCT</th>\n",
       "      <td>10505</td>\n",
       "      <td>10505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCATTCGTGACCG</th>\n",
       "      <td>8775</td>\n",
       "      <td>8775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGTGGTTGTAGCCCTA</th>\n",
       "      <td>10412</td>\n",
       "      <td>10412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGTGGTTGTTGAAGGC</th>\n",
       "      <td>11811</td>\n",
       "      <td>11811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGTGTTAGTTGGTAAG</th>\n",
       "      <td>6249</td>\n",
       "      <td>6249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGTGTTAGTTTGCGAT</th>\n",
       "      <td>2804</td>\n",
       "      <td>2804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGTGTTGAGCCTATCT</th>\n",
       "      <td>13163</td>\n",
       "      <td>13163</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10883 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  n_genes  n_counts_UMIs\n",
       "cell_id                                 \n",
       "AAACCAAAGGGTAGCA    16001          16001\n",
       "AAACCAAAGTAAGGGT     7864           7864\n",
       "AAACCATTCAGGTAGG    15610          15610\n",
       "AAACCATTCCAGCCCT    10505          10505\n",
       "AAACCATTCGTGACCG     8775           8775\n",
       "...                   ...            ...\n",
       "TGTGGTTGTAGCCCTA    10412          10412\n",
       "TGTGGTTGTTGAAGGC    11811          11811\n",
       "TGTGTTAGTTGGTAAG     6249           6249\n",
       "TGTGTTAGTTTGCGAT     2804           2804\n",
       "TGTGTTGAGCCTATCT    13163          13163\n",
       "\n",
       "[10883 rows x 2 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "embedAdGF.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5b08e165-2a1b-49a8-a19d-0c96a7e9c4fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "embedAdGF.write_h5ad(os.path.join(output_path, output_file))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f664cf36-8eff-43ed-9465-35c4d94c36e4",
   "metadata": {},
   "source": [
    "## HSC landscape embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "09ad6a6f-76d2-4a5d-93ff-cb7c00a64594",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ensemble_id</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>n_cells</th>\n",
       "      <th>highly_variable</th>\n",
       "      <th>means</th>\n",
       "      <th>dispersions</th>\n",
       "      <th>dispersions_norm</th>\n",
       "      <th>B_mem</th>\n",
       "      <th>B_naive</th>\n",
       "      <th>...</th>\n",
       "      <th>Mega</th>\n",
       "      <th>Mono</th>\n",
       "      <th>MDSC</th>\n",
       "      <th>Neut</th>\n",
       "      <th>Osteo</th>\n",
       "      <th>Plasma</th>\n",
       "      <th>pDC</th>\n",
       "      <th>Retic</th>\n",
       "      <th>Stromal</th>\n",
       "      <th>cluster_genes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>1653.0</td>\n",
       "      <td>39930</td>\n",
       "      <td>False</td>\n",
       "      <td>0.041392</td>\n",
       "      <td>2.130549</td>\n",
       "      <td>-0.184673</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>53.0</td>\n",
       "      <td>42149</td>\n",
       "      <td>False</td>\n",
       "      <td>0.001226</td>\n",
       "      <td>0.618202</td>\n",
       "      <td>-0.620396</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>471452.0</td>\n",
       "      <td>36875</td>\n",
       "      <td>False</td>\n",
       "      <td>10.314141</td>\n",
       "      <td>20.391302</td>\n",
       "      <td>0.641015</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>A2ML1</td>\n",
       "      <td>ENSG00000166535</td>\n",
       "      <td>6200.0</td>\n",
       "      <td>49012</td>\n",
       "      <td>True</td>\n",
       "      <td>0.392610</td>\n",
       "      <td>8.034155</td>\n",
       "      <td>1.516218</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A3GALT2</th>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>ENSG00000184389</td>\n",
       "      <td>1324.0</td>\n",
       "      <td>46926</td>\n",
       "      <td>True</td>\n",
       "      <td>0.784540</td>\n",
       "      <td>10.645084</td>\n",
       "      <td>2.268454</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>ENSG00000203995</td>\n",
       "      <td>129.0</td>\n",
       "      <td>26942</td>\n",
       "      <td>False</td>\n",
       "      <td>0.000869</td>\n",
       "      <td>-0.580755</td>\n",
       "      <td>-0.965828</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>28036.0</td>\n",
       "      <td>36992</td>\n",
       "      <td>False</td>\n",
       "      <td>0.225987</td>\n",
       "      <td>0.316922</td>\n",
       "      <td>-0.707198</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>90723.0</td>\n",
       "      <td>41568</td>\n",
       "      <td>False</td>\n",
       "      <td>0.383848</td>\n",
       "      <td>1.318699</td>\n",
       "      <td>-0.418576</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>24856.0</td>\n",
       "      <td>28154</td>\n",
       "      <td>False</td>\n",
       "      <td>3.753854</td>\n",
       "      <td>14.622520</td>\n",
       "      <td>0.324226</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ENSG00000036549</td>\n",
       "      <td>1238.0</td>\n",
       "      <td>43157</td>\n",
       "      <td>True</td>\n",
       "      <td>0.044078</td>\n",
       "      <td>4.729982</td>\n",
       "      <td>0.564251</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19070 rows Ã— 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene_name      ensemble_id  n_counts  n_cells  highly_variable  \\\n",
       "gene_name                                                                  \n",
       "A1BG           A1BG  ENSG00000121410    1653.0    39930            False   \n",
       "A1CF           A1CF  ENSG00000148584      53.0    42149            False   \n",
       "A2M             A2M  ENSG00000175899  471452.0    36875            False   \n",
       "A2ML1         A2ML1  ENSG00000166535    6200.0    49012             True   \n",
       "A3GALT2     A3GALT2  ENSG00000184389    1324.0    46926             True   \n",
       "...             ...              ...       ...      ...              ...   \n",
       "ZYG11A       ZYG11A  ENSG00000203995     129.0    26942            False   \n",
       "ZYG11B       ZYG11B  ENSG00000162378   28036.0    36992            False   \n",
       "ZYX             ZYX  ENSG00000159840   90723.0    41568            False   \n",
       "ZZEF1         ZZEF1  ENSG00000074755   24856.0    28154            False   \n",
       "ZZZ3           ZZZ3  ENSG00000036549    1238.0    43157             True   \n",
       "\n",
       "               means  dispersions  dispersions_norm  B_mem  B_naive  ...  \\\n",
       "gene_name                                                            ...   \n",
       "A1BG        0.041392     2.130549         -0.184673  False    False  ...   \n",
       "A1CF        0.001226     0.618202         -0.620396  False    False  ...   \n",
       "A2M        10.314141    20.391302          0.641015  False    False  ...   \n",
       "A2ML1       0.392610     8.034155          1.516218  False    False  ...   \n",
       "A3GALT2     0.784540    10.645084          2.268454  False    False  ...   \n",
       "...              ...          ...               ...    ...      ...  ...   \n",
       "ZYG11A      0.000869    -0.580755         -0.965828  False    False  ...   \n",
       "ZYG11B      0.225987     0.316922         -0.707198  False    False  ...   \n",
       "ZYX         0.383848     1.318699         -0.418576  False    False  ...   \n",
       "ZZEF1       3.753854    14.622520          0.324226  False    False  ...   \n",
       "ZZZ3        0.044078     4.729982          0.564251  False    False  ...   \n",
       "\n",
       "            Mega   Mono   MDSC   Neut  Osteo  Plasma    pDC  Retic  Stromal  \\\n",
       "gene_name                                                                     \n",
       "A1BG       False  False  False  False  False   False  False  False    False   \n",
       "A1CF       False  False  False  False  False   False  False  False    False   \n",
       "A2M        False  False  False  False  False   False  False  False    False   \n",
       "A2ML1      False  False  False  False  False   False  False  False    False   \n",
       "A3GALT2    False  False  False  False  False   False  False  False    False   \n",
       "...          ...    ...    ...    ...    ...     ...    ...    ...      ...   \n",
       "ZYG11A     False  False  False  False  False   False  False  False    False   \n",
       "ZYG11B     False  False  False  False  False   False  False  False    False   \n",
       "ZYX        False  False  False  False  False   False  False  False    False   \n",
       "ZZEF1      False  False  False  False  False   False  False  False    False   \n",
       "ZZZ3       False  False  False  False  False   False  False  False    False   \n",
       "\n",
       "           cluster_genes  \n",
       "gene_name                 \n",
       "A1BG               False  \n",
       "A1CF               False  \n",
       "A2M                False  \n",
       "A2ML1              False  \n",
       "A3GALT2            False  \n",
       "...                  ...  \n",
       "ZYG11A             False  \n",
       "ZYG11B             False  \n",
       "ZYX                False  \n",
       "ZZEF1              False  \n",
       "ZZZ3               False  \n",
       "\n",
       "[19070 rows x 28 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata2.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "6dce4d05-5acf-4772-a7ef-219b73a10647",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_counts</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>dataset</th>\n",
       "      <th>organ_tissue</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>new_cluster</th>\n",
       "      <th>cell_label</th>\n",
       "      <th>cluster_str</th>\n",
       "      <th>UMAP 1</th>\n",
       "      <th>UMAP 2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGGTTACCT_iHSC</th>\n",
       "      <td>6558.0</td>\n",
       "      <td>iHSC</td>\n",
       "      <td>iHSC</td>\n",
       "      <td>nan</td>\n",
       "      <td>12480</td>\n",
       "      <td>1</td>\n",
       "      <td>Reprogrammed Cells</td>\n",
       "      <td>D2</td>\n",
       "      <td>8.989130</td>\n",
       "      <td>4.937538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGTTGAAGT_iHSC</th>\n",
       "      <td>5488.0</td>\n",
       "      <td>iHSC</td>\n",
       "      <td>iHSC</td>\n",
       "      <td>nan</td>\n",
       "      <td>12413</td>\n",
       "      <td>1</td>\n",
       "      <td>Reprogrammed Cells</td>\n",
       "      <td>D2</td>\n",
       "      <td>7.586400</td>\n",
       "      <td>4.889987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGTTGTCGT_iHSC</th>\n",
       "      <td>4330.0</td>\n",
       "      <td>iHSC</td>\n",
       "      <td>iHSC</td>\n",
       "      <td>nan</td>\n",
       "      <td>11906</td>\n",
       "      <td>10</td>\n",
       "      <td>Reprogrammed Cells</td>\n",
       "      <td>D11</td>\n",
       "      <td>8.027761</td>\n",
       "      <td>3.311177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAGAAGCGT_iHSC</th>\n",
       "      <td>3442.0</td>\n",
       "      <td>iHSC</td>\n",
       "      <td>iHSC</td>\n",
       "      <td>nan</td>\n",
       "      <td>11861</td>\n",
       "      <td>10</td>\n",
       "      <td>Reprogrammed Cells</td>\n",
       "      <td>D11</td>\n",
       "      <td>8.584245</td>\n",
       "      <td>2.674468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAGGAGGTT_iHSC</th>\n",
       "      <td>14427.0</td>\n",
       "      <td>iHSC</td>\n",
       "      <td>iHSC</td>\n",
       "      <td>nan</td>\n",
       "      <td>13436</td>\n",
       "      <td>1</td>\n",
       "      <td>Reprogrammed Cells</td>\n",
       "      <td>D2</td>\n",
       "      <td>9.509144</td>\n",
       "      <td>4.923519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGACTTCGCAATCG-2_old1_BMMC_HSPC</th>\n",
       "      <td>3616.0</td>\n",
       "      <td>MPP</td>\n",
       "      <td>old1_BMMC_HSPC</td>\n",
       "      <td>nan</td>\n",
       "      <td>13075</td>\n",
       "      <td>8</td>\n",
       "      <td>Multipotent Progenitor</td>\n",
       "      <td>D9</td>\n",
       "      <td>5.472989</td>\n",
       "      <td>7.853297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGACTTCGTTAAGC-2_old1_BMMC_HSPC</th>\n",
       "      <td>1262.0</td>\n",
       "      <td>MPP</td>\n",
       "      <td>old1_BMMC_HSPC</td>\n",
       "      <td>nan</td>\n",
       "      <td>12312</td>\n",
       "      <td>2</td>\n",
       "      <td>Multipotent Progenitor</td>\n",
       "      <td>D3</td>\n",
       "      <td>4.184635</td>\n",
       "      <td>4.982112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGAGTCAAGGAATC-2_old1_BMMC_HSPC</th>\n",
       "      <td>4720.0</td>\n",
       "      <td>MPP</td>\n",
       "      <td>old1_BMMC_HSPC</td>\n",
       "      <td>nan</td>\n",
       "      <td>13563</td>\n",
       "      <td>2</td>\n",
       "      <td>Multipotent Progenitor</td>\n",
       "      <td>D3</td>\n",
       "      <td>4.448498</td>\n",
       "      <td>7.588126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTGCAGAGAGCC-2_old1_BMMC_HSPC</th>\n",
       "      <td>5012.0</td>\n",
       "      <td>MPP</td>\n",
       "      <td>old1_BMMC_HSPC</td>\n",
       "      <td>nan</td>\n",
       "      <td>13628</td>\n",
       "      <td>2</td>\n",
       "      <td>Multipotent Progenitor</td>\n",
       "      <td>D3</td>\n",
       "      <td>4.706429</td>\n",
       "      <td>7.555507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCTAGCGCCTTT-2_old1_BMMC_HSPC</th>\n",
       "      <td>5103.0</td>\n",
       "      <td>MPP</td>\n",
       "      <td>old1_BMMC_HSPC</td>\n",
       "      <td>nan</td>\n",
       "      <td>13724</td>\n",
       "      <td>0</td>\n",
       "      <td>Multipotent Progenitor</td>\n",
       "      <td>D1</td>\n",
       "      <td>6.413480</td>\n",
       "      <td>7.787901</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54347 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   n_counts cell_type         dataset  \\\n",
       "AAACCCAAGGTTACCT_iHSC                6558.0      iHSC            iHSC   \n",
       "AAACCCAAGTTGAAGT_iHSC                5488.0      iHSC            iHSC   \n",
       "AAACCCAAGTTGTCGT_iHSC                4330.0      iHSC            iHSC   \n",
       "AAACCCACAGAAGCGT_iHSC                3442.0      iHSC            iHSC   \n",
       "AAACCCACAGGAGGTT_iHSC               14427.0      iHSC            iHSC   \n",
       "...                                     ...       ...             ...   \n",
       "TTTGACTTCGCAATCG-2_old1_BMMC_HSPC    3616.0       MPP  old1_BMMC_HSPC   \n",
       "TTTGACTTCGTTAAGC-2_old1_BMMC_HSPC    1262.0       MPP  old1_BMMC_HSPC   \n",
       "TTTGAGTCAAGGAATC-2_old1_BMMC_HSPC    4720.0       MPP  old1_BMMC_HSPC   \n",
       "TTTGGTGCAGAGAGCC-2_old1_BMMC_HSPC    5012.0       MPP  old1_BMMC_HSPC   \n",
       "TTTGTCTAGCGCCTTT-2_old1_BMMC_HSPC    5103.0       MPP  old1_BMMC_HSPC   \n",
       "\n",
       "                                  organ_tissue  n_genes new_cluster  \\\n",
       "AAACCCAAGGTTACCT_iHSC                      nan    12480           1   \n",
       "AAACCCAAGTTGAAGT_iHSC                      nan    12413           1   \n",
       "AAACCCAAGTTGTCGT_iHSC                      nan    11906          10   \n",
       "AAACCCACAGAAGCGT_iHSC                      nan    11861          10   \n",
       "AAACCCACAGGAGGTT_iHSC                      nan    13436           1   \n",
       "...                                        ...      ...         ...   \n",
       "TTTGACTTCGCAATCG-2_old1_BMMC_HSPC          nan    13075           8   \n",
       "TTTGACTTCGTTAAGC-2_old1_BMMC_HSPC          nan    12312           2   \n",
       "TTTGAGTCAAGGAATC-2_old1_BMMC_HSPC          nan    13563           2   \n",
       "TTTGGTGCAGAGAGCC-2_old1_BMMC_HSPC          nan    13628           2   \n",
       "TTTGTCTAGCGCCTTT-2_old1_BMMC_HSPC          nan    13724           0   \n",
       "\n",
       "                                               cell_label cluster_str  \\\n",
       "AAACCCAAGGTTACCT_iHSC                  Reprogrammed Cells          D2   \n",
       "AAACCCAAGTTGAAGT_iHSC                  Reprogrammed Cells          D2   \n",
       "AAACCCAAGTTGTCGT_iHSC                  Reprogrammed Cells         D11   \n",
       "AAACCCACAGAAGCGT_iHSC                  Reprogrammed Cells         D11   \n",
       "AAACCCACAGGAGGTT_iHSC                  Reprogrammed Cells          D2   \n",
       "...                                                   ...         ...   \n",
       "TTTGACTTCGCAATCG-2_old1_BMMC_HSPC  Multipotent Progenitor          D9   \n",
       "TTTGACTTCGTTAAGC-2_old1_BMMC_HSPC  Multipotent Progenitor          D3   \n",
       "TTTGAGTCAAGGAATC-2_old1_BMMC_HSPC  Multipotent Progenitor          D3   \n",
       "TTTGGTGCAGAGAGCC-2_old1_BMMC_HSPC  Multipotent Progenitor          D3   \n",
       "TTTGTCTAGCGCCTTT-2_old1_BMMC_HSPC  Multipotent Progenitor          D1   \n",
       "\n",
       "                                     UMAP 1    UMAP 2  \n",
       "AAACCCAAGGTTACCT_iHSC              8.989130  4.937538  \n",
       "AAACCCAAGTTGAAGT_iHSC              7.586400  4.889987  \n",
       "AAACCCAAGTTGTCGT_iHSC              8.027761  3.311177  \n",
       "AAACCCACAGAAGCGT_iHSC              8.584245  2.674468  \n",
       "AAACCCACAGGAGGTT_iHSC              9.509144  4.923519  \n",
       "...                                     ...       ...  \n",
       "TTTGACTTCGCAATCG-2_old1_BMMC_HSPC  5.472989  7.853297  \n",
       "TTTGACTTCGTTAAGC-2_old1_BMMC_HSPC  4.184635  4.982112  \n",
       "TTTGAGTCAAGGAATC-2_old1_BMMC_HSPC  4.448498  7.588126  \n",
       "TTTGGTGCAGAGAGCC-2_old1_BMMC_HSPC  4.706429  7.555507  \n",
       "TTTGTCTAGCGCCTTT-2_old1_BMMC_HSPC  6.413480  7.787901  \n",
       "\n",
       "[54347 rows x 10 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata2.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "216421f3-d841-4ffe-8de5-ed8d883a2498",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output_directory='/scratch/indikar_root/indikar1/shared_data/HSC/'\n",
      "l212: output_path='/scratch/indikar_root/indikar1/shared_data/HSC/'\n",
      "l213: type(output_path)=<class 'str'>\n",
      "l214: (output_path==None)=False\n",
      "Tokenizing the dataset\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "cb90fd07b0824ed3adbdfbd0c05c14da",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Map (num_proc=16):   0%|          | 0/54347 [00:00<?, ? examples/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5a4ff174f10f45779ad433a45a602d10",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Saving the dataset (0/1 shards):   0%|          | 0/54347 [00:00<?, ? examples/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/nfs/turbo/umms-indikar/shared/projects/geneformer/geneformer-12L-30M/\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f2531f6e5ec04dbc81fd3f789a2c274a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/5435 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "output_path = '/scratch/indikar_root/indikar1/shared_data/HSC/'\n",
    "output_file = 'geneformer_no_perturbed_jpic_9_12_2024.h5ad'\n",
    "# Embed the AnnData object using Geneformer\n",
    "embedAdGF = embed(adata2, 'geneformer',\n",
    "                  output_directory=output_path,\n",
    "                  output_file=output_file,\n",
    "                  genenameloc='ensemble_id',\n",
    "                  n_counts_column='n_counts'\n",
    "                 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0b624d06-8d3b-4568-91a6-6beb2205536d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_counts</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>dataset</th>\n",
       "      <th>organ_tissue</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>new_cluster</th>\n",
       "      <th>cell_label</th>\n",
       "      <th>cluster_str</th>\n",
       "      <th>UMAP 1</th>\n",
       "      <th>UMAP 2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGGTTACCT_iHSC</th>\n",
       "      <td>6558.0</td>\n",
       "      <td>iHSC</td>\n",
       "      <td>iHSC</td>\n",
       "      <td>nan</td>\n",
       "      <td>12480</td>\n",
       "      <td>1</td>\n",
       "      <td>Reprogrammed Cells</td>\n",
       "      <td>D2</td>\n",
       "      <td>8.989130</td>\n",
       "      <td>4.937538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGTTGAAGT_iHSC</th>\n",
       "      <td>5488.0</td>\n",
       "      <td>iHSC</td>\n",
       "      <td>iHSC</td>\n",
       "      <td>nan</td>\n",
       "      <td>12413</td>\n",
       "      <td>1</td>\n",
       "      <td>Reprogrammed Cells</td>\n",
       "      <td>D2</td>\n",
       "      <td>7.586400</td>\n",
       "      <td>4.889987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGTTGTCGT_iHSC</th>\n",
       "      <td>4330.0</td>\n",
       "      <td>iHSC</td>\n",
       "      <td>iHSC</td>\n",
       "      <td>nan</td>\n",
       "      <td>11906</td>\n",
       "      <td>10</td>\n",
       "      <td>Reprogrammed Cells</td>\n",
       "      <td>D11</td>\n",
       "      <td>8.027761</td>\n",
       "      <td>3.311177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAGAAGCGT_iHSC</th>\n",
       "      <td>3442.0</td>\n",
       "      <td>iHSC</td>\n",
       "      <td>iHSC</td>\n",
       "      <td>nan</td>\n",
       "      <td>11861</td>\n",
       "      <td>10</td>\n",
       "      <td>Reprogrammed Cells</td>\n",
       "      <td>D11</td>\n",
       "      <td>8.584245</td>\n",
       "      <td>2.674468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACAGGAGGTT_iHSC</th>\n",
       "      <td>14427.0</td>\n",
       "      <td>iHSC</td>\n",
       "      <td>iHSC</td>\n",
       "      <td>nan</td>\n",
       "      <td>13436</td>\n",
       "      <td>1</td>\n",
       "      <td>Reprogrammed Cells</td>\n",
       "      <td>D2</td>\n",
       "      <td>9.509144</td>\n",
       "      <td>4.923519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGACTTCGCAATCG-2_old1_BMMC_HSPC</th>\n",
       "      <td>3616.0</td>\n",
       "      <td>MPP</td>\n",
       "      <td>old1_BMMC_HSPC</td>\n",
       "      <td>nan</td>\n",
       "      <td>13075</td>\n",
       "      <td>8</td>\n",
       "      <td>Multipotent Progenitor</td>\n",
       "      <td>D9</td>\n",
       "      <td>5.472989</td>\n",
       "      <td>7.853297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGACTTCGTTAAGC-2_old1_BMMC_HSPC</th>\n",
       "      <td>1262.0</td>\n",
       "      <td>MPP</td>\n",
       "      <td>old1_BMMC_HSPC</td>\n",
       "      <td>nan</td>\n",
       "      <td>12312</td>\n",
       "      <td>2</td>\n",
       "      <td>Multipotent Progenitor</td>\n",
       "      <td>D3</td>\n",
       "      <td>4.184635</td>\n",
       "      <td>4.982112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGAGTCAAGGAATC-2_old1_BMMC_HSPC</th>\n",
       "      <td>4720.0</td>\n",
       "      <td>MPP</td>\n",
       "      <td>old1_BMMC_HSPC</td>\n",
       "      <td>nan</td>\n",
       "      <td>13563</td>\n",
       "      <td>2</td>\n",
       "      <td>Multipotent Progenitor</td>\n",
       "      <td>D3</td>\n",
       "      <td>4.448498</td>\n",
       "      <td>7.588126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTGCAGAGAGCC-2_old1_BMMC_HSPC</th>\n",
       "      <td>5012.0</td>\n",
       "      <td>MPP</td>\n",
       "      <td>old1_BMMC_HSPC</td>\n",
       "      <td>nan</td>\n",
       "      <td>13628</td>\n",
       "      <td>2</td>\n",
       "      <td>Multipotent Progenitor</td>\n",
       "      <td>D3</td>\n",
       "      <td>4.706429</td>\n",
       "      <td>7.555507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCTAGCGCCTTT-2_old1_BMMC_HSPC</th>\n",
       "      <td>5103.0</td>\n",
       "      <td>MPP</td>\n",
       "      <td>old1_BMMC_HSPC</td>\n",
       "      <td>nan</td>\n",
       "      <td>13724</td>\n",
       "      <td>0</td>\n",
       "      <td>Multipotent Progenitor</td>\n",
       "      <td>D1</td>\n",
       "      <td>6.413480</td>\n",
       "      <td>7.787901</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54347 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   n_counts cell_type         dataset  \\\n",
       "AAACCCAAGGTTACCT_iHSC                6558.0      iHSC            iHSC   \n",
       "AAACCCAAGTTGAAGT_iHSC                5488.0      iHSC            iHSC   \n",
       "AAACCCAAGTTGTCGT_iHSC                4330.0      iHSC            iHSC   \n",
       "AAACCCACAGAAGCGT_iHSC                3442.0      iHSC            iHSC   \n",
       "AAACCCACAGGAGGTT_iHSC               14427.0      iHSC            iHSC   \n",
       "...                                     ...       ...             ...   \n",
       "TTTGACTTCGCAATCG-2_old1_BMMC_HSPC    3616.0       MPP  old1_BMMC_HSPC   \n",
       "TTTGACTTCGTTAAGC-2_old1_BMMC_HSPC    1262.0       MPP  old1_BMMC_HSPC   \n",
       "TTTGAGTCAAGGAATC-2_old1_BMMC_HSPC    4720.0       MPP  old1_BMMC_HSPC   \n",
       "TTTGGTGCAGAGAGCC-2_old1_BMMC_HSPC    5012.0       MPP  old1_BMMC_HSPC   \n",
       "TTTGTCTAGCGCCTTT-2_old1_BMMC_HSPC    5103.0       MPP  old1_BMMC_HSPC   \n",
       "\n",
       "                                  organ_tissue  n_genes new_cluster  \\\n",
       "AAACCCAAGGTTACCT_iHSC                      nan    12480           1   \n",
       "AAACCCAAGTTGAAGT_iHSC                      nan    12413           1   \n",
       "AAACCCAAGTTGTCGT_iHSC                      nan    11906          10   \n",
       "AAACCCACAGAAGCGT_iHSC                      nan    11861          10   \n",
       "AAACCCACAGGAGGTT_iHSC                      nan    13436           1   \n",
       "...                                        ...      ...         ...   \n",
       "TTTGACTTCGCAATCG-2_old1_BMMC_HSPC          nan    13075           8   \n",
       "TTTGACTTCGTTAAGC-2_old1_BMMC_HSPC          nan    12312           2   \n",
       "TTTGAGTCAAGGAATC-2_old1_BMMC_HSPC          nan    13563           2   \n",
       "TTTGGTGCAGAGAGCC-2_old1_BMMC_HSPC          nan    13628           2   \n",
       "TTTGTCTAGCGCCTTT-2_old1_BMMC_HSPC          nan    13724           0   \n",
       "\n",
       "                                               cell_label cluster_str  \\\n",
       "AAACCCAAGGTTACCT_iHSC                  Reprogrammed Cells          D2   \n",
       "AAACCCAAGTTGAAGT_iHSC                  Reprogrammed Cells          D2   \n",
       "AAACCCAAGTTGTCGT_iHSC                  Reprogrammed Cells         D11   \n",
       "AAACCCACAGAAGCGT_iHSC                  Reprogrammed Cells         D11   \n",
       "AAACCCACAGGAGGTT_iHSC                  Reprogrammed Cells          D2   \n",
       "...                                                   ...         ...   \n",
       "TTTGACTTCGCAATCG-2_old1_BMMC_HSPC  Multipotent Progenitor          D9   \n",
       "TTTGACTTCGTTAAGC-2_old1_BMMC_HSPC  Multipotent Progenitor          D3   \n",
       "TTTGAGTCAAGGAATC-2_old1_BMMC_HSPC  Multipotent Progenitor          D3   \n",
       "TTTGGTGCAGAGAGCC-2_old1_BMMC_HSPC  Multipotent Progenitor          D3   \n",
       "TTTGTCTAGCGCCTTT-2_old1_BMMC_HSPC  Multipotent Progenitor          D1   \n",
       "\n",
       "                                     UMAP 1    UMAP 2  \n",
       "AAACCCAAGGTTACCT_iHSC              8.989130  4.937538  \n",
       "AAACCCAAGTTGAAGT_iHSC              7.586400  4.889987  \n",
       "AAACCCAAGTTGTCGT_iHSC              8.027761  3.311177  \n",
       "AAACCCACAGAAGCGT_iHSC              8.584245  2.674468  \n",
       "AAACCCACAGGAGGTT_iHSC              9.509144  4.923519  \n",
       "...                                     ...       ...  \n",
       "TTTGACTTCGCAATCG-2_old1_BMMC_HSPC  5.472989  7.853297  \n",
       "TTTGACTTCGTTAAGC-2_old1_BMMC_HSPC  4.184635  4.982112  \n",
       "TTTGAGTCAAGGAATC-2_old1_BMMC_HSPC  4.448498  7.588126  \n",
       "TTTGGTGCAGAGAGCC-2_old1_BMMC_HSPC  4.706429  7.555507  \n",
       "TTTGTCTAGCGCCTTT-2_old1_BMMC_HSPC  6.413480  7.787901  \n",
       "\n",
       "[54347 rows x 10 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "embedAdGF.obs = adata2.obs\n",
    "embedAdGF.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8e752cc3-b8e6-4795-946f-e2b012cae1bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "embedAdGF.write_h5ad(os.path.join(output_path, output_file))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e8d8447-cb68-4ae6-b314-5167fa58705e",
   "metadata": {},
   "source": [
    "## Perturbe and Embed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1bb00d76-a18a-4157-b18c-a06a0511a6b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>ensemble_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000160072</th>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>1</td>\n",
       "      <td>1471764</td>\n",
       "      <td>1497848</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000160072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142611</th>\n",
       "      <td>PRDM16</td>\n",
       "      <td>1</td>\n",
       "      <td>3069167</td>\n",
       "      <td>3438621</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000157933</th>\n",
       "      <td>SKI</td>\n",
       "      <td>1</td>\n",
       "      <td>2228318</td>\n",
       "      <td>2310213</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000157933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142655</th>\n",
       "      <td>PEX14</td>\n",
       "      <td>1</td>\n",
       "      <td>10472287</td>\n",
       "      <td>10630758</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000149527</th>\n",
       "      <td>PLCH2</td>\n",
       "      <td>1</td>\n",
       "      <td>2425979</td>\n",
       "      <td>2505532</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000149527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000187191</th>\n",
       "      <td>DAZ3</td>\n",
       "      <td>Y</td>\n",
       "      <td>24763068</td>\n",
       "      <td>24813492</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000187191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000188120</th>\n",
       "      <td>DAZ1</td>\n",
       "      <td>Y</td>\n",
       "      <td>23129354</td>\n",
       "      <td>23199010</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000188120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000099721</th>\n",
       "      <td>AMELY</td>\n",
       "      <td>Y</td>\n",
       "      <td>6865917</td>\n",
       "      <td>6911752</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000099721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000242389</th>\n",
       "      <td>RBMY1E</td>\n",
       "      <td>Y</td>\n",
       "      <td>21903617</td>\n",
       "      <td>21918042</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000242389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000172352</th>\n",
       "      <td>CDY1B</td>\n",
       "      <td>Y</td>\n",
       "      <td>24045228</td>\n",
       "      <td>24048019</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000172352</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19393 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_name Chromosome     Start       End Strand  \\\n",
       "gene_id                                                           \n",
       "ENSG00000160072    ATAD3B          1   1471764   1497848      +   \n",
       "ENSG00000142611    PRDM16          1   3069167   3438621      +   \n",
       "ENSG00000157933       SKI          1   2228318   2310213      +   \n",
       "ENSG00000142655     PEX14          1  10472287  10630758      +   \n",
       "ENSG00000149527     PLCH2          1   2425979   2505532      +   \n",
       "...                   ...        ...       ...       ...    ...   \n",
       "ENSG00000187191      DAZ3          Y  24763068  24813492      -   \n",
       "ENSG00000188120      DAZ1          Y  23129354  23199010      -   \n",
       "ENSG00000099721     AMELY          Y   6865917   6911752      -   \n",
       "ENSG00000242389    RBMY1E          Y  21903617  21918042      -   \n",
       "ENSG00000172352     CDY1B          Y  24045228  24048019      -   \n",
       "\n",
       "                     ensemble_id  \n",
       "gene_id                           \n",
       "ENSG00000160072  ENSG00000160072  \n",
       "ENSG00000142611  ENSG00000142611  \n",
       "ENSG00000157933  ENSG00000157933  \n",
       "ENSG00000142655  ENSG00000142655  \n",
       "ENSG00000149527  ENSG00000149527  \n",
       "...                          ...  \n",
       "ENSG00000187191  ENSG00000187191  \n",
       "ENSG00000188120  ENSG00000188120  \n",
       "ENSG00000099721  ENSG00000099721  \n",
       "ENSG00000242389  ENSG00000242389  \n",
       "ENSG00000172352  ENSG00000172352  \n",
       "\n",
       "[19393 rows x 6 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata1.var['ensemble_id'] = adata1.var.index\n",
    "adata1.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "dc4a91f9-e49e-4119-83db-614b64ad56ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "def iterate_perturb_id_counts(adata, tf_list, scalar_list):\n",
    "    \"\"\"\n",
    "    Applies perturbations to the expression data of specified transcription factors across multiple scalars \n",
    "    and stores the resulting AnnData objects in a dictionary.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Iterates over a list of scalar values.\n",
    "    2. For each scalar, creates a copy of the AnnData object to preserve the original data.\n",
    "    3. Applies the `perturb_counts` function to scale the expression data of genes listed in `tf_list` by\n",
    "       the maximum gene expression of each cell and the current scalar.\n",
    "    4. Stores the perturbed AnnData object in a dictionary with the scalar as the key.\n",
    "\n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols (transcription factors) to be perturbed.\n",
    "    scalar_list (list): A list of scalar values for scaling the gene expression.\n",
    "    adata (AnnData): The AnnData object containing gene expression data (cells x genes).\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary where keys are scalar values and values are the corresponding perturbed AnnData objects.\n",
    "    \"\"\"\n",
    "    \n",
    "    adata_dict = {}\n",
    "    \n",
    "    for scalar in scalar_list:\n",
    "        # Create a copy of the AnnData object for each scalar value\n",
    "        adata_temp = adata.copy()\n",
    "        \n",
    "        # Apply perturb_counts to the copied AnnData object\n",
    "        perturbed_adata = perturb_id_counts(adata_temp, tf_list, scalar) # JP change this line\n",
    "        \n",
    "        # Store the perturbed AnnData object in the dictionary with scalar as the key\n",
    "        adata_dict[scalar] = perturbed_adata\n",
    "    \n",
    "    return adata_dict\n",
    "\n",
    "def perturb_id_counts(adata, tf_list, scalar): \n",
    "    \"\"\"\n",
    "    Applies a perturbation to the expression data of specific genes in an AnnData object.\n",
    "\n",
    "    This function performs the following steps:\n",
    "    1. Computes the maximum gene expression level for each cell.\n",
    "    2. Applies a scaling operation to the expression levels of genes listed in `tf_list`.\n",
    "       - Each entry of these genes in the matrix is multiplied by the maximum expression level \n",
    "         of its respective cell and a specified scalar value.\n",
    "    3. Updates the AnnData object with new columns:\n",
    "       - 'scaled': A boolean column indicating whether each gene is in the `tf_list`.\n",
    "       - 'scaled_by': Contains the scaling factor used for each gene (the product of the maximum \n",
    "         expression level of each cell and the scalar), or `1` if the gene was not in `tf_list`.\n",
    "    \n",
    "    Parameters:\n",
    "    tf_list (list): A list of gene symbols to be perturbed.\n",
    "    scalar (float): The scalar value used to scale the expression levels.\n",
    "    adata (AnnData): The AnnData object containing gene expression data.\n",
    "\n",
    "    Returns:\n",
    "    AnnData: The updated AnnData object with applied perturbations and new columns.\n",
    "    \"\"\"\n",
    "\n",
    "    # Create a boolean mask for genes in tf_list\n",
    "    gene_mask = adata.var['gene_name'].isin(tf_list)\n",
    "    \n",
    "    # Save the original state of the parameter objects, in case some tfs do not translate (failsafe)\n",
    "    original_X = adata.X.copy()\n",
    "    original_gene_mask = gene_mask.copy()\n",
    "    \n",
    "    # Compute maximum expression level of each cell\n",
    "    max_exp = np.max(adata.X, axis=1)\n",
    "\n",
    "    \"\"\"This is new today. v \"\"\"\n",
    "    # Raise an error if any of the gene names in tf_list do not match column names (we will manually update these in adata):\n",
    "    missing_genes = [gene for gene in tf_list if gene not in adata.var['gene_name'].values]\n",
    "    \n",
    "    if missing_genes:\n",
    "        # Restore original parameter objects\n",
    "        adata.X = original_X\n",
    "        gene_mask = original_gene_mask\n",
    "        raise ValueError(f\"Genes {missing_genes} not found in anndata object\")\n",
    "\n",
    "    else:    \n",
    "        \n",
    "        # Apply the scaling operation to the specified genes\n",
    "        adata.X[:, gene_mask] = (max_exp * scalar).reshape(-1, 1)\n",
    "        \n",
    "        # Add/Update 'scaled' column in var\n",
    "        adata.var['scaled'] = gene_mask\n",
    "        \n",
    "        # Add/Update 'scaled_by' column in var\n",
    "        adata.var['scaled_by'] = scalar  # Default value for genes not in tf_list\n",
    "\n",
    "        # Add/Update 'scaled_by' column in var\n",
    "        adata.obs['U'] = scalar  # Default value for genes not in tf_list\n",
    "    # adata.var = \n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "69bbdd79-68d9-44f5-b712-25a06c80b65e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['GATA2'],\n",
       " ['GFI1B'],\n",
       " ['STAT5A'],\n",
       " ['REL'],\n",
       " ['FOS'],\n",
       " ['GATA2', 'GFI1B', 'STAT5A', 'REL', 'FOS']]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_directory = \"/scratch/indikar_root/indikar1/shared_data/single_cell_fibroblast/scanpy/\"\n",
    "TFlist = ['GATA2', 'GFI1B', 'STAT5A', 'REL', 'FOS']\n",
    "scalars = [0.25, 0.5, 0.75, 1.001]\n",
    "\n",
    "TFcombos = [[tf] for tf in TFlist]\n",
    "TFcombos.append(TFlist)\n",
    "TFcombos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "45053fda-a304-4f7c-9ab6-105f78321a28",
   "metadata": {},
   "outputs": [],
   "source": [
    "TFs = ['GATA2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "fcbeec3c-daaf-4cb8-a652-89d9fbb0a986",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GATA2'"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"_\".join(TFs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5028cbb1-9917-4952-a8a5-63e5fcc0a193",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/scratch/indikar_root/indikar1/shared_data/single_cell_fibroblast/scanpy/GATA2.h5ad already exists: continuing to next TF!\n",
      "/scratch/indikar_root/indikar1/shared_data/single_cell_fibroblast/scanpy/GFI1B.h5ad already exists: continuing to next TF!\n",
      "/scratch/indikar_root/indikar1/shared_data/single_cell_fibroblast/scanpy/STAT5A.h5ad already exists: continuing to next TF!\n",
      "/scratch/indikar_root/indikar1/shared_data/single_cell_fibroblast/scanpy/REL.h5ad already exists: continuing to next TF!\n",
      "output_path='/scratch/indikar_root/indikar1/shared_data/single_cell_fibroblast/scanpy/FOS.h5ad'\n",
      "Perturbing TFs: ['FOS']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Beginning geneformer embeddings\n",
      "output_directory='/scratch/indikar_root/indikar1/shared_data/single_cell_fibroblast/scanpy/FOS.h5ad'\n",
      "l212: output_path='/scratch/indikar_root/indikar1/shared_data/single_cell_fibroblast/scanpy/FOS.h5ad'\n",
      "l213: type(output_path)=<class 'str'>\n",
      "l214: (output_path==None)=False\n",
      "Dataset is pretokenized\n",
      "/nfs/turbo/umms-indikar/shared/projects/geneformer/geneformer-12L-30M/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/torch/_utils.py:831: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()\n",
      "  return self.fget.__get__(instance, owner)()\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "312d94e8146449ac902f61edd6b18265",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4354 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output_path='/scratch/indikar_root/indikar1/shared_data/single_cell_fibroblast/scanpy/GATA2_GFI1B_STAT5A_REL_FOS.h5ad'\n",
      "Perturbing TFs: ['GATA2', 'GFI1B', 'STAT5A', 'REL', 'FOS']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Beginning geneformer embeddings\n",
      "output_directory='/scratch/indikar_root/indikar1/shared_data/single_cell_fibroblast/scanpy/GATA2_GFI1B_STAT5A_REL_FOS.h5ad'\n",
      "l212: output_path='/scratch/indikar_root/indikar1/shared_data/single_cell_fibroblast/scanpy/GATA2_GFI1B_STAT5A_REL_FOS.h5ad'\n",
      "l213: type(output_path)=<class 'str'>\n",
      "l214: (output_path==None)=False\n",
      "Tokenizing the dataset\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/views.py:144: RuntimeWarning: invalid value encountered in divide\n",
      "  results = super().__array_ufunc__(\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1c145d10ccd546b18917cff37d3ca706",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Map (num_proc=16):   0%|          | 0/43532 [00:00<?, ? examples/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f84a65a7a467464f9355e8d076d62e06",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Saving the dataset (0/1 shards):   0%|          | 0/43532 [00:00<?, ? examples/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/nfs/turbo/umms-indikar/shared/projects/geneformer/geneformer-12L-30M/\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "955eb882f8414338a1e598e7d5831fd2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4354 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for TFs in TFcombos:\n",
    "    TFs_str = str(TFs)\n",
    "    if isinstance(TFs, list):\n",
    "        TFs_str = \"_\".join(TFs)\n",
    "    output_path = os.path.join(output_directory, f\"{TFs_str}.h5ad\")\n",
    "    \n",
    "    # Check if the file already exists\n",
    "    if os.path.exists(output_path):\n",
    "        print(f\"{output_path} already exists: continuing to next TF!\")\n",
    "        continue\n",
    "\n",
    "    print(f\"{output_path=}\")\n",
    "\n",
    "    print(f\"Perturbing TFs: {TFs}\")\n",
    "    adataDict = iterate_perturb_id_counts(adata1.copy(), TFs, scalars)\n",
    "\n",
    "    # Concatenate all AnnData objects along the observations axis\n",
    "    concatenated_adata = ad.concat(list(adataDict.values()), axis=0)\n",
    "\n",
    "    # Copy var information from the original AnnData\n",
    "    concatenated_adata.var = adata1.var.copy()\n",
    "\n",
    "    # Print before embeddings\n",
    "    print('Beginning geneformer embeddings')\n",
    "    \n",
    "    embedAdGF = embed(concatenated_adata, 'geneformer',\n",
    "                      output_directory=output_path,\n",
    "                      output_file=output_path,\n",
    "                      genenameloc='ensemble_id',\n",
    "                      n_counts_column='n_counts'\n",
    "                     )\n",
    "    \n",
    "    # Save the concatenated AnnData object to the file\n",
    "    embedAdGF.write_h5ad(output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7c00dd72-6558-46cf-bb28-bff5a342ffc8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "adata = sc.read_h5ad('/scratch/indikar_root/indikar1/shared_data/single_cell_fibroblast/scanpy/GATA2_GFI1B_STAT5A_REL_FOS.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "bcbec135-6086-4250-b285-b9f7a8b89fc6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output_directory='/scratch/indikar_root/indikar1/shared_data/single_cell_fibroblast/scanpy/GATA2.h5ad'\n",
      "l212: output_path='/scratch/indikar_root/indikar1/shared_data/single_cell_fibroblast/scanpy/GATA2.h5ad'\n",
      "l213: type(output_path)=<class 'str'>\n",
      "l214: (output_path==None)=False\n",
      "Tokenizing the dataset\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpic/.local/lib/python3.11/site-packages/anndata/_core/views.py:144: RuntimeWarning: invalid value encountered in divide\n",
      "  results = super().__array_ufunc__(\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f9970e3ad6cd428f84c38d86caa75a5a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Map (num_proc=16):   0%|          | 0/43532 [00:00<?, ? examples/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8629f8b996554f61b3b54c2085a509b1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Saving the dataset (0/1 shards):   0%|          | 0/43532 [00:00<?, ? examples/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/nfs/turbo/umms-indikar/shared/projects/geneformer/geneformer-12L-30M/\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4bb1e2be71064a569ba2635faf1532a5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4354 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "embedAdGF = embed(concatenated_adata, 'geneformer',\n",
    "                  output_directory=output_path,\n",
    "                  output_file=output_path,\n",
    "                  genenameloc='ensemble_id',\n",
    "                  n_counts_column='n_counts'\n",
    "                 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "81683582-61e5-43af-8c38-3b74a3678c8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the concatenated AnnData object to the file\n",
    "embedAdGF.write_h5ad(output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "45cca88e-e420-4a34-a135-f8b2066cd174",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/scratch/indikar_root/indikar1/shared_data/single_cell_fibroblast/scanpy/GATA2.h5ad'"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "e52944a3-4803-4194-b734-a553ece7b5db",
   "metadata": {},
   "outputs": [],
   "source": [
    "concatenated_adata.var = adata1.var.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "99187d92-8931-4126-bc7e-b554fd213a42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>ensemble_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000160072</th>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>1</td>\n",
       "      <td>1471764</td>\n",
       "      <td>1497848</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000160072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142611</th>\n",
       "      <td>PRDM16</td>\n",
       "      <td>1</td>\n",
       "      <td>3069167</td>\n",
       "      <td>3438621</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000157933</th>\n",
       "      <td>SKI</td>\n",
       "      <td>1</td>\n",
       "      <td>2228318</td>\n",
       "      <td>2310213</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000157933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142655</th>\n",
       "      <td>PEX14</td>\n",
       "      <td>1</td>\n",
       "      <td>10472287</td>\n",
       "      <td>10630758</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000149527</th>\n",
       "      <td>PLCH2</td>\n",
       "      <td>1</td>\n",
       "      <td>2425979</td>\n",
       "      <td>2505532</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000149527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000187191</th>\n",
       "      <td>DAZ3</td>\n",
       "      <td>Y</td>\n",
       "      <td>24763068</td>\n",
       "      <td>24813492</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000187191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000188120</th>\n",
       "      <td>DAZ1</td>\n",
       "      <td>Y</td>\n",
       "      <td>23129354</td>\n",
       "      <td>23199010</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000188120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000099721</th>\n",
       "      <td>AMELY</td>\n",
       "      <td>Y</td>\n",
       "      <td>6865917</td>\n",
       "      <td>6911752</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000099721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000242389</th>\n",
       "      <td>RBMY1E</td>\n",
       "      <td>Y</td>\n",
       "      <td>21903617</td>\n",
       "      <td>21918042</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000242389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000172352</th>\n",
       "      <td>CDY1B</td>\n",
       "      <td>Y</td>\n",
       "      <td>24045228</td>\n",
       "      <td>24048019</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000172352</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19393 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_name Chromosome     Start       End Strand  \\\n",
       "gene_id                                                           \n",
       "ENSG00000160072    ATAD3B          1   1471764   1497848      +   \n",
       "ENSG00000142611    PRDM16          1   3069167   3438621      +   \n",
       "ENSG00000157933       SKI          1   2228318   2310213      +   \n",
       "ENSG00000142655     PEX14          1  10472287  10630758      +   \n",
       "ENSG00000149527     PLCH2          1   2425979   2505532      +   \n",
       "...                   ...        ...       ...       ...    ...   \n",
       "ENSG00000187191      DAZ3          Y  24763068  24813492      -   \n",
       "ENSG00000188120      DAZ1          Y  23129354  23199010      -   \n",
       "ENSG00000099721     AMELY          Y   6865917   6911752      -   \n",
       "ENSG00000242389    RBMY1E          Y  21903617  21918042      -   \n",
       "ENSG00000172352     CDY1B          Y  24045228  24048019      -   \n",
       "\n",
       "                     ensemble_id  \n",
       "gene_id                           \n",
       "ENSG00000160072  ENSG00000160072  \n",
       "ENSG00000142611  ENSG00000142611  \n",
       "ENSG00000157933  ENSG00000157933  \n",
       "ENSG00000142655  ENSG00000142655  \n",
       "ENSG00000149527  ENSG00000149527  \n",
       "...                          ...  \n",
       "ENSG00000187191  ENSG00000187191  \n",
       "ENSG00000188120  ENSG00000188120  \n",
       "ENSG00000099721  ENSG00000099721  \n",
       "ENSG00000242389  ENSG00000242389  \n",
       "ENSG00000172352  ENSG00000172352  \n",
       "\n",
       "[19393 rows x 6 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concatenated_adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "ad9cc3a6-179d-4a84-a522-37587f2c9ed7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>ensemble_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000160072</th>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>1</td>\n",
       "      <td>1471764</td>\n",
       "      <td>1497848</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000160072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142611</th>\n",
       "      <td>PRDM16</td>\n",
       "      <td>1</td>\n",
       "      <td>3069167</td>\n",
       "      <td>3438621</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000157933</th>\n",
       "      <td>SKI</td>\n",
       "      <td>1</td>\n",
       "      <td>2228318</td>\n",
       "      <td>2310213</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000157933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000142655</th>\n",
       "      <td>PEX14</td>\n",
       "      <td>1</td>\n",
       "      <td>10472287</td>\n",
       "      <td>10630758</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000142655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000149527</th>\n",
       "      <td>PLCH2</td>\n",
       "      <td>1</td>\n",
       "      <td>2425979</td>\n",
       "      <td>2505532</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000149527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000187191</th>\n",
       "      <td>DAZ3</td>\n",
       "      <td>Y</td>\n",
       "      <td>24763068</td>\n",
       "      <td>24813492</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000187191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000188120</th>\n",
       "      <td>DAZ1</td>\n",
       "      <td>Y</td>\n",
       "      <td>23129354</td>\n",
       "      <td>23199010</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000188120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000099721</th>\n",
       "      <td>AMELY</td>\n",
       "      <td>Y</td>\n",
       "      <td>6865917</td>\n",
       "      <td>6911752</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000099721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000242389</th>\n",
       "      <td>RBMY1E</td>\n",
       "      <td>Y</td>\n",
       "      <td>21903617</td>\n",
       "      <td>21918042</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000242389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000172352</th>\n",
       "      <td>CDY1B</td>\n",
       "      <td>Y</td>\n",
       "      <td>24045228</td>\n",
       "      <td>24048019</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000172352</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19393 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_name Chromosome     Start       End Strand  \\\n",
       "gene_id                                                           \n",
       "ENSG00000160072    ATAD3B          1   1471764   1497848      +   \n",
       "ENSG00000142611    PRDM16          1   3069167   3438621      +   \n",
       "ENSG00000157933       SKI          1   2228318   2310213      +   \n",
       "ENSG00000142655     PEX14          1  10472287  10630758      +   \n",
       "ENSG00000149527     PLCH2          1   2425979   2505532      +   \n",
       "...                   ...        ...       ...       ...    ...   \n",
       "ENSG00000187191      DAZ3          Y  24763068  24813492      -   \n",
       "ENSG00000188120      DAZ1          Y  23129354  23199010      -   \n",
       "ENSG00000099721     AMELY          Y   6865917   6911752      -   \n",
       "ENSG00000242389    RBMY1E          Y  21903617  21918042      -   \n",
       "ENSG00000172352     CDY1B          Y  24045228  24048019      -   \n",
       "\n",
       "                     ensemble_id  \n",
       "gene_id                           \n",
       "ENSG00000160072  ENSG00000160072  \n",
       "ENSG00000142611  ENSG00000142611  \n",
       "ENSG00000157933  ENSG00000157933  \n",
       "ENSG00000142655  ENSG00000142655  \n",
       "ENSG00000149527  ENSG00000149527  \n",
       "...                          ...  \n",
       "ENSG00000187191  ENSG00000187191  \n",
       "ENSG00000188120  ENSG00000188120  \n",
       "ENSG00000099721  ENSG00000099721  \n",
       "ENSG00000242389  ENSG00000242389  \n",
       "ENSG00000172352  ENSG00000172352  \n",
       "\n",
       "[19393 rows x 6 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# adata1.var['ensemble_id'] = adata1.var.index\n",
    "adata1.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "25647b60-efc7-463b-9bbb-8e88af9753be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes</th>\n",
       "      <th>U</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGGGTAGCA</th>\n",
       "      <td>16001</td>\n",
       "      <td>0.250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCAAAGTAAGGGT</th>\n",
       "      <td>7864</td>\n",
       "      <td>0.250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCATTCAGGTAGG</th>\n",
       "      <td>15610</td>\n",
       "      <td>0.250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCATTCCAGCCCT</th>\n",
       "      <td>10505</td>\n",
       "      <td>0.250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCATTCGTGACCG</th>\n",
       "      <td>8775</td>\n",
       "      <td>0.250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGTGGTTGTAGCCCTA</th>\n",
       "      <td>10412</td>\n",
       "      <td>1.001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGTGGTTGTTGAAGGC</th>\n",
       "      <td>11811</td>\n",
       "      <td>1.001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGTGTTAGTTGGTAAG</th>\n",
       "      <td>6249</td>\n",
       "      <td>1.001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGTGTTAGTTTGCGAT</th>\n",
       "      <td>2804</td>\n",
       "      <td>1.001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGTGTTGAGCCTATCT</th>\n",
       "      <td>13163</td>\n",
       "      <td>1.001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43532 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  n_genes      U\n",
       "cell_id                         \n",
       "AAACCAAAGGGTAGCA    16001  0.250\n",
       "AAACCAAAGTAAGGGT     7864  0.250\n",
       "AAACCATTCAGGTAGG    15610  0.250\n",
       "AAACCATTCCAGCCCT    10505  0.250\n",
       "AAACCATTCGTGACCG     8775  0.250\n",
       "...                   ...    ...\n",
       "TGTGGTTGTAGCCCTA    10412  1.001\n",
       "TGTGGTTGTTGAAGGC    11811  1.001\n",
       "TGTGTTAGTTGGTAAG     6249  1.001\n",
       "TGTGTTAGTTTGCGAT     2804  1.001\n",
       "TGTGTTGAGCCTATCT    13163  1.001\n",
       "\n",
       "[43532 rows x 2 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concatenated_adata.obs"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
